{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "center"
   },
   "source": [
    "# Assemble the final Dataframe from tarpmir bindingsites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in bindingsite & Ensembl data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:09:14.154242Z",
     "start_time": "2020-10-18T22:09:12.017874Z"
    },
    "cell_style": "center"
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from helper_fcts import *\n",
    "path_tarp = 'C:/Users/Lena/Documents/Master big files/TarPmiR_Windows/data/'\n",
    "path_gdc = 'C:/Users/Lena/Documents/Master big files/GDC_download_legacy/'\n",
    "path2 = 'C:/Users/Lena/Documents/auf mega gesyncte Dokumente/Uni/Master/Semester 5/Masterarbeit/'\n",
    "fig_path = 'Dokumente/plots/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:10:24.395113Z",
     "start_time": "2020-10-18T22:10:03.887413Z"
    }
   },
   "outputs": [],
   "source": [
    "all_mapping, ids = read_in_ids2(path_tarp+'all_mapping_ids.fasta') # mapping gene, transcript, exon ids, all\n",
    "chrom_exon_starts = read_in_fasta2(path_tarp+'all_exon_start.fasta')\n",
    "chrom_exon_ends = read_in_fasta2(path_tarp+'all_exon_end.fasta')\n",
    "exon_starts, exon_ends = calc_exon_data(chrom_exon_starts, chrom_exon_ends)#TODO remove this & instead work with exon_info? slower :(\n",
    "binding_sites, bs_starts, bs_ends = parse_tarp_bs(path_tarp+'new_input/input_miRNA.fasta_exampleA1BG_mRNA.txt.bp')\n",
    "\n",
    "#df_main: bs_id - chr_start - chr_end - miRNA_name - bs_seed - bs_len - tarp_output - transcript_id\n",
    "#df_gene: strand - chrom - gene_id - gene_name - transcript_id (1-N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:10:37.912103Z",
     "start_time": "2020-10-18T22:10:35.733919Z"
    }
   },
   "outputs": [],
   "source": [
    "#parse miRNA and mRNA seed from sequence\n",
    "df_main = binding_sites.copy()\n",
    "miRNA_sequences = parse_seq(path_tarp+'new_input/input_miRNA.fasta')\n",
    "df_main['miRNA_seed'] = df_main.apply(lambda row: miRNA_sequences[row['miRNA']], axis=1)\n",
    "mRNA_sequences = parse_seq(path_tarp+'new_input/exampleA1BG_mRNA.txt')\n",
    "df_main['mRNA_bs_seq'] = df_main.apply(lambda row: mRNA_sequences[row['mRNA']][row['bs_start']:row['bs_end']], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:10:52.459149Z",
     "start_time": "2020-10-18T22:10:49.567017Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>miRNA_seed</th>\n",
       "      <th>mRNA_bs_seq</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>bs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-15.4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003371</td>\n",
       "      <td>0.368</td>\n",
       "      <td>-0.568203</td>\n",
       "      <td>0.198163</td>\n",
       "      <td>-7.757047</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>396</td>\n",
       "      <td>416</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AACTACAGCTGCAGCTACCG</td>\n",
       "      <td>58352138</td>\n",
       "      <td>58352158</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>-23.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.426</td>\n",
       "      <td>0.844482</td>\n",
       "      <td>-0.095065</td>\n",
       "      <td>-5.585164</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>52</td>\n",
       "      <td>75</td>\n",
       "      <td>CTGTACAGCCTCCTAGCTTTCC</td>\n",
       "      <td>GGGTGGCCCAGGAGCCTGTGCAC</td>\n",
       "      <td>58350645</td>\n",
       "      <td>58351407</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-21.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.353</td>\n",
       "      <td>1.253080</td>\n",
       "      <td>0.015255</td>\n",
       "      <td>-8.613887</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>775</td>\n",
       "      <td>794</td>\n",
       "      <td>TGAGGTAGTAGGTTGTGTGGTT</td>\n",
       "      <td>ACTACACCTGCCGCTACCG</td>\n",
       "      <td>58352987</td>\n",
       "      <td>58353006</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-18.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.294</td>\n",
       "      <td>1.302461</td>\n",
       "      <td>-0.092658</td>\n",
       "      <td>-9.582385</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>126</td>\n",
       "      <td>143</td>\n",
       "      <td>TGAGGTAGTAGGTTGTGTGGTT</td>\n",
       "      <td>GGCCGCTACCGCTGCCG</td>\n",
       "      <td>58351458</td>\n",
       "      <td>58351475</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-19.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>0.397</td>\n",
       "      <td>0.062334</td>\n",
       "      <td>0.643183</td>\n",
       "      <td>-8.600034</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>560</td>\n",
       "      <td>580</td>\n",
       "      <td>TGAGGTAGTAGGTTGTGTGGTT</td>\n",
       "      <td>GGCTGCACCACCACCGCCTG</td>\n",
       "      <td>58352400</td>\n",
       "      <td>58352420</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "0    hsa-let-7a-5p  ENST00000600966             0.692308   -15.4     0   \n",
       "1  hsa-let-7a-2-3p  ENST00000600966             0.923077   -23.2     1   \n",
       "2    hsa-let-7b-5p  ENST00000600966             0.692308   -21.3     0   \n",
       "3    hsa-let-7b-5p  ENST00000600966             0.692308   -18.5     0   \n",
       "4    hsa-let-7b-5p  ENST00000600966             0.846154   -19.2     0   \n",
       "\n",
       "   accessibility  AU_content  PhyloP_Stem  PyloP_Flanking      m/e   ...  \\\n",
       "0       0.003371       0.368    -0.568203        0.198163 -7.757047  ...   \n",
       "1       0.000094       0.426     0.844482       -0.095065 -5.585164  ...   \n",
       "2       0.000009       0.353     1.253080        0.015255 -8.613887  ...   \n",
       "3       0.000001       0.294     1.302461       -0.092658 -9.582385  ...   \n",
       "4       0.000174       0.397     0.062334        0.643183 -8.600034  ...   \n",
       "\n",
       "   position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "0                                        10                       6   \n",
       "1                                         1                       6   \n",
       "2                                         1                       7   \n",
       "3                                         1                       6   \n",
       "4                                        12                       8   \n",
       "\n",
       "   difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \\\n",
       "0                                                  1        396     416   \n",
       "1                                                  1         52      75   \n",
       "2                                                  2        775     794   \n",
       "3                                                  1        126     143   \n",
       "4                                                  3        560     580   \n",
       "\n",
       "               miRNA_seed              mRNA_bs_seq  chrom_bs_start  \\\n",
       "0  TGAGGTAGTAGGTTGTATAGTT     AACTACAGCTGCAGCTACCG        58352138   \n",
       "1  CTGTACAGCCTCCTAGCTTTCC  GGGTGGCCCAGGAGCCTGTGCAC        58350645   \n",
       "2  TGAGGTAGTAGGTTGTGTGGTT      ACTACACCTGCCGCTACCG        58352987   \n",
       "3  TGAGGTAGTAGGTTGTGTGGTT        GGCCGCTACCGCTGCCG        58351458   \n",
       "4  TGAGGTAGTAGGTTGTGTGGTT     GGCTGCACCACCACCGCCTG        58352400   \n",
       "\n",
       "  chrom_bs_end bs_id  \n",
       "0     58352158     0  \n",
       "1     58351407     1  \n",
       "2     58353006     2  \n",
       "3     58351475     3  \n",
       "4     58352420     4  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#translate genome position relative to transcript to chromosome\n",
    "def pos_to_chrom(pos, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid not in exon_starts: print(tid,'not in exon_starts')\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if pos >= exon_starts[tid][eid] and pos <= exon_ends[tid][eid]:\n",
    "                diff = pos - exon_starts[tid][eid]\n",
    "                chrom_pos = chrom_exon_starts[tid][eid] + diff\n",
    "                return chrom_pos\n",
    "        else:\n",
    "            print(eid,'not in exon_starts')\n",
    "            return None\n",
    "df_main['chrom_bs_start'] = df_main.apply(lambda row: pos_to_chrom(row.bs_start, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "df_main['chrom_bs_end'] = df_main.apply(lambda row: pos_to_chrom(row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "df_main['bs_id'] = df_main.index\n",
    "df_main.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:17:42.615053Z",
     "start_time": "2020-10-18T22:17:41.195505Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id_version</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>ensembl_transcript_id_version</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id ensembl_gene_id_version ensembl_transcript_id  \\\n",
       "0  ENSG00000210049       ENSG00000210049.1       ENST00000387314   \n",
       "1  ENSG00000211459       ENSG00000211459.2       ENST00000389680   \n",
       "2  ENSG00000210077       ENSG00000210077.1       ENST00000387342   \n",
       "3  ENSG00000210082       ENSG00000210082.2       ENST00000387347   \n",
       "4  ENSG00000209082       ENSG00000209082.1       ENST00000386347   \n",
       "\n",
       "  ensembl_transcript_id_version chromosome_name  strand  start_position  \\\n",
       "0             ENST00000387314.1              MT       1             577   \n",
       "1             ENST00000389680.2              MT       1             648   \n",
       "2             ENST00000387342.1              MT       1            1602   \n",
       "3             ENST00000387347.2              MT       1            1671   \n",
       "4             ENST00000386347.1              MT       1            3230   \n",
       "\n",
       "   end_position  \n",
       "0           647  \n",
       "1          1601  \n",
       "2          1670  \n",
       "3          3229  \n",
       "4          3304  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_gene: strand - chrom - gene_id - gene_name - transcript_id (1-N)\n",
    "df_gene = pd.read_csv(path2+'gene_infos.csv', dtype={'ensembl_gene_id':str,'ensembl_gene_id_version':str,'ensembl_transcript_id':str, 'ensembl_transcript_id_version':str, 'chromosome_name':'category', 'strand':'int8', 'start_position':int, 'end_position':int})\n",
    "del df_gene['Unnamed: 0']\n",
    "df_gene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get TCGA data from GDC API "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:16:01.774905Z",
     "start_time": "2020-10-18T22:16:00.756680Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:16:41.184687Z",
     "start_time": "2020-10-18T22:16:01.300Z"
    }
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def c_split_chrom_column(chrom):\n",
    "    chrom_list = chrom.split(':')\n",
    "    chromosome_name = str(chrom_list[0][3:])\n",
    "    strand = 1 if chrom_list[2] == '+' else -1\n",
    "    chrom_exon_start = int(chrom_list[1].split('-')[0])\n",
    "    chrom_exon_end = int(chrom_list[1].split('-')[1])\n",
    "    return chromosome_name,strand,chrom_exon_start,chrom_exon_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:16:41.200240Z",
     "start_time": "2020-10-18T22:16:01.700Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def list_to_df(list_, nr_columns, data_type=None):\n",
    "    data = []\n",
    "    remove_last_columns = 2 if data_type != None else 0\n",
    "    nr_rows = int((len(list_)-1)/nr_columns)\n",
    "    for row in range(nr_rows):\n",
    "        start = row*nr_columns\n",
    "        end = (row+1)*nr_columns\n",
    "        data.append(list_[start:end-remove_last_columns])\n",
    "    df = pd.DataFrame(data[1:], columns = data[0]) #use first line as header\n",
    "    if data_type=='miRNA':\n",
    "        df = df.astype({'miRNA_ID':str, 'read_count':int})#, 'reads_per_million_miRNA_mapped':float, 'cross-mapped': 'category'})\n",
    "    elif data_type=='exon':\n",
    "        df = df.astype({'exon':str, 'raw_counts':int})#, 'reads_per_million_miRNA_mapped':float, 'cross-mapped': 'category'})\n",
    "        df[['chromosome_name','strand','chrom_exon_start','chrom_exon_end']] = df.exon.apply(lambda x: pd.Series(c_split_chrom_column(str(x))))\n",
    "        df[['raw_counts','chrom_exon_start','chrom_exon_end']] = df[['raw_counts','chrom_exon_start','chrom_exon_end']].apply(pd.to_numeric,downcast='unsigned')\n",
    "        df[['chromosome_name','strand']] = df[['chromosome_name','strand']].astype('str').astype('category')\n",
    "        del df['exon']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.092550Z",
     "start_time": "2020-10-18T22:50:52.800Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#legacy API, get list of files which contain miRNA expression per case_id there\n",
    "import requests\n",
    "import json\n",
    "import re\n",
    "import gzip\n",
    "\n",
    "def download_files_from_legacy(data_type, size, primary_site,write = False):\n",
    "    files_data_type = 'miRNA gene quantification' if data_type == 'miRNA' else 'Exon quantification'\n",
    "    fields = [\n",
    "        \"file_name\",\n",
    "        \"cases.disease_type\",\n",
    "        \"cases.case_id\",\n",
    "        \"data_category\",\n",
    "        \"data_type\",\n",
    "        \"cases.project.primary_site\"\n",
    "        ]\n",
    "\n",
    "    fields = \",\".join(fields)\n",
    "\n",
    "    files_endpt = \"https://api.gdc.cancer.gov/legacy/files\"\n",
    "\n",
    "    # This set of filters is nested under an 'and' operator.\n",
    "    filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\":[\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"cases.project.primary_site\",\n",
    "                \"value\": [primary_site]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.data_type\",\n",
    "                \"value\": [files_data_type]\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "\n",
    "    # A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "    params = {\n",
    "        \"filters\": filters,\n",
    "        \"fields\": fields,\n",
    "        \"format\": \"TSV\",\n",
    "        \"size\": str(size)\n",
    "        }\n",
    "\n",
    "    # The parameters are passed to 'json' rather than 'params' in this case\n",
    "    response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "    df_files = list_to_df(response.content.decode(\"utf-8\").replace(\"\\r\\n\", '\\t').split('\\t'),7)\n",
    "    all_dfs = {}\n",
    "\n",
    "    for i,row in df_files.iterrows():\n",
    "        file_id = row.id\n",
    "        case_id = row['cases.0.case_id']\n",
    "        data_endpt = f\"https://api.gdc.cancer.gov/legacy/data/{file_id}\"\n",
    "\n",
    "        response = requests.get(data_endpt, headers = {\"Content-Type\": \"application/json\"})\n",
    "        response_head_cd = response.headers[\"Content-Disposition\"]\n",
    "        data = str(response.content,'utf-8')\n",
    "        data = data.replace('\\n', '\\t').split('\\t')\n",
    "        new_df = list_to_df(data,4,data_type)\n",
    "        if write:\n",
    "            file_name = f'{path_gdc}{data_type}/{(re.findall(\"filename=(.+)\", response_head_cd)[0])}'\n",
    "            if not file_name.exists():\n",
    "                new_df.to_feather(file_name)\n",
    "        #id_dict = ('.').join(file_name.split('.')[0:3])\n",
    "        #all_dfs[id_dict] = new_df\n",
    "        print(i+1,'/',len(df_files),'done.')\n",
    "    #return all_dfs\n",
    "primary_site='Kidney'\n",
    "download_files_from_legacy('miRNA',3000,primary_site,write=True) #2029 #70min\n",
    "#download_files_from_legacy('exon',2000,primary_site,write=True) #1588 #35h\n",
    "#takes long & needs internet (DONE: all mirna_dfs (but weird only 873 items but 2019 files, maybe case_id not id), 248 / 1588 exon_dfs)\n",
    "#normal 1min 44s - #cython 1min 42s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:12:30.124823Z",
     "start_time": "2020-10-18T22:11:13.316631Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1588 2013\n"
     ]
    }
   ],
   "source": [
    "#read in downloaded counts from feather files\n",
    "#TODO remove later when all data there\n",
    "import os\n",
    "n_all_mirna_dfs = {}\n",
    "dir_ = os.path.join(path_gdc,'mirna')\n",
    "for i, file in enumerate(os.listdir(dir_)):\n",
    "    #print(file)\n",
    "    filename = ('.').join(file.split('.')[0:3])\n",
    "    n_all_mirna_dfs[filename] = pd.read_feather(os.path.join(dir_,file))\n",
    "    \n",
    "n_all_exon_dfs = {}\n",
    "dir_ = os.path.join(path_gdc,'exon')\n",
    "for i, file in enumerate(os.listdir(dir_)):\n",
    "    #print(file)\n",
    "    filename = ('.').join(file.split('.')[0:3])\n",
    "    n_all_exon_dfs[filename] = pd.read_feather(os.path.join(dir_,file))\n",
    "print(len(n_all_exon_dfs),len(n_all_mirna_dfs)) #yes this is right, read in from path\n",
    "all_mirna_dfs = n_all_mirna_dfs\n",
    "all_exon_dfs = n_all_exon_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:08:29.746576Z",
     "start_time": "2020-10-13T08:08:26.600307Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2029 873\n",
      "1588 889\n"
     ]
    }
   ],
   "source": [
    "#print, proof that caseid not unique\n",
    "#TODO remove later when all data there\n",
    "#find id for files - not only caseid: caseid + \n",
    "def print_files_from_legacy(data_type, size, primary_site,write = False):\n",
    "    files_data_type = 'miRNA gene quantification' if data_type == 'miRNA' else 'Exon quantification'\n",
    "    fields = [\n",
    "        \"file_name\",\n",
    "        \"cases.disease_type\",\n",
    "        \"cases.case_id\",\n",
    "        \"data_category\",\n",
    "        \"data_type\",\n",
    "        \"cases.project.primary_site\"\n",
    "        ]\n",
    "\n",
    "    fields = \",\".join(fields)\n",
    "\n",
    "    files_endpt = \"https://api.gdc.cancer.gov/legacy/files\"\n",
    "\n",
    "    # This set of filters is nested under an 'and' operator.\n",
    "    filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\":[\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"cases.project.primary_site\",\n",
    "                \"value\": [primary_site]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.data_type\",\n",
    "                \"value\": [files_data_type]\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "\n",
    "    # A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "    params = {\n",
    "        \"filters\": filters,\n",
    "        \"fields\": fields,\n",
    "        \"format\": \"TSV\",\n",
    "        \"size\": str(size)\n",
    "        }\n",
    "\n",
    "    # The parameters are passed to 'json' rather than 'params' in this case\n",
    "    response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "    df_files = list_to_df(response.content.decode(\"utf-8\").replace(\"\\r\\n\", '\\t').split('\\t'),7)\n",
    "    all_dfs = {}\n",
    "    jo = []\n",
    "    for i,row in df_files.iterrows():\n",
    "        file_id = row.id\n",
    "        case_id = row['cases.0.case_id']\n",
    "        #print(i+1,file_id,case_id)\n",
    "        jo.append(case_id)\n",
    "    print(len(jo),len(set(jo)))\n",
    "    \n",
    "primary_site='Kidney'\n",
    "all_mirna_dfs = print_files_from_legacy('miRNA',3000,'Kidney') #2029 #70min\n",
    "all_exon_dfs = print_files_from_legacy('exon',2000,'Kidney') #1588 #53h\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#save dictionaries to file\n",
    "import pickle\n",
    "path = 'intermed_save'\n",
    "#f = open(path+\"all_mirna_dfs.pkl\",\"wb\")\n",
    "#pickle.dump(all_mirna_dfs,f)\n",
    "#f.close()\n",
    "path = 'intermed_save'\n",
    "#f = open(path+\"all_exon_dfs.pkl\",\"wb\")\n",
    "#pickle.dump(all_exon_dfs,f)\n",
    "#f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T07:59:18.114128Z",
     "start_time": "2020-10-13T07:57:59.785935Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "87 87\n"
     ]
    }
   ],
   "source": [
    "#load dictionaries from file\n",
    "import pickle\n",
    "with open(path_gdc+'miRNA/all_mirna_dfs.pkl', 'rb') as handle:\n",
    "    all_mirna_dfs = pickle.load(handle)\n",
    "with open(path_gdc+'exon/all_exon_dfs.pkl', 'rb') as handle:\n",
    "    all_exon_dfs = pickle.load(handle)\n",
    "print(len(all_exon_dfs),len(all_mirna_dfs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:19:30.847655Z",
     "start_time": "2020-10-18T22:19:30.508993Z"
    }
   },
   "outputs": [],
   "source": [
    "#TODO for every case do\n",
    "#for _,row in df_exons.iterrows():\n",
    "#case_id =  row['cases.0.case_id']\n",
    "\n",
    "case_id = 'unc.edu.84675636-1d8c-474a-a4bf-baffcc36b5bf'\n",
    "important_counts = df_main[['bs_id','miRNA','mRNA','chrom_bs_start','chrom_bs_end']].copy()\n",
    "important_counts = important_counts.merge(df_gene[['ensembl_transcript_id','chromosome_name','strand']], left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "important_counts[['chromosome_name','strand','miRNA','mRNA']] = important_counts[['chromosome_name','strand','miRNA','mRNA']].astype('str').astype('category')\n",
    "del important_counts['ensembl_transcript_id']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## translate TCGA exon counts to GRCh38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:24:48.414251Z",
     "start_time": "2020-10-18T22:24:48.298344Z"
    }
   },
   "outputs": [],
   "source": [
    "exon_counts = all_exon_dfs[case_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write exoncounts to BED file for translating GRCh37 to GRCh38\n",
    "\n",
    "file = open('chrom_pos_hg19_'+primary_site+'.bed', 'w')\n",
    "for row in exon_counts.iterrows():\n",
    "    line = row[0]\n",
    "    sep = '\\t'\n",
    "    strand = '+' if (row[1].strand=='1') else '-'\n",
    "    start = row[1].chrom_exon_start if strand == '+' else row[1].chrom_exon_end\n",
    "    end = row[1].chrom_exon_end if strand == '+' else row[1].chrom_exon_start\n",
    "    string = f'chr{row[1].chromosome_name}{sep}{start}{sep}{end}{sep}{line}{sep}0{sep}{strand}\\n' \n",
    "    #chr7  127471196  127472363  Pos1  0  +\n",
    "    file.write(string)\n",
    "file.close()\n",
    "#TODO then upload to website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os #pyliftover doesnt work cause doesnt work with ranges\n",
    "import subprocess\n",
    "if os.name == 'posix':\n",
    "    bashCommand = \"liftOver chrom_pos_hg19_\"+primary_site+\".bed hg19ToHg37.over.chain.gz chrom_pos_translated.bed unlifted.bed\"\n",
    "    process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)\n",
    "    output, error = process.communicate()\n",
    "    print(output, error)\n",
    "elif os.name == 'nt':\n",
    "    print(\"liftOver doesn't exist for Windows, manually upload BED file to website\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:20:21.725496Z",
     "start_time": "2020-10-18T22:20:06.558124Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12595</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12646</td>\n",
       "      <td>12697</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name  chrom_exon_start  chrom_exon_end strand\n",
       "index                                                         \n",
       "0                   1             11874           12227      1\n",
       "1                   1             12595           12721      1\n",
       "2                   1             12613           12721      1\n",
       "3                   1             12646           12697      1\n",
       "4                   1             13221           14409      1"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#then read in new liftover exoncounts in GRCh38 from files\n",
    "df_liftover = pd.read_csv('chrom_pos_translated.bed', sep=\"\\t\",names=['chromosome_name','chrom_exon_start','chrom_exon_end','index','0','strand'])\n",
    "df_liftover = df_liftover.set_index('index')\n",
    "df_liftover.chromosome_name = df_liftover.apply(lambda row: row.chromosome_name[3:], axis=1)\n",
    "df_liftover.strand = df_liftover.apply(lambda row: '1' if (row.strand=='+') else '-1', axis=1)\n",
    "del df_liftover['0']\n",
    "df_liftover.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:26:35.739119Z",
     "start_time": "2020-10-18T22:26:35.569777Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>raw_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>1</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12595</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12646</td>\n",
       "      <td>12697</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name  chrom_exon_start  chrom_exon_end strand  raw_counts\n",
       "index                                                                     \n",
       "0                   1             11874           12227      1        15.0\n",
       "1                   1             12595           12721      1         3.0\n",
       "2                   1             12613           12721      1         3.0\n",
       "3                   1             12646           12697      1         2.0\n",
       "4                   1             13221           14409      1        67.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# changes of exon_counts (GRCh37) to new_exon_counts(GRCh38)\n",
    "new_exon_counts = df_liftover.join(exon_counts,rsuffix='r',lsuffix='l')#['raw_counts'])\n",
    "#print('indices, which got lost during translation with liftover:',set(range(239885))-set(new_exon_counts.index))\n",
    "new_exon_counts = df_liftover.join(exon_counts['raw_counts'])\n",
    "new_exon_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map exons to bs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:51:10.878073Z",
     "start_time": "2020-10-18T22:50:11.327509Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>raw_counts</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>58352928</td>\n",
       "      <td>58353197</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216603</td>\n",
       "      <td>2</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58352987</td>\n",
       "      <td>58353006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>58351391</td>\n",
       "      <td>58351687</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216600</td>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58351458</td>\n",
       "      <td>58351475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>58352283</td>\n",
       "      <td>58352555</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216602</td>\n",
       "      <td>4</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58352400</td>\n",
       "      <td>58352420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>58351391</td>\n",
       "      <td>58351687</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216600</td>\n",
       "      <td>5</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58351648</td>\n",
       "      <td>58351684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>58351391</td>\n",
       "      <td>58351687</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216600</td>\n",
       "      <td>8</td>\n",
       "      <td>hsa-let-7d-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58351399</td>\n",
       "      <td>58351420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome_name  chrom_exon_start  chrom_exon_end strand  raw_counts  \\\n",
       "0              19          58352928        58353197     -1           0   \n",
       "1              19          58351391        58351687     -1           0   \n",
       "2              19          58352283        58352555     -1           0   \n",
       "3              19          58351391        58351687     -1           0   \n",
       "4              19          58351391        58351687     -1           0   \n",
       "\n",
       "   exon_id  bs_id          miRNA             mRNA  chrom_bs_start  \\\n",
       "0   216603      2  hsa-let-7b-5p  ENST00000600966        58352987   \n",
       "1   216600      3  hsa-let-7b-5p  ENST00000600966        58351458   \n",
       "2   216602      4  hsa-let-7b-5p  ENST00000600966        58352400   \n",
       "3   216600      5  hsa-let-7b-3p  ENST00000600966        58351648   \n",
       "4   216600      8  hsa-let-7d-5p  ENST00000600966        58351399   \n",
       "\n",
       "   chrom_bs_end  \n",
       "0      58353006  \n",
       "1      58351475  \n",
       "2      58352420  \n",
       "3      58351684  \n",
       "4      58351420  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#for a large data set, you will likely get a significant speed increase by creating an index for any column(s) used in the join condition. \n",
    "#https://stackoverflow.com/questions/30627968/merge-pandas-dataframes-where-one-value-is-between-two-others\n",
    "#instead use SQL\n",
    "import sqlite3\n",
    "conn = sqlite3.connect(':memory:') #Make the database in memory\n",
    "\n",
    "#write the tables\n",
    "important_counts.to_sql('bs', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "new_exon_counts.to_sql('counts', conn, index=True, index_label='exon_id', dtype={\"raw_counts\": 'INTEGER', \"exon_id\": 'INTEGER'})\n",
    "qry = '''\n",
    "    select\n",
    "        counts.chromosome_name, counts.chrom_exon_start, counts.chrom_exon_end, counts.strand, counts.raw_counts, counts.exon_id, bs.bs_id, bs.miRNA, bs.mRNA, bs.chrom_bs_start, bs.chrom_bs_end\n",
    "    from\n",
    "        bs inner join counts on\n",
    "        bs.chromosome_name = counts.chromosome_name and bs.strand = counts.strand and bs.chrom_bs_start >= counts.chrom_exon_start and bs.chrom_bs_end <= counts.chrom_exon_end\n",
    "    '''\n",
    "bs_exons = pd.read_sql_query(qry, conn) #only return bs with minimum 1 exon_id cause inner join (else left join)\n",
    "bs_exons.head()\n",
    "#TODO filter out all binding sites that have more than 1 exon_id: none for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T00:31:04.522842Z",
     "start_time": "2020-10-13T00:31:04.432995Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#visualize amount exons by bindingsite | amount bindingsites by exon\n",
    "import sqlite3\n",
    "conn = sqlite3.connect(':memory:') #Make the database in memory\n",
    "\n",
    "#write the tables\n",
    "important_counts.to_sql('bs', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "new_exon_counts.to_sql('counts', conn, index=True, index_label='exon_id', dtype={\"exon_id\": 'INTEGER'})\n",
    "qry = '''\n",
    "    select\n",
    "        counts.chromosome_name, counts.chrom_exon_start, counts.chrom_exon_end, counts.strand, counts.raw_counts, counts.exon_id, bs.bs_id, bs.miRNA, bs.mRNA, bs.chrom_bs_start, bs.chrom_bs_end\n",
    "    from\n",
    "        bs left join counts on\n",
    "        bs.chromosome_name = counts.chromosome_name and bs.strand = counts.strand and bs.chrom_bs_start >= counts.chrom_exon_start and bs.chrom_bs_end <= counts.chrom_exon_end\n",
    "    '''\n",
    "bs_exons2 = pd.read_sql_query(qry, conn)\n",
    "bs_exons2.head()\n",
    "#bs_exons2.groupby('bs_id').count().raw_counts.hist()\n",
    "#bs_exons2.groupby('exon_id').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map miRNA to bs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.058588Z",
     "start_time": "2020-10-18T22:52:42.986240Z"
    }
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'24ea3cd5-c2e6-44f7-9658-78af7befeef7'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-29-ecc3fe5af74e>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m#map miRNA ID to mature ID\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mmirna_counts\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mall_mirna_dfs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mcase_id\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mtrans\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfrom_dict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mread_in_miRNA\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpath2\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'miRNA2mature.fasta'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0morient\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'index'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;31m#,columns=[\"mirna\", \"mature\"])\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mtrans\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtrans\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m{\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;34m\"miRNA_ID\"\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: '24ea3cd5-c2e6-44f7-9658-78af7befeef7'"
     ]
    }
   ],
   "source": [
    "#map miRNA ID to mature ID\n",
    "mirna_counts = all_mirna_dfs[case_id]\n",
    "\n",
    "trans = pd.DataFrame.from_dict(read_in_miRNA(path2+'miRNA2mature.fasta'), orient='index')#,columns=[\"mirna\", \"mature\"])\n",
    "trans = trans.rename(columns={0: \"miRNA_ID\"})\n",
    "trans[['miRNA_ID']] = trans[['miRNA_ID']].astype('str').astype('category')\n",
    "\n",
    "#map miRNA to miRNA family\n",
    "mirna_families = read_in_mirna_families(path_tarp+'mapping_miRNA_families.fasta')\n",
    "trans = trans.join(mirna_families,how='inner')\n",
    "\n",
    "bs_mirnas = bs_exons[['exon_id','raw_counts','miRNA']].join(trans,on='miRNA').merge(mirna_counts[['miRNA_ID','read_count']],on='miRNA_ID')\n",
    "#bs_mirnas.drop(['miRNA','miRNA_ID'], axis=1, inplace=True)\n",
    "bs_mirnas = bs_mirnas.rename(columns={'read_count': \"mirna_read_count\",'raw_counts':'exon_raw_counts','miRNA family name':'mirna_family'})\n",
    "bs_mirnas[['mirna_read_count']] = bs_mirnas[['mirna_read_count']].astype('int')#TODO do this early before saving\n",
    "bs_mirnas.head()\n",
    "\n",
    "#2693 with hsa in miRNA2mature / 49156 overall in miRNA2mature (also other species eg cel)\n",
    "#2656 if inner join cause only 2656 families\n",
    "#TODO check if miRNA counts from TCGA are more different miRNAs?\n",
    "#TODO is inner right? are some in important counts None?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T01:30:28.905033Z",
     "start_time": "2020-10-13T01:30:28.803071Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_raw_counts</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>mirna_read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>let-7</td>\n",
       "      <td>55647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7d-5p</td>\n",
       "      <td>hsa-let-7d</td>\n",
       "      <td>let-7</td>\n",
       "      <td>3767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>let-7</td>\n",
       "      <td>11424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>let-7</td>\n",
       "      <td>11424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-15a-3p</td>\n",
       "      <td>hsa-mir-15a</td>\n",
       "      <td>mir-15</td>\n",
       "      <td>2258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12562</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-5692a</td>\n",
       "      <td>hsa-mir-5692a-2</td>\n",
       "      <td>mir-5692</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12563</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-548az-5p</td>\n",
       "      <td>hsa-mir-548az</td>\n",
       "      <td>mir-548</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12566</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-8062</td>\n",
       "      <td>hsa-mir-8062</td>\n",
       "      <td>hsa-miR-8062</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12571</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-8064</td>\n",
       "      <td>hsa-mir-8064</td>\n",
       "      <td>hsa-miR-8064</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12578</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1257</td>\n",
       "      <td>hsa-mir-1257</td>\n",
       "      <td>mir-1257</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2129 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_id  exon_raw_counts             miRNA         miRNA_ID  \\\n",
       "7       216595                0     hsa-let-7b-5p       hsa-let-7b   \n",
       "14      216595                0     hsa-let-7d-5p       hsa-let-7d   \n",
       "25      216595                0     hsa-let-7e-3p       hsa-let-7e   \n",
       "33      216595                0     hsa-let-7e-3p       hsa-let-7e   \n",
       "46      216595                0    hsa-miR-15a-3p      hsa-mir-15a   \n",
       "...        ...              ...               ...              ...   \n",
       "12562   216595                0     hsa-miR-5692a  hsa-mir-5692a-2   \n",
       "12563   216595                0  hsa-miR-548az-5p    hsa-mir-548az   \n",
       "12566   216595                0      hsa-miR-8062     hsa-mir-8062   \n",
       "12571   216595                0      hsa-miR-8064     hsa-mir-8064   \n",
       "12578   216595                0      hsa-miR-1257     hsa-mir-1257   \n",
       "\n",
       "       mirna_family  mirna_read_count  \n",
       "7             let-7             55647  \n",
       "14            let-7              3767  \n",
       "25            let-7             11424  \n",
       "33            let-7             11424  \n",
       "46           mir-15              2258  \n",
       "...             ...               ...  \n",
       "12562      mir-5692                 0  \n",
       "12563       mir-548                 0  \n",
       "12566  hsa-miR-8062                 0  \n",
       "12571  hsa-miR-8064                 0  \n",
       "12578      mir-1257                 0  \n",
       "\n",
       "[2129 rows x 6 columns]"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs_mirnas[bs_mirnas.exon_id ==216595]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T01:29:10.977928Z",
     "start_time": "2020-10-13T01:29:10.892977Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_raw_counts</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>mirna_read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3365</th>\n",
       "      <td>216603</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371</th>\n",
       "      <td>216596</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exon_id  exon_raw_counts            miRNA      miRNA_ID  \\\n",
       "3365   216603                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3368   216595                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3370   216595                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3371   216596                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "\n",
       "         mirna_family  mirna_read_count  \n",
       "3365  hsa-miR-1229-5p                 4  \n",
       "3368  hsa-miR-1229-5p                 4  \n",
       "3370  hsa-miR-1229-5p                 4  \n",
       "3371  hsa-miR-1229-5p                 4  "
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs_mirnas[bs_mirnas.mirna_family =='hsa-miR-1229-5p']\n",
    "\n",
    "# 1 right? count is added for exonid 216595, but do i already add it before ?\n",
    "# 2 also : exon_id + mirna double (3368 & 3370) wrong!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.071562Z",
     "start_time": "2020-10-18T22:50:38.700Z"
    }
   },
   "outputs": [],
   "source": [
    "#output: for each case_id for each bs_id: interesting miRNA expression + exon expression\n",
    "#unstack family counts from 2 columns to several columns (mirna_family=name,mirna_read_count=value)\n",
    "pivoted = bs_mirnas.copy()\n",
    "pivoted = pivoted.pivot_table(values='mirna_read_count', index=['exon_id', 'exon_raw_counts'], columns='mirna_family', aggfunc='sum', fill_value=0)\n",
    "pivoted#.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualize amount exons by bindingsite\n",
    "#bs_exons.groupby('bs_id').count().raw_counts.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:40:44.254071Z",
     "start_time": "2020-10-13T08:40:44.166104Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8, 1680) (8,)\n"
     ]
    }
   ],
   "source": [
    "#TODO check - is nr mirna downloaded also here? if not: mistake\n",
    "print(X.shape,y[0].shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Elastic Net Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.081556Z",
     "start_time": "2020-10-18T22:50:43.000Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.getsizeof(pivoted)#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:42:05.340886Z",
     "start_time": "2020-10-13T08:42:05.198969Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0. 0. 0. ... 0. 0. 0.]\n",
      "0.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Lena\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\linear_model\\_coordinate_descent.py:529: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 0.0, tolerance: 0.0\n",
      "  model = cd_fast.enet_coordinate_descent(\n"
     ]
    }
   ],
   "source": [
    "#regression btw miRNA expression, exon expression\n",
    "import sklearn\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.datasets import make_regression\n",
    "#faster if np.array(x, order='F')\n",
    "#If you are interested in controlling the L1 and L2 penalty separately, keep in mind that this is equivalent to:\n",
    "#a * L1 + b * L2 where: alpha = a + b and l1_ratio = a / (a + b)\n",
    "X = np.array(pivoted)\n",
    "y = np.array(pivoted.index.get_level_values('exon_raw_counts'))\n",
    "\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "print(regr.coef_)\n",
    "print(regr.intercept_)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get exonid for exon for database later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO this is not necessary anymore if the new exon counts work!!!!\n",
    "#instead try directly with ensembl start & end match - if it doesnt work \n",
    "#-> try download from https://www.gencodegenes.org/human/ (GRCh38.p13) if same reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:42:06.292710Z",
     "start_time": "2020-10-18T22:30:34.453194Z"
    }
   },
   "outputs": [],
   "source": [
    "#read in exon ids + starts and ends from GENCODE GTF file (ca 12min)\n",
    "#import the file into a local sqlite3 file-based database\n",
    "import gffutils\n",
    "fn = gffutils.example_filename(path2+'gencode.v19.annotation.gtf_withproteinids')\n",
    "db = gffutils.create_db(fn, ':memory:', keep_order=True, disable_infer_genes=True, disable_infer_transcripts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:28:13.015971Z",
     "start_time": "2020-10-18T22:28:12.951921Z"
    }
   },
   "outputs": [],
   "source": [
    "new_exon_counts['exonname'] = np.nan "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:48:41.381066Z",
     "start_time": "2020-10-18T22:48:41.296112Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\texon\t11869\t12227\t.\t+\t.\tgene_id \"ENSG00000223972.4\"; transcript_id \"ENST00000456328.2\"; gene_type \"pseudogene\"; gene_status \"KNOWN\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_status \"KNOWN\"; transcript_name \"DDX11L1-002\"; level \"2\"; havana_gene \"OTTHUMG00000000961.2\"; tag \"basic\"; havana_transcript \"OTTHUMT00000362751.1\"; exon_number \"1\"; exon_id \"ENSE00002234944.1\";\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'Feature' object has no attribute 'chromosome'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-27-4f7f5b7e5e1c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mexon\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mdb\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfeatures_of_type\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'exon'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mexon\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m     \u001b[0mis_same\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mnew_exon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchromosome_name\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mexon\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchromosome\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mnew_exon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchrom_exon_start\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mexon\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mnew_exon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchrom_exon_end\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mexon\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstop\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mnew_exon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstrand\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mexon\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstrand\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m     \u001b[0mnew_exon_counts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mis_same\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexonname\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mexon\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexon_id\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnew_exon_counts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mis_same\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'Feature' object has no attribute 'chromosome'"
     ]
    }
   ],
   "source": [
    "for exon in db.features_of_type('exon'):\n",
    "    print(exon)#TODO\n",
    "    is_same = (new_exon_counts.chromosome_name == exon.chromosome) and (new_exon_counts.chrom_exon_start == exon.start) and (new_exon_counts.chrom_exon_end == exon.stop) and (new_exon_counts.strand == exon.strand)\n",
    "    new_exon_counts[is_same].exonname = exon.exon_id\n",
    "    print(len(new_exon_counts[is_same]))\n",
    "    #exons = list(db.children(gene, featuretype='exon'))\n",
    "    #first_3_exons = exons[:3]\n",
    "    #last_3_exons = exons[-3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#exon line\n",
    "#chr1\tHAVANA\texon\t11869\t12227\t.\t+\t.\tgene_id \"ENSG00000223972.4\"; \n",
    "#transcript_id \"ENST00000456328.2\"; gene_type \"pseudogene\"; gene_status \"KNOWN\"; \n",
    "#gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_status \"KNOWN\"; \n",
    "#transcript_name \"DDX11L1-002\"; exon_number 1; exon_id \"ENSE00002234944.1\"; level 2; tag \"basic\"; \n",
    "#havana_gene \"OTTHUMG00000000961.2\"; havana_transcript \"OTTHUMT00000362751.1\";"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "Seqvec basic Seq ligand binding.ipynb",
   "provenance": []
  },
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python (thesis)",
   "language": "python",
   "name": "thesis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "581.6px",
    "left": "235px",
    "top": "158.533px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
