{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "center"
   },
   "source": [
    "# Assemble the final Dataframe from tarpmir bindingsites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T17:05:28.291804Z",
     "start_time": "2020-10-02T17:05:28.053945Z"
    },
    "cell_style": "center"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from helper_fcts import *\n",
    "path_tarp = 'C:/Users/Lena/Documents/Master big files/TarPmiR_Windows/data/'\n",
    "path_gdc = 'C:/Users/Lena/Documents/Master big files/GDC_download_legacy/'\n",
    "path2 = 'C:/Users/Lena/Documents/auf mega gesyncte Dokumente/Uni/Master/Semester 5/Masterarbeit/'\n",
    "fig_path = 'Dokumente/plots/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T15:44:31.809343Z",
     "start_time": "2020-10-03T15:44:11.479538Z"
    }
   },
   "outputs": [],
   "source": [
    "all_mapping, ids = read_in_ids2(path_tarp+'all_mapping_ids.fasta') # mapping gene, transcript, exon ids, all\n",
    "chrom_exon_starts = read_in_fasta2(path_tarp+'all_exon_start.fasta')\n",
    "chrom_exon_ends = read_in_fasta2(path_tarp+'all_exon_end.fasta')\n",
    "exon_starts, exon_ends = calc_exon_data(chrom_exon_starts, chrom_exon_ends)#TODO remove this & instead work with exon_info? slower :(\n",
    "binding_sites, bs_starts, bs_ends = parse_tarp_bs(path_tarp+'new_input/input_miRNA.fasta_exampleA1BG_mRNA.txt.bp')\n",
    "exon_info = pd.read_feather(path2+'exon_level_info.feather', use_threads=True)\n",
    "exon_info = exon_info.rename(columns={'exon_chrom_start': 'chrom_exon_start','exon_chrom_end':'chrom_exon_end'})\n",
    "exon_info[['chrom_exon_start','chrom_exon_end']] = exon_info[['chrom_exon_start','chrom_exon_end']].apply(pd.to_numeric,downcast='unsigned')\n",
    "exon_info[['ensembl_gene_id','ensembl_transcript_id','ensembl_exon_id','chromosome_name','strand']] = exon_info[['ensembl_gene_id','ensembl_transcript_id','ensembl_exon_id','chromosome_name','strand']].astype('str').astype('category')\n",
    "\n",
    "#TODO here\n",
    "#several dataframes\n",
    "#0) df_main: bs_id - chr_start - chr_end - miRNA_name - bs_seed - bs_len - tarp_output - transcript_id\n",
    "#1) df_exon: bs_id - exon_id (cause N-N)\n",
    "#2) df_gene: strand - chrom - gene_id - gene_name - transcript_id (1-N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T19:52:10.543474Z",
     "start_time": "2020-10-03T19:41:26.653061Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-251-6fa8e56a8aeb>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     22\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mstarts\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mends\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     23\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 24\u001b[1;33m \u001b[0mexon_info\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'exon_start'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexon_info\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'exon_end'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcalc_exon_start\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mexon_info\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     25\u001b[0m \u001b[0mexon_info\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-251-6fa8e56a8aeb>\u001b[0m in \u001b[0;36mcalc_exon_start\u001b[1;34m(exon_info)\u001b[0m\n\u001b[0;32m      3\u001b[0m     \u001b[0mends\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mzeros\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mexon_info\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mint\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0mtid\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mexon_info\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mensembl_transcript_id\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0munique\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m         \u001b[0mtids\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mexon_info\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mexon_info\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mensembl_transcript_id\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mtid\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mby\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'exon_chrom_start'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m         \u001b[0mlast_exon_end\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtids\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\common.py\u001b[0m in \u001b[0;36mnew_method\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     62\u001b[0m         \u001b[0mother\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mitem_from_zerodim\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mother\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 64\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mother\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     65\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     66\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mnew_method\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\__init__.py\u001b[0m in \u001b[0;36mwrapper\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m    527\u001b[0m         \u001b[0mrvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mextract_array\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mother\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mextract_numpy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    528\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 529\u001b[1;33m         \u001b[0mres_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcomparison_op\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    530\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    531\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0m_construct_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mres_values\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mres_name\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py\u001b[0m in \u001b[0;36mcomparison_op\u001b[1;34m(left, right, op)\u001b[0m\n\u001b[0;32m    245\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    246\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlvalues\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 247\u001b[1;33m         \u001b[0mres_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcomp_method_OBJECT_ARRAY\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mop\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrvalues\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    248\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    249\u001b[0m     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py\u001b[0m in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[1;34m(op, x, y)\u001b[0m\n\u001b[0;32m     56\u001b[0m     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     57\u001b[0m         \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlibops\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mscalar_compare\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 58\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     59\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     60\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#skip, takes too long, > 10min -> use dicts instead \n",
    "def calc_exon_start(exon_info):\n",
    "    starts = np.zeros(len(exon_info),dtype=int)\n",
    "    ends = np.zeros(len(exon_info),dtype=int)\n",
    "    for tid in exon_info.ensembl_transcript_id.unique():\n",
    "        tids = exon_info[exon_info.ensembl_transcript_id == tid].sort_values(by='exon_chrom_start')\n",
    "        last_exon_end = 0\n",
    "        for i in range(len(tids)):\n",
    "            row = tids[i:i+1]\n",
    "            this_pos = tids.index.get_loc(row.index[0]) #get position of exon within transcript\n",
    "            ov_pos = exon_info.index.get_loc(row.index[0])\n",
    "            if this_pos == 0:\n",
    "                starts[ov_pos] = 0\n",
    "                end = (row.exon_chrom_end - row.exon_chrom_start).values[0]\n",
    "                ends[ov_pos] = end\n",
    "                last_exon_end = end\n",
    "            else:\n",
    "                start = last_exon_end + 1  #get end of exon before\n",
    "                starts[ov_pos] = start\n",
    "                end = start + (row.exon_chrom_end - row.exon_chrom_start).values[0]\n",
    "                last_exon_end = end\n",
    "                ends[ov_pos] = end\n",
    "    return starts, ends\n",
    "\n",
    "#exon_info['exon_start'], exon_info['exon_end'] = calc_exon_start(exon_info)\n",
    "#exon_info\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse seed & translate genome positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T16:26:51.156326Z",
     "start_time": "2020-10-02T16:26:49.145049Z"
    }
   },
   "outputs": [],
   "source": [
    "df_main = binding_sites.copy()\n",
    "miRNA_sequences = parse_seq(path_tarp+'new_input/input_miRNA.fasta')\n",
    "df_main['miRNA_seed'] = df_main.apply(lambda row: miRNA_sequences[row['miRNA']], axis=1)\n",
    "mRNA_sequences = parse_seq(path_tarp+'new_input/exampleA1BG_mRNA.txt')\n",
    "df_main['mRNA_bs_seq'] = df_main.apply(lambda row: mRNA_sequences[row['mRNA']][row['bs_start']:row['bs_end']], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T16:26:59.874070Z",
     "start_time": "2020-10-02T16:26:56.565549Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>miRNA_seed</th>\n",
       "      <th>mRNA_bs_seq</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>bs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-15.4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003371</td>\n",
       "      <td>0.368</td>\n",
       "      <td>-0.568203</td>\n",
       "      <td>0.198163</td>\n",
       "      <td>-7.757047</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>396</td>\n",
       "      <td>416</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AACTACAGCTGCAGCTACCG</td>\n",
       "      <td>58352138</td>\n",
       "      <td>58352158</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>-23.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.426</td>\n",
       "      <td>0.844482</td>\n",
       "      <td>-0.095065</td>\n",
       "      <td>-5.585164</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>52</td>\n",
       "      <td>75</td>\n",
       "      <td>CTGTACAGCCTCCTAGCTTTCC</td>\n",
       "      <td>GGGTGGCCCAGGAGCCTGTGCAC</td>\n",
       "      <td>58350645</td>\n",
       "      <td>58351407</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-21.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.353</td>\n",
       "      <td>1.253080</td>\n",
       "      <td>0.015255</td>\n",
       "      <td>-8.613887</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>775</td>\n",
       "      <td>794</td>\n",
       "      <td>TGAGGTAGTAGGTTGTGTGGTT</td>\n",
       "      <td>ACTACACCTGCCGCTACCG</td>\n",
       "      <td>58352987</td>\n",
       "      <td>58353006</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-18.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.294</td>\n",
       "      <td>1.302461</td>\n",
       "      <td>-0.092658</td>\n",
       "      <td>-9.582385</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>126</td>\n",
       "      <td>143</td>\n",
       "      <td>TGAGGTAGTAGGTTGTGTGGTT</td>\n",
       "      <td>GGCCGCTACCGCTGCCG</td>\n",
       "      <td>58351458</td>\n",
       "      <td>58351475</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-19.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>0.397</td>\n",
       "      <td>0.062334</td>\n",
       "      <td>0.643183</td>\n",
       "      <td>-8.600034</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>560</td>\n",
       "      <td>580</td>\n",
       "      <td>TGAGGTAGTAGGTTGTGTGGTT</td>\n",
       "      <td>GGCTGCACCACCACCGCCTG</td>\n",
       "      <td>58352400</td>\n",
       "      <td>58352420</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "0    hsa-let-7a-5p  ENST00000600966             0.692308   -15.4     0   \n",
       "1  hsa-let-7a-2-3p  ENST00000600966             0.923077   -23.2     1   \n",
       "2    hsa-let-7b-5p  ENST00000600966             0.692308   -21.3     0   \n",
       "3    hsa-let-7b-5p  ENST00000600966             0.692308   -18.5     0   \n",
       "4    hsa-let-7b-5p  ENST00000600966             0.846154   -19.2     0   \n",
       "\n",
       "   accessibility  AU_content  PhyloP_Stem  PyloP_Flanking      m/e   ...  \\\n",
       "0       0.003371       0.368    -0.568203        0.198163 -7.757047  ...   \n",
       "1       0.000094       0.426     0.844482       -0.095065 -5.585164  ...   \n",
       "2       0.000009       0.353     1.253080        0.015255 -8.613887  ...   \n",
       "3       0.000001       0.294     1.302461       -0.092658 -9.582385  ...   \n",
       "4       0.000174       0.397     0.062334        0.643183 -8.600034  ...   \n",
       "\n",
       "   position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "0                                        10                       6   \n",
       "1                                         1                       6   \n",
       "2                                         1                       7   \n",
       "3                                         1                       6   \n",
       "4                                        12                       8   \n",
       "\n",
       "   difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \\\n",
       "0                                                  1        396     416   \n",
       "1                                                  1         52      75   \n",
       "2                                                  2        775     794   \n",
       "3                                                  1        126     143   \n",
       "4                                                  3        560     580   \n",
       "\n",
       "               miRNA_seed              mRNA_bs_seq  chrom_bs_start  \\\n",
       "0  TGAGGTAGTAGGTTGTATAGTT     AACTACAGCTGCAGCTACCG        58352138   \n",
       "1  CTGTACAGCCTCCTAGCTTTCC  GGGTGGCCCAGGAGCCTGTGCAC        58350645   \n",
       "2  TGAGGTAGTAGGTTGTGTGGTT      ACTACACCTGCCGCTACCG        58352987   \n",
       "3  TGAGGTAGTAGGTTGTGTGGTT        GGCCGCTACCGCTGCCG        58351458   \n",
       "4  TGAGGTAGTAGGTTGTGTGGTT     GGCTGCACCACCACCGCCTG        58352400   \n",
       "\n",
       "  chrom_bs_end bs_id  \n",
       "0     58352158     0  \n",
       "1     58351407     1  \n",
       "2     58353006     2  \n",
       "3     58351475     3  \n",
       "4     58352420     4  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def pos_to_chrom(pos, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid not in exon_starts: print(tid,'not in exon_starts')\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if pos >= exon_starts[tid][eid] and pos <= exon_ends[tid][eid]:\n",
    "                diff = pos - exon_starts[tid][eid]\n",
    "                chrom_pos = chrom_exon_starts[tid][eid] + diff\n",
    "                return chrom_pos\n",
    "        else:\n",
    "            print(eid,'not in exon_starts')\n",
    "            return None\n",
    "df_main['chrom_bs_start'] = df_main.apply(lambda row: pos_to_chrom(row.bs_start, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "df_main['chrom_bs_end'] = df_main.apply(lambda row: pos_to_chrom(row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "df_main['bs_id'] = df_main.index\n",
    "df_main.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse from mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T19:57:33.198039Z",
     "start_time": "2020-10-03T19:57:30.939200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00003224625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003509215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003142302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003185372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003509215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24709</th>\n",
       "      <td>24709</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24710</th>\n",
       "      <td>24710</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24711</th>\n",
       "      <td>24711</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24712</th>\n",
       "      <td>24712</td>\n",
       "      <td>ENSE00003024001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24713</th>\n",
       "      <td>24713</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26043 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       bs_id          exon_id\n",
       "0          0  ENSE00003224625\n",
       "1          1  ENSE00003509215\n",
       "1          1  ENSE00003142302\n",
       "2          2  ENSE00003185372\n",
       "3          3  ENSE00003509215\n",
       "...      ...              ...\n",
       "24709  24709  ENSE00003207101\n",
       "24710  24710  ENSE00003207101\n",
       "24711  24711  ENSE00003207101\n",
       "24712  24712  ENSE00003024001\n",
       "24713  24713  ENSE00003207101\n",
       "\n",
       "[26043 rows x 2 columns]"
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_exon = mapping bs_id - exon_id\n",
    "def get_exon_ids(transcript_id, chrom_bs_start, chrom_bs_end, chrom_exon_starts, chrom_exon_ends):\n",
    "    exon_ids = []\n",
    "    for eid in chrom_exon_starts[transcript_id]:\n",
    "        chrom_exon_start = chrom_exon_starts[transcript_id][eid]\n",
    "        chrom_exon_end = chrom_exon_ends[transcript_id][eid]\n",
    "        #TODO strand beachten? look at results? do they make sense?\n",
    "        if not((chrom_bs_start <= chrom_exon_start and chrom_bs_end <= chrom_exon_start) or (chrom_bs_start >= chrom_exon_end and chrom_bs_end >= chrom_exon_end)):\n",
    "            exon_ids.append(eid)\n",
    "    return exon_ids\n",
    "\n",
    "df_exon = df_main[['mRNA','bs_id','chrom_bs_start','chrom_bs_end']].copy()\n",
    "df_exon['exon_id'] = df_exon.apply(lambda row: get_exon_ids(row.mRNA, row.chrom_bs_start, row.chrom_bs_end, chrom_exon_starts, chrom_exon_ends), axis=1)\n",
    "df_exon = df_exon.explode('exon_id')\n",
    "df_exon['exon_id'] = df_exon['exon_id'].astype('category')\n",
    "df_exon.head()\n",
    "#df_exon['chrom_exon_start'] = df_exon.apply(lambda row: chrom_exon_starts[row.mRNA][row.exon_id], axis=1)\n",
    "#df_exon['chrom_exon_end'] = df_exon.apply(lambda row: chrom_exon_ends[row.mRNA][row.exon_id], axis=1)\n",
    "#df_exon['exon_len'] = df_exon.apply(lambda row: row.exon_chrom_end - row.exon_chrom_start, axis=1)\n",
    "del df_exon['chrom_bs_start']\n",
    "del df_exon['chrom_bs_end']\n",
    "del df_exon['mRNA']\n",
    "#TODO remove code that we dont need! dont add columns to delete them later!\n",
    "df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T16:30:58.780491Z",
     "start_time": "2020-10-02T16:30:57.793360Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id_version</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>ensembl_transcript_id_version</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id ensembl_gene_id_version ensembl_transcript_id  \\\n",
       "0  ENSG00000210049       ENSG00000210049.1       ENST00000387314   \n",
       "1  ENSG00000211459       ENSG00000211459.2       ENST00000389680   \n",
       "2  ENSG00000210077       ENSG00000210077.1       ENST00000387342   \n",
       "3  ENSG00000210082       ENSG00000210082.2       ENST00000387347   \n",
       "4  ENSG00000209082       ENSG00000209082.1       ENST00000386347   \n",
       "\n",
       "  ensembl_transcript_id_version chromosome_name  strand  start_position  \\\n",
       "0             ENST00000387314.1              MT       1             577   \n",
       "1             ENST00000389680.2              MT       1             648   \n",
       "2             ENST00000387342.1              MT       1            1602   \n",
       "3             ENST00000387347.2              MT       1            1671   \n",
       "4             ENST00000386347.1              MT       1            3230   \n",
       "\n",
       "   end_position  \n",
       "0           647  \n",
       "1          1601  \n",
       "2          1670  \n",
       "3          3229  \n",
       "4          3304  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_gene: strand - chrom - gene_id - gene_name - transcript_id (1-N)\n",
    "df_gene = pd.read_csv(path2+'gene_infos.csv', dtype={'ensembl_gene_id':str,'ensembl_gene_id_version':str,'ensembl_transcript_id':str, 'ensembl_transcript_id_version':str, 'chromosome_name':'category', 'strand':'int8', 'start_position':int, 'end_position':int})\n",
    "del df_gene['Unnamed: 0']\n",
    "df_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T15:21:15.077692Z",
     "start_time": "2020-10-03T15:21:14.993058Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "bs_id                  int64\n",
       "exon_id             category\n",
       "exon_chrom_start       int64\n",
       "exon_chrom_end         int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get TCGA data from GDC API "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T13:36:57.219795Z",
     "start_time": "2020-10-03T13:36:56.480687Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T13:36:51.522193Z",
     "start_time": "2020-10-03T13:36:51.422002Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Cell magic `%%cython` not found.\n"
     ]
    }
   ],
   "source": [
    "%%cython\n",
    "cpdef new_split_chrom_column(Py_UNICODE* chrom):\n",
    "    cdef Py_UNICODE* chromosome_name\n",
    "    cdef Py_UNICODE* chrom_list[1]\n",
    "    cdef int strand\n",
    "    cdef int chrom_exon_start\n",
    "    cdef int chrom_exon_end\n",
    "    chrom_list = chrom.split(':')\n",
    "    chromosome_name = chrom[0][3:]\n",
    "    strand = 1 if chrom[2] == '+' else -1\n",
    "    chrom_exon_start = int(chrom[1].split('-')[0])\n",
    "    chrom_exon_end = int(chrom[1].split('-')[1])\n",
    "    return chromosome_name,strand,chrom_exon_start,chrom_exon_end\n",
    "#cython try of implementing, doesn't work in ipython, wrong version? not installed yet?\n",
    "#slow! 39.2 ms ± 3.29 ms per loop (mean ± std. dev. of 7 runs, 10 loops each) with cython\n",
    "# 36.4 ms ± 795 µs per loop (mean ± std. dev. of 7 runs, 10 loops each) normal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T21:10:16.717667Z",
     "start_time": "2020-10-03T21:10:16.527366Z"
    }
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def c_split_chrom_column(chrom):\n",
    "    chrom_list = chrom.split(':')\n",
    "    chromosome_name = str(chrom_list[0][3:])\n",
    "    strand = 1 if chrom_list[2] == '+' else -1\n",
    "    chrom_exon_start = int(chrom_list[1].split('-')[0])\n",
    "    chrom_exon_end = int(chrom_list[1].split('-')[1])\n",
    "    return chromosome_name,strand,chrom_exon_start,chrom_exon_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T21:10:35.107527Z",
     "start_time": "2020-10-03T21:10:34.998203Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#define list_to_df function\n",
    "\n",
    "def split_chrom_column(chrom):\n",
    "    chrom = chrom.split(':')\n",
    "    chromosome_name = str(chrom[0][3:])\n",
    "    strand = 1 if chrom[2] == '+' else -1\n",
    "    chrom_exon_start = int(chrom[1].split('-')[0])\n",
    "    chrom_exon_end = int(chrom[1].split('-')[1])\n",
    "    return chromosome_name,strand,chrom_exon_start,chrom_exon_end\n",
    "    \n",
    "def list_to_df(list_, nr_columns,data_type=None):\n",
    "    data = []\n",
    "    remove_last_columns = 2 if data_type != None else 0\n",
    "    nr_rows = int((len(list_)-1)/nr_columns)\n",
    "    for row in range(nr_rows):\n",
    "        start = row*nr_columns\n",
    "        end = (row+1)*nr_columns\n",
    "        data.append(list_[start:end-remove_last_columns])\n",
    "    df = pd.DataFrame(data[1:], columns = data[0]) #use first line as header\n",
    "    if data_type=='miRNA':\n",
    "        df = df.astype({'miRNA_ID':str, 'read_count':int})#, 'reads_per_million_miRNA_mapped':float, 'cross-mapped': 'category'})\n",
    "    elif data_type=='exon':\n",
    "        df = df.astype({'exon':str, 'raw_counts':int})#, 'reads_per_million_miRNA_mapped':float, 'cross-mapped': 'category'})\n",
    "        df[['chromosome_name','strand','chrom_exon_start','chrom_exon_end']] = df.exon.apply(lambda x: pd.Series(c_split_chrom_column(str(x))))\n",
    "        df[['raw_counts','chrom_exon_start','chrom_exon_end']] = df[['raw_counts','chrom_exon_start','chrom_exon_end']].apply(pd.to_numeric,downcast='unsigned')\n",
    "        df[['chromosome_name','strand']] = df[['chromosome_name','strand']].astype('str').astype('category')\n",
    "        del df['exon']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T21:12:17.937809Z",
     "start_time": "2020-10-03T21:10:52.437544Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 / 1 done.\n"
     ]
    }
   ],
   "source": [
    "#legacy API, get list of files which contain miRNA expression per case_id there\n",
    "import requests\n",
    "import json\n",
    "import re\n",
    "import gzip\n",
    "\n",
    "def download_files_from_legacy(data_type, size, primary_site='Kidney',write = False):\n",
    "    files_data_type = 'miRNA gene quantification' if data_type == 'miRNA' else 'Exon quantification'\n",
    "    fields = [\n",
    "        \"file_name\",\n",
    "        \"cases.disease_type\",\n",
    "        \"cases.case_id\",\n",
    "        \"data_category\",\n",
    "        \"data_type\",\n",
    "        \"cases.project.primary_site\"\n",
    "        ]\n",
    "\n",
    "    fields = \",\".join(fields)\n",
    "\n",
    "    files_endpt = \"https://api.gdc.cancer.gov/legacy/files\"\n",
    "\n",
    "    # This set of filters is nested under an 'and' operator.\n",
    "    filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\":[\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"cases.project.primary_site\",\n",
    "                \"value\": [primary_site]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.data_type\",\n",
    "                \"value\": [files_data_type]\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "\n",
    "    # A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "    params = {\n",
    "        \"filters\": filters,\n",
    "        \"fields\": fields,\n",
    "        \"format\": \"TSV\",\n",
    "        \"size\": str(size)\n",
    "        }\n",
    "\n",
    "    # The parameters are passed to 'json' rather than 'params' in this case\n",
    "    response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "    df_files = list_to_df(response.content.decode(\"utf-8\").replace(\"\\r\\n\", '\\t').split('\\t'),7)\n",
    "    all_dfs = {}\n",
    "\n",
    "    for i,row in df_files.iterrows():\n",
    "        file_id = row.id\n",
    "        case_id = row['cases.0.case_id']\n",
    "        data_endpt = f\"https://api.gdc.cancer.gov/legacy/data/{file_id}\"\n",
    "\n",
    "        response = requests.get(data_endpt, headers = {\"Content-Type\": \"application/json\"})\n",
    "        response_head_cd = response.headers[\"Content-Disposition\"]\n",
    "        data = str(response.content,'utf-8')\n",
    "        data = data.replace('\\n', '\\t').split('\\t')\n",
    "        new_df = list_to_df(data,4,data_type)\n",
    "        if write:\n",
    "            file_name = f'{path_gdc}{data_type}/{(re.findall(\"filename=(.+)\", response_head_cd)[0])}'\n",
    "            new_df.to_feather(file_name)\n",
    "        all_dfs[case_id] = new_df\n",
    "        print(i+1,'/',len(df_files),'done.')\n",
    "    return all_dfs\n",
    "\n",
    "#all_mirna_dfs = download_files_from_legacy('miRNA',1,write=True)#,3000,write=True) #2029 #70min\n",
    "all_exon_dfs = download_files_from_legacy('exon',1,write=True)#2000,write=True) #1588 #53h\n",
    "#TODO execute, takes loooooong & needs internet\n",
    "#normal 1min 44s  1.38 s\n",
    "#cython 1min 42s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#save dictionaries to file\n",
    "import pickle\n",
    "path = 'intermed_save'\n",
    "#f = open(path+\"all_mirna_dfs.pkl\",\"wb\")\n",
    "#pickle.dump(all_mirna_dfs,f)\n",
    "#f.close()\n",
    "path = 'intermed_save'\n",
    "#f = open(path+\"all_exon_dfs.pkl\",\"wb\")\n",
    "#pickle.dump(all_exon_dfs,f)\n",
    "#f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T17:08:56.844945Z",
     "start_time": "2020-10-02T17:07:47.624796Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "87 1\n"
     ]
    }
   ],
   "source": [
    "#load dictionaries from file\n",
    "import pickle\n",
    "#with open(path_gdc+'miRNA/all_mirna_dfs.pkl', 'rb') as handle:\n",
    " #   all_mirna_dfs = pickle.load(handle)\n",
    "with open(path_gdc+'exon/all_exon_dfs.pkl', 'rb') as handle:\n",
    "    all_exon_dfs = pickle.load(handle)\n",
    "print(len(all_exon_dfs),len(all_mirna_dfs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T16:47:24.499866Z",
     "start_time": "2020-10-02T16:47:23.296253Z"
    }
   },
   "outputs": [],
   "source": [
    "#for every case do\n",
    "#for _,row in df_exons.iterrows():\n",
    "#case_id =  row['cases.0.case_id']\n",
    "\n",
    "case_id = '24ea3cd5-c2e6-44f7-9658-78af7befeef7'\n",
    "\n",
    "important_counts = df_main[['bs_id','miRNA','mRNA','chrom_bs_start','chrom_bs_end']].copy()\n",
    "important_counts = important_counts.merge(df_gene[['ensembl_transcript_id','chromosome_name','strand']], left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "important_counts[['chromosome_name','strand','miRNA','mRNA']] = important_counts[['chromosome_name','strand','miRNA','mRNA']].astype('str').astype('category')\n",
    "del important_counts['ensembl_transcript_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T14:15:16.310279Z",
     "start_time": "2020-10-03T14:15:16.194356Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mRNA</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0</td>\n",
       "      <td>58352138</td>\n",
       "      <td>58352158</td>\n",
       "      <td>ENSE00003224625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>1</td>\n",
       "      <td>58350645</td>\n",
       "      <td>58351407</td>\n",
       "      <td>ENSE00003509215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>1</td>\n",
       "      <td>58350645</td>\n",
       "      <td>58351407</td>\n",
       "      <td>ENSE00003142302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>2</td>\n",
       "      <td>58352987</td>\n",
       "      <td>58353006</td>\n",
       "      <td>ENSE00003185372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>3</td>\n",
       "      <td>58351458</td>\n",
       "      <td>58351475</td>\n",
       "      <td>ENSE00003509215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24709</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24709</td>\n",
       "      <td>58347425</td>\n",
       "      <td>58347451</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24710</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24710</td>\n",
       "      <td>58347375</td>\n",
       "      <td>58347395</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24711</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24711</td>\n",
       "      <td>58347505</td>\n",
       "      <td>58347518</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24712</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24712</td>\n",
       "      <td>58346909</td>\n",
       "      <td>58346925</td>\n",
       "      <td>ENSE00003024001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24713</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24713</td>\n",
       "      <td>58347513</td>\n",
       "      <td>58347543</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26043 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  mRNA  bs_id  chrom_bs_start  chrom_bs_end          exon_id\n",
       "0      ENST00000600966      0        58352138      58352158  ENSE00003224625\n",
       "1      ENST00000600966      1        58350645      58351407  ENSE00003509215\n",
       "1      ENST00000600966      1        58350645      58351407  ENSE00003142302\n",
       "2      ENST00000600966      2        58352987      58353006  ENSE00003185372\n",
       "3      ENST00000600966      3        58351458      58351475  ENSE00003509215\n",
       "...                ...    ...             ...           ...              ...\n",
       "24709  ENST00000598345  24709        58347425      58347451  ENSE00003207101\n",
       "24710  ENST00000598345  24710        58347375      58347395  ENSE00003207101\n",
       "24711  ENST00000598345  24711        58347505      58347518  ENSE00003207101\n",
       "24712  ENST00000598345  24712        58346909      58346925  ENSE00003024001\n",
       "24713  ENST00000598345  24713        58347513      58347543  ENSE00003207101\n",
       "\n",
       "[26043 rows x 5 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T16:27:05.551294Z",
     "start_time": "2020-10-02T16:27:04.749585Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>miRNA_seed</th>\n",
       "      <th>mRNA_bs_seq</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>bs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1583</th>\n",
       "      <td>hsa-miR-3162-3p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-23.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000609</td>\n",
       "      <td>0.353</td>\n",
       "      <td>0.024639</td>\n",
       "      <td>-0.085003</td>\n",
       "      <td>-3.937574</td>\n",
       "      <td>...</td>\n",
       "      <td>59</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>356</td>\n",
       "      <td>429</td>\n",
       "      <td>TCCCTACCCCTCCACTCCCCA</td>\n",
       "      <td>GGAGGATGTGGAGGCCACCTTTCCAGTCCATCAGCCTGGCAACTAC...</td>\n",
       "      <td>58352098</td>\n",
       "      <td>58352171</td>\n",
       "      <td>1583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2779</th>\n",
       "      <td>hsa-miR-4436b-3p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-19.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.368</td>\n",
       "      <td>-0.014578</td>\n",
       "      <td>0.472495</td>\n",
       "      <td>-12.255123</td>\n",
       "      <td>...</td>\n",
       "      <td>64</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>150</td>\n",
       "      <td>222</td>\n",
       "      <td>CAGGGCAGGAAGAAGTGGACAA</td>\n",
       "      <td>TTGTCCACAGGATGGACCCAGCTGAGCAAGCTCCTGGAGCTGACAG...</td>\n",
       "      <td>58351482</td>\n",
       "      <td>58351554</td>\n",
       "      <td>2779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3032</th>\n",
       "      <td>hsa-miR-6081</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>-25.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.089396</td>\n",
       "      <td>0.273189</td>\n",
       "      <td>-7.912362</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>752</td>\n",
       "      <td>823</td>\n",
       "      <td>AGGAGCAGTGCCGGCCAAGGCGCC</td>\n",
       "      <td>GGTGGCCCTGGGGGATGGAGGTCACTACACCTGCCGCTACCGGCTG...</td>\n",
       "      <td>58352964</td>\n",
       "      <td>58353035</td>\n",
       "      <td>3032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3162</th>\n",
       "      <td>hsa-miR-6722-3p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-28.3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.382</td>\n",
       "      <td>-0.065482</td>\n",
       "      <td>-0.350447</td>\n",
       "      <td>-4.794414</td>\n",
       "      <td>...</td>\n",
       "      <td>69</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>217</td>\n",
       "      <td>294</td>\n",
       "      <td>TGCAGGGGTCGGGTGGGCCAGG</td>\n",
       "      <td>CCTGGCTCTCGATGGCGCCAGTGTCCTGGATCACCCCCGGCCTGAA...</td>\n",
       "      <td>58351549</td>\n",
       "      <td>58351626</td>\n",
       "      <td>3162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3322</th>\n",
       "      <td>hsa-miR-6762-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.397</td>\n",
       "      <td>0.269960</td>\n",
       "      <td>0.138124</td>\n",
       "      <td>-3.937574</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>680</td>\n",
       "      <td>763</td>\n",
       "      <td>CGGGGCCATGGAGCAGCCTGTGT</td>\n",
       "      <td>GCGCGGGGAGAAAGAGCTGCTGGTACCCAGGAGCAGCACCAGCCCA...</td>\n",
       "      <td>58352520</td>\n",
       "      <td>58352975</td>\n",
       "      <td>3322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "1583   hsa-miR-3162-3p  ENST00000600966             0.692308   -23.2     1   \n",
       "2779  hsa-miR-4436b-3p  ENST00000600966             0.615385   -19.2     1   \n",
       "3032      hsa-miR-6081  ENST00000600966             0.769231   -25.8     0   \n",
       "3162   hsa-miR-6722-3p  ENST00000600966             0.538462   -28.3     1   \n",
       "3322   hsa-miR-6762-5p  ENST00000600966             0.615385   -25.9     1   \n",
       "\n",
       "      accessibility  AU_content  PhyloP_Stem  PyloP_Flanking       m/e   ...  \\\n",
       "1583       0.000609       0.353     0.024639       -0.085003  -3.937574  ...   \n",
       "2779       0.000008       0.368    -0.014578        0.472495 -12.255123  ...   \n",
       "3032       0.000001       0.338     0.089396        0.273189  -7.912362  ...   \n",
       "3162       0.000055       0.382    -0.065482       -0.350447  -4.794414  ...   \n",
       "3322       0.000005       0.397     0.269960        0.138124  -3.937574  ...   \n",
       "\n",
       "      position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "1583                                        59                       7   \n",
       "2779                                        64                       8   \n",
       "3032                                        27                       7   \n",
       "3162                                        69                       8   \n",
       "3322                                         1                       8   \n",
       "\n",
       "      difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \\\n",
       "1583                                                  1        356     429   \n",
       "2779                                                  0        150     222   \n",
       "3032                                                  1        752     823   \n",
       "3162                                                  0        217     294   \n",
       "3322                                                  0        680     763   \n",
       "\n",
       "                    miRNA_seed  \\\n",
       "1583     TCCCTACCCCTCCACTCCCCA   \n",
       "2779    CAGGGCAGGAAGAAGTGGACAA   \n",
       "3032  AGGAGCAGTGCCGGCCAAGGCGCC   \n",
       "3162    TGCAGGGGTCGGGTGGGCCAGG   \n",
       "3322   CGGGGCCATGGAGCAGCCTGTGT   \n",
       "\n",
       "                                            mRNA_bs_seq  chrom_bs_start  \\\n",
       "1583  GGAGGATGTGGAGGCCACCTTTCCAGTCCATCAGCCTGGCAACTAC...        58352098   \n",
       "2779  TTGTCCACAGGATGGACCCAGCTGAGCAAGCTCCTGGAGCTGACAG...        58351482   \n",
       "3032  GGTGGCCCTGGGGGATGGAGGTCACTACACCTGCCGCTACCGGCTG...        58352964   \n",
       "3162  CCTGGCTCTCGATGGCGCCAGTGTCCTGGATCACCCCCGGCCTGAA...        58351549   \n",
       "3322  GCGCGGGGAGAAAGAGCTGCTGGTACCCAGGAGCAGCACCAGCCCA...        58352520   \n",
       "\n",
       "     chrom_bs_end bs_id  \n",
       "1583     58352171  1583  \n",
       "2779     58351554  2779  \n",
       "3032     58353035  3032  \n",
       "3162     58351626  3162  \n",
       "3322     58352975  3322  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAEWCAYAAABMoxE0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3dfbhVZZ3/8fdH8AE0EhINAT1WpKFTKkiYTlmaYqZYo0WjSY3GZFbar5lCp0m7Gvrp5K/SKS00Ex9GJfKBdCwNny4bFY/PIJIkKCcQ8BEwRcDv749171hs9jnsc9bZZ+/N+byua197rXutda/vXvuc/d3rvte+lyICMzOzrtqq3gGYmVlzcyIxM7NCnEjMzKwQJxIzMyvEicTMzApxIjEzs0KcSAxJcyUdUu846knSpyQtlrRa0n4Vloek97Sz7QmSbuvifltS3X3T/K2SJnalrp4k6SxJl9ag3i9Iure7661y35dL+o967LvZOZFs4SQtknRYWdlG/6wRsXdE3LWZejb6wNsCnQ98NSJ2iIhHOrNhRFwdEYd3RxARcWRETOuOumopIn4QEafUO46uqmfC2hI5kVhDaIAEtTswt84x9JgGON62BXEisY3OWiSNkdQqaaWkZZJ+lFa7Jz2/kpp/DpS0laTvSHpW0nJJV0h6e67ek9KyFyX9e9l+zpE0Q9JVklYCX0j7vk/SK5KWSvqppG1y9YWkr0h6WtIqSd+X9O60zUpJ0/Prl73GirFK2lbSaqAP8JikP3dwqD4h6RlJL0j6oaStUt0bfbtNcX45xfmypJ9JUlrWR9L5qY5ngKPK4rxL0in5etP6L0taKOnI3Lp7SLonHYs/pP1c1c7rP0RSm6RvS3oe+FU6JpMl/Tm9R9MlDerE+3dVbt1jlDWRvpJew/tyyxZJ+hdJj0t6VdJ1krbr4Djn495L0u2SXpI0X9JncssuT6/5lnQMHpD07tzyw9M2r0q6SNLdkk5Jsf0cODD9Lb+S2+XASvUp8+P0t/Nqei37VPMaeoWI8GMLfgCLgMPKyr4A3FtpHeA+4PNpegdgbJpuAQLom9vun4AFwLvSutcDV6ZlI4HVwMHANmRNR2tz+zknzR9L9oWmHzAKGAv0TfubB5yR218AM4EBwN7AGmBW2v/bgSeBie0ch3ZjzdX9ng6OYwB3AoOA3YA/Aae0czwDuBnYMa27AhiXln0ZeAoYnuq6M39cgbvK6l0LfIks0Z0KLAGUe6/OT8f3YGAlcFU78R8CrAPOA7ZNx/sM4H5gWCr7BXBNJ96/q9L0e4HXgI8DWwPfSsd6m9zf12xg1/Sa5wFfbifOvx1LYHtgMfDF9DexP/ACsHdafjnwEjAmLb8auDYt2ykdj0+nZaen+Cu+Z1XUdwTwUHpPBbwPGFLv/+9GefiMpHe4MX1TfCV9+7qog3XXAu+RtFNErI6I+ztY9wTgRxHxTESsBs4EJihrNjkO+G1E3BsRbwLfJfvAzLsvIm6MiLci4vWIeCgi7o+IdRGxiOyD7SNl25wXESsjYi4wB7gt7f9V4FZgk47yKmKt1nkR8VJEPAf8BPhcB+ueGxGvpHXvBPZN5Z8BfhIRiyPiJeD/bmafz0bEJRGxHpgGDAF2kbQbcADw3Yh4MyLuJUuyHXkLODsi1kTE68A/A/8WEW0RsYYsORzXifev5LPALRFxe0SsJUs6/YAP5da5MCKWpNf829zx6MgngUUR8av0N/Ew8JsUW8n1ETE7ItaRffCX6v0EMDcirk/LLgSer2Kf7dW3FngbsBdZIp8XEUurqK9XcCLpHY6NiB1LD+ArHax7Mtk3zKckPSjpkx2suyvwbG7+WbJvcrukZYtLCyLir8CLZdsvzs9Ieq+kmyU9r6y56wdk3yzzluWmX68wv0MXYq1WPt5nU53tyX9o/TUX10bHpSymDutJx5BU167AS7my8vgqWRERb+TmdwduyH3BmAesp/r3r2SjYxsRb6Vth1Z6HWx8PDqyO/DBsi9BJwDvrKLe8vgDaKtinxXri4g7gJ8CPwOWSZoqaUAV9fUKTiS2kYh4OiI+B+xM1gwyQ9L2VP42uoTsn71kN7Lmk2XAUrImEwAk9QPeUb67svmLyZp9RkTEAOAssmaE7tBRrNUaXrb9ki7EsbRCPV2xFBgkqX+ubHh7Kyflx3sxcGT+S0ZEbBcRf6G6969ko2Ob+oOGA3+p7qW0azFwd1l8O0TEqVVsWx6/8vO0f3bVroi4MCJGkTWrvhf4187WsaVyIrGNSDpR0uD0rbLUCbmerJ3/LbI+hpJrgG+kTt8dyM4grkvNAjOAoyV9SFkH+PfYfFJ4G1m79mpJe5H1CXSXjmKt1r9KGihpOFmb+3VdiGM68HVJwyQNBCZ3oQ4i4lmgFThH0jaSDgSO7mQ1PwemSNodQNJgSePTss68f9OBoyQdKmlr4Jtk/Vf/28l4yt0MvFfS5yVtnR4H5DvyO3AL8HeSjk1Ndaex8ZnMMmCY2rk4o1za7wfT63sNeIPs/8JwIrFNjQPmKruS6QJgQkS8kZo2pgB/TM0MY4HLgCvJruhaSPbP9TWA1IfxNeBasm+Hq4DlZB8w7fkX4B/TupfQtQ/q9rQbayfcRNbh+ijZB9UvuxDHJcDvgceAh8k6/bvqBOBAsian/yA7Xh0d33IXkPWr3CZpFVnH+wehc+9fRMwHTgT+i6wz/Gjg6NS30mURsQo4HJhAdtbzPBsuFtjcti8AxwP/SXZ8RpIl3lL8d5Bd7v28pBeqCGcA2Xv3Mlkz3otkfUHGhqs/zGoqnQW8QtZstbDe8WyJJF0HPBURZ9eg7qZ+/5Rdqt0GnBARd9Y7ni2Nz0isZiQdLal/6mM5H3iC7FJQ6wapueXdyn4PMg4YD9zYjfU39fsn6QhJO0ralg39bR1dhWhd5ERitTSerEliCTCCrJnMp8Dd551kvztZTXZ566nRyeFdNqPZ378DgT+zobnt2HTZs3UzN22ZmVkhPiMxM7NCet3AbTvttFO0tLTUOwwzs6by0EMPvRARgyst63WJpKWlhdbW1nqHYWbWVCS1OwqDm7bMzKwQJxIzMyvEicTMzApxIjEzs0KcSMzMrBAnEjMzK8SJxMzMCnEiMTOzQpxIzMyskF73y/Zm1TL5lrrsd9G5R9Vlv2bWPHxGYmZmhTiRmJlZIU4kZmZWiBOJmZkV4kRiZmaFOJGYmVkhTiRmZlaIE4mZmRXiRGJmZoU4kZiZWSFOJGZmVogTiZmZFeJEYmZmhTiRmJlZIU4kZmZWiBOJmZkV4kRiZmaFOJGYmVkhTiRmZlaIE4mZmRVSs0Qi6TJJyyXNyZUNknS7pKfT88DcsjMlLZA0X9IRufJRkp5Iyy6UpFS+raTrUvkDklpq9VrMzKx9tTwjuRwYV1Y2GZgVESOAWWkeSSOBCcDeaZuLJPVJ21wMTAJGpEepzpOBlyPiPcCPgfNq9krMzKxdNUskEXEP8FJZ8XhgWpqeBhybK782ItZExEJgATBG0hBgQETcFxEBXFG2TamuGcChpbMVMzPrOT3dR7JLRCwFSM87p/KhwOLcem2pbGiaLi/faJuIWAe8Cryj0k4lTZLUKql1xYoV3fRSzMwMGqezvdKZRHRQ3tE2mxZGTI2I0RExevDgwV0M0czMKunpRLIsNVeRnpen8jZgeG69YcCSVD6sQvlG20jqC7ydTZvSzMysxno6kcwEJqbpicBNufIJ6UqsPcg61Wen5q9Vksam/o+TyrYp1XUccEfqRzEzsx7Ut1YVS7oGOATYSVIbcDZwLjBd0snAc8DxABExV9J04ElgHXBaRKxPVZ1KdgVYP+DW9AD4JXClpAVkZyITavVazMysfTVLJBHxuXYWHdrO+lOAKRXKW4F9KpS/QUpEZmZWP43S2W5mZk3KicTMzApxIjEzs0KcSMzMrBAnEjMzK8SJxMzMCnEiMTOzQpxIzMysECcSMzMrxInEzMwKcSIxM7NCnEjMzKwQJxIzMyvEicTMzApxIjEzs0KcSMzMrBAnEjMzK8SJxMzMCnEiMTOzQpxIzMysECcSMzMrxInEzMwKcSIxM7NCnEjMzKwQJxIzMyvEicTMzApxIjEzs0LqkkgkfUPSXElzJF0jaTtJgyTdLunp9Dwwt/6ZkhZImi/piFz5KElPpGUXSlI9Xo+ZWW/W44lE0lDg68DoiNgH6ANMACYDsyJiBDArzSNpZFq+NzAOuEhSn1TdxcAkYER6jOvBl2JmZtSvaasv0E9SX6A/sAQYD0xLy6cBx6bp8cC1EbEmIhYCC4AxkoYAAyLivogI4IrcNmZm1kN6PJFExF+A84HngKXAqxFxG7BLRCxN6ywFdk6bDAUW56poS2VD03R5+SYkTZLUKql1xYoV3flyzMx6vXo0bQ0kO8vYA9gV2F7SiR1tUqEsOijftDBiakSMjojRgwcP7mzIZmbWgXo0bR0GLIyIFRGxFrge+BCwLDVXkZ6Xp/XbgOG57YeRNYW1penycjMz60H1SCTPAWMl9U9XWR0KzANmAhPTOhOBm9L0TGCCpG0l7UHWqT47NX+tkjQ21XNSbhszM+shfXt6hxHxgKQZwMPAOuARYCqwAzBd0slkyeb4tP5cSdOBJ9P6p0XE+lTdqcDlQD/g1vQwM7Me1OOJBCAizgbOLiteQ3Z2Umn9KcCUCuWtwD7dHqCZmVXNv2w3M7NCnEjMzKwQJxIzMyvEicTMzApxIjEzs0KcSMzMrBAnEjMzK8SJxMzMCnEiMTOzQqpKJJL863EzM6uo2jOSn0uaLekrknasaURmZtZUqkokEXEwcALZcO6tkv5b0sdrGpmZmTWFqvtIIuJp4DvAt4GPABdKekrSp2sVnJmZNb5q+0jeL+nHZPcN+RhwdES8L03/uIbxmZlZg6t2GPmfApcAZ0XE66XCiFgi6Ts1iczMzJpCtYnkE8DrpRtKSdoK2C4i/hoRV9YsOjMza3jV9pH8gewuhCX9U5mZmfVy1SaS7SJidWkmTfevTUhmZtZMqk0kr0navzQjaRTwegfrm5lZL1FtH8kZwK8lLUnzQ4DP1iYkMzNrJlUlkoh4UNJewJ6AgKciYm1NIzMzs6ZQ7RkJwAFAS9pmP0lExBU1icrMzJpGVYlE0pXAu4FHgfWpOAAnEjOzXq7aM5LRwMiIiFoGY2Zmzafaq7bmAO+sZSBmZtacqj0j2Ql4UtJsYE2pMCKOqUlUZmbWNKpNJOfUMggzM2te1V7+e7ek3YEREfEHSf2BPrUNzczMmkG1w8h/CZgB/CIVDQVu7OpOJe0oaUa6n8k8SQdKGiTpdklPp+eBufXPlLRA0nxJR+TKR0l6Ii27UJK6GpOZmXVNtZ3tpwEHASvhbze52rnAfi8AfhcRewEfILvPyWRgVkSMAGaleSSNBCYAewPjgIsklc6GLgYmASPSY1yBmMzMrAuqTSRrIuLN0oykvmS/I+k0SQOADwO/BIiINyPiFWA8MC2tNg04Nk2PB66NiDURsRBYAIyRNAQYEBH3pcuSr8htY2ZmPaTaRHK3pLOAfule7b8GftvFfb4LWAH8StIjki6VtD2wS0QsBUjPpTOeocDi3PZtqWxomi4v34SkSZJaJbWuWLGii2GbmVkl1SaSyWQf/k8A/wz8D9n927uiL7A/cHFE7Ae8lupvT6V+j+igfNPCiKkRMToiRg8ePLiz8ZqZWQeqvWrrLbJb7V7SDftsA9oi4oE0P4MskSyTNCQilqZmq+W59Yfnth8GLEnlwyqUWzdqmXxL3fa96Nyj6rZvM6tetVdtLZT0TPmjKzuMiOeBxZL2TEWHAk8CM4GJqWwicFOanglMkLStpD3IOtVnp+avVZLGpqu1TsptY2ZmPaQzY22VbAccDwwqsN+vAVdL2gZ4BvgiWVKbLulk4Lm0DyJirqTpZMlmHXBa6d7xwKnA5WS3Ab41PczMrAdV27T1YlnRTyTdC3y3KzuNiEfZODmVHNrO+lOAKRXKW4F9uhKDmZl1j2qHkd8/N7sVWRJ4W00iMjOzplJt09b/y02vAxYBn+n2aMzMrOlU27T10VoHYmZmzanapq3/09HyiPhR94RjZmbNpjNXbR1AdikuwNHAPWz8i3MzM+uFOnNjq/0jYhWApHOAX0fEKbUKzMzMmkO1Q6TsBryZm38TaOn2aMzMrOlUe0ZyJTBb0g1k41l9imy0XTMz6+WqvWpriqRbgb9PRV+MiEdqF5aZmTWLapu2APoDKyPiAqAtjXtlZma9XLWDNp4NfBs4MxVtDVxVq6DMzKx5VHtG8ingGLJ7hxARS/AQKWZmRvWJ5M10O9sASHc0NDMzqzqRTJf0C2BHSV8C/kD33OTKzMya3Gav2ko3jboO2AtYCewJfDcibq9xbGZm1gQ2m0giIiTdGBGjACcPMzPbSLVNW/dLOqCmkZiZWVOq9pftHwW+LGkR2ZVbIjtZeX+tAjMzs+bQYSKRtFtEPAcc2UPxmJlZk9ncGcmNZKP+PivpNxHxDz0RlJmZNY/N9ZEoN/2uWgZiZmbNaXOJJNqZNjMzAzbftPUBSSvJzkz6pWnY0Nk+oKbRmZlZw+swkUREn54KxMzMmlNnhpE3MzPbhBOJmZkV4kRiZmaF1C2RSOoj6RFJN6f5QZJul/R0eh6YW/dMSQskzZd0RK58lKQn0rIL0wCTZmbWg+p5RnI6MC83PxmYFREjgFlpHkkjgQnA3sA44CJJpYsALgYmASPSY1zPhG5mZiV1SSSShgFHAZfmiscD09L0NODYXPm1EbEmIhYCC4AxkoYAAyLivnTTrSty25iZWQ+p1xnJT4BvAW/lynaJiKUA6XnnVD4UWJxbry2VDU3T5eVmZtaDejyRSPoksDwiHqp2kwpl0UF5pX1OktQqqXXFihVV7tbMzKpRjzOSg4Bj0pD01wIfk3QVsCw1V5Gel6f124Dhue2HAUtS+bAK5ZuIiKkRMToiRg8ePLg7X4uZWa/X44kkIs6MiGER0ULWiX5HRJwIzAQmptUmAjel6ZnABEnbStqDrFN9dmr+WiVpbLpa66TcNmZm1kOqvbFVTzgXmC7pZOA54HiAiJgraTrwJLAOOC0i1qdtTgUuB/oBt6aHmZn1oLomkoi4C7grTb8IHNrOelOAKRXKW4F9ahehmZltjn/ZbmZmhTiRmJlZIU4kZmZWiBOJmZkV4kRiZmaFOJGYmVkhTiRmZlaIE4mZmRXiRGJmZoU4kZiZWSFOJGZmVogTiZmZFeJEYmZmhTiRmJlZIU4kZmZWiBOJmZkV4kRiZmaFOJGYmVkhTiRmZlaIE4mZmRXiRGJmZoU4kZiZWSFOJGZmVogTiZmZFeJEYmZmhTiRmJlZIX3rHYBZe1om31KX/S4696i67NesWfmMxMzMCunxMxJJw4ErgHcCbwFTI+ICSYOA64AWYBHwmYh4OW1zJnAysB74ekT8PpWPAi4H+gH/A5weEVGr2Ov1DdnMrJHV44xkHfDNiHgfMBY4TdJIYDIwKyJGALPSPGnZBGBvYBxwkaQ+qa6LgUnAiPQY15MvxMzM6pBIImJpRDycplcB84ChwHhgWlptGnBsmh4PXBsRayJiIbAAGCNpCDAgIu5LZyFX5LYxM7MeUtc+EkktwH7AA8AuEbEUsmQD7JxWGwoszm3WlsqGpuny8kr7mSSpVVLrihUruvMlmJn1enVLJJJ2AH4DnBERKztatUJZdFC+aWHE1IgYHRGjBw8e3PlgzcysXXVJJJK2JksiV0fE9al4WWquIj0vT+VtwPDc5sOAJal8WIVyMzPrQT2eSCQJ+CUwLyJ+lFs0E5iYpicCN+XKJ0jaVtIeZJ3qs1Pz1ypJY1OdJ+W2MTOzHlKPHyQeBHweeELSo6nsLOBcYLqkk4HngOMBImKupOnAk2RXfJ0WEevTdqey4fLfW9PDzMx6UI8nkoi4l8r9GwCHtrPNFGBKhfJWYJ/ui87MzDrLv2w3M7NCnEjMzKwQJxIzMyvEo/+alfGow2ad4zMSMzMrxInEzMwKcSIxM7NCnEjMzKwQJxIzMyvEicTMzArx5b9mDaKet3L2pcdWhM9IzMysECcSMzMrxInEzMwKcSIxM7NCnEjMzKwQJxIzMyvEicTMzApxIjEzs0KcSMzMrBAnEjMzK8SJxMzMCnEiMTOzQpxIzMysECcSMzMrxInEzMwKcSIxM7NCfGMrM6vbTbV8Q60tQ9OfkUgaJ2m+pAWSJtc7HjOz3qapz0gk9QF+BnwcaAMelDQzIp6sb2RmVg3fXnjL0NSJBBgDLIiIZwAkXQuMB5xIzKxD9Uxi9VKr5NnsiWQosDg33wZ8sHwlSZOASWl2taT53bT/nYAXuqmuWmqGOB1j93CM3WOLjFHnFdrf7u0taPZEogplsUlBxFRgarfvXGqNiNHdXW93a4Y4HWP3cIzdwzF2TrN3trcBw3Pzw4AldYrFzKxXavZE8iAwQtIekrYBJgAz6xyTmVmv0tRNWxGxTtJXgd8DfYDLImJuD4bQ7c1lNdIMcTrG7uEYu4dj7ARFbNKlYGZmVrVmb9oyM7M6cyIxM7NCnEiqJOkyScslzcmVDZJ0u6Sn0/PAOsc4XNKdkuZJmivp9EaLU9J2kmZLeizF+L1GizEXax9Jj0i6uRFjlLRI0hOSHpXU2qAx7ihphqSn0t/lgQ0Y457pGJYeKyWd0YBxfiP9z8yRdE36X2qIGJ1Iqnc5MK6sbDIwKyJGALPSfD2tA74ZEe8DxgKnSRpJY8W5BvhYRHwA2BcYJ2lsg8VYcjowLzffiDF+NCL2zf2eoNFivAD4XUTsBXyA7Hg2VIwRMT8dw32BUcBfgRtooDglDQW+DoyOiH3ILi6a0DAxRoQfVT6AFmBObn4+MCRNDwHm1zvGsnhvIhuHrCHjBPoDD5ONRtBQMZL9JmkW8DHg5kZ8v4FFwE5lZQ0TIzAAWEi6qKcRY6wQ8+HAHxstTjaM4jGI7Grbm1OsDRGjz0iK2SUilgKk553rHM/fSGoB9gMeoMHiTE1GjwLLgdsjouFiBH4CfAt4K1fWaDEGcJukh9IwQNBYMb4LWAH8KjURXipp+waLsdwE4Jo03TBxRsRfgPOB54ClwKsRcVujxOhEsgWStAPwG+CMiFhZ73jKRcT6yJoRhgFjJO1T75jyJH0SWB4RD9U7ls04KCL2B44ka8b8cL0DKtMX2B+4OCL2A16j/k1t7Uo/aj4G+HW9YymX+j7GA3sAuwLbSzqxvlFt4ERSzDJJQwDS8/I6x4OkrcmSyNURcX0qbrg4ASLiFeAusr6nRorxIOAYSYuAa4GPSbqKxoqRiFiSnpeTtemPobFibAPa0hknwAyyxNJIMeYdCTwcEcvSfCPFeRiwMCJWRMRa4HrgQ40SoxNJMTOBiWl6IlmfRN1IEvBLYF5E/Ci3qGHilDRY0o5puh/ZP8hTNFCMEXFmRAyLiBaypo47IuJEGihGSdtLeltpmqy9fA4NFGNEPA8slrRnKjqU7BYPDRNjmc+xoVkLGivO54Cxkvqn//NDyS5caIwY69V51GwPsj+wpcBasm9aJwPvIOuQfTo9D6pzjAeTtZs/DjyaHp9opDiB9wOPpBjnAN9N5Q0TY1m8h7Chs71hYiTrf3gsPeYC/9ZoMaZ49gVa0/t9IzCw0WJMcfYHXgTenitrqDiB75F96ZoDXAls2ygxeogUMzMrxE1bZmZWiBOJmZkV4kRiZmaFOJGYmVkhTiRmZlaIE4k1JUkt+ZGYy5Zdmgar7Ex9q9PzrpJmdEeM3UHS/3ZTPZdLOq476iqr96zcdLvviW3ZnEhsixMRp0TEk13cdklE1OIDt0u3tY6ID3V3LN3srM2vYls6JxJrZn0lTZP0eLrnRX8ASXdJGp2mV0uaku5/cr+kXVL5HpLuk/SgpO+XKsx/q5b0BUnXS/pdut/Df+bWO1nSn9K+LpH00/LgJJ0jaaqk24Ar0q/6f5P2+aCkg9J6g9O9JB6W9AtJz0raqRR/epakH6Z7UTwh6bOp/JAUQ+meH1enXz63S9IoSXenwR5/nxti4y5J5ym7X8yfJP19Ku8vaXo6ztdJekDSaEnnAv2U3cPj6lR9n3Q85kq6LY1eYFs4JxJrZnsCUyPi/cBK4CsV1tkeuD+y+5/cA3wplV9ANpjgAcDzHexjX+CzwN8Bn1V287BdgX8nu+fLx4G9Oth+FDA+Iv4x7fPHaZ//AFya1jmbbBiW/cnGzNqtQj2fTrF8gGxYmR+WEgDZKM9nACPJfvF+UHvBKBuL7b+A4yJiFHAZMCW3St+IGJPqOzuVfQV4OR3n76fXRERMBl6P7F4eJ6R1RwA/i4i9gVfS67QtnBOJNbPFEfHHNH0V2RAx5d4ku3cDwENk95SB7MO2NK7SlR3sY1ZEvBoRb5CNE7U72eCId0fES5ENoNfRaLEzI+L1NH0Y8FNlQ+jPBAak8bIOJhsckoj4HfByhXoOBq6JbOTkZcDdwAFp2eyIaIuIt8iGxWmpsH3JnsA+wO0pju+QjcJcUhroM3+s8vHNIRvupD0LI+LRCnXYFqxL7bZmDaJ8fJ9K4/2sjQ3jAK1n47/5asYHWpObLm3fYdNRmddy01sBB+YSC/C3wTY3p6N1KsXYUT1zI+LAzdSVr6czr7c8Fjdt9QI+I7Fmtpuk0gfi54B7O7HtH8lG9gU4oaMVK5gNfETSwNSJXm3zzW3AV0szkvZNk/cCn0llh5MNbFjuHrKmtT6SBgMfTnF01nxgcOm4Sdpa0t6b2SYf30iyZr6Stam5zHoxJxJrZvOAiZIeJ7sF6cWd2PZ0sptBPQi8vTM7jexudT8gu/vkH8iavF6tYtOvA6NTp/WTwJdT+feAwyU9THZPjKXAqrJtbyBrUnoMuAP4VmTDtHdKRLwJHAecJ+kxsqawzV0ZdhFZ8nkc+HaKo/R6pwKP5zrbrRfy6L9mXSBph4hYnc5IbgAui4gbuljXtsD6iFiXzhQujuwOkg1BUh9g64h4Q9K7yYYrf29KSmbuIzHronMkHQZsR9ZkdWOBunYDpkvaiuzigC9tZv2e1h+4MzVhCTjVScTyfEZiZmaFuI/EzMwKcSIxM7NCnEjMzKwQJxIzMyvEicTMzAr5/wjyIxoAAAAESURBVFr3Pm4V1zq2AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ax = df_main['binding_region_length'].plot(kind='hist',title='Histogram of binding region lengths')\n",
    "ax.set_xlabel('binding region length')\n",
    "df_main[df_main['binding_region_length'] > 70].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T14:22:18.127002Z",
     "start_time": "2020-10-03T14:22:17.393877Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "94.64 % (23389 / 24714) bs_ids have 1 exon\n",
      "5.35 % (1321 / 24714) bs_ids have 2 exons\n",
      "0.02 % (4 / 24714) bs_ids have 3 exons\n",
      "0.0 % (0 / 24714) bs_ids have > 3 exons\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAEICAYAAABfz4NwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAbF0lEQVR4nO3df5CV1Z3n8fdnAJWF+BPtBUFbA8kEsUpXJG50U2zMrEbH0p3VDcYoJhimXDPRLSdZkq3dTGZCFdZE3dXEzJJVAYdVibqD0VgTVu24ZhWCrgkCIXYUBUEIgkgTNbb73T/O6dTD5faPe/vHc7vv51V1i3vP85zT3+fwvf19ft2+igjMzMz+qOwAzMysMbggmJkZ4IJgZmaZC4KZmQEuCGZmlrkgmJkZ4IJgZmaZC8IAkXS1pKfLjqOSpJA0tew4qpHUJumasuOw2jjXRy4XBDOzGkhaIunb/RyjIYuqC0IJlHjubcRzrh9M0uiyY+iO/6PqIGmKpIck/VbSm5K+W1j2HUl7JL0i6TOF9jZJCyX9DPgdcLKkT0j6uaS9+d9PVKz/bUn/R1KHpB9JOkbScklv5/Vb+xjyBZJelrRL0t92vUElTZX00/zzd0m6vw/b/seSVknaLWmTpH9bWLZE0vckPSppn6TVkj5cWP4nkn6Vf953AfUxfitJs+Z6LmS3StqZ+/xS0gxJ84ErgK91xZrXXyDpNznvN0j614Wxrpb0szzebuB+4O+Af57HeKuP2zb4IsKPGh7AKOAXwK3AOOAw4BzgauB94Et5nWuBbYByvzbgNeAUYDTQAuwBrsyvL8+vjyms3w58GDgC2AD8Gvh0Xn8ZcHcf4g3gSeBo4IQ8xjV52b3AfyTtGBwGnNPLWOOALcAXcgz/DNgFnJKXLwF2A7Py8uXAfXnZBOBt4FJgDPDvgc6uWPxovEeT5/p5wHPAkaQdl48BEwt5/u2K9S8DJuXxPwvsL6x/dc71v8jbMza3PV32/3Hlw0cItZtF+o//akTsj4h3I6LrXOCrEfGDiPgAWApMJL0ZuiyJiPUR0Qn8K+CliLgnIjoj4l7gV8BFhfXvjojfRMRe4DHgNxHxv3L/HwKn9zHmmyJid0S8BvwX0hsS0pv6RGBSxXZ050+BzRFxd475eeBB0i/5Lg9FxJoc43LgtNx+AbAhIh6IiPdzHG/0MX4rRzPn+vvAh4A/JhW6jRGxvbuVI+KHEbEtIv5fRNwPvESavy7bIuL2vP3v9HFbhpwLQu2mkN4MnVWW/eEXXET8Lj8dX1i+pfB8EvBqRf9XgeMLr3cUnr9T5XVx7J4Uf+6r+WcDfI2097NG0npJX+xlnBOBj0t6q+tBOnz+p4V1ir/kf1eIcVIxjki7TsW4rPE0ba5HxBPAd4HvATskLZZ0eHfrS7pK0guF98UM0lFxtbgalgtC7bYAJ9R5Yaj4t8a3kX7BFp0AvF5vYD2YUvEztgFExBsR8aWImAT8OXCHer5tbwvw04g4svAYHxHX9iGG7cU4JKkiLms8zZzrRMRtEXEG6dTXR4Cvdi0qrifpROAHwJdJp8GOBF7kwGtkld8z0JDfO+CCULs1pF9uiySNk3SYpLPrGOfHwEckfU7SaEmfBaYDjwxksNlXJR0laQpwPemiFpIukzQ5r7OHlKQf9DDOIznmKyWNyY8zJX2sDzE8Cpwi6c/yL5ivcOCRhTWeps31nNcflzSGdD3g3cL6O4CTC6uPy+P9Nvf9AukIoSc7gMmSDqll4wabC0KN8jnTi4CppAtnW0kXkWod503SOfkbgTdJh7R/GhG7Bi7aP1hJukD2AukX8525/UxgtaQO4GHg+oh4pYeY95HOB88h7Xm9AdwEHNpbAHm7LgMWkbZ3GvCzOrfHhkAz5zpwOGmvfw/p1NObwHfysjuB6fn00D9ExAbgZuAZ0i/6U+k9t58A1gNvSBqMeahL110BZmbW5HyEYGZmQLon1oYxSf+CdJveQSKir3dmDNp4ZgPFuT74fMrIzMyAYXyEMGHChGhtba26bP/+/YwbN25oA2pAnoekp3l47rnndkXEsUMcUl2c873zPCT15vywLQitra2sXbu26rK2tjZmz549tAE1IM9D0tM8SKr8wFTDcs73zvOQ1JvzvqhsZmaAC4KZmWUuCGZmBgzjawg9Wff6Xq5e8GhNfTYvunCQojEbfM55Gwg+QjAzM8AFwczMMhcEMzMDXBDMzCxzQTAzM8AFwczMMhcEMzMDXBDMzCxzQTAzM6APBUHSFElPStooab2k63P70ZJWSXop/3tUoc/XJbVL2iTpvEL7GZLW5WW3SVJuP1TS/bl9taTWgd9Us75xzluz6ssRQidwY0R8DDgLuE7SdGAB8HhETAMez6/Jy+YApwDnA3dIGpXH+j4wn/QF69PycoB5wJ6ImArcSvridrOyOOetKfVaECJie0Q8n5/vAzYCxwMXA0vzakuBS/Lzi4H7IuK9iHgFaAdmSZoIHB4Rz0T6mrZlFX26xnoAOLdrT8psqDnnrVnV9Mft8mHt6cBqoCUitkN6A0k6Lq92PPBsodvW3PZ+fl7Z3tVnSx6rU9Je4BhgV8XPn0/a26KlpYW2traqcbaMhRtP7axl07odazjr6OgYkdtVq/7Mg3N+eHHOJ/XOQ58LgqTxwIPADRHxdg87M9UWRA/tPfU5sCFiMbAYYObMmdHdNwLdvnwlN6+r7Q+5br6i+ljDmb89Kql3Hpzzw49zPql3Hvp0l5GkMaQ3xvKIeCg378iHxOR/d+b2rcCUQvfJwLbcPrlK+wF9JI0GjgB217oxZgPFOW/NqC93GQm4E9gYEbcUFj0MzM3P5wIrC+1z8l0UJ5EupK3Jh9r7JJ2Vx7yqok/XWJcCT+RzrmZDzjlvzaovx5hnA1cC6yS9kNu+ASwCVkiaB7wGXAYQEeslrQA2kO7WuC4iPsj9rgWWAGOBx/ID0pvvHkntpL2kOf3cLrP+cM5bU+q1IETE01Q/3wlwbjd9FgILq7SvBWZUaX+X/OYyK5tz3pqVP6lsZmaAC4KZmWUuCGZmBrggmJlZ5oJgZmaAC4KZmWUuCGZmBrggmJlZ5oJgZmaAC4KZmWUuCGZmBrggmJlZ5oJgZmaAC4KZmWUuCGZmBrggmJlZ5oJgZmaAC4KZmWUuCGZmBrggmJlZ5oJgZmaAC4KZmWUuCGZmBrggmJlZ5oJgZmaAC4KZmWUuCGZmBrggmJlZ5oJgZmZAHwqCpLsk7ZT0YqHtaEmrJL2U/z2qsOzrktolbZJ0XqH9DEnr8rLbJCm3Hyrp/ty+WlLrwG6iWe2c99aM+nKEsAQ4v6JtAfB4REwDHs+vkTQdmAOckvvcIWlU7vN9YD4wLT+6xpwH7ImIqcCtwE31bozZAFqC896aTK8FISKeAnZXNF8MLM3PlwKXFNrvi4j3IuIVoB2YJWkicHhEPBMRASyr6NM11gPAuV17UWZlcd5bMxpdZ7+WiNgOEBHbJR2X248Hni2stzW3vZ+fV7Z39dmSx+qUtBc4BthV+UMlzSftbdHS0kJbW1v14MbCjad21rRB3Y01nHV0dIzI7arVAM7DkOe9c742zvmk3nmotyB0p9oeTvTQ3lOfgxsjFgOLAWbOnBmzZ8+uGsTty1dy87raNm3zFdXHGs7a2trobo6ayRDMw6DlvXO+Ns75pN55qPcuox35cJj8787cvhWYUlhvMrAtt0+u0n5AH0mjgSM4+FDdrBE4721Eq7cgPAzMzc/nAisL7XPyHRQnkS6ircmH2fsknZXPk15V0adrrEuBJ/L5VrNG47y3Ea3XY0xJ9wKzgQmStgLfBBYBKyTNA14DLgOIiPWSVgAbgE7guoj4IA91LenOjbHAY/kBcCdwj6R20h7SnAHZMrN+cN5bM+q1IETE5d0sOreb9RcCC6u0rwVmVGl/l/zGMmsUzntrRv6kspmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAS4IZmaWuSCYmRnggmBmZtnosgOwwbPu9b1cveDRmvpsXnThIEVjNvic8/3jIwQzMwMaqCBIOl/SJkntkhaUHY/ZUHDeWyNpiIIgaRTwPeAzwHTgcknTy43KbHA5763RNERBAGYB7RHxckT8HrgPuLjkmMwGm/PeGkqjXFQ+HthSeL0V+HjlSpLmA/Pzyw5Jm7oZbwKwq5YAdFMtaw8bnoekp3k4cSgDqdBr3jvna+Z5SOrK+UYpCKrSFgc1RCwGFvc6mLQ2ImYORGDDmechaeB56DXvnfO18Twk9c5Do5wy2gpMKbyeDGwrKRazoeK8t4bSKAXh58A0SSdJOgSYAzxcckxmg815bw2lIU4ZRUSnpC8D/wiMAu6KiPX9GLLXQ+wm4XlIGnIeBjjvG3IbS+B5SOqaB0UcdKrezMyaUKOcMjIzs5K5IJiZGTDCCoKkuyTtlPRi2bGURdIUSU9K2ihpvaTry46pLJIOk7RG0i/yXHyr7JgGmnM+cd4n/c35EXUNQdIngQ5gWUTMKDueMkiaCEyMiOclfQh4DrgkIjaUHNqQkyRgXER0SBoDPA1cHxHPlhzagHHOJ877pL85P6KOECLiKWB32XGUKSK2R8Tz+fk+YCPpE7FNJ5KO/HJMfoycPSCc812c90l/c35EFQQ7kKRW4HRgdbmRlEfSKEkvADuBVRHRtHPRLJo97/uT8y4II5Sk8cCDwA0R8XbZ8ZQlIj6IiNNInwKeJalpT6s0A+d9/3LeBWEEyucOHwSWR8RDZcfTCCLiLaANOL/kUGyQOO8PVE/OuyCMMPmi0p3Axoi4pca+J0jqyH+nv9ryv5L09wMR51CQdKykI/PzscCngV+VG5X1l6RWSZFztUPS5vzlQn/I+9y2Q9K4Qr9rJLVVjCVJL0saERef+5vzI6ogSLoXeAb4qKStkuaVHVMJzgauBD4l6YX8uKAvHSPitYgYHxEfDG6IQ2Yi8KSkX5L+btCqiHik5JgGVJPn/JERMR64FPgmhbwHJgFjgd5uP/0kcBxwsqQzBzPYIdKvnB9Rt53a4JL0V8DUiPh82bHYyCNpM+kb5K4EPkz6wqBvAEuAc0gXiS8DjgBeAcZERGfuuwb4YUT8bWGsvwO+BpwcEW9Jugb4fETMLvzMu4BDScVjW0R8ebC3s5GNqCOE4UzSJEkPSvqtpFckfUXS0Xmv76K8znil7969Kr/+hKSfS9qb//1EYbw2SX8j6WeS9kn6iaQJvcTQdSg+Or8+SdJPc/9VpC/dMBtM/wb4E+AjwEXAY6SiMIH0++orlR0knQXMANorFq0lnUP/y2o/SNI/IR1dLM+POUp/dbZpuSA0AEl/BPwI+AXp3ulzgRuAM4EvAj+QdBxwK/BCRCyTdDTwKHAbcAxwC/CopGMKQ38O+ALpkPgQunlj9OB/kD7gMwH4G2BuXRto1ne3R8SOiHgd+N/A6oj4vxHxHvA/SbeTdtkl6R3SKbM7gH+oMt5/Bv5C0rFVlv0Z8B7wE+AR0l9/vnDgNmX4cUFoDGcCx0bEX0fE7yPiZeAHwJyI+AnwQ+BxUrL+ee5zIfBSRNwTEZ0RcS/p4tFFhXHvjohfR8Q7wArgtL4GJOmEHNd/ioj38gegftTP7TTrzY7C83eqvB5feD0hv/5LYDbpQ1gHiIgXSb/sF1T5WXOBFfn98x7wEE2+0+OC0BhOBCZJeqvrQTpMbsnLF5MOie+OiDdz2yTg1YpxXuXAT2e+UXj+Ow58M/VmErAnIvZXjG/WMPI99zcD7wL/rpvVvgl8icJ7Q9Jk4FPA5yW9IekN0umjC3o7tTqSuSA0hi3AKxFxZOHxoYi4IN8C+t+AZcC1kqbmPts4+MuyTwBeH6CYtgNHFW/by+ObNaJFwNckHVa5ICLagfs58PrDlcCvgY+SjpxPI1232ApcPujRNigXhMawBnhb0n+QNDZ/9HxGvg3uG3mdLwLfAZblIvFj4COSPidptKTPAtNJh8f9FhGvki7KfUvSIZLO4cDTUWaN5FFgD+lIoJq/Boo7N3OBOyLijeKDdGdS05428m2nDULSJOBm4F+SboPbRLqItgA4MyLacyF4CvhxRCzMv6T/KzCVdIfF9RHxdB6vDfj7iPjv+fXVwDURcU4PMbRSuJ1P0snAUtKFvGdyTEf6tlOzkckFwczMAJ8yMjOzzAWhyUi6ovA3YIqP9WXHZmbl8ikjMzMD0ifzhqUJEyZEa2tr1WX79+9n3LhxVZc1E89D0tM8PPfcc7siotqnWBuOc753noek3pwftgWhtbWVtWvXVl3W1tbG7NmzhzagBuR5SHqaB0nD5sN2zvneeR6SenPe1xDMzAxwQTAzs8wFwczMgGF8DaEn617fy9ULHq2pz+ZFTf1Xb22Yc87bQPARgpmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAS4IZmaWuSCYmRnggmBmZpkLgpmZAX0oCJKmSHpS0kZJ6yVdn9uPlrRK0kv536MKfb4uqV3SJknnFdrPkLQuL7tNknL7oZLuz+2r83f7mpXCOW/Nqi9HCJ3AjRHxMeAs4DpJ00lf/v54REwDHs+vycvmAKcA5wN35C+HB/g+MB+Ylh/n5/Z5wJ6ImArcCtw0ANtmVi/nvDWlXgtCRGyPiOfz833ARuB44GJgaV5tKXBJfn4xcF9EvBcRrwDtwCxJE4HDI+KZSF/TtqyiT9dYDwDndu1JmQ0157w1q5r+uF0+rD0dWA20RMR2SG8gScfl1Y4Hni1025rb3s/PK9u7+mzJY3VK2gscA+yq+PnzSXtbtLS00NbWVjXOlrFw46mdtWxat2MNZx0dHSNyu2rVn3lwzg8vzvmk3nnoc0GQNB54ELghIt7uYWem2oLoob2nPgc2RCwGFgPMnDkzuvtGoNuXr+TmdbX9IdfNV1Qfazjzt0cl9c6Dc374cc4n9c5Dn+4ykjSG9MZYHhEP5eYd+ZCY/O/O3L4VmFLoPhnYltsnV2k/oI+k0cARwO5aN8ZsoDjnrRn15S4jAXcCGyPilsKih4G5+flcYGWhfU6+i+Ik0oW0NflQe5+ks/KYV1X06RrrUuCJfM7VbMg5561Z9eUY82zgSmCdpBdy2zeARcAKSfOA14DLACJivaQVwAbS3RrXRcQHud+1wBJgLPBYfkB6890jqZ20lzSnn9tl1h/OeWtKvRaEiHia6uc7Ac7tps9CYGGV9rXAjCrt75LfXGZlc85bs/Inlc3MDHBBMDOzzAXBzMwAFwQzM8tcEMzMDHBBMDOzzAXBzMwAFwQzM8tcEMzMDHBBMDOzzAXBzMwAFwQzM8tcEMzMDHBBMDOzzAXBzMwAFwQzM8tcEMzMDHBBMDOzzAXBzMwAFwQzM8tcEMzMDHBBMDOzzAXBzMwAFwQzM8tcEMzMDHBBMDOzzAXBzMwAFwQzM8tcEMzMDOhDQZB0l6Sdkl4stB0taZWkl/K/RxWWfV1Su6RNks4rtJ8haV1edpsk5fZDJd2f21dLah3YTTSrnfPemlFfjhCWAOdXtC0AHo+IacDj+TWSpgNzgFNynzskjcp9vg/MB6blR9eY84A9ETEVuBW4qd6NMRtAS3DeW5PptSBExFPA7ormi4Gl+flS4JJC+30R8V5EvAK0A7MkTQQOj4hnIiKAZRV9usZ6ADi3ay/KrCzOe2tGo+vs1xIR2wEiYruk43L78cCzhfW25rb38/PK9q4+W/JYnZL2AscAuyp/qKT5pL0tWlpaaGtrqx7cWLjx1M6aNqi7sYazjo6OEbldtRrAeRjyvHfO18Y5n9Q7D/UWhO5U28OJHtp76nNwY8RiYDHAzJkzY/bs2VWDuH35Sm5eV9umbb6i+ljDWVtbG93NUTMZgnkYtLx3ztfGOZ/UOw/13mW0Ix8Ok//dmdu3AlMK600GtuX2yVXaD+gjaTRwBAcfqps1Aue9jWj1FoSHgbn5+VxgZaF9Tr6D4iTSRbQ1+TB7n6Sz8nnSqyr6dI11KfBEPt9q1mic9zai9XqMKeleYDYwQdJW4JvAImCFpHnAa8BlABGxXtIKYAPQCVwXER/koa4l3bkxFngsPwDuBO6R1E7aQ5ozIFtm1g/Oe2tGvRaEiLi8m0XndrP+QmBhlfa1wIwq7e+S31hmjcJ5b83In1Q2MzPABcHMzDIXBDMzA1wQzMwsc0EwMzPABcHMzDIXBDMzA1wQzMwsc0EwMzPABcHMzDIXBDMzA1wQzMwsc0EwMzPABcHMzDIXBDMzA1wQzMwsc0EwMzPABcHMzDIXBDMzA1wQzMwsc0EwMzPABcHMzDIXBDMzA1wQzMwsc0EwMzPABcHMzDIXBDMzA1wQzMwsG112ADZ41r2+l6sXPFpTn82LLhykaMwGn3O+f3yEYGZmQAMVBEnnS9okqV3SgrLjMRsKzntrJA1RECSNAr4HfAaYDlwuaXq5UZkNLue9NZqGKAjALKA9Il6OiN8D9wEXlxyT2WBz3ltDaZSLyscDWwqvtwIfr1xJ0nxgfn7ZIWlTN+NNAHbVEoBuqmXtYcPzkPQ0DycOZSAVes1753zNPA9JXTnfKAVBVdrioIaIxcDiXgeT1kbEzIEIbDjzPCQNPA+95r1zvjaeh6TeeWiUU0ZbgSmF15OBbSXFYjZUnPfWUBqlIPwcmCbpJEmHAHOAh0uOyWywOe+toTTEKaOI6JT0ZeAfgVHAXRGxvh9D9nqI3SQ8D0lDzsMA531DbmMJPA9JXfOgiINO1ZuZWRNqlFNGZmZWMhcEMzMDRlhBkHSXpJ2SXiw7lrJImiLpSUkbJa2XdH3ZMZVF0mGS1kj6RZ6Lb5Ud00BzzifO+6S/OT+iriFI+iTQASyLiBllx1MGSROBiRHxvKQPAc8Bl0TEhpJDG3KSBIyLiA5JY4Cngesj4tmSQxswzvnEeZ/0N+dH1BFCRDwF7C47jjJFxPaIeD4/3wdsJH0itulE0pFfjsmPkbMHhHO+i/M+6W/Oj6iCYAeS1AqcDqwuN5LySBol6QVgJ7AqIpp2LppFs+d9f3LeBWGEkjQeeBC4ISLeLjueskTEBxFxGulTwLMkNe1plWbgvO9fzrsgjED53OGDwPKIeKjseBpBRLwFtAHnlxyKDRLn/YHqyXkXhBEmX1S6E9gYEbeUHU+ZJB0r6cj8fCzwaeBX5UZlg8F5n/Q350dUQZB0L/AM8FFJWyXNKzumEpwNXAl8StIL+XFB2UGVZCLwpKRfkv5u0KqIeKTkmAaUc/4PnPdJv3J+RN12amZm9RtRRwhmZlY/FwQzMwNcEMzMLHNBMDMzwAXBzMwyFwQzMwNcEMzMLPv/PP4LkpQNMvIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#visualize number bindingsites with more than 1 exon\n",
    "df_exon.groupby('bs_id').count().hist()\n",
    "#['exon_id'].agg('count').astype('int64').hist()\n",
    "a = df_exon.groupby('bs_id').count()\n",
    "print(round((len(a[a.exon_id == 1])/len(a))*100,2),'% ('+str(len(a[a.exon_id == 1]))+' / '+str(len(a))+') bs_ids have 1 exon')#23389 Zeilen\n",
    "print(round((len(a[a.exon_id == 2])/len(a))*100,2),'% ('+str(len(a[a.exon_id == 2]))+' / '+str(len(a))+') bs_ids have 2 exons')#1321 Zeilen\n",
    "print(round((len(a[a.exon_id == 3])/len(a))*100,2),'% ('+str(len(a[a.exon_id == 3]))+' / '+str(len(a))+') bs_ids have 3 exons')#4 Zeilen\n",
    "print(round((len(a[a.exon_id > 3])/len(a))*100,2),'% ('+str(len(a[a.exon_id > 3]))+' / '+str(len(a))+') bs_ids have > 3 exons')#0 Zeilen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T15:18:52.460735Z",
     "start_time": "2020-10-03T15:18:52.305059Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mRNA</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>0</td>\n",
       "      <td>58352138</td>\n",
       "      <td>58352158</td>\n",
       "      <td>ENSE00003224625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>1</td>\n",
       "      <td>58350645</td>\n",
       "      <td>58351407</td>\n",
       "      <td>ENSE00003509215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>1</td>\n",
       "      <td>58350645</td>\n",
       "      <td>58351407</td>\n",
       "      <td>ENSE00003142302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>2</td>\n",
       "      <td>58352987</td>\n",
       "      <td>58353006</td>\n",
       "      <td>ENSE00003185372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>3</td>\n",
       "      <td>58351458</td>\n",
       "      <td>58351475</td>\n",
       "      <td>ENSE00003509215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24709</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24709</td>\n",
       "      <td>58347425</td>\n",
       "      <td>58347451</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24710</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24710</td>\n",
       "      <td>58347375</td>\n",
       "      <td>58347395</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24711</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24711</td>\n",
       "      <td>58347505</td>\n",
       "      <td>58347518</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24712</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24712</td>\n",
       "      <td>58346909</td>\n",
       "      <td>58346925</td>\n",
       "      <td>ENSE00003024001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24713</th>\n",
       "      <td>ENST00000598345</td>\n",
       "      <td>24713</td>\n",
       "      <td>58347513</td>\n",
       "      <td>58347543</td>\n",
       "      <td>ENSE00003207101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26043 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  mRNA  bs_id  chrom_bs_start  chrom_bs_end          exon_id\n",
       "0      ENST00000600966      0        58352138      58352158  ENSE00003224625\n",
       "1      ENST00000600966      1        58350645      58351407  ENSE00003509215\n",
       "1      ENST00000600966      1        58350645      58351407  ENSE00003142302\n",
       "2      ENST00000600966      2        58352987      58353006  ENSE00003185372\n",
       "3      ENST00000600966      3        58351458      58351475  ENSE00003509215\n",
       "...                ...    ...             ...           ...              ...\n",
       "24709  ENST00000598345  24709        58347425      58347451  ENSE00003207101\n",
       "24710  ENST00000598345  24710        58347375      58347395  ENSE00003207101\n",
       "24711  ENST00000598345  24711        58347505      58347518  ENSE00003207101\n",
       "24712  ENST00000598345  24712        58346909      58346925  ENSE00003024001\n",
       "24713  ENST00000598345  24713        58347513      58347543  ENSE00003207101\n",
       "\n",
       "[26043 rows x 5 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter out all binding sites that have more than 1 exon_id: 24714 -> 23389 bindingsites\n",
    "important_counts2 = important_counts[df_exon.groupby('bs_id').count().merge(important_counts,on='bs_id').exon_id==1]\n",
    "important_counts2\n",
    "exon_counts\n",
    "df_exon #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write exoncounts to BED file for translating GRCh37 to GRCh38\n",
    "\n",
    "file = open('chrom_pos_hg19.bed', 'w')\n",
    "for row in exon_counts.iterrows():\n",
    "    line = row[0]\n",
    "    sep = '\\t'\n",
    "    strand = '+' if (row[1].strand=='1') else '-'\n",
    "    start = row[1].chrom_exon_start if strand == '+' else row[1].chrom_exon_end\n",
    "    end = row[1].chrom_exon_end if strand == '+' else row[1].chrom_exon_start\n",
    "    string = f'chr{row[1].chromosome_name}{sep}{start}{sep}{end}{sep}{line}{sep}0{sep}{strand}\\n' \n",
    "    #chr7  127471196  127472363  Pos1  0  +\n",
    "    file.write(string)\n",
    "file.close()\n",
    "#TODO then upload to website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-04T20:08:54.578226Z",
     "start_time": "2020-10-04T20:08:41.784481Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12595</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12646</td>\n",
       "      <td>12697</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name  chrom_exon_start  chrom_exon_end strand\n",
       "index                                                         \n",
       "0                   1             11874           12227      1\n",
       "1                   1             12595           12721      1\n",
       "2                   1             12613           12721      1\n",
       "3                   1             12646           12697      1\n",
       "4                   1             13221           14409      1"
      ]
     },
     "execution_count": 439,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#then read in new liftover exoncounts in GRCh38 from files\n",
    "df_liftover = pd.read_csv('chrom_pos_translated.bed', sep=\"\\t\",names=['chromosome_name','chrom_exon_start','chrom_exon_end','index','0','strand'])\n",
    "df_liftover = df_liftover.set_index('index')\n",
    "df_liftover.chromosome_name = df_liftover.apply(lambda row: row.chromosome_name[3:], axis=1)\n",
    "df_liftover.strand = df_liftover.apply(lambda row: '1' if (row.strand=='+') else '-1', axis=1)\n",
    "del df_liftover['0']\n",
    "df_liftover.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-04T20:16:56.496796Z",
     "start_time": "2020-10-04T20:16:56.296272Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>raw_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12595</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12646</td>\n",
       "      <td>12697</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name  chrom_exon_start  chrom_exon_end strand  raw_counts\n",
       "index                                                                     \n",
       "0                   1             11874           12227      1          50\n",
       "1                   1             12595           12721      1           0\n",
       "2                   1             12613           12721      1           0\n",
       "3                   1             12646           12697      1           0\n",
       "4                   1             13221           14409      1          70"
      ]
     },
     "execution_count": 455,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# changes of old to new df\n",
    "new_exon_counts = df_liftover.join(exon_counts,rsuffix='r',lsuffix='l')#['raw_counts'])\n",
    "new_exon_counts[(new_exon_counts.chrom_exon_startr != new_exon_counts.chrom_exon_startl)&(new_exon_counts.chrom_exon_startr != new_exon_counts.chrom_exon_endl)]\n",
    "new_exon_counts = df_liftover.join(exon_counts['raw_counts'])\n",
    "new_exon_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map exons to bs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-04T20:27:35.787877Z",
     "start_time": "2020-10-04T20:27:35.636672Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-457-efd565344068>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  new_exon_counts[(exon_counts.chromosome_name == '19') & (exon_counts.strand == '-1') & (exon_counts.chrom_exon_start > 58361900) & (exon_counts.chrom_exon_start < 58362900)] #& ( exon_counts.chrom_exon_start < 58352300 )]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>raw_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>216496</th>\n",
       "      <td>19</td>\n",
       "      <td>57848731</td>\n",
       "      <td>57850561</td>\n",
       "      <td>-1</td>\n",
       "      <td>9768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name  chrom_exon_start  chrom_exon_end strand  raw_counts\n",
       "index                                                                      \n",
       "216496              19          57848731        57850561     -1        9768"
      ]
     },
     "execution_count": 457,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_exon_counts[(exon_counts.chromosome_name == '19') & (exon_counts.strand == '-1') & (exon_counts.chrom_exon_start > 58361900) & (exon_counts.chrom_exon_start < 58362900)] #& ( exon_counts.chrom_exon_start < 58352300 )]\n",
    "#exon_counts.head() #1830 #58,354,357-58,362,848"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-04T20:32:46.711663Z",
     "start_time": "2020-10-04T20:32:12.900Z"
    }
   },
   "outputs": [],
   "source": [
    "df_exon[(df_exon.chromosome_name == '19') & (df_exon.strand == '-1') & (df_exon.chrom_exon_start > 58361900) & (df_exon.chrom_exon_start < 58362900)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-04T20:27:49.699565Z",
     "start_time": "2020-10-04T20:27:49.552409Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>ensembl_exon_id</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59442</th>\n",
       "      <td>ENSG00000268230</td>\n",
       "      <td>ENST00000599109</td>\n",
       "      <td>ENSE00002304388</td>\n",
       "      <td>58362677</td>\n",
       "      <td>58362751</td>\n",
       "      <td>-1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59480</th>\n",
       "      <td>ENSG00000174586</td>\n",
       "      <td>ENST00000311044</td>\n",
       "      <td>ENSE00001376549</td>\n",
       "      <td>58362677</td>\n",
       "      <td>58362751</td>\n",
       "      <td>-1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59483</th>\n",
       "      <td>ENSG00000174586</td>\n",
       "      <td>ENST00000425453</td>\n",
       "      <td>ENSE00003052652</td>\n",
       "      <td>58362677</td>\n",
       "      <td>58362848</td>\n",
       "      <td>-1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59486</th>\n",
       "      <td>ENSG00000174586</td>\n",
       "      <td>ENST00000595763</td>\n",
       "      <td>ENSE00003176780</td>\n",
       "      <td>58362677</td>\n",
       "      <td>58362825</td>\n",
       "      <td>-1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id ensembl_transcript_id  ensembl_exon_id  \\\n",
       "59442  ENSG00000268230       ENST00000599109  ENSE00002304388   \n",
       "59480  ENSG00000174586       ENST00000311044  ENSE00001376549   \n",
       "59483  ENSG00000174586       ENST00000425453  ENSE00003052652   \n",
       "59486  ENSG00000174586       ENST00000595763  ENSE00003176780   \n",
       "\n",
       "       chrom_exon_start  chrom_exon_end strand chromosome_name  \n",
       "59442          58362677        58362751     -1              19  \n",
       "59480          58362677        58362751     -1              19  \n",
       "59483          58362677        58362848     -1              19  \n",
       "59486          58362677        58362825     -1              19  "
      ]
     },
     "execution_count": 458,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_info#[exon_info.ensembl_exon_id == 'ENSE00003224625'] #74 #171\n",
    "#find version of exon_id to see if i can compare - TGCA\n",
    "exon_info[(exon_info.chromosome_name == '19') & (exon_info.strand == '-1') & (exon_info.chrom_exon_start > 58361900) & (exon_info.chrom_exon_start < 58362900)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-04T20:23:37.605742Z",
     "start_time": "2020-10-04T20:23:37.505485Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>raw_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12595</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12646</td>\n",
       "      <td>12697</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name  chrom_exon_start  chrom_exon_end strand  raw_counts\n",
       "index                                                                     \n",
       "0                   1             11874           12227      1          50\n",
       "1                   1             12595           12721      1           0\n",
       "2                   1             12613           12721      1           0\n",
       "3                   1             12646           12697      1           0\n",
       "4                   1             13221           14409      1          70"
      ]
     },
     "execution_count": 456,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_exon_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2020-10-04T20:33:44.300Z"
    }
   },
   "outputs": [],
   "source": [
    "#merge bs with exon counts, real slow, TODO cython, 22.34 - lookup when\n",
    "\n",
    "def is_in_exon(row, exon_counts, df_exon_all,exon_info):\n",
    "    bs_start = row.chrom_bs_start\n",
    "    bs_end = row.chrom_bs_end\n",
    "    df_exon = df_exon_all[df_exon_all.bs_id == row.bs_id]\n",
    "    df_exon = df_exon.merge(exon_info, left_on='exon_id', right_on='ensembl_exon_id') #this works only for 1 exon per bs\n",
    "    here = exon_counts.loc[(exon_counts['chromosome_name'] == df_exon.chromosome_name.values[0]) & (exon_counts['strand'] == df_exon.strand.values[0]) & (exon_counts['chrom_exon_start'] == df_exon.chrom_exon_start.values[0]) & (exon_counts['chrom_exon_end'] == df_exon.chrom_exon_end.values[0])]\n",
    "    #print(here)\n",
    "    count_list = str(list(here['raw_counts']))[1:-1]\n",
    "    return count_list if not count_list == '' else None\n",
    "\n",
    "#takes long! for 1 case id: 2 min \n",
    "#exon_counts = all_exon_dfs['aba26e6b-f11c-4ae6-a0d6-85bdd2060e8f'] #df#all_exon_dfs[case_id]\n",
    "important_counts['exon'] = important_counts.apply(lambda row: is_in_exon(row,new_exon_counts,df_exon,exon_info), axis=1)\n",
    "bs_exons = pd.DataFrame(important_counts.exons.str.split(',').tolist(),index=important_counts.bs_id)\n",
    "bs_exons = bs_exons.rename(columns={0: \"exon_raw_count\"})\n",
    "bs_exons.head()\n",
    "#TODO takes fuuuuuckin long, instead join on strand, chromosome_name, start, end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%cython\n",
    "def is_in_exon(row, exon_counts):\n",
    "    bs_start = row.chrom_bs_start\n",
    "    bs_end = row.chrom_bs_end\n",
    "    here = exon_counts.loc[(exon_counts['chromosome_name'] == row.chromosome_name) & (exon_counts['strand'] == row.strand) & (exon_counts['chrom_exon_start'] <= bs_start) & (exon_counts['chrom_exon_end'] >= bs_end)]\n",
    "    count_list = str(list(here['raw_counts']))[1:-1]\n",
    "    return count_list if not count_list == '' else None\n",
    "#merge bs (important_counts) with exon counts (new_exon_counts)\n",
    "#takes too long! for 1 case id: x min \n",
    "important_counts['exon'] = important_counts.apply(lambda row: is_in_exon(row,new_exon_counts), axis=1)\n",
    "bs_exons = pd.DataFrame(important_counts.exons.str.split(',').tolist(),index=important_counts.bs_id)\n",
    "bs_exons = bs_exons.rename(columns={0: \"exon_raw_count\"})\n",
    "bs_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T13:53:50.825661Z",
     "start_time": "2020-10-03T13:53:50.709779Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>raw_counts</th>\n",
       "      <th>median_length_normalized</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>61</td>\n",
       "      <td>0.9036827</td>\n",
       "      <td>1.21214019174369</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>0.8333333</td>\n",
       "      <td>0.886222027370111</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12595</td>\n",
       "      <td>12721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>0.8055556</td>\n",
       "      <td>0.96803495535554</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>0.8823529</td>\n",
       "      <td>1.62332015590391</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12646</td>\n",
       "      <td>12697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>287</td>\n",
       "      <td>0.9166667</td>\n",
       "      <td>1.69795556537075</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239317</th>\n",
       "      <td>499</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>113.230944638157</td>\n",
       "      <td>M_rCRS</td>\n",
       "      <td>1</td>\n",
       "      <td>12207</td>\n",
       "      <td>12237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239318</th>\n",
       "      <td>49206</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>149.971431353018</td>\n",
       "      <td>M_rCRS</td>\n",
       "      <td>1</td>\n",
       "      <td>12337</td>\n",
       "      <td>14644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239319</th>\n",
       "      <td>181297</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>1131.6001082413</td>\n",
       "      <td>M_rCRS</td>\n",
       "      <td>1</td>\n",
       "      <td>14761</td>\n",
       "      <td>15887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239320</th>\n",
       "      <td>1587</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>60.3436362818982</td>\n",
       "      <td>M_rCRS</td>\n",
       "      <td>1</td>\n",
       "      <td>15998</td>\n",
       "      <td>16182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239321</th>\n",
       "      <td>2386</td>\n",
       "      <td>0.9974093</td>\n",
       "      <td>43.3696335881373</td>\n",
       "      <td>M_rCRS</td>\n",
       "      <td>1</td>\n",
       "      <td>16183</td>\n",
       "      <td>16569</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>239322 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        raw_counts median_length_normalized               RPKM  \\\n",
       "0               61                0.9036827   1.21214019174369   \n",
       "1               16                0.8333333  0.886222027370111   \n",
       "2               15                0.8055556   0.96803495535554   \n",
       "3               12                0.8823529   1.62332015590391   \n",
       "4              287                0.9166667   1.69795556537075   \n",
       "...            ...                      ...                ...   \n",
       "239317         499                1.0000000   113.230944638157   \n",
       "239318       49206                1.0000000   149.971431353018   \n",
       "239319      181297                1.0000000    1131.6001082413   \n",
       "239320        1587                1.0000000   60.3436362818982   \n",
       "239321        2386                0.9974093   43.3696335881373   \n",
       "\n",
       "       chromosome_name strand  chrom_exon_start  chrom_exon_end  \n",
       "0                    1      1             11874           12227  \n",
       "1                    1      1             12595           12721  \n",
       "2                    1      1             12613           12721  \n",
       "3                    1      1             12646           12697  \n",
       "4                    1      1             13221           14409  \n",
       "...                ...    ...               ...             ...  \n",
       "239317          M_rCRS      1             12207           12237  \n",
       "239318          M_rCRS      1             12337           14644  \n",
       "239319          M_rCRS      1             14761           15887  \n",
       "239320          M_rCRS      1             15998           16182  \n",
       "239321          M_rCRS      1             16183           16569  \n",
       "\n",
       "[239322 rows x 7 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "important_counts\n",
    "exon_counts\n",
    "#df_exon #TODO add id version? (biomart)\n",
    "#first match transcript for far smaller searchspace, then match positions of exon\n",
    "#get ['exon_id'] = \n",
    "#important_counts: bs_id, transcript_id, exon_id, mirna    TODO if > 1 exon: remove row\n",
    "#exon_id count      from exon_counts (or also counts in important counts) \n",
    "#exon id (mature) mirna_id family count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map miRNA to bs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T13:41:05.787625Z",
     "start_time": "2020-10-03T13:41:05.410819Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                           miRNA_ID\n",
      "cel-let-7-5p              cel-let-7\n",
      "cel-let-7-3p              cel-let-7\n",
      "cel-lin-4-5p              cel-lin-4\n",
      "cel-lin-4-3p              cel-lin-4\n",
      "cel-miR-1-3p              cel-mir-1\n",
      "...                             ...\n",
      "gga-miR-1784b-5p      gga-mir-1784b\n",
      "gga-miR-1784b-3p      gga-mir-1784b\n",
      "mdo-miR-7385g-3p    mdo-mir-7385g-1\n",
      "mdo-miR-7385g-2-3p  mdo-mir-7385g-2\n",
      "NA                      vvi-MIR828b\n",
      "\n",
      "[49156 rows x 1 columns]\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"['exon_id'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-94-fca7684735e3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[0mtrans\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtrans\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmirna_families\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mhow\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'inner'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 13\u001b[1;33m \u001b[0mbs_mirnas\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mimportant_counts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'exon_id'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'miRNA'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtrans\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mon\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'miRNA'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmirna_counts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'miRNA_ID'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'read_count'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mon\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'miRNA_ID'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     14\u001b[0m \u001b[1;31m#print(bs_mirnas)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[0mbs_mirnas\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'miRNA'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'miRNA_ID'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   2804\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2805\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2806\u001b[1;33m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2807\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2808\u001b[0m         \u001b[1;31m# take() does not accept boolean indexers\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[1;34m(self, key, axis, raise_missing)\u001b[0m\n\u001b[0;32m   1550\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0max\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1551\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1552\u001b[1;33m         self._validate_read_indexer(\n\u001b[0m\u001b[0;32m   1553\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mo\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_axis_number\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mraise_missing\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1554\u001b[0m         )\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[1;34m(self, key, indexer, axis, raise_missing)\u001b[0m\n\u001b[0;32m   1644\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mname\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"loc\"\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1645\u001b[0m                 \u001b[0mnot_found\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mset\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0mset\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0max\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1646\u001b[1;33m                 \u001b[1;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf\"{not_found} not in index\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1647\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1648\u001b[0m             \u001b[1;31m# we skip the warning on Categorical/Interval\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: \"['exon_id'] not in index\""
     ]
    }
   ],
   "source": [
    "#map miRNA ID to mature ID\n",
    "mirna_counts = all_mirna_dfs['45bdcfd6-1e3f-4be8-b843-ae949e8e43eb']#case_id]\n",
    "\n",
    "trans = pd.DataFrame.from_dict(read_in_miRNA(path2+'miRNA2mature.fasta'), orient='index')#,columns=[\"mirna\", \"mature\"])\n",
    "trans = trans.rename(columns={0: \"miRNA_ID\"})\n",
    "trans[['miRNA_ID']] = trans[['miRNA_ID']].astype('str').astype('category')\n",
    "print(trans)\n",
    "\n",
    "#map miRNA to miRNA family\n",
    "mirna_families = read_in_mirna_families(path_tarp+'mapping_miRNA_families.fasta')\n",
    "trans = trans.join(mirna_families,how='inner')\n",
    "\n",
    "bs_mirnas = important_counts[['exon_id','miRNA']].join(trans,on='miRNA').merge(mirna_counts[['miRNA_ID','read_count']],on='miRNA_ID')\n",
    "#print(bs_mirnas)\n",
    "bs_mirnas.drop(['miRNA','miRNA_ID'], axis=1, inplace=True)\n",
    "bs_mirnas = bs_mirnas.rename(columns={'read_count': \"mirna_read_count\"})\n",
    "bs_mirnas.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-03T13:21:17.881609Z",
     "start_time": "2020-10-03T13:21:17.781335Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon</th>\n",
       "      <th>raw_counts</th>\n",
       "      <th>median_length_normalized</th>\n",
       "      <th>RPKM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:11874-12227:+</td>\n",
       "      <td>61</td>\n",
       "      <td>0.9036827</td>\n",
       "      <td>1.21214019174369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:12595-12721:+</td>\n",
       "      <td>16</td>\n",
       "      <td>0.8333333</td>\n",
       "      <td>0.886222027370111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:12613-12721:+</td>\n",
       "      <td>15</td>\n",
       "      <td>0.8055556</td>\n",
       "      <td>0.96803495535554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:12646-12697:+</td>\n",
       "      <td>12</td>\n",
       "      <td>0.8823529</td>\n",
       "      <td>1.62332015590391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:13221-14409:+</td>\n",
       "      <td>287</td>\n",
       "      <td>0.9166667</td>\n",
       "      <td>1.69795556537075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239317</th>\n",
       "      <td>chrM_rCRS:12207-12237:+</td>\n",
       "      <td>499</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>113.230944638157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239318</th>\n",
       "      <td>chrM_rCRS:12337-14644:+</td>\n",
       "      <td>49206</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>149.971431353018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239319</th>\n",
       "      <td>chrM_rCRS:14761-15887:+</td>\n",
       "      <td>181297</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>1131.6001082413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239320</th>\n",
       "      <td>chrM_rCRS:15998-16182:+</td>\n",
       "      <td>1587</td>\n",
       "      <td>1.0000000</td>\n",
       "      <td>60.3436362818982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239321</th>\n",
       "      <td>chrM_rCRS:16183-16569:+</td>\n",
       "      <td>2386</td>\n",
       "      <td>0.9974093</td>\n",
       "      <td>43.3696335881373</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>239322 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           exon raw_counts median_length_normalized  \\\n",
       "0            chr1:11874-12227:+         61                0.9036827   \n",
       "1            chr1:12595-12721:+         16                0.8333333   \n",
       "2            chr1:12613-12721:+         15                0.8055556   \n",
       "3            chr1:12646-12697:+         12                0.8823529   \n",
       "4            chr1:13221-14409:+        287                0.9166667   \n",
       "...                         ...        ...                      ...   \n",
       "239317  chrM_rCRS:12207-12237:+        499                1.0000000   \n",
       "239318  chrM_rCRS:12337-14644:+      49206                1.0000000   \n",
       "239319  chrM_rCRS:14761-15887:+     181297                1.0000000   \n",
       "239320  chrM_rCRS:15998-16182:+       1587                1.0000000   \n",
       "239321  chrM_rCRS:16183-16569:+       2386                0.9974093   \n",
       "\n",
       "                     RPKM  \n",
       "0        1.21214019174369  \n",
       "1       0.886222027370111  \n",
       "2        0.96803495535554  \n",
       "3        1.62332015590391  \n",
       "4        1.69795556537075  \n",
       "...                   ...  \n",
       "239317   113.230944638157  \n",
       "239318   149.971431353018  \n",
       "239319    1131.6001082413  \n",
       "239320   60.3436362818982  \n",
       "239321   43.3696335881373  \n",
       "\n",
       "[239322 rows x 4 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans \n",
    "important_counts\n",
    "mirna_counts\n",
    "exon_counts\n",
    "#2693 with hsa in miRNA2mature / 49156 overall in miRNA2mature (also other species eg cel)\n",
    "#2656 if inner join cause only 2656 families\n",
    "#TODO check if miRNA counts from TCGA are more different miRNAs?\n",
    "#TODO is inner right? are some in important counts None?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-02T16:32:02.644533Z",
     "start_time": "2020-10-02T16:32:02.544292Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'bs_exons' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-20-67c2df6327f5>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m#output: for each case_id for each bs_id: interesting miRNA expression + exon expression\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;31m#then regression btw miRNA expression, exon expression\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m \u001b[0mboth\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mbs_exons\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mbs_mirnas\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mon\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'bs_id'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'outer'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdf_exon\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mon\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'bs_id'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mboth\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexon_raw_count\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m<\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mboth\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Error: Number of exon count rows with None:'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mboth\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0mboth\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexon_raw_count\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'/'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mboth\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'bs_exons' is not defined"
     ]
    }
   ],
   "source": [
    "#output: for each case_id for each bs_id: interesting miRNA expression + exon expression\n",
    "#then regression btw miRNA expression, exon expression\n",
    "both = bs_exons.merge(bs_mirnas, on='bs_id', how='outer').merge(df_exon, on='bs_id')\n",
    "if both.exon_raw_count.count() < len(both):\n",
    "    print('Error: Number of exon count rows with None:',len(both) - both.exon_raw_count.count(),'/',len(both))\n",
    "    #this is because we dont have all the bs predictions yet?\n",
    "if both.mirna_read_count.count() < len(both):\n",
    "    print('Error: Number of mirna count rows with None:',both[pd.isnull(both['mirna_read_count']) & pd.isnull(both['exon_raw_count'])])\n",
    "    #this is because for this case we dont information on this mirna in TGCA?\n",
    "both = both.dropna()\n",
    "both[['exon_raw_count','mirna_read_count']] = both[['exon_raw_count','mirna_read_count']].astype('int')\n",
    "both.head()\n",
    "\n",
    "both_sum = both.drop_duplicates(subset='exon_id')[['exon_raw_count','exon_id']].merge(both.groupby(['exon_id'])['mirna_read_count'].agg('sum'),on='exon_id')\n",
    "both_sum[['exon_raw_count','mirna_read_count']] = both_sum[['exon_raw_count','mirna_read_count']].astype('int')\n",
    "both_sum = both_sum.rename(columns={'mirna_read_count': 'mirna_read_count_sum'})\n",
    "both_sum.head()\n",
    "#both.head()\n",
    "#TODO check - is nr mirna downloaded also here? if not: mistake\n",
    "#TODo step both is unnecessary, remove and directly merge 2 df as in slides (in case summing up is right)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Elastic Net Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-26T15:36:04.383771Z",
     "start_time": "2020-09-26T15:36:04.236607Z"
    }
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Expected 2D array, got 1D array instead:\narray=[1320. 1490.  300.].\nReshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-147-8df5da2da401>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mboth2\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0me\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mboth2\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mm\u001b[0m\u001b[1;31m#both.exon_raw_count, both.mirna_read_count\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[0mregr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mElasticNet\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrandom_state\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 9\u001b[1;33m \u001b[0mregr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     10\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mregr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcoef_\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;31m#X{ndarray, sparse matrix} of (n_samples, n_features) data\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\linear_model\\_coordinate_descent.py\u001b[0m in \u001b[0;36mfit\u001b[1;34m(self, X, y, sample_weight, check_input)\u001b[0m\n\u001b[0;32m    757\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mcheck_input\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    758\u001b[0m             \u001b[0mX_copied\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcopy_X\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfit_intercept\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 759\u001b[1;33m             X, y = self._validate_data(X, y, accept_sparse='csc',\n\u001b[0m\u001b[0;32m    760\u001b[0m                                        \u001b[0morder\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'F'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    761\u001b[0m                                        \u001b[0mdtype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfloat64\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfloat32\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\base.py\u001b[0m in \u001b[0;36m_validate_data\u001b[1;34m(self, X, y, reset, validate_separately, **check_params)\u001b[0m\n\u001b[0;32m    430\u001b[0m                 \u001b[0my\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mcheck_y_params\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    431\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 432\u001b[1;33m                 \u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcheck_X_y\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mcheck_params\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    433\u001b[0m             \u001b[0mout\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    434\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\utils\\validation.py\u001b[0m in \u001b[0;36minner_f\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     70\u001b[0m                           FutureWarning)\n\u001b[0;32m     71\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m{\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msig\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0margs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 72\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     73\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     74\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\utils\\validation.py\u001b[0m in \u001b[0;36mcheck_X_y\u001b[1;34m(X, y, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, estimator)\u001b[0m\n\u001b[0;32m    793\u001b[0m         \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"y cannot be None\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    794\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 795\u001b[1;33m     X = check_array(X, accept_sparse=accept_sparse,\n\u001b[0m\u001b[0;32m    796\u001b[0m                     \u001b[0maccept_large_sparse\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maccept_large_sparse\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    797\u001b[0m                     \u001b[0mdtype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0morder\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0morder\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\utils\\validation.py\u001b[0m in \u001b[0;36minner_f\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     70\u001b[0m                           FutureWarning)\n\u001b[0;32m     71\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m{\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msig\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0margs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 72\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     73\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     74\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\utils\\validation.py\u001b[0m in \u001b[0;36mcheck_array\u001b[1;34m(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator)\u001b[0m\n\u001b[0;32m    617\u001b[0m             \u001b[1;31m# If input is 1D raise error\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    618\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0marray\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 619\u001b[1;33m                 raise ValueError(\n\u001b[0m\u001b[0;32m    620\u001b[0m                     \u001b[1;34m\"Expected 2D array, got 1D array instead:\\narray={}.\\n\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    621\u001b[0m                     \u001b[1;34m\"Reshape your data either using array.reshape(-1, 1) if \"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Expected 2D array, got 1D array instead:\narray=[1320. 1490.  300.].\nReshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample."
     ]
    }
   ],
   "source": [
    "import sklearn\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.datasets import make_regression\n",
    "#faster if np.array(x, order='F')\n",
    "#If you are interested in controlling the L1 and L2 penalty separately, keep in mind that this is equivalent to:\n",
    "#a * L1 + b * L2 where: alpha = a + b and l1_ratio = a / (a + b)\n",
    "X, y = both_sum.mirna_read_count, both_sum.exon_raw_count\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "print(regr.coef_)\n",
    "\n",
    "#X{ndarray, sparse matrix} of (n_samples, n_features) data\n",
    "#y{ndarray, sparse matrix} of shape (n_samples,) or (n_samples, n_targets) target, auto cast to x\n",
    "\n",
    "#print(regr.intercept_)\n",
    "#print(regr.predict([[0, 0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-09-24T21:28:45.043481Z",
     "start_time": "2020-09-24T21:28:44.896344Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 5 entries, 0 to 4\n",
      "Data columns (total 8 columns):\n",
      " #   Column           Non-Null Count  Dtype   \n",
      "---  ------           --------------  -----   \n",
      " 0   bs_id            5 non-null      int64   \n",
      " 1   miRNA            5 non-null      category\n",
      " 2   mRNA             5 non-null      category\n",
      " 3   chrom_bs_start   5 non-null      int64   \n",
      " 4   chrom_bs_end     5 non-null      int64   \n",
      " 5   chromosome_name  5 non-null      category\n",
      " 6   strand           5 non-null      category\n",
      " 7   exons            5 non-null      object  \n",
      "dtypes: category(4), int64(3), object(1)\n",
      "memory usage: 250.0 KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       exons\n",
       "count      5\n",
       "unique     1\n",
       "top         \n",
       "freq       5"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#improve memory of DataFrame\n",
    "#converting object variable of type string to categorical -> reduce memory\n",
    "print(important_counts.info(memory_usage='deep'))\n",
    "a = important_counts.select_dtypes(include=['object']).copy()\n",
    "a.describe()"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "Seqvec basic Seq ligand binding.ipynb",
   "provenance": []
  },
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python (thesis)",
   "language": "python",
   "name": "thesis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "581.6px",
    "left": "235px",
    "top": "158.533px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
