{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "center"
   },
   "source": [
    "# Assemble the final Dataframe from tarpmir bindingsites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in bindingsite & Ensembl data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-27T12:43:04.072027Z",
     "start_time": "2020-10-27T12:43:03.334014Z"
    },
    "cell_style": "center"
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from helper_fcts import *\n",
    "import sqlite3\n",
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "test = False\n",
    "path = Path('data')\n",
    "tcga_path = Path(path/'KICH')\n",
    "path_tarp = Path(path/'tarp-bs') if test else Path('../data')\n",
    "path_input = Path(path/'tarp-bs/cdna') if test else Path('../data/all_cdna_subsets')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test bindingsite\n",
    "bs = parse_tarp_bs(path/'tarp-bs'/'input_miRNA.fasta_cdna6247.fasta.bp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T21:08:25.447218Z",
     "start_time": "2020-10-25T21:08:19.206772Z"
    }
   },
   "outputs": [],
   "source": [
    "#read in df_gene: gene_id, g_id_v, transcript_id, t_id_v, chromosome, strand, start, end (1-N)\n",
    "df_gene = pd.read_csv(path/'gene_infos.csv', dtype={'ensembl_gene_id':str,'ensembl_gene_id_version':str,'ensembl_transcript_id':str, 'ensembl_transcript_id_version':str, 'chromosome_name':'category', 'strand':'int8', 'start_position':int, 'end_position':int})\n",
    "del df_gene['Unnamed: 0']\n",
    "df_gene[['strand']] = df_gene[['strand']].astype('str').astype('category')\n",
    "\n",
    "#executed once, read in from pickled files\n",
    "exon_positions = read_in_exon_pos(path/'exon_positions.csv') # mapping of transcript ids to exon ids, chrom_exon_starts, exon_starts, exon_ends\n",
    "\n",
    "#df_exon = calc_exon_data(path) # exon starts and ends per transcript + exon id\n",
    "#df_exon = df_exon.merge(df_gene, left_on = 'transcript_id', right_on = 'ensembl_transcript_id', how='left')\n",
    "#df_exon[['chrom_exon_start','chrom_exon_end','start_position','end_position']] = df_exon[['chrom_exon_start','chrom_exon_end','start_position','end_position']].astype('Int64')\n",
    "#df_exon[['strand']] = df_exon[['strand']].astype('str').astype('category')\n",
    "\n",
    "#TODO put overall more function in helper\n",
    "\n",
    "#len(df_exon[df_exon.ensembl_transcript_id.isna()].transcript_id.unique())\n",
    "#371 transcripts are not in df_gene, TODO, why? which transcripts are in tarpmir predictions\n",
    "#TODO instead of joining df_gene, exon_info -> directly read in from R feather file & also use that for calculating chrom pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 819237 entries, 0 to 819236\n",
      "Data columns (total 5 columns):\n",
      " #   Column            Non-Null Count   Dtype   \n",
      "---  ------            --------------   -----   \n",
      " 0   exon_id           819237 non-null  object  \n",
      " 1   chrom_exon_start  819237 non-null  Int64   \n",
      " 2   chrom_exon_end    819237 non-null  Int64   \n",
      " 3   chromosome_name   817389 non-null  category\n",
      " 4   strand            817389 non-null  category\n",
      "dtypes: Int64(2), category(2), object(1)\n",
      "memory usage: 28.9+ MB\n"
     ]
    }
   ],
   "source": [
    "#IMPORTANT\n",
    "df_exon = calc_exon_data(path) # exon starts and ends per transcript + exon id\n",
    "df_exon = df_exon.merge(df_gene[['ensembl_transcript_id','chromosome_name','strand']], left_on = 'transcript_id', right_on = 'ensembl_transcript_id', how='left')\n",
    "df_exon[['chrom_exon_start','chrom_exon_end']] = df_exon[['chrom_exon_start','chrom_exon_end']].astype('Int64')\n",
    "del df_exon['transcript_id']\n",
    "del df_exon['ensembl_transcript_id']\n",
    "df_exon[['strand']] = df_exon[['strand']].astype('str').astype('category')\n",
    "df_exon[['exon_id']] = df_exon[['exon_id']].astype('str').astype('category')\n",
    "df_exon.to_parquet(path/'ensembl_exon_info.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#IMPORTANT\n",
    "df_tran = df_gene[['ensembl_transcript_id','chromosome_name','strand']]\n",
    "df_tran[['ensembl_transcript_id']] = df_tran[['ensembl_transcript_id']].astype('str').astype('category')\n",
    "df_tran.to_parquet(path/'ensembl_transcript_info.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in bs + make subset \n",
    "bs = pd.read_parquet(path/'bs_all.parquet') #983499270 rows\n",
    "bs_new = bs[['miRNA','mRNA','bs_start','bs_end']].copy()\n",
    "bs_new['miRNA'] = pd.Categorical(bs_new.miRNA)\n",
    "bs_new['mRNA'] = pd.Categorical(bs_new.mRNA)\n",
    "bs_new.info()\n",
    "bs_new.to_parquet(path/'bs_subset.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>number_of_pairings</th>\n",
       "      <th>binding_region_length</th>\n",
       "      <th>longest_consecutive_pairings</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000476331</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>1</td>\n",
       "      <td>1.777982e-03</td>\n",
       "      <td>0.426</td>\n",
       "      <td>1.394372</td>\n",
       "      <td>1.710881</td>\n",
       "      <td>-4.026374</td>\n",
       "      <td>14.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>301</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000476331</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>-18.5</td>\n",
       "      <td>0</td>\n",
       "      <td>2.694336e-04</td>\n",
       "      <td>0.456</td>\n",
       "      <td>0.275755</td>\n",
       "      <td>1.061554</td>\n",
       "      <td>-13.549385</td>\n",
       "      <td>16.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>423</td>\n",
       "      <td>441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000476331</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>-18.6</td>\n",
       "      <td>0</td>\n",
       "      <td>6.177285e-05</td>\n",
       "      <td>0.559</td>\n",
       "      <td>0.251558</td>\n",
       "      <td>0.853162</td>\n",
       "      <td>-16.125569</td>\n",
       "      <td>16.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>158</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000476331</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-21.7</td>\n",
       "      <td>0</td>\n",
       "      <td>5.099209e-09</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.233710</td>\n",
       "      <td>-0.052141</td>\n",
       "      <td>-6.514540</td>\n",
       "      <td>19.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>11</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000476331</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>1</td>\n",
       "      <td>1.777982e-03</td>\n",
       "      <td>0.426</td>\n",
       "      <td>1.018645</td>\n",
       "      <td>1.724342</td>\n",
       "      <td>-4.026374</td>\n",
       "      <td>14.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>318</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983499265</th>\n",
       "      <td>hsa-miR-12125</td>\n",
       "      <td>ENST00000511501</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-16.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.360765e-06</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-8.350718</td>\n",
       "      <td>12.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>81</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983499266</th>\n",
       "      <td>hsa-miR-12126</td>\n",
       "      <td>ENST00000511501</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-16.3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.420751e-06</td>\n",
       "      <td>0.368</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-11.418273</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983499267</th>\n",
       "      <td>hsa-miR-12127</td>\n",
       "      <td>ENST00000511501</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>1</td>\n",
       "      <td>1.328924e-04</td>\n",
       "      <td>0.412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-7.781775</td>\n",
       "      <td>19.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983499268</th>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000511501</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-15.7</td>\n",
       "      <td>1</td>\n",
       "      <td>3.936156e-04</td>\n",
       "      <td>0.618</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-9.061245</td>\n",
       "      <td>16.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>190</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983499269</th>\n",
       "      <td>hsa-miR-12131</td>\n",
       "      <td>ENST00000511501</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>-18.5</td>\n",
       "      <td>0</td>\n",
       "      <td>9.873809e-06</td>\n",
       "      <td>0.412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-6.193851</td>\n",
       "      <td>14.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>24</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>983499270 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     miRNA             mRNA  binding_probability  energy  \\\n",
       "0            hsa-let-7a-5p  ENST00000476331             1.000000   -18.4   \n",
       "1          hsa-let-7a-2-3p  ENST00000476331             0.692308   -18.5   \n",
       "2          hsa-let-7a-2-3p  ENST00000476331             0.769231   -18.6   \n",
       "3            hsa-let-7b-5p  ENST00000476331             0.615385   -21.7   \n",
       "4            hsa-let-7b-5p  ENST00000476331             1.000000   -18.4   \n",
       "...                    ...              ...                  ...     ...   \n",
       "983499265    hsa-miR-12125  ENST00000511501             0.538462   -16.1   \n",
       "983499266    hsa-miR-12126  ENST00000511501             1.000000   -16.3   \n",
       "983499267    hsa-miR-12127  ENST00000511501             0.923077   -20.6   \n",
       "983499268    hsa-miR-12129  ENST00000511501             0.538462   -15.7   \n",
       "983499269    hsa-miR-12131  ENST00000511501             0.769231   -18.5   \n",
       "\n",
       "           seed  accessibility  AU_content  PhyloP_Stem  PyloP_Flanking  \\\n",
       "0             1   1.777982e-03       0.426     1.394372        1.710881   \n",
       "1             0   2.694336e-04       0.456     0.275755        1.061554   \n",
       "2             0   6.177285e-05       0.559     0.251558        0.853162   \n",
       "3             0   5.099209e-09       0.500     0.233710       -0.052141   \n",
       "4             1   1.777982e-03       0.426     1.018645        1.724342   \n",
       "...         ...            ...         ...          ...             ...   \n",
       "983499265     0   1.360765e-06       0.338     0.000000        0.000000   \n",
       "983499266     1   1.420751e-06       0.368     0.000000        0.000000   \n",
       "983499267     1   1.328924e-04       0.412     0.000000        0.000000   \n",
       "983499268     1   3.936156e-04       0.618     0.000000        0.000000   \n",
       "983499269     0   9.873809e-06       0.412     0.000000        0.000000   \n",
       "\n",
       "                m/e   number_of_pairings  binding_region_length  \\\n",
       "0          -4.026374                14.0                   17.0   \n",
       "1         -13.549385                16.0                   18.0   \n",
       "2         -16.125569                16.0                   23.0   \n",
       "3          -6.514540                19.0                   24.0   \n",
       "4          -4.026374                14.0                   17.0   \n",
       "...              ...                 ...                    ...   \n",
       "983499265  -8.350718                12.0                   16.0   \n",
       "983499266 -11.418273                 9.0                   10.0   \n",
       "983499267  -7.781775                19.0                   24.0   \n",
       "983499268  -9.061245                16.0                   21.0   \n",
       "983499269  -6.193851                14.0                   18.0   \n",
       "\n",
       "           longest_consecutive_pairings  \\\n",
       "0                                   9.0   \n",
       "1                                  10.0   \n",
       "2                                   8.0   \n",
       "3                                   9.0   \n",
       "4                                   9.0   \n",
       "...                                 ...   \n",
       "983499265                           5.0   \n",
       "983499266                           9.0   \n",
       "983499267                          11.0   \n",
       "983499268                           8.0   \n",
       "983499269                           6.0   \n",
       "\n",
       "           position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "0                                               7.0                     1.0   \n",
       "1                                               1.0                     6.0   \n",
       "2                                               8.0                     7.0   \n",
       "3                                              15.0                     8.0   \n",
       "4                                               7.0                     1.0   \n",
       "...                                             ...                     ...   \n",
       "983499265                                       1.0                     0.0   \n",
       "983499266                                       1.0                     0.0   \n",
       "983499267                                      13.0                     5.0   \n",
       "983499268                                      13.0                     5.0   \n",
       "983499269                                       7.0                     4.0   \n",
       "\n",
       "           difference_of_pairings_between_seed_and_3prime_end  bs_start  \\\n",
       "0                                                        6.0        301   \n",
       "1                                                        1.0        423   \n",
       "2                                                        3.0        158   \n",
       "3                                                        3.0         11   \n",
       "4                                                        6.0        318   \n",
       "...                                                      ...        ...   \n",
       "983499265                                                6.0         81   \n",
       "983499266                                                6.0         18   \n",
       "983499267                                                1.0         42   \n",
       "983499268                                                3.0        190   \n",
       "983499269                                                2.0         24   \n",
       "\n",
       "           bs_end  \n",
       "0             340  \n",
       "1             441  \n",
       "2             203  \n",
       "3              35  \n",
       "4             340  \n",
       "...           ...  \n",
       "983499265      97  \n",
       "983499266      28  \n",
       "983499267      66  \n",
       "983499268     211  \n",
       "983499269      72  \n",
       "\n",
       "[983499270 rows x 18 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs\n",
    "#TODO commit\n",
    "#TODO prepare kidney data as categorical & uint & parquet format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read bs\n"
     ]
    }
   ],
   "source": [
    "#split bs in 10 parts\n",
    "for i in range(1,9):\n",
    "    bs = pd.read_parquet(path/'bs_all.parquet') #983499270 rows\n",
    "    bs = bs[(i-1)*100000000:i*100000000]\n",
    "    print('read bs')\n",
    "    compute_chrom_pos(bs, path/('bs_chrom_pos'+str(i)+'.parquet'), exon_positions)\n",
    "    print('saved parquet',i)#387.26 GB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:07:50.384154Z",
     "start_time": "2020-10-21T23:07:44.980201Z"
    }
   },
   "outputs": [],
   "source": [
    "#parse miRNA and mRNA seed from sequence\n",
    "miRNA_sequences = parse_seq(path_tarp/'input_miRNA.fasta') #2661 miRNAs \n",
    "mRNA_sequences = pd.read_parquet(path/'mRNA_all.parquet')\n",
    "print('All miRNAs and mRNAs were read into pandas.')\n",
    "\n",
    "bs['miRNA_seed'] = bs.apply(lambda row: miRNA_sequences[row['miRNA']], axis=1)\n",
    "bs['mRNA_bs_seq'] = bs.apply(lambda row: mRNA_sequences[row['mRNA']][row['bs_start']:row['bs_end']], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:20:02.121012Z",
     "start_time": "2020-10-25T14:20:01.486989Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:20:20.279598Z",
     "start_time": "2020-10-25T14:20:20.189820Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def pos_to_chrom(pos, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if pos >= exon_starts[tid][eid] and pos <= exon_ends[tid][eid]:\n",
    "                diff = pos - exon_starts[tid][eid]\n",
    "                chrom_pos = chrom_exon_starts[tid][eid] + diff\n",
    "                return chrom_pos\n",
    "        else: print('Error: Either',tid,'or',eid,'not in exon_starts or exon_ends')\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:20:37.810692Z",
     "start_time": "2020-10-25T14:20:37.716745Z"
    }
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def get_eid(start, end, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if (start >= exon_starts[tid][eid]) and (start <= exon_ends[tid][eid]) and (end >= exon_starts[tid][eid]) and (end <= exon_ends[tid][eid]):\n",
    "                return eid\n",
    "        else: print('Error: Either',tid,'or',eid,'not in exon_starts or exon_ends')\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:21:08.552003Z",
     "start_time": "2020-10-25T14:20:55.441548Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#translate genome position relative to transcript to chromosome\n",
    "#fastest, always puts bs start and end even if bs spans several exons, but only puts exonid if only 1 exon\n",
    "bs['chrom_bs_start'] = bs.apply(lambda row: pos_to_chrom(row.bs_start, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['chrom_bs_end'] = bs.apply(lambda row: pos_to_chrom(row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['exon_id'] = bs.apply(lambda row: get_eid(row.bs_start, row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['bs_id'] = bs.index\n",
    "bs.to_parquet(path/'bs_chrom_pos.parquet')\n",
    "bs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in GDC data (processed by Xena) from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n#!wget https://tcga.xenahubs.net/download/unc_v2_exon_hg19_probe_TCGA #unnecessary cause some exons missing\\n!wget https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/HiSeqV2_exon.gz\\n!wget https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/miRNA_HiSeq_gene.gz\\n!gunzip HiSeqV2_exon.gz\\n!gunzip miRNA_HiSeq_gene.gz\\n!mv HiSeqV2_exon data/KICH\\n!mv miRNA_HiSeq_gene data/KICH\\n!mv unc_v2_exon_hg19_probe_TCGA data/KICH'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#download GDC ChGR37 data from web, once, TCGA Kidney Chromophobe (KICH)\n",
    "#https://xenabrowser.net/datapages/?hub=https://tcga.xenahubs.net:443\n",
    "'''\n",
    "#!wget https://tcga.xenahubs.net/download/unc_v2_exon_hg19_probe_TCGA #unnecessary cause some exons missing\n",
    "!wget https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/HiSeqV2_exon.gz\n",
    "!wget https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/miRNA_HiSeq_gene.gz\n",
    "!gunzip HiSeqV2_exon.gz\n",
    "!gunzip miRNA_HiSeq_gene.gz\n",
    "!mv HiSeqV2_exon data/KICH\n",
    "!mv miRNA_HiSeq_gene data/KICH\n",
    "!mv unc_v2_exon_hg19_probe_TCGA data/KICH'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%cython\n",
    "def split_chrom_column(chrom):\n",
    "    chrom_list = chrom.split(':')\n",
    "    chromosome_name = str(chrom_list[0])#[3:])\n",
    "    strand = 1 if chrom_list[2] == '+' else -1\n",
    "    chrom_exon_start = int(chrom_list[1].split('-')[0])\n",
    "    chrom_exon_end = int(chrom_list[1].split('-')[1])\n",
    "    return chromosome_name,strand,chrom_exon_start,chrom_exon_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#IMPORTANT\n",
    "mirna_counts = pd.read_csv(tcga_path/'miRNA_HiSeq_gene', delimiter='\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "102885 values are NaN out of 172530 values (= 0.5963310728568945 %)\n"
     ]
    }
   ],
   "source": [
    "#TODO many values NaN!!!\n",
    "count_nans = mirna_counts.isnull().sum().sum()\n",
    "print(count_nans,'values are NaN out of',1917*90,'values (=',(count_nans/(1917*90))*100,'%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       86\n",
       "1       76\n",
       "2       82\n",
       "3       78\n",
       "4       63\n",
       "        ..\n",
       "1912    82\n",
       "1913    63\n",
       "1914    63\n",
       "1915    74\n",
       "1916     2\n",
       "Length: 1917, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_counts.isnull().sum(axis=1) #amount of NaNs per miRNA\n",
    "mirna_counts.isnull().sum() #amount of NaNs per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mature_name</th>\n",
       "      <th>mirna_family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIMAT0000062</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>let-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIMAT0004481</td>\n",
       "      <td>hsa-let-7a-3p</td>\n",
       "      <td>let-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>let-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIMAT0000063</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>let-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>let-7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     mature_acc      mature_name mirna_family\n",
       "0  MIMAT0000062    hsa-let-7a-5p        let-7\n",
       "1  MIMAT0004481    hsa-let-7a-3p        let-7\n",
       "2  MIMAT0010195  hsa-let-7a-2-3p        let-7\n",
       "3  MIMAT0000063    hsa-let-7b-5p        let-7\n",
       "4  MIMAT0004482    hsa-let-7b-3p        let-7"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#IMPORTANT\n",
    "trans = pd.read_feather(path/'mature2families.feather')\n",
    "trans[['mature_name']] = trans[['mature_name']].astype('category')\n",
    "mirna_counts = mirna_counts.merge(trans, left_on='sample',right_on='mature_acc', how='left')\n",
    "del mirna_counts['mature_acc']\n",
    "mirna_counts.to_parquet(path/'KICH'/'miRNA_counts.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>TCGA-KN-8419-11</th>\n",
       "      <th>TCGA-KL-8332-01</th>\n",
       "      <th>TCGA-KN-8423-01</th>\n",
       "      <th>TCGA-KO-8405-01</th>\n",
       "      <th>TCGA-KL-8337-01</th>\n",
       "      <th>TCGA-KN-8424-11</th>\n",
       "      <th>TCGA-KN-8422-11</th>\n",
       "      <th>TCGA-KL-8334-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "      <th>mature_name</th>\n",
       "      <th>mirna_family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>MIMAT0000062</td>\n",
       "      <td>14.422574</td>\n",
       "      <td>14.82746</td>\n",
       "      <td>14.880758</td>\n",
       "      <td>14.793954</td>\n",
       "      <td>14.541135</td>\n",
       "      <td>15.14849</td>\n",
       "      <td>14.417657</td>\n",
       "      <td>14.382524</td>\n",
       "      <td>14.991116</td>\n",
       "      <td>...</td>\n",
       "      <td>15.348145</td>\n",
       "      <td>14.600445</td>\n",
       "      <td>14.474857</td>\n",
       "      <td>14.568523</td>\n",
       "      <td>14.486888</td>\n",
       "      <td>15.229448</td>\n",
       "      <td>14.887372</td>\n",
       "      <td>14.794259</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>let-7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample  TCGA-KN-8426-01  TCGA-KN-8419-11  TCGA-KL-8332-01  \\\n",
       "692  MIMAT0000062        14.422574         14.82746        14.880758   \n",
       "\n",
       "     TCGA-KN-8423-01  TCGA-KO-8405-01  TCGA-KL-8337-01  TCGA-KN-8424-11  \\\n",
       "692        14.793954        14.541135         15.14849        14.417657   \n",
       "\n",
       "     TCGA-KN-8422-11  TCGA-KL-8334-01  ...  TCGA-KL-8323-01  TCGA-KN-8429-11  \\\n",
       "692        14.382524        14.991116  ...        15.348145        14.600445   \n",
       "\n",
       "     TCGA-KO-8413-01  TCGA-KO-8411-01  TCGA-KL-8332-11  TCGA-KL-8335-01  \\\n",
       "692        14.474857        14.568523        14.486888        15.229448   \n",
       "\n",
       "     TCGA-KL-8339-11  TCGA-KO-8415-01    mature_name  mirna_family  \n",
       "692        14.887372        14.794259  hsa-let-7a-5p         let-7  \n",
       "\n",
       "[1 rows x 92 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_counts[mirna_counts.mature_name=='hsa-let-7a-5p']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "73451"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_counts.count(0).sum() #I cant just replace Nan with 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirna_counts.group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frame['c'] = frame.a.fillna(0) + frame.b.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirna_counts.to_parquet(path/'KICH'/'miRNA_counts.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-27T13:09:50.676110Z",
     "start_time": "2020-10-27T12:43:13.778366Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read from file\n"
     ]
    }
   ],
   "source": [
    "#IMPORTANT\n",
    "f = Path(path/'exon_counts_kidney.feather')\n",
    "if f.is_file():\n",
    "    print('read from file')\n",
    "    exon_counts = pd.read_feather(f)\n",
    "else:\n",
    "    print('translate exon count positions from GRCh37 to GRCh38')\n",
    "    exon_counts = pd.read_csv(tcga_path/'HiSeqV2_exon', delimiter='\t')\n",
    "    exon_counts[['chrom_old','strand_old','start_old','end_old']] = exon_counts['sample'].apply(lambda x: pd.Series(split_chrom_column(str(x))))\n",
    "    #save positions to BED input file\n",
    "    exon_counts['score'] = 0\n",
    "    exon_counts.to_csv('data/exon_hg19_pos.bed', sep='\\t', columns=['chrom_old','start_old','end_old','sample','score','strand_old'], header=False, index=False)\n",
    "    #translate positions from GRCh37 to GRCh38\n",
    "    !../liftOver data/exon_hg19_pos.bed data/hg19ToHg38.over.chain.gz data/exon_hg38_pos.bed data/unlifted.bed\n",
    "    #read in translated positions from liftOver output BED file\n",
    "    translated_positions = pd.read_csv('data/exon_hg38_pos.bed', delimiter='\t', names=['chromosome_name','chrom_exon_start','chrom_exon_end','id','score','strand'])\n",
    "    exon_counts = exon_counts.merge(translated_positions[['chromosome_name','chrom_exon_start','chrom_exon_end','id','strand']], left_on='sample', right_on='id', how='inner',validate='1:1')\n",
    "    exon_counts.drop(['sample','chrom_old','strand_old','start_old','end_old','score'], axis=1, inplace=True)\n",
    "    #transform strand {-,+} to {-1,+1}\n",
    "    exon_counts['strand'] = exon_counts.apply(lambda row: -1 if row.strand == '-' else 1, axis=1)\n",
    "    exon_counts['chromosome_name'] = exon_counts['chromosome_name'].str[3:]\n",
    "    exon_counts[['chromosome_name','strand']] = exon_counts[['chromosome_name','strand']].astype('str').astype('category')\n",
    "    exon_counts[['chrom_exon_start','chrom_exon_end']] = exon_counts[['chrom_exon_start','chrom_exon_end']].astype('Int64')\n",
    "\n",
    "    #TODO option 2: get exonid per exon count from Ensembl (compare)\n",
    "    #exon_counts2 = exon_counts.merge(df_exon[['exon_id','chromosome_name','strand','chrom_exon_start','chrom_exon_end']], left_on = ['chrom','strand','chromStart','chromEnd'], right_on = ['chromosome_name','strand','chrom_exon_start','chrom_exon_end'], how='left')\n",
    "\n",
    "    caseids = list(set(mirna_counts.columns).intersection(list(exon_counts.columns)))\n",
    "    exon_counts.to_feather(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ensembl_exons' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-219c14ab71a1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#df_exon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mensembl_exons\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;31m#exon_counts['exon_id'] = exon_counts.merge(ensembl_exons, left_on='mature_acc', right_on='sample', how='inner')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ensembl_exons' is not defined"
     ]
    }
   ],
   "source": [
    "#IMPORTANT \n",
    "#TODO merge to check how many exon start + end after translation right / wrong\n",
    "#alternative: first exon_id then translate\n",
    "#TODO change slides\n",
    "#chromosome_name strand chrom_exon_start chrom_exon_end\n",
    "exon_counts['exon_id'] = exon_counts.merge(df_exon, in sqlite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSE00001544501</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSE00001544499</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSE00001544498</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSE00001544497</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSE00002006242</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819232</th>\n",
       "      <td>ENSE00003815305</td>\n",
       "      <td>128348163</td>\n",
       "      <td>128348438</td>\n",
       "      <td>CHR_HSCHR3_9_CTG2_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819233</th>\n",
       "      <td>ENSE00003827353</td>\n",
       "      <td>128343834</td>\n",
       "      <td>128344022</td>\n",
       "      <td>CHR_HSCHR3_9_CTG2_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819234</th>\n",
       "      <td>ENSE00003819782</td>\n",
       "      <td>128348163</td>\n",
       "      <td>128348718</td>\n",
       "      <td>CHR_HSCHR3_9_CTG2_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819235</th>\n",
       "      <td>ENSE00003835089</td>\n",
       "      <td>128188319</td>\n",
       "      <td>128188643</td>\n",
       "      <td>CHR_HSCHR3_9_CTG2_1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819236</th>\n",
       "      <td>ENSE00003831432</td>\n",
       "      <td>128200992</td>\n",
       "      <td>128201362</td>\n",
       "      <td>CHR_HSCHR3_9_CTG2_1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>819237 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                exon_id  chrom_exon_start  chrom_exon_end  \\\n",
       "0       ENSE00001544501               577             647   \n",
       "1       ENSE00001544499               648            1601   \n",
       "2       ENSE00001544498              1602            1670   \n",
       "3       ENSE00001544497              1671            3229   \n",
       "4       ENSE00002006242              3230            3304   \n",
       "...                 ...               ...             ...   \n",
       "819232  ENSE00003815305         128348163       128348438   \n",
       "819233  ENSE00003827353         128343834       128344022   \n",
       "819234  ENSE00003819782         128348163       128348718   \n",
       "819235  ENSE00003835089         128188319       128188643   \n",
       "819236  ENSE00003831432         128200992       128201362   \n",
       "\n",
       "            chromosome_name strand  \n",
       "0                        MT      1  \n",
       "1                        MT      1  \n",
       "2                        MT      1  \n",
       "3                        MT      1  \n",
       "4                        MT      1  \n",
       "...                     ...    ...  \n",
       "819232  CHR_HSCHR3_9_CTG2_1      1  \n",
       "819233  CHR_HSCHR3_9_CTG2_1      1  \n",
       "819234  CHR_HSCHR3_9_CTG2_1      1  \n",
       "819235  CHR_HSCHR3_9_CTG2_1     -1  \n",
       "819236  CHR_HSCHR3_9_CTG2_1     -1  \n",
       "\n",
       "[819237 rows x 5 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect(':memory:')\n",
    "#write the tables\n",
    "df_exon.to_sql('df_exon', conn, index=False, dtype={\"chrom_exon_start\": 'INTEGER', \"chrom_exon_end\": 'INTEGER'})\n",
    "exon_counts.to_sql('exon_counts', conn, index=True, index_label='exon_id', dtype={caseid : 'INTEGER' for caseid in caseids}.update({\"chrom_exon_start\": 'INTEGER', \"chrom_exon_end\": 'INTEGER'}))\n",
    "qry = '''\n",
    "    CREATE TABLE n_counts AS\n",
    "    select\n",
    "        *\n",
    "    from\n",
    "        exon_counts inner join df_exon\n",
    "        on\n",
    "        df_exon.chromosome_name = exon_counts.chromosome_name and df_exon.strand = exon_counts.strand and df_exon.chrom_bs_start >= exon_counts.chrom_exon_start and df_exon.chrom_bs_end <= exon_counts.chrom_exon_end\n",
    "    '''\n",
    "pd.read_sql_query(qry,conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#IMPORTANT\n",
    "n_exon_counts = exon_counts.copy()\n",
    "n_exon_counts = pd.melt(n_exon_counts, id_vars=[\"chromosome_name\", \"chrom_exon_start\",\"chrom_exon_end\",\"id\",\"strand\"], \n",
    "                  var_name=\"patient_id\", value_name=\"exon_count\")\n",
    "n_exon_counts['patient_id'] = n_exon_counts['patient_id'].astype('category')\n",
    "del n_exon_counts['id']\n",
    "n_exon_counts.info() #560 MB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 21730436 entries, 0 to 21730435\n",
      "Data columns (total 6 columns):\n",
      " #   Column            Dtype   \n",
      "---  ------            -----   \n",
      " 0   chromosome_name   category\n",
      " 1   chrom_exon_start  int64   \n",
      " 2   chrom_exon_end    int64   \n",
      " 3   strand            category\n",
      " 4   patient_id        category\n",
      " 5   exon_count        float64 \n",
      "dtypes: category(3), float64(1), int64(2)\n",
      "memory usage: 559.5 MB\n"
     ]
    }
   ],
   "source": [
    "n_exon_counts.to_parquet(path/'KICH'/'exon_counts.parquet') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>TCGA-KL-8346-01</th>\n",
       "      <th>TCGA-KN-8422-01</th>\n",
       "      <th>TCGA-KN-8431-11</th>\n",
       "      <th>TCGA-KN-8430-11</th>\n",
       "      <th>TCGA-KM-8440-01</th>\n",
       "      <th>TCGA-KO-8414-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KM-8639-01</th>\n",
       "      <th>TCGA-KO-8415-11</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8443-01</th>\n",
       "      <th>TCGA-KM-8442-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>id</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.1156</td>\n",
       "      <td>1.7396</td>\n",
       "      <td>2.0641</td>\n",
       "      <td>3.1457</td>\n",
       "      <td>3.0436</td>\n",
       "      <td>2.2209</td>\n",
       "      <td>1.8213</td>\n",
       "      <td>1.9092</td>\n",
       "      <td>2.9139</td>\n",
       "      <td>3.1049</td>\n",
       "      <td>...</td>\n",
       "      <td>2.2369</td>\n",
       "      <td>2.4951</td>\n",
       "      <td>2.1010</td>\n",
       "      <td>3.0617</td>\n",
       "      <td>0.7424</td>\n",
       "      <td>3</td>\n",
       "      <td>51973965</td>\n",
       "      <td>51974630</td>\n",
       "      <td>chr3:52007981-52008646:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>215728030</td>\n",
       "      <td>215728384</td>\n",
       "      <td>chr1:215901372-215901726:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.3884</td>\n",
       "      <td>0.2219</td>\n",
       "      <td>0.2050</td>\n",
       "      <td>0.9168</td>\n",
       "      <td>0.6069</td>\n",
       "      <td>0.4547</td>\n",
       "      <td>0.1571</td>\n",
       "      <td>0.3857</td>\n",
       "      <td>0.7535</td>\n",
       "      <td>0.8673</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1602</td>\n",
       "      <td>0.3507</td>\n",
       "      <td>0.7154</td>\n",
       "      <td>0.4353</td>\n",
       "      <td>0.0966</td>\n",
       "      <td>17</td>\n",
       "      <td>51191601</td>\n",
       "      <td>51193016</td>\n",
       "      <td>chr17:49268962-49270377:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.5290</td>\n",
       "      <td>0.7196</td>\n",
       "      <td>1.3094</td>\n",
       "      <td>2.3074</td>\n",
       "      <td>1.6894</td>\n",
       "      <td>1.5829</td>\n",
       "      <td>1.0835</td>\n",
       "      <td>1.9041</td>\n",
       "      <td>1.2131</td>\n",
       "      <td>1.7214</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6381</td>\n",
       "      <td>1.5227</td>\n",
       "      <td>1.7786</td>\n",
       "      <td>1.4157</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4</td>\n",
       "      <td>146292929</td>\n",
       "      <td>146292980</td>\n",
       "      <td>chr4:147214081-147214132:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4721</td>\n",
       "      <td>0.1284</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>2.4087</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0189</td>\n",
       "      <td>0.1334</td>\n",
       "      <td>4.8648</td>\n",
       "      <td>0.0900</td>\n",
       "      <td>1</td>\n",
       "      <td>15747546</td>\n",
       "      <td>15747980</td>\n",
       "      <td>chr1:16074041-16074475:+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238791</th>\n",
       "      <td>2.4455</td>\n",
       "      <td>5.1078</td>\n",
       "      <td>2.9468</td>\n",
       "      <td>5.4212</td>\n",
       "      <td>5.1299</td>\n",
       "      <td>4.5371</td>\n",
       "      <td>3.5954</td>\n",
       "      <td>4.4245</td>\n",
       "      <td>4.4484</td>\n",
       "      <td>5.5820</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0518</td>\n",
       "      <td>4.9067</td>\n",
       "      <td>3.8206</td>\n",
       "      <td>7.3459</td>\n",
       "      <td>4.1126</td>\n",
       "      <td>3</td>\n",
       "      <td>51978259</td>\n",
       "      <td>51978374</td>\n",
       "      <td>chr3:52012275-52012390:+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238792</th>\n",
       "      <td>1.5701</td>\n",
       "      <td>0.5500</td>\n",
       "      <td>0.3451</td>\n",
       "      <td>2.2783</td>\n",
       "      <td>1.7366</td>\n",
       "      <td>0.6813</td>\n",
       "      <td>0.5851</td>\n",
       "      <td>1.4460</td>\n",
       "      <td>2.4188</td>\n",
       "      <td>1.8648</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8498</td>\n",
       "      <td>1.0774</td>\n",
       "      <td>0.8774</td>\n",
       "      <td>1.2410</td>\n",
       "      <td>0.2725</td>\n",
       "      <td>6</td>\n",
       "      <td>43357126</td>\n",
       "      <td>43357765</td>\n",
       "      <td>chr6:43324864-43325503:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238793</th>\n",
       "      <td>0.4810</td>\n",
       "      <td>0.4219</td>\n",
       "      <td>0.3687</td>\n",
       "      <td>0.7313</td>\n",
       "      <td>0.7666</td>\n",
       "      <td>0.7646</td>\n",
       "      <td>2.4586</td>\n",
       "      <td>0.9557</td>\n",
       "      <td>1.1452</td>\n",
       "      <td>1.9724</td>\n",
       "      <td>...</td>\n",
       "      <td>1.3415</td>\n",
       "      <td>1.8842</td>\n",
       "      <td>0.8646</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.9102</td>\n",
       "      <td>17</td>\n",
       "      <td>36168630</td>\n",
       "      <td>36168695</td>\n",
       "      <td>chr17:34495988-34496053:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238794</th>\n",
       "      <td>2.6944</td>\n",
       "      <td>2.2891</td>\n",
       "      <td>1.8186</td>\n",
       "      <td>4.0291</td>\n",
       "      <td>4.0015</td>\n",
       "      <td>3.7968</td>\n",
       "      <td>2.3984</td>\n",
       "      <td>2.6564</td>\n",
       "      <td>4.2824</td>\n",
       "      <td>3.6833</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0631</td>\n",
       "      <td>1.8938</td>\n",
       "      <td>2.9015</td>\n",
       "      <td>2.3518</td>\n",
       "      <td>1.1468</td>\n",
       "      <td>1</td>\n",
       "      <td>120833602</td>\n",
       "      <td>120833774</td>\n",
       "      <td>chr1:148010884-148011056:-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238795</th>\n",
       "      <td>6.8409</td>\n",
       "      <td>5.8767</td>\n",
       "      <td>6.3420</td>\n",
       "      <td>6.0730</td>\n",
       "      <td>5.8084</td>\n",
       "      <td>6.7076</td>\n",
       "      <td>5.0113</td>\n",
       "      <td>6.8595</td>\n",
       "      <td>6.8408</td>\n",
       "      <td>6.0389</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0402</td>\n",
       "      <td>6.0639</td>\n",
       "      <td>6.8604</td>\n",
       "      <td>4.9554</td>\n",
       "      <td>3.4175</td>\n",
       "      <td>18</td>\n",
       "      <td>12353005</td>\n",
       "      <td>12353158</td>\n",
       "      <td>chr18:12353004-12353157:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>238796 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        TCGA-KN-8419-01  TCGA-KL-8346-01  TCGA-KN-8422-01  TCGA-KN-8431-11  \\\n",
       "0                2.1156           1.7396           2.0641           3.1457   \n",
       "1                0.0000           0.0000           0.0000           0.0251   \n",
       "2                0.3884           0.2219           0.2050           0.9168   \n",
       "3                2.5290           0.7196           1.3094           2.3074   \n",
       "4                0.0000           0.0000           0.0000           1.4721   \n",
       "...                 ...              ...              ...              ...   \n",
       "238791           2.4455           5.1078           2.9468           5.4212   \n",
       "238792           1.5701           0.5500           0.3451           2.2783   \n",
       "238793           0.4810           0.4219           0.3687           0.7313   \n",
       "238794           2.6944           2.2891           1.8186           4.0291   \n",
       "238795           6.8409           5.8767           6.3420           6.0730   \n",
       "\n",
       "        TCGA-KN-8430-11  TCGA-KM-8440-01  TCGA-KO-8414-01  TCGA-KL-8323-01  \\\n",
       "0                3.0436           2.2209           1.8213           1.9092   \n",
       "1                0.0233           0.0000           0.0000           0.0000   \n",
       "2                0.6069           0.4547           0.1571           0.3857   \n",
       "3                1.6894           1.5829           1.0835           1.9041   \n",
       "4                0.1284           0.0000           0.0000           0.0000   \n",
       "...                 ...              ...              ...              ...   \n",
       "238791           5.1299           4.5371           3.5954           4.4245   \n",
       "238792           1.7366           0.6813           0.5851           1.4460   \n",
       "238793           0.7666           0.7646           2.4586           0.9557   \n",
       "238794           4.0015           3.7968           2.3984           2.6564   \n",
       "238795           5.8084           6.7076           5.0113           6.8595   \n",
       "\n",
       "        TCGA-KM-8639-01  TCGA-KO-8415-11  ...  TCGA-KO-8408-01  \\\n",
       "0                2.9139           3.1049  ...           2.2369   \n",
       "1                0.0000           0.0000  ...           0.0000   \n",
       "2                0.7535           0.8673  ...           0.1602   \n",
       "3                1.2131           1.7214  ...           0.6381   \n",
       "4                0.0374           2.4087  ...           0.0000   \n",
       "...                 ...              ...  ...              ...   \n",
       "238791           4.4484           5.5820  ...           5.0518   \n",
       "238792           2.4188           1.8648  ...           0.8498   \n",
       "238793           1.1452           1.9724  ...           1.3415   \n",
       "238794           4.2824           3.6833  ...           3.0631   \n",
       "238795           6.8408           6.0389  ...           6.0402   \n",
       "\n",
       "        TCGA-KM-8443-01  TCGA-KM-8442-01  TCGA-KL-8332-11  TCGA-KL-8327-01  \\\n",
       "0                2.4951           2.1010           3.0617           0.7424   \n",
       "1                0.0000           0.0000           0.0000           0.0000   \n",
       "2                0.3507           0.7154           0.4353           0.0966   \n",
       "3                1.5227           1.7786           1.4157           0.0000   \n",
       "4                0.0189           0.1334           4.8648           0.0900   \n",
       "...                 ...              ...              ...              ...   \n",
       "238791           4.9067           3.8206           7.3459           4.1126   \n",
       "238792           1.0774           0.8774           1.2410           0.2725   \n",
       "238793           1.8842           0.8646           0.0000           1.9102   \n",
       "238794           1.8938           2.9015           2.3518           1.1468   \n",
       "238795           6.0639           6.8604           4.9554           3.4175   \n",
       "\n",
       "        chromosome_name  chrom_exon_start  chrom_exon_end  \\\n",
       "0                     3          51973965        51974630   \n",
       "1                     1         215728030       215728384   \n",
       "2                    17          51191601        51193016   \n",
       "3                     4         146292929       146292980   \n",
       "4                     1          15747546        15747980   \n",
       "...                 ...               ...             ...   \n",
       "238791                3          51978259        51978374   \n",
       "238792                6          43357126        43357765   \n",
       "238793               17          36168630        36168695   \n",
       "238794                1         120833602       120833774   \n",
       "238795               18          12353005        12353158   \n",
       "\n",
       "                                id  strand  \n",
       "0         chr3:52007981-52008646:-      -1  \n",
       "1       chr1:215901372-215901726:-      -1  \n",
       "2        chr17:49268962-49270377:-      -1  \n",
       "3       chr4:147214081-147214132:-      -1  \n",
       "4         chr1:16074041-16074475:+       1  \n",
       "...                            ...     ...  \n",
       "238791    chr3:52012275-52012390:+       1  \n",
       "238792    chr6:43324864-43325503:-      -1  \n",
       "238793   chr17:34495988-34496053:-      -1  \n",
       "238794  chr1:148010884-148011056:-       1  \n",
       "238795   chr18:12353004-12353157:-      -1  \n",
       "\n",
       "[238796 rows x 96 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map miRNA to counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:22:01.482945Z",
     "start_time": "2020-10-25T14:22:01.013851Z"
    }
   },
   "outputs": [],
   "source": [
    "#get subset counts of bs and merge with df_gene\n",
    "counts = bs[['bs_id', 'miRNA', 'mRNA', 'chrom_bs_start', 'chrom_bs_end']].merge(df_gene[['ensembl_transcript_id','chromosome_name','strand']], left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "\n",
    "#mapping mature ID to miRNA family\n",
    "#trans = pd.read_feather(path/'mature2families.feather')\n",
    "#counts = counts.merge(trans, left_on='miRNA',right_on='mature_name', how='left')\n",
    "\n",
    "mirna_counts\n",
    "trans = pd.read_feather(path/'mature2families.feather')\n",
    "mirna_counts = mirna_counts.merge(trans, left_on='miRNA',right_on='mature_name', how='left')\n",
    "\n",
    "#alternative 0: dont join with mirnaids\n",
    "#counts.drop(['ensembl_transcript_id', 'mature_name'], axis=1, inplace=True)\n",
    "\n",
    "#! alternative 1: for all caseids at the same time !\n",
    "counts = counts.merge(mirna_counts, left_on='mature_acc', right_on='sample', how='inner')\n",
    "counts.drop(['ensembl_transcript_id', 'mature_name', 'sample'], axis=1, inplace=True)\n",
    "\n",
    "# alternative 2: for 1 caseids at a time\n",
    "for caseid in []:#[caseids[0]]:\n",
    "    counts = counts.merge(mirna_counts[['sample',caseid]], left_on='mature_acc', right_on='sample', how='left')\n",
    "    counts.drop(['ensembl_transcript_id', 'mature_name', 'sample'], axis=1, inplace=True)\n",
    "\n",
    "#counts[['chromosome_name','strand','miRNA','mRNA']] = counts[['chromosome_name','strand','miRNA','mRNA']].astype('str')#.astype('category')\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9595</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000469694</td>\n",
       "      <td>240462246</td>\n",
       "      <td>240462269</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19235</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000670658</td>\n",
       "      <td>195454546</td>\n",
       "      <td>195454591</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23207</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178063</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52767</th>\n",
       "      <td>61977</td>\n",
       "      <td>hsa-miR-542-3p</td>\n",
       "      <td>ENST00000414717</td>\n",
       "      <td>11446564</td>\n",
       "      <td>11446583</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0003389</td>\n",
       "      <td>mir-542</td>\n",
       "      <td>9.029325</td>\n",
       "      <td>...</td>\n",
       "      <td>7.874904</td>\n",
       "      <td>8.344636</td>\n",
       "      <td>7.057697</td>\n",
       "      <td>8.625030</td>\n",
       "      <td>7.208745</td>\n",
       "      <td>7.794783</td>\n",
       "      <td>8.570388</td>\n",
       "      <td>8.508410</td>\n",
       "      <td>9.049981</td>\n",
       "      <td>8.300311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52768</th>\n",
       "      <td>65657</td>\n",
       "      <td>hsa-miR-542-3p</td>\n",
       "      <td>ENST00000570409</td>\n",
       "      <td>3651006</td>\n",
       "      <td>3651031</td>\n",
       "      <td>16</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0003389</td>\n",
       "      <td>mir-542</td>\n",
       "      <td>9.029325</td>\n",
       "      <td>...</td>\n",
       "      <td>7.874904</td>\n",
       "      <td>8.344636</td>\n",
       "      <td>7.057697</td>\n",
       "      <td>8.625030</td>\n",
       "      <td>7.208745</td>\n",
       "      <td>7.794783</td>\n",
       "      <td>8.570388</td>\n",
       "      <td>8.508410</td>\n",
       "      <td>9.049981</td>\n",
       "      <td>8.300311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52769</th>\n",
       "      <td>62648</td>\n",
       "      <td>hsa-miR-3611</td>\n",
       "      <td>ENST00000414717</td>\n",
       "      <td>11426846</td>\n",
       "      <td>11426866</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0017988</td>\n",
       "      <td>hsa-miR-3611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.170100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.180300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52770</th>\n",
       "      <td>62772</td>\n",
       "      <td>hsa-miR-3910</td>\n",
       "      <td>ENST00000414717</td>\n",
       "      <td>11364682</td>\n",
       "      <td>11364729</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0018184</td>\n",
       "      <td>mir-3910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52771</th>\n",
       "      <td>69960</td>\n",
       "      <td>hsa-miR-208b-3p</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38474539</td>\n",
       "      <td>38474563</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0004960</td>\n",
       "      <td>mir-208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>52772 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end  \\\n",
       "0          0  hsa-let-7a-2-3p  ENST00000576537         1578440       1578466   \n",
       "1          1  hsa-let-7a-2-3p  ENST00000576537         1578561       1578581   \n",
       "2       9595  hsa-let-7a-2-3p  ENST00000469694       240462246     240462269   \n",
       "3      19235  hsa-let-7a-2-3p  ENST00000670658       195454546     195454591   \n",
       "4      23207  hsa-let-7a-2-3p  ENST00000474430       155178044     155178063   \n",
       "...      ...              ...              ...             ...           ...   \n",
       "52767  61977   hsa-miR-542-3p  ENST00000414717        11446564      11446583   \n",
       "52768  65657   hsa-miR-542-3p  ENST00000570409         3651006       3651031   \n",
       "52769  62648     hsa-miR-3611  ENST00000414717        11426846      11426866   \n",
       "52770  62772     hsa-miR-3910  ENST00000414717        11364682      11364729   \n",
       "52771  69960  hsa-miR-208b-3p  ENST00000216014        38474539      38474563   \n",
       "\n",
       "      chromosome_name  strand    mature_acc  mirna_family  TCGA-KN-8426-01  \\\n",
       "0                  17      -1  MIMAT0010195         let-7         2.893871   \n",
       "1                  17      -1  MIMAT0010195         let-7         2.893871   \n",
       "2                   2       1  MIMAT0010195         let-7         2.893871   \n",
       "3                   2       1  MIMAT0010195         let-7         2.893871   \n",
       "4                   1       1  MIMAT0010195         let-7         2.893871   \n",
       "...               ...     ...           ...           ...              ...   \n",
       "52767               3       1  MIMAT0003389       mir-542         9.029325   \n",
       "52768              16      -1  MIMAT0003389       mir-542         9.029325   \n",
       "52769               3       1  MIMAT0017988  hsa-miR-3611              NaN   \n",
       "52770               3       1  MIMAT0018184      mir-3910              NaN   \n",
       "52771              22       1  MIMAT0004960       mir-208              NaN   \n",
       "\n",
       "       ...  TCGA-KO-8408-01  TCGA-KM-8438-01  TCGA-KL-8323-01  \\\n",
       "0      ...         2.829471         2.015161         2.594806   \n",
       "1      ...         2.829471         2.015161         2.594806   \n",
       "2      ...         2.829471         2.015161         2.594806   \n",
       "3      ...         2.829471         2.015161         2.594806   \n",
       "4      ...         2.829471         2.015161         2.594806   \n",
       "...    ...              ...              ...              ...   \n",
       "52767  ...         7.874904         8.344636         7.057697   \n",
       "52768  ...         7.874904         8.344636         7.057697   \n",
       "52769  ...              NaN              NaN              NaN   \n",
       "52770  ...              NaN              NaN              NaN   \n",
       "52771  ...              NaN              NaN              NaN   \n",
       "\n",
       "       TCGA-KN-8429-11  TCGA-KO-8413-01  TCGA-KO-8411-01  TCGA-KL-8332-11  \\\n",
       "0             2.644861         1.786991         1.854876         2.824365   \n",
       "1             2.644861         1.786991         1.854876         2.824365   \n",
       "2             2.644861         1.786991         1.854876         2.824365   \n",
       "3             2.644861         1.786991         1.854876         2.824365   \n",
       "4             2.644861         1.786991         1.854876         2.824365   \n",
       "...                ...              ...              ...              ...   \n",
       "52767         8.625030         7.208745         7.794783         8.570388   \n",
       "52768         8.625030         7.208745         7.794783         8.570388   \n",
       "52769         0.170100              NaN              NaN              NaN   \n",
       "52770              NaN              NaN              NaN              NaN   \n",
       "52771              NaN              NaN              NaN              NaN   \n",
       "\n",
       "       TCGA-KL-8335-01  TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "0             2.395032         2.791374         2.819371  \n",
       "1             2.395032         2.791374         2.819371  \n",
       "2             2.395032         2.791374         2.819371  \n",
       "3             2.395032         2.791374         2.819371  \n",
       "4             2.395032         2.791374         2.819371  \n",
       "...                ...              ...              ...  \n",
       "52767         8.508410         9.049981         8.300311  \n",
       "52768         8.508410         9.049981         8.300311  \n",
       "52769         0.180300              NaN              NaN  \n",
       "52770              NaN              NaN              NaN  \n",
       "52771              NaN              NaN              NaN  \n",
       "\n",
       "[52772 rows x 98 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T14:29:43.996902Z",
     "start_time": "2020-10-23T14:29:43.876970Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1917"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#TODO put in presentation, analysis mirna, how many in xena data\n",
    "#TODO find out : if na should i use it as 0????\n",
    "counts['TCGA-KN-8426-01'].isna().sum() #for caseid TCGA-KN-8426-01: 56.035 bs are thrown away \n",
    "len(list(counts[counts['TCGA-KN-8426-01'].isna()].mature_acc.unique())) #we don't have the counts for 1948 mirnas in bs\n",
    "len(list(counts[~ counts['TCGA-KN-8426-01'].isna()].mature_acc.unique())) #we have counts for 658 mirnas\n",
    "\n",
    "len(list(mirna_counts['sample'].unique())) #1917 = absolute nr mirnas from TCGA-KN-8426-01\n",
    "len(list(mirna_counts[~ mirna_counts['TCGA-KN-8426-01'].isna()]['sample'].unique())) #681 = not na nr mirnas from TCGA-KN-8426-01"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map exons to counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:23:02.738089Z",
     "start_time": "2020-10-25T14:22:50.007963Z"
    }
   },
   "outputs": [],
   "source": [
    "#for a large data set, you will likely get a significant speed increase by creating an index for any column(s) used in the join condition. \n",
    "#https://stackoverflow.com/questions/30627968/merge-pandas-dataframes-where-one-value-is-between-two-others\n",
    "#instead use SQL\n",
    "conn = sqlite3.connect('data/db.db')#':memory:') #Make the database in memory\n",
    "c = conn.cursor()\n",
    "\n",
    "#write the tables\n",
    "counts.to_sql('counts', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "exon_counts.to_sql('exon_counts', conn, index=True, index_label='exon_id', dtype={caseid : 'INTEGER' for caseid in caseids}.update({\"chrom_exon_start\": 'INTEGER', \"chrom_exon_end\": 'INTEGER'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T14:54:28.603611Z",
     "start_time": "2020-10-23T14:54:21.994290Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238791</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238792</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238793</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238794</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238795</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>238796 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name\n",
       "0                    3\n",
       "1                    1\n",
       "2                   17\n",
       "3                    4\n",
       "4                    1\n",
       "...                ...\n",
       "238791               3\n",
       "238792               6\n",
       "238793              17\n",
       "238794               1\n",
       "238795              18\n",
       "\n",
       "[238796 rows x 1 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print sqlite3 table \n",
    "qry = '''\n",
    "    select exon_counts.chromosome_name from exon_counts\n",
    "    '''\n",
    "#c.execute(qry)\n",
    "#conn.commit()\n",
    "#c.fetchall()\n",
    "pd.read_sql_query(qry,conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:22:28.922131Z",
     "start_time": "2020-10-25T14:22:28.509946Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#DELETE table from sqlite3 DB\n",
    "c.execute(\"DROP TABLE counts;\")\n",
    "conn.commit()\n",
    "c.execute(\"DROP TABLE exon_counts;\")\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:10:08.661651Z",
     "start_time": "2020-10-25T14:09:05.000Z"
    }
   },
   "outputs": [],
   "source": [
    "#TODO change column name case_id to column name case_id_mirna so it doesnt overlap with exon\n",
    "#instead of joining like i do & repeating the exon info -> only add exon_id + pivot / addup per exon + add exon counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2020-10-25T14:13:28.400Z"
    }
   },
   "outputs": [],
   "source": [
    "qry = '''\n",
    "    CREATE TABLE n_counts AS\n",
    "    select\n",
    "        *\n",
    "    from\n",
    "        ( \n",
    "        select exon_counts.chromosome_name, exon_counts.chrom_exon_start, exon_counts.chrom_exon_end, exon_counts.strand, exon_counts.id\n",
    "        from exon_counts \n",
    "        ) as temp\n",
    "        inner join\n",
    "        counts\n",
    "        on\n",
    "        counts.chromosome_name = temp.chromosome_name and counts.strand = temp.strand and counts.chrom_bs_start >= temp.chrom_exon_start and counts.chrom_bs_end <= temp.chrom_exon_end\n",
    "    '''\n",
    "c.execute(qry)\n",
    "conn.commit()\n",
    "#TODO this finds the exon id for every bs \n",
    "#52772 -> 18736 bindingsites\n",
    "#only returns bs with minimum 1 exon_id cause inner join (else left join)\n",
    "#automatically filters out all binding sites that have more than 1 exon_id\n",
    "#if u dont want that use outer join -> actually wrong, it uses the first instead of leaving it out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "qry = '''\n",
    "    select * from n_counts\n",
    "    '''\n",
    "n_counts = pd.read_sql_query(qry,conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T12:39:09.497431Z",
     "start_time": "2020-10-25T12:39:09.067541Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>id</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23204</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178054</td>\n",
       "      <td>155178091</td>\n",
       "      <td>...</td>\n",
       "      <td>14.502304</td>\n",
       "      <td>14.302077</td>\n",
       "      <td>15.348145</td>\n",
       "      <td>14.600445</td>\n",
       "      <td>14.474857</td>\n",
       "      <td>14.568523</td>\n",
       "      <td>14.486888</td>\n",
       "      <td>15.229448</td>\n",
       "      <td>14.887372</td>\n",
       "      <td>14.794259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23207</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178063</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23219</td>\n",
       "      <td>hsa-let-7c-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178063</td>\n",
       "      <td>...</td>\n",
       "      <td>2.714853</td>\n",
       "      <td>2.455017</td>\n",
       "      <td>2.155998</td>\n",
       "      <td>3.673442</td>\n",
       "      <td>2.676288</td>\n",
       "      <td>3.043974</td>\n",
       "      <td>4.407776</td>\n",
       "      <td>2.500585</td>\n",
       "      <td>3.844001</td>\n",
       "      <td>2.977714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23229</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178066</td>\n",
       "      <td>...</td>\n",
       "      <td>2.523778</td>\n",
       "      <td>3.836892</td>\n",
       "      <td>2.846723</td>\n",
       "      <td>4.056371</td>\n",
       "      <td>2.882109</td>\n",
       "      <td>3.414638</td>\n",
       "      <td>3.685095</td>\n",
       "      <td>2.885547</td>\n",
       "      <td>3.698584</td>\n",
       "      <td>3.274872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23230</td>\n",
       "      <td>hsa-let-7f-5p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178054</td>\n",
       "      <td>155178091</td>\n",
       "      <td>...</td>\n",
       "      <td>12.768630</td>\n",
       "      <td>13.254969</td>\n",
       "      <td>13.261290</td>\n",
       "      <td>12.433737</td>\n",
       "      <td>12.189768</td>\n",
       "      <td>12.644354</td>\n",
       "      <td>12.628264</td>\n",
       "      <td>13.546305</td>\n",
       "      <td>12.972790</td>\n",
       "      <td>12.892615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18731</th>\n",
       "      <td>17</td>\n",
       "      <td>1576597</td>\n",
       "      <td>1576720</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17:1479891-1480014:-</td>\n",
       "      <td>3197</td>\n",
       "      <td>hsa-miR-4433b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576618</td>\n",
       "      <td>1576639</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.214300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18732</th>\n",
       "      <td>17</td>\n",
       "      <td>1576597</td>\n",
       "      <td>1576720</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17:1479891-1480014:-</td>\n",
       "      <td>3201</td>\n",
       "      <td>hsa-miR-1273h-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576685</td>\n",
       "      <td>1576706</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.194100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18733</th>\n",
       "      <td>17</td>\n",
       "      <td>1576597</td>\n",
       "      <td>1576720</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17:1479891-1480014:-</td>\n",
       "      <td>3217</td>\n",
       "      <td>hsa-miR-7850-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576613</td>\n",
       "      <td>1576660</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0.170100</td>\n",
       "      <td>0.316200</td>\n",
       "      <td>0.264600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.408745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18734</th>\n",
       "      <td>17</td>\n",
       "      <td>1576597</td>\n",
       "      <td>1576720</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17:1479891-1480014:-</td>\n",
       "      <td>3230</td>\n",
       "      <td>hsa-miR-7976</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576619</td>\n",
       "      <td>1576638</td>\n",
       "      <td>...</td>\n",
       "      <td>0.346500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.194100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.180300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.218800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18735</th>\n",
       "      <td>17</td>\n",
       "      <td>1576597</td>\n",
       "      <td>1576720</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr17:1479891-1480014:-</td>\n",
       "      <td>3271</td>\n",
       "      <td>hsa-miR-8079</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576697</td>\n",
       "      <td>1576720</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18736 rows × 103 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name  chrom_exon_start  chrom_exon_end strand  \\\n",
       "0                   1         155178002       155178255      1   \n",
       "1                   1         155178002       155178255      1   \n",
       "2                   1         155178002       155178255      1   \n",
       "3                   1         155178002       155178255      1   \n",
       "4                   1         155178002       155178255      1   \n",
       "...               ...               ...             ...    ...   \n",
       "18731              17           1576597         1576720     -1   \n",
       "18732              17           1576597         1576720     -1   \n",
       "18733              17           1576597         1576720     -1   \n",
       "18734              17           1576597         1576720     -1   \n",
       "18735              17           1576597         1576720     -1   \n",
       "\n",
       "                               id  bs_id             miRNA             mRNA  \\\n",
       "0      chr1:155150478-155150731:+  23204     hsa-let-7a-5p  ENST00000474430   \n",
       "1      chr1:155150478-155150731:+  23207   hsa-let-7a-2-3p  ENST00000474430   \n",
       "2      chr1:155150478-155150731:+  23219     hsa-let-7c-3p  ENST00000474430   \n",
       "3      chr1:155150478-155150731:+  23229     hsa-let-7e-3p  ENST00000474430   \n",
       "4      chr1:155150478-155150731:+  23230     hsa-let-7f-5p  ENST00000474430   \n",
       "...                           ...    ...               ...              ...   \n",
       "18731     chr17:1479891-1480014:-   3197  hsa-miR-4433b-5p  ENST00000576537   \n",
       "18732     chr17:1479891-1480014:-   3201  hsa-miR-1273h-3p  ENST00000576537   \n",
       "18733     chr17:1479891-1480014:-   3217   hsa-miR-7850-5p  ENST00000576537   \n",
       "18734     chr17:1479891-1480014:-   3230      hsa-miR-7976  ENST00000576537   \n",
       "18735     chr17:1479891-1480014:-   3271      hsa-miR-8079  ENST00000576537   \n",
       "\n",
       "       chrom_bs_start  chrom_bs_end  ... TCGA-KO-8408-01  TCGA-KM-8438-01  \\\n",
       "0           155178054     155178091  ...       14.502304        14.302077   \n",
       "1           155178044     155178063  ...        2.829471         2.015161   \n",
       "2           155178044     155178063  ...        2.714853         2.455017   \n",
       "3           155178044     155178066  ...        2.523778         3.836892   \n",
       "4           155178054     155178091  ...       12.768630        13.254969   \n",
       "...               ...           ...  ...             ...              ...   \n",
       "18731         1576618       1576639  ...             NaN         0.214300   \n",
       "18732         1576685       1576706  ...             NaN              NaN   \n",
       "18733         1576613       1576660  ...             NaN              NaN   \n",
       "18734         1576619       1576638  ...        0.346500              NaN   \n",
       "18735         1576697       1576720  ...             NaN              NaN   \n",
       "\n",
       "      TCGA-KL-8323-01 TCGA-KN-8429-11  TCGA-KO-8413-01  TCGA-KO-8411-01  \\\n",
       "0           15.348145       14.600445        14.474857        14.568523   \n",
       "1            2.594806        2.644861         1.786991         1.854876   \n",
       "2            2.155998        3.673442         2.676288         3.043974   \n",
       "3            2.846723        4.056371         2.882109         3.414638   \n",
       "4           13.261290       12.433737        12.189768        12.644354   \n",
       "...               ...             ...              ...              ...   \n",
       "18731             NaN             NaN              NaN              NaN   \n",
       "18732        0.194100             NaN              NaN              NaN   \n",
       "18733        0.365156        0.170100         0.316200         0.264600   \n",
       "18734        0.194100             NaN              NaN              NaN   \n",
       "18735             NaN             NaN              NaN              NaN   \n",
       "\n",
       "       TCGA-KL-8332-11  TCGA-KL-8335-01  TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "0            14.486888        15.229448        14.887372        14.794259  \n",
       "1             2.824365         2.395032         2.791374         2.819371  \n",
       "2             4.407776         2.500585         3.844001         2.977714  \n",
       "3             3.685095         2.885547         3.698584         3.274872  \n",
       "4            12.628264        13.546305        12.972790        12.892615  \n",
       "...                ...              ...              ...              ...  \n",
       "18731              NaN              NaN              NaN              NaN  \n",
       "18732              NaN              NaN              NaN              NaN  \n",
       "18733              NaN              NaN              NaN         0.408745  \n",
       "18734              NaN         0.180300              NaN         0.218800  \n",
       "18735              NaN              NaN              NaN              NaN  \n",
       "\n",
       "[18736 rows x 103 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T00:31:04.522842Z",
     "start_time": "2020-10-13T00:31:04.432995Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9595</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000469694</td>\n",
       "      <td>240462246</td>\n",
       "      <td>240462269</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "      <td>2</td>\n",
       "      <td>240462191</td>\n",
       "      <td>240462582</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2:241401608-241401999:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23207</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178063</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32328</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000412954</td>\n",
       "      <td>133435201</td>\n",
       "      <td>133435231</td>\n",
       "      <td>6</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "      <td>6</td>\n",
       "      <td>133435078</td>\n",
       "      <td>133435570</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr6:133756216-133756708:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32329</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000412954</td>\n",
       "      <td>133435386</td>\n",
       "      <td>133435423</td>\n",
       "      <td>6</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "      <td>6</td>\n",
       "      <td>133435078</td>\n",
       "      <td>133435570</td>\n",
       "      <td>-1</td>\n",
       "      <td>chr6:133756216-133756708:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>39797</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000526202</td>\n",
       "      <td>75760745</td>\n",
       "      <td>75760767</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "      <td>13</td>\n",
       "      <td>75760444</td>\n",
       "      <td>75761038</td>\n",
       "      <td>1</td>\n",
       "      <td>chr13:76334580-76335174:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 103 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end  \\\n",
       "0   9595  hsa-let-7a-2-3p  ENST00000469694       240462246     240462269   \n",
       "1  23207  hsa-let-7a-2-3p  ENST00000474430       155178044     155178063   \n",
       "2  32328  hsa-let-7a-2-3p  ENST00000412954       133435201     133435231   \n",
       "3  32329  hsa-let-7a-2-3p  ENST00000412954       133435386     133435423   \n",
       "4  39797  hsa-let-7a-2-3p  ENST00000526202        75760745      75760767   \n",
       "\n",
       "   chromosome_name  strand    mature_acc mirna_family  TCGA-KN-8426-01  ...  \\\n",
       "0                2       1  MIMAT0010195        let-7         2.893871  ...   \n",
       "1                1       1  MIMAT0010195        let-7         2.893871  ...   \n",
       "2                6      -1  MIMAT0010195        let-7         2.893871  ...   \n",
       "3                6      -1  MIMAT0010195        let-7         2.893871  ...   \n",
       "4               13       1  MIMAT0010195        let-7         2.893871  ...   \n",
       "\n",
       "   TCGA-KO-8411-01  TCGA-KL-8332-11  TCGA-KL-8335-01  TCGA-KL-8339-11  \\\n",
       "0         1.854876         2.824365         2.395032         2.791374   \n",
       "1         1.854876         2.824365         2.395032         2.791374   \n",
       "2         1.854876         2.824365         2.395032         2.791374   \n",
       "3         1.854876         2.824365         2.395032         2.791374   \n",
       "4         1.854876         2.824365         2.395032         2.791374   \n",
       "\n",
       "   TCGA-KO-8415-01  chromosome_name  chrom_exon_start  chrom_exon_end  strand  \\\n",
       "0         2.819371                2         240462191       240462582       1   \n",
       "1         2.819371                1         155178002       155178255       1   \n",
       "2         2.819371                6         133435078       133435570      -1   \n",
       "3         2.819371                6         133435078       133435570      -1   \n",
       "4         2.819371               13          75760444        75761038       1   \n",
       "\n",
       "                           id  \n",
       "0  chr2:241401608-241401999:+  \n",
       "1  chr1:155150478-155150731:+  \n",
       "2  chr6:133756216-133756708:-  \n",
       "3  chr6:133756216-133756708:-  \n",
       "4   chr13:76334580-76335174:+  \n",
       "\n",
       "[5 rows x 103 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#visualize amount exons by bindingsite | amount bindingsites by exon\n",
    "#TODO maybe also include other cases eg 1 & 5\n",
    "qry = '''SELECT * FROM counts inner join (\n",
    "        select exon_counts.chromosome_name, exon_counts.chrom_exon_start, exon_counts.chrom_exon_end, exon_counts.strand, exon_counts.id\n",
    "        from exon_counts ) as temp\n",
    "        on\n",
    "        counts.chromosome_name = temp.chromosome_name and counts.strand = temp.strand and ((counts.chrom_bs_start >= temp.chrom_exon_start and counts.chrom_bs_start <= temp.chrom_exon_end) or (counts.chrom_bs_end >= temp.chrom_exon_start and counts.chrom_bs_end <= temp.chrom_exon_end))\n",
    "    '''\n",
    "bs_exons2 = pd.read_sql_query(qry, conn)\n",
    "bs_exons2\n",
    "#bs_exons2.groupby('bs_id').count().raw_counts.hist()\n",
    "#bs_exons2.groupby('exon_id').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1576707</td>\n",
       "      <td>1578261</td>\n",
       "      <td>1576597</td>\n",
       "      <td>1576720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1576707</td>\n",
       "      <td>1578261</td>\n",
       "      <td>1578250</td>\n",
       "      <td>1578323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>215124742</td>\n",
       "      <td>215169216</td>\n",
       "      <td>215169199</td>\n",
       "      <td>215169359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>215124742</td>\n",
       "      <td>215169216</td>\n",
       "      <td>215124633</td>\n",
       "      <td>215124750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>75800869</td>\n",
       "      <td>75804292</td>\n",
       "      <td>75800684</td>\n",
       "      <td>75800882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24862</th>\n",
       "      <td>75853378</td>\n",
       "      <td>75855273</td>\n",
       "      <td>75853092</td>\n",
       "      <td>75853388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24864</th>\n",
       "      <td>75804538</td>\n",
       "      <td>75817180</td>\n",
       "      <td>75804289</td>\n",
       "      <td>75804541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24865</th>\n",
       "      <td>75804538</td>\n",
       "      <td>75817180</td>\n",
       "      <td>75817161</td>\n",
       "      <td>75817278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24874</th>\n",
       "      <td>75761038</td>\n",
       "      <td>75795419</td>\n",
       "      <td>75760444</td>\n",
       "      <td>75761038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24875</th>\n",
       "      <td>75761038</td>\n",
       "      <td>75795419</td>\n",
       "      <td>75795401</td>\n",
       "      <td>75795431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6148 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom_bs_start  chrom_bs_end  chrom_exon_start  chrom_exon_end\n",
       "12            1576707       1578261           1576597         1576720\n",
       "13            1576707       1578261           1578250         1578323\n",
       "15          215124742     215169216         215169199       215169359\n",
       "16          215124742     215169216         215124633       215124750\n",
       "39           75800869      75804292          75800684        75800882\n",
       "...               ...           ...               ...             ...\n",
       "24862        75853378      75855273          75853092        75853388\n",
       "24864        75804538      75817180          75804289        75804541\n",
       "24865        75804538      75817180          75817161        75817278\n",
       "24874        75761038      75795419          75760444        75761038\n",
       "24875        75761038      75795419          75795401        75795431\n",
       "\n",
       "[6148 rows x 4 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#bs_exons2[['chrom_bs_start','chrom_bs_end','chrom_exon_start','chrom_exon_end']][~((bs_exons2.chrom_exon_start<=bs_exons2.chrom_bs_start) & (bs_exons2.chrom_exon_end>=bs_exons2.chrom_bs_end))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pivot tables & put all case_id counts into one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T14:36:05.887125Z",
     "start_time": "2020-10-23T14:36:05.763198Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9595</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000469694</td>\n",
       "      <td>240462246</td>\n",
       "      <td>240462269</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19235</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000670658</td>\n",
       "      <td>195454546</td>\n",
       "      <td>195454591</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23207</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178063</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end  \\\n",
       "0      0  hsa-let-7a-2-3p  ENST00000576537         1578440       1578466   \n",
       "1      1  hsa-let-7a-2-3p  ENST00000576537         1578561       1578581   \n",
       "2   9595  hsa-let-7a-2-3p  ENST00000469694       240462246     240462269   \n",
       "3  19235  hsa-let-7a-2-3p  ENST00000670658       195454546     195454591   \n",
       "4  23207  hsa-let-7a-2-3p  ENST00000474430       155178044     155178063   \n",
       "\n",
       "  chromosome_name  strand    mature_acc mirna_family  TCGA-KN-8426-01  ...  \\\n",
       "0              17      -1  MIMAT0010195        let-7         2.893871  ...   \n",
       "1              17      -1  MIMAT0010195        let-7         2.893871  ...   \n",
       "2               2       1  MIMAT0010195        let-7         2.893871  ...   \n",
       "3               2       1  MIMAT0010195        let-7         2.893871  ...   \n",
       "4               1       1  MIMAT0010195        let-7         2.893871  ...   \n",
       "\n",
       "   TCGA-KO-8408-01  TCGA-KM-8438-01  TCGA-KL-8323-01  TCGA-KN-8429-11  \\\n",
       "0         2.829471         2.015161         2.594806         2.644861   \n",
       "1         2.829471         2.015161         2.594806         2.644861   \n",
       "2         2.829471         2.015161         2.594806         2.644861   \n",
       "3         2.829471         2.015161         2.594806         2.644861   \n",
       "4         2.829471         2.015161         2.594806         2.644861   \n",
       "\n",
       "   TCGA-KO-8413-01  TCGA-KO-8411-01  TCGA-KL-8332-11  TCGA-KL-8335-01  \\\n",
       "0         1.786991         1.854876         2.824365         2.395032   \n",
       "1         1.786991         1.854876         2.824365         2.395032   \n",
       "2         1.786991         1.854876         2.824365         2.395032   \n",
       "3         1.786991         1.854876         2.824365         2.395032   \n",
       "4         1.786991         1.854876         2.824365         2.395032   \n",
       "\n",
       "   TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "0         2.791374         2.819371  \n",
       "1         2.791374         2.819371  \n",
       "2         2.791374         2.819371  \n",
       "3         2.791374         2.819371  \n",
       "4         2.791374         2.819371  \n",
       "\n",
       "[5 rows x 98 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T15:24:42.060464Z",
     "start_time": "2020-10-23T15:24:41.970516Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>TCGA-KL-8346-01</th>\n",
       "      <th>TCGA-KN-8422-01</th>\n",
       "      <th>TCGA-KN-8431-11</th>\n",
       "      <th>TCGA-KN-8430-11</th>\n",
       "      <th>TCGA-KM-8440-01</th>\n",
       "      <th>TCGA-KO-8414-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KM-8639-01</th>\n",
       "      <th>TCGA-KO-8415-11</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8443-01</th>\n",
       "      <th>TCGA-KM-8442-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>id</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.1156</td>\n",
       "      <td>1.7396</td>\n",
       "      <td>2.0641</td>\n",
       "      <td>3.1457</td>\n",
       "      <td>3.0436</td>\n",
       "      <td>2.2209</td>\n",
       "      <td>1.8213</td>\n",
       "      <td>1.9092</td>\n",
       "      <td>2.9139</td>\n",
       "      <td>3.1049</td>\n",
       "      <td>...</td>\n",
       "      <td>2.2369</td>\n",
       "      <td>2.4951</td>\n",
       "      <td>2.1010</td>\n",
       "      <td>3.0617</td>\n",
       "      <td>0.7424</td>\n",
       "      <td>3</td>\n",
       "      <td>51973965</td>\n",
       "      <td>51974630</td>\n",
       "      <td>chr3:52007981-52008646:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>215728030</td>\n",
       "      <td>215728384</td>\n",
       "      <td>chr1:215901372-215901726:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.3884</td>\n",
       "      <td>0.2219</td>\n",
       "      <td>0.2050</td>\n",
       "      <td>0.9168</td>\n",
       "      <td>0.6069</td>\n",
       "      <td>0.4547</td>\n",
       "      <td>0.1571</td>\n",
       "      <td>0.3857</td>\n",
       "      <td>0.7535</td>\n",
       "      <td>0.8673</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1602</td>\n",
       "      <td>0.3507</td>\n",
       "      <td>0.7154</td>\n",
       "      <td>0.4353</td>\n",
       "      <td>0.0966</td>\n",
       "      <td>17</td>\n",
       "      <td>51191601</td>\n",
       "      <td>51193016</td>\n",
       "      <td>chr17:49268962-49270377:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.5290</td>\n",
       "      <td>0.7196</td>\n",
       "      <td>1.3094</td>\n",
       "      <td>2.3074</td>\n",
       "      <td>1.6894</td>\n",
       "      <td>1.5829</td>\n",
       "      <td>1.0835</td>\n",
       "      <td>1.9041</td>\n",
       "      <td>1.2131</td>\n",
       "      <td>1.7214</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6381</td>\n",
       "      <td>1.5227</td>\n",
       "      <td>1.7786</td>\n",
       "      <td>1.4157</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4</td>\n",
       "      <td>146292929</td>\n",
       "      <td>146292980</td>\n",
       "      <td>chr4:147214081-147214132:-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4721</td>\n",
       "      <td>0.1284</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>2.4087</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0189</td>\n",
       "      <td>0.1334</td>\n",
       "      <td>4.8648</td>\n",
       "      <td>0.0900</td>\n",
       "      <td>1</td>\n",
       "      <td>15747546</td>\n",
       "      <td>15747980</td>\n",
       "      <td>chr1:16074041-16074475:+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   TCGA-KN-8419-01  TCGA-KL-8346-01  TCGA-KN-8422-01  TCGA-KN-8431-11  \\\n",
       "0           2.1156           1.7396           2.0641           3.1457   \n",
       "1           0.0000           0.0000           0.0000           0.0251   \n",
       "2           0.3884           0.2219           0.2050           0.9168   \n",
       "3           2.5290           0.7196           1.3094           2.3074   \n",
       "4           0.0000           0.0000           0.0000           1.4721   \n",
       "\n",
       "   TCGA-KN-8430-11  TCGA-KM-8440-01  TCGA-KO-8414-01  TCGA-KL-8323-01  \\\n",
       "0           3.0436           2.2209           1.8213           1.9092   \n",
       "1           0.0233           0.0000           0.0000           0.0000   \n",
       "2           0.6069           0.4547           0.1571           0.3857   \n",
       "3           1.6894           1.5829           1.0835           1.9041   \n",
       "4           0.1284           0.0000           0.0000           0.0000   \n",
       "\n",
       "   TCGA-KM-8639-01  TCGA-KO-8415-11  ...  TCGA-KO-8408-01  TCGA-KM-8443-01  \\\n",
       "0           2.9139           3.1049  ...           2.2369           2.4951   \n",
       "1           0.0000           0.0000  ...           0.0000           0.0000   \n",
       "2           0.7535           0.8673  ...           0.1602           0.3507   \n",
       "3           1.2131           1.7214  ...           0.6381           1.5227   \n",
       "4           0.0374           2.4087  ...           0.0000           0.0189   \n",
       "\n",
       "   TCGA-KM-8442-01  TCGA-KL-8332-11  TCGA-KL-8327-01  chromosome_name  \\\n",
       "0           2.1010           3.0617           0.7424                3   \n",
       "1           0.0000           0.0000           0.0000                1   \n",
       "2           0.7154           0.4353           0.0966               17   \n",
       "3           1.7786           1.4157           0.0000                4   \n",
       "4           0.1334           4.8648           0.0900                1   \n",
       "\n",
       "   chrom_exon_start  chrom_exon_end                          id  strand  \n",
       "0          51973965        51974630    chr3:52007981-52008646:-      -1  \n",
       "1         215728030       215728384  chr1:215901372-215901726:-      -1  \n",
       "2          51191601        51193016   chr17:49268962-49270377:-      -1  \n",
       "3         146292929       146292980  chr4:147214081-147214132:-      -1  \n",
       "4          15747546        15747980    chr1:16074041-16074475:+       1  \n",
       "\n",
       "[5 rows x 96 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO idea 1: do seperate tables for all caseids + after pivot put all different caseids sqlite tables together into one\n",
    "#TODO idea 2: 3D\n",
    "#TODO debug does this do it right?\n",
    "#pivot counts (which only has bs & mirna) and afterwards merge with exon counts\n",
    "#INSERT INTO artists_backup \n",
    "#SELECT ArtistId, Name\n",
    "#FROM artists;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.071562Z",
     "start_time": "2020-10-18T22:50:38.700Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"10\" halign=\"left\">TCGA-KL-8323-01</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">strand:1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>hsa-miR-1229-3p</th>\n",
       "      <th>hsa-miR-1229-5p</th>\n",
       "      <th>hsa-miR-1234-3p</th>\n",
       "      <th>hsa-miR-1268b</th>\n",
       "      <th>hsa-miR-1269b</th>\n",
       "      <th>hsa-miR-1273h-3p</th>\n",
       "      <th>hsa-miR-1295b-3p</th>\n",
       "      <th>hsa-miR-1295b-5p</th>\n",
       "      <th>hsa-miR-1539</th>\n",
       "      <th>hsa-miR-1973</th>\n",
       "      <th>...</th>\n",
       "      <th>mir-934</th>\n",
       "      <th>mir-935</th>\n",
       "      <th>mir-937</th>\n",
       "      <th>mir-938</th>\n",
       "      <th>mir-939</th>\n",
       "      <th>mir-940</th>\n",
       "      <th>mir-942</th>\n",
       "      <th>mir-943</th>\n",
       "      <th>mir-95</th>\n",
       "      <th>mir-96</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr12:49760688-49761370:+</th>\n",
       "      <th>12</th>\n",
       "      <th>49366905</th>\n",
       "      <th>49367587</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr12:49765011-49765073:+</th>\n",
       "      <th>12</th>\n",
       "      <th>49371228</th>\n",
       "      <th>49371290</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr12:49835513-49835599:+</th>\n",
       "      <th>12</th>\n",
       "      <th>49441730</th>\n",
       "      <th>49441816</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr12:49854543-49854820:+</th>\n",
       "      <th>12</th>\n",
       "      <th>49460760</th>\n",
       "      <th>49461037</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5181</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr13:76334580-76335174:+</th>\n",
       "      <th>13</th>\n",
       "      <th>75760444</th>\n",
       "      <th>75761038</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:11406133-11406208:+</th>\n",
       "      <th>3</th>\n",
       "      <th>11364659</th>\n",
       "      <th>11364734</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:11421446-11421526:+</th>\n",
       "      <th>3</th>\n",
       "      <th>11379972</th>\n",
       "      <th>11380052</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:11468278-11468400:+</th>\n",
       "      <th>3</th>\n",
       "      <th>11426804</th>\n",
       "      <th>11426926</th>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr6:133756216-133756708:-</th>\n",
       "      <th>6</th>\n",
       "      <th>133435078</th>\n",
       "      <th>133435570</th>\n",
       "      <th>-1</th>\n",
       "      <td>0.1941</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5181</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1941</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr6:133760466-133760619:-</th>\n",
       "      <th>6</th>\n",
       "      <th>133439328</th>\n",
       "      <th>133439481</th>\n",
       "      <th>-1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 115971 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                  TCGA-KL-8323-01  \\\n",
       "mirna_family                                                                      hsa-miR-1229-3p   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                   \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1               0.0000   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1               0.0000   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1               0.0000   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1               0.0000   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1               0.0000   \n",
       "...                                                                                           ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1               0.0000   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1               0.0000   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1               0.0000   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1              0.1941   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1              0.0000   \n",
       "\n",
       "                                                                                                   \\\n",
       "mirna_family                                                                      hsa-miR-1229-5p   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                   \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1             0.000000   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1             0.365156   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1             0.000000   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1             0.365156   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1             0.000000   \n",
       "...                                                                                           ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1             0.365156   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1             0.000000   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1             0.000000   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1            0.365156   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1            0.000000   \n",
       "\n",
       "                                                                                                   \\\n",
       "mirna_family                                                                      hsa-miR-1234-3p   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                   \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1                    0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1                    0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1                    0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1                    0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1                    0   \n",
       "...                                                                                           ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1                    0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1                    0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1                    0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1                   0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1                   0   \n",
       "\n",
       "                                                                                                 \\\n",
       "mirna_family                                                                      hsa-miR-1268b   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                 \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1             0.0000   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1             0.0000   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1             0.0000   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1             0.5181   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1             0.0000   \n",
       "...                                                                                         ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1             0.0000   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1             0.0000   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1             0.0000   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1            0.5181   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1            0.0000   \n",
       "\n",
       "                                                                                                 \\\n",
       "mirna_family                                                                      hsa-miR-1269b   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                 \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1                  0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1                  0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1                  0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1                  0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1                  0   \n",
       "...                                                                                         ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1                  0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1                  0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1                  0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1                 0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1                 0   \n",
       "\n",
       "                                                                                                    \\\n",
       "mirna_family                                                                      hsa-miR-1273h-3p   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                    \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1                0.0000   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1                0.0000   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1                0.0000   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1                0.0000   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1                0.0000   \n",
       "...                                                                                            ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1                0.0000   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1                0.0000   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1                0.0000   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1               0.1941   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1               0.0000   \n",
       "\n",
       "                                                                                                    \\\n",
       "mirna_family                                                                      hsa-miR-1295b-3p   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                    \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1                     0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1                     0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1                     0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1                     0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1                     0   \n",
       "...                                                                                            ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1                     0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1                     0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1                     0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1                    0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1                    0   \n",
       "\n",
       "                                                                                                    \\\n",
       "mirna_family                                                                      hsa-miR-1295b-5p   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                    \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1                     0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1                     0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1                     0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1                     0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1                     0   \n",
       "...                                                                                            ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1                     0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1                     0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1                     0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1                    0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1                    0   \n",
       "\n",
       "                                                                                                \\\n",
       "mirna_family                                                                      hsa-miR-1539   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1                 0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1                 0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1                 0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1                 0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1                 0   \n",
       "...                                                                                        ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1                 0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1                 0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1                 0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1                0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1                0   \n",
       "\n",
       "                                                                                                \\\n",
       "mirna_family                                                                      hsa-miR-1973   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand                \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1                 0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1                 0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1                 0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1                 0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1                 0   \n",
       "...                                                                                        ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1                 0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1                 0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1                 0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1                0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1                0   \n",
       "\n",
       "                                                                                   ...  \\\n",
       "mirna_family                                                                       ...   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand  ...   \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1       ...   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1       ...   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1       ...   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1       ...   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1       ...   \n",
       "...                                                                                ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1       ...   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1       ...   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1       ...   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1      ...   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1      ...   \n",
       "\n",
       "                                                                                  strand:1  \\\n",
       "mirna_family                                                                       mir-934   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand            \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1             0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1             0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1             1   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1             0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1             0   \n",
       "...                                                                                    ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1             0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1             0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1             0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1           -1   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1            0   \n",
       "\n",
       "                                                                                           \\\n",
       "mirna_family                                                                      mir-935   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand           \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1            0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1            0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1            0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1            0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1            0   \n",
       "...                                                                                   ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1            0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1            0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1            0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1          -1   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1           0   \n",
       "\n",
       "                                                                                           \\\n",
       "mirna_family                                                                      mir-937   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand           \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1            0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1            1   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1            0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1            0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1            0   \n",
       "...                                                                                   ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1            0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1            1   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1            0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1          -2   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1          -1   \n",
       "\n",
       "                                                                                           \\\n",
       "mirna_family                                                                      mir-938   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand           \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1            0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1            0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1            0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1            0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1            0   \n",
       "...                                                                                   ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1            0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1            0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1            0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1           0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1           0   \n",
       "\n",
       "                                                                                           \\\n",
       "mirna_family                                                                      mir-939   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand           \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1            1   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1            0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1            0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1            1   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1            0   \n",
       "...                                                                                   ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1            1   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1            0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1            0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1          -3   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1           0   \n",
       "\n",
       "                                                                                           \\\n",
       "mirna_family                                                                      mir-940   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand           \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1            1   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1            0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1            0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1            0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1            0   \n",
       "...                                                                                   ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1            0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1            0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1            0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1          -3   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1           0   \n",
       "\n",
       "                                                                                           \\\n",
       "mirna_family                                                                      mir-942   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand           \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1            0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1            0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1            0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1            0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1            0   \n",
       "...                                                                                   ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1            0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1            0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1            0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1          -1   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1           0   \n",
       "\n",
       "                                                                                           \\\n",
       "mirna_family                                                                      mir-943   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand           \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1            1   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1            0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1            0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1            0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1            1   \n",
       "...                                                                                   ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1            0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1            0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1            0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1          -2   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1           0   \n",
       "\n",
       "                                                                                          \\\n",
       "mirna_family                                                                      mir-95   \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand          \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1           0   \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1           0   \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1           0   \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1           0   \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1           1   \n",
       "...                                                                                  ...   \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1           0   \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1           0   \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1           0   \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1          0   \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1          0   \n",
       "\n",
       "                                                                                          \n",
       "mirna_family                                                                      mir-96  \n",
       "id                         chromosome_name chrom_exon_start chrom_exon_end strand         \n",
       "chr12:49760688-49761370:+  12              49366905         49367587       1           0  \n",
       "chr12:49765011-49765073:+  12              49371228         49371290       1           0  \n",
       "chr12:49835513-49835599:+  12              49441730         49441816       1           0  \n",
       "chr12:49854543-49854820:+  12              49460760         49461037       1           0  \n",
       "chr13:76334580-76335174:+  13              75760444         75761038       1           0  \n",
       "...                                                                                  ...  \n",
       "chr3:11406133-11406208:+   3               11364659         11364734       1           0  \n",
       "chr3:11421446-11421526:+   3               11379972         11380052       1           0  \n",
       "chr3:11468278-11468400:+   3               11426804         11426926       1           0  \n",
       "chr6:133756216-133756708:- 6               133435078        133435570      -1         -1  \n",
       "chr6:133760466-133760619:- 6               133439328        133439481      -1          0  \n",
       "\n",
       "[63 rows x 115971 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#output: for each case_id for each bs_id: interesting miRNA expression + exon expression\n",
    "#unstack family counts from 2 columns to several columns (mirna_family=name,mirna_read_count=value)\n",
    "pivoted = n_counts.copy()\n",
    "#pivoted = pivoted.pivot_table(values='mirna_read_count', index=['exon_id', 'exon_raw_counts'], columns='mirna_family', aggfunc='sum', fill_value=0)\n",
    "pivoted = pivoted.pivot_table(index=['id','chromosome_name','chrom_exon_start','chrom_exon_end','strand'], columns='mirna_family', aggfunc='sum', fill_value=0)\n",
    "pivoted#.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>id</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23204</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178054</td>\n",
       "      <td>155178091</td>\n",
       "      <td>...</td>\n",
       "      <td>14.502304</td>\n",
       "      <td>14.302077</td>\n",
       "      <td>15.348145</td>\n",
       "      <td>14.600445</td>\n",
       "      <td>14.474857</td>\n",
       "      <td>14.568523</td>\n",
       "      <td>14.486888</td>\n",
       "      <td>15.229448</td>\n",
       "      <td>14.887372</td>\n",
       "      <td>14.794259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23207</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178063</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23219</td>\n",
       "      <td>hsa-let-7c-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178063</td>\n",
       "      <td>...</td>\n",
       "      <td>2.714853</td>\n",
       "      <td>2.455017</td>\n",
       "      <td>2.155998</td>\n",
       "      <td>3.673442</td>\n",
       "      <td>2.676288</td>\n",
       "      <td>3.043974</td>\n",
       "      <td>4.407776</td>\n",
       "      <td>2.500585</td>\n",
       "      <td>3.844001</td>\n",
       "      <td>2.977714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23229</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178044</td>\n",
       "      <td>155178066</td>\n",
       "      <td>...</td>\n",
       "      <td>2.523778</td>\n",
       "      <td>3.836892</td>\n",
       "      <td>2.846723</td>\n",
       "      <td>4.056371</td>\n",
       "      <td>2.882109</td>\n",
       "      <td>3.414638</td>\n",
       "      <td>3.685095</td>\n",
       "      <td>2.885547</td>\n",
       "      <td>3.698584</td>\n",
       "      <td>3.274872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>155178002</td>\n",
       "      <td>155178255</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>23230</td>\n",
       "      <td>hsa-let-7f-5p</td>\n",
       "      <td>ENST00000474430</td>\n",
       "      <td>155178054</td>\n",
       "      <td>155178091</td>\n",
       "      <td>...</td>\n",
       "      <td>12.768630</td>\n",
       "      <td>13.254969</td>\n",
       "      <td>13.261290</td>\n",
       "      <td>12.433737</td>\n",
       "      <td>12.189768</td>\n",
       "      <td>12.644354</td>\n",
       "      <td>12.628264</td>\n",
       "      <td>13.546305</td>\n",
       "      <td>12.972790</td>\n",
       "      <td>12.892615</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 103 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome_name  chrom_exon_start  chrom_exon_end strand  \\\n",
       "0               1         155178002       155178255      1   \n",
       "1               1         155178002       155178255      1   \n",
       "2               1         155178002       155178255      1   \n",
       "3               1         155178002       155178255      1   \n",
       "4               1         155178002       155178255      1   \n",
       "\n",
       "                           id  bs_id            miRNA             mRNA  \\\n",
       "0  chr1:155150478-155150731:+  23204    hsa-let-7a-5p  ENST00000474430   \n",
       "1  chr1:155150478-155150731:+  23207  hsa-let-7a-2-3p  ENST00000474430   \n",
       "2  chr1:155150478-155150731:+  23219    hsa-let-7c-3p  ENST00000474430   \n",
       "3  chr1:155150478-155150731:+  23229    hsa-let-7e-3p  ENST00000474430   \n",
       "4  chr1:155150478-155150731:+  23230    hsa-let-7f-5p  ENST00000474430   \n",
       "\n",
       "   chrom_bs_start  chrom_bs_end  ... TCGA-KO-8408-01  TCGA-KM-8438-01  \\\n",
       "0       155178054     155178091  ...       14.502304        14.302077   \n",
       "1       155178044     155178063  ...        2.829471         2.015161   \n",
       "2       155178044     155178063  ...        2.714853         2.455017   \n",
       "3       155178044     155178066  ...        2.523778         3.836892   \n",
       "4       155178054     155178091  ...       12.768630        13.254969   \n",
       "\n",
       "  TCGA-KL-8323-01 TCGA-KN-8429-11  TCGA-KO-8413-01  TCGA-KO-8411-01  \\\n",
       "0       15.348145       14.600445        14.474857        14.568523   \n",
       "1        2.594806        2.644861         1.786991         1.854876   \n",
       "2        2.155998        3.673442         2.676288         3.043974   \n",
       "3        2.846723        4.056371         2.882109         3.414638   \n",
       "4       13.261290       12.433737        12.189768        12.644354   \n",
       "\n",
       "   TCGA-KL-8332-11  TCGA-KL-8335-01  TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "0        14.486888        15.229448        14.887372        14.794259  \n",
       "1         2.824365         2.395032         2.791374         2.819371  \n",
       "2         4.407776         2.500585         3.844001         2.977714  \n",
       "3         3.685095         2.885547         3.698584         3.274872  \n",
       "4        12.628264        13.546305        12.972790        12.892615  \n",
       "\n",
       "[5 rows x 103 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_counts.head() #18736 rows × 103 columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>hsa-miR-1229-3p</th>\n",
       "      <th>hsa-miR-1229-5p</th>\n",
       "      <th>hsa-miR-1234-3p</th>\n",
       "      <th>hsa-miR-1268b</th>\n",
       "      <th>hsa-miR-1269b</th>\n",
       "      <th>hsa-miR-1273h-3p</th>\n",
       "      <th>hsa-miR-1295b-3p</th>\n",
       "      <th>hsa-miR-1295b-5p</th>\n",
       "      <th>...</th>\n",
       "      <th>mir-934</th>\n",
       "      <th>mir-935</th>\n",
       "      <th>mir-937</th>\n",
       "      <th>mir-938</th>\n",
       "      <th>mir-939</th>\n",
       "      <th>mir-940</th>\n",
       "      <th>mir-942</th>\n",
       "      <th>mir-943</th>\n",
       "      <th>mir-95</th>\n",
       "      <th>mir-96</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:155150478-155150731:+</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1941</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>4.243361</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6:133760466-133760619:-</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2:241404033-241404163:+</td>\n",
       "      <td>4.8354</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.049261</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22:38878501-38879444:+</td>\n",
       "      <td>2.6066</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5181</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1941</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.451873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17:1481544-1481617:-</td>\n",
       "      <td>6.8482</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr13:76334580-76335174:+</td>\n",
       "      <td>1.7193</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.005966</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1:179335539-179335741:+</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5181</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1941</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.451873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr12:49854543-49854820:+</td>\n",
       "      <td>2.7075</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5181</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.365156</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr13:76369537-76369567:+</td>\n",
       "      <td>4.6895</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr17:1479891-1480014:-</td>\n",
       "      <td>6.4498</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1941</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.049261</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.451873</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 1249 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            id  TCGA-KL-8323-01  hsa-miR-1229-3p  \\\n",
       "0   chr1:155150478-155150731:+           0.0000           0.1941   \n",
       "1   chr6:133760466-133760619:-           0.0000           0.0000   \n",
       "2   chr2:241404033-241404163:+           4.8354           0.0000   \n",
       "3    chr22:38878501-38879444:+           2.6066           0.0000   \n",
       "4      chr17:1481544-1481617:-           6.8482           0.0000   \n",
       "..                         ...              ...              ...   \n",
       "58   chr13:76334580-76335174:+           1.7193           0.0000   \n",
       "59  chr1:179335539-179335741:+           0.0000           0.0000   \n",
       "60   chr12:49854543-49854820:+           2.7075           0.0000   \n",
       "61   chr13:76369537-76369567:+           4.6895           0.0000   \n",
       "62     chr17:1479891-1480014:-           6.4498           0.0000   \n",
       "\n",
       "    hsa-miR-1229-5p  hsa-miR-1234-3p  hsa-miR-1268b  hsa-miR-1269b  \\\n",
       "0          0.000000                0         0.0000              0   \n",
       "1          0.000000                0         0.0000              0   \n",
       "2          0.000000                0         0.0000              0   \n",
       "3          0.365156                0         0.5181              0   \n",
       "4          0.000000                0         0.0000              0   \n",
       "..              ...              ...            ...            ...   \n",
       "58         0.000000                0         0.0000              0   \n",
       "59         0.365156                0         0.5181              0   \n",
       "60         0.365156                0         0.5181              0   \n",
       "61         0.000000                0         0.0000              0   \n",
       "62         0.000000                0         0.0000              0   \n",
       "\n",
       "    hsa-miR-1273h-3p  hsa-miR-1295b-3p  hsa-miR-1295b-5p  ...  mir-934  \\\n",
       "0             0.0000                 0                 0  ...   0.0000   \n",
       "1             0.0000                 0                 0  ...   0.0000   \n",
       "2             0.0000                 0                 0  ...   0.0000   \n",
       "3             0.0000                 0                 0  ...   0.1941   \n",
       "4             0.0000                 0                 0  ...   0.0000   \n",
       "..               ...               ...               ...  ...      ...   \n",
       "58            0.0000                 0                 0  ...   0.0000   \n",
       "59            0.0000                 0                 0  ...   0.1941   \n",
       "60            0.0000                 0                 0  ...   0.0000   \n",
       "61            0.0000                 0                 0  ...   0.0000   \n",
       "62            0.1941                 0                 0  ...   0.0000   \n",
       "\n",
       "    mir-935  mir-937  mir-938   mir-939   mir-940   mir-942  mir-943  \\\n",
       "0         0        0        0  0.000000  0.365156  4.243361        0   \n",
       "1         0        0        0  0.000000  0.000000  0.000000        0   \n",
       "2         0        0        0  0.000000  0.000000  4.049261        0   \n",
       "3         0        0        0  0.365156  0.000000  0.000000        0   \n",
       "4         0        0        0  0.000000  0.000000  0.000000        0   \n",
       "..      ...      ...      ...       ...       ...       ...      ...   \n",
       "58        0        0        0  0.000000  0.000000  0.000000        0   \n",
       "59        0        0        0  0.000000  0.000000  0.000000        0   \n",
       "60        0        0        0  0.365156  0.000000  0.000000        0   \n",
       "61        0        0        0  0.000000  0.000000  0.000000        0   \n",
       "62        0        0        0  0.000000  0.000000  4.049261        0   \n",
       "\n",
       "      mir-95    mir-96  \n",
       "0   0.000000  0.000000  \n",
       "1   0.000000  0.000000  \n",
       "2   0.000000  0.000000  \n",
       "3   0.000000  4.451873  \n",
       "4   0.000000  0.000000  \n",
       "..       ...       ...  \n",
       "58  1.005966  0.000000  \n",
       "59  0.000000  4.451873  \n",
       "60  0.000000  0.000000  \n",
       "61  0.000000  0.000000  \n",
       "62  0.000000  4.451873  \n",
       "\n",
       "[63 rows x 1249 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[51.60572298  6.61539042 18.14517162 ...  0.          0.\n",
      "   0.        ]\n",
      " [38.6407133   0.          0.         ...  3.44881894  0.\n",
      "   0.        ]\n",
      " [ 4.74323022  2.20186765  0.         ...  0.          0.\n",
      "   0.        ]\n",
      " ...\n",
      " [38.51844668  1.10582756 15.23751325 ...  3.44881894  0.\n",
      "   0.        ]\n",
      " [ 0.          0.          0.         ...  0.          0.\n",
      "   0.        ]\n",
      " [22.92590321  5.69809075 29.2371607  ...  0.          0.\n",
      "   0.        ]] ['let-7' 'mir-15' 'mir-17' ... 'mir-1270' 'mir-3118' 'mir-5701']\n"
     ]
    }
   ],
   "source": [
    "caseid = 'TCGA-KL-8323-01'\n",
    "case_data = exon_counts[['id',caseid]].merge(pivoted[caseid], how='inner', left_on='id', right_on='id')\n",
    "y = np.array(case_data[caseid])\n",
    "mirnas = trans.mirna_family.drop_duplicates().values # 1868 / 2656\n",
    "for mirna in mirnas:\n",
    "    if mirna not in case_data:\n",
    "        case_data[mirna] = 0\n",
    "X = case_data[mirnas].values\n",
    "#TODO check whether there are some mirnas in GDC data which we end up not using as I have no entry in family dictionary for them -> add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualize amount exons by bindingsite\n",
    "#bs_exons.groupby('bs_id').count().raw_counts.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:40:44.254071Z",
     "start_time": "2020-10-13T08:40:44.166104Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63, 1868) (63,)\n"
     ]
    }
   ],
   "source": [
    "#TODO check - is nr mirna downloaded also here? if not: mistake\n",
    "print(X.shape,y.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Elastic Net Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.081556Z",
     "start_time": "2020-10-18T22:50:43.000Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.getsizeof(pivoted)#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:42:05.340886Z",
     "start_time": "2020-10-13T08:42:05.198969Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-0.01084431 -0.00921378  0.01006737 ... -0.          0.\n",
      "  0.        ]\n",
      "3.436325606218843\n"
     ]
    }
   ],
   "source": [
    "#regression btw miRNA expression, exon expression\n",
    "import sklearn\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.datasets import make_regression\n",
    "#faster if np.array(x, order='F')\n",
    "#If you are interested in controlling the L1 and L2 penalty separately, keep in mind that this is equivalent to:\n",
    "#a * L1 + b * L2 where: alpha = a + b and l1_ratio = a / (a + b)\n",
    "\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "print(regr.coef_)\n",
    "print(regr.intercept_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "Seqvec basic Seq ligand binding.ipynb",
   "provenance": []
  },
  "hide_input": false,
  "kernelspec": {
   "display_name": "thesis",
   "language": "python",
   "name": "thesis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "581.6px",
    "left": "235px",
    "top": "158.533px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
