{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "center"
   },
   "source": [
    "# Assemble the final Dataframe from tarpmir bindingsites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in bindingsite & Ensembl data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-27T12:43:04.072027Z",
     "start_time": "2020-10-27T12:43:03.334014Z"
    },
    "cell_style": "center"
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from helper_fcts import *\n",
    "import sqlite3\n",
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "path = Path('data')\n",
    "tcga_path = Path(path/'KICH')\n",
    "path_tarp = Path(path/'tarp-bs')\n",
    "#ann_path = Path('C:/Users/Lena/Documents/Master big files/manual_GDC_download')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T21:08:25.447218Z",
     "start_time": "2020-10-25T21:08:19.206772Z"
    }
   },
   "outputs": [],
   "source": [
    "#read in df_gene: gene_id, g_id_v, transcript_id, t_id_v, chromosome, strand, start, end (1-N)\n",
    "df_gene = pd.read_csv(path/'gene_infos.csv', dtype={'ensembl_gene_id':str,'ensembl_gene_id_version':str,'ensembl_transcript_id':str, 'ensembl_transcript_id_version':str, 'chromosome_name':'category', 'strand':'int8', 'start_position':int, 'end_position':int})\n",
    "del df_gene['Unnamed: 0']\n",
    "df_gene[['strand']] = df_gene[['strand']].astype('int').astype('Int64')\n",
    "\n",
    "#executed once, read in from pickled files\n",
    "ids = read_in_ids(path/'all_mapping_ids.fasta') # mapping of transcript ids to exon ids\n",
    "chrom_exon_starts, exon_starts, exon_ends, df_exon = calc_exon_data(path) # exon starts and ends per transcript + exon id\n",
    "\n",
    "df_exon = df_exon.merge(df_gene, left_on = 'transcript_id', right_on = 'ensembl_transcript_id', how='left')\n",
    "df_exon[['chrom_exon_start','chrom_exon_end','start_position','end_position']] = df_exon[['chrom_exon_start','chrom_exon_end','start_position','end_position']].astype('Int64')\n",
    "df_exon[['strand']] = df_exon[['strand']].astype('str').astype('category')\n",
    "\n",
    "#len(df_exon[df_exon.ensembl_transcript_id.isna()].transcript_id.unique())\n",
    "#371 transcripts are not in df_gene, TODO, why? which transcripts are in tarpmir predictions\n",
    "#TODO instead of joining df_gene, exon_info -> directly read in from R feather file & also use that for calculating chrom pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:08:53.720453Z",
     "start_time": "2020-10-21T23:08:52.933037Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All predicted bindingsites were read into pandas.\n"
     ]
    }
   ],
   "source": [
    "#read in bs\n",
    "seperate_bs = []\n",
    "for filename in os.listdir(path_tarp):\n",
    "    binding_sites = parse_tarp_bs(path_tarp/filename)\n",
    "    seperate_bs.append(binding_sites)\n",
    "bs = pd.concat(seperate_bs, axis=0, ignore_index=True) #14 388 + 10 326 = 24 714 rows\n",
    "bs.to_feather(path/'bs.feather')\n",
    "print('All predicted bindingsites were read into pandas.')\n",
    "\n",
    "# alternative: set bs from feather\n",
    "#bs = pd.read_feather(path/'bs.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:07:50.384154Z",
     "start_time": "2020-10-21T23:07:44.980201Z"
    }
   },
   "outputs": [],
   "source": [
    "#parse miRNA and mRNA seed from sequence\n",
    "miRNA_sequences = parse_seq(path/'input_miRNA.fasta')#2661 miRNAs \n",
    "bs['miRNA_seed'] = bs.apply(lambda row: miRNA_sequences[row['miRNA']], axis=1)\n",
    "mRNA_sequences = parse_seq(path/'cdna6247.fasta') #TODO change to bigger file with all mRNAs or iterate over all files\n",
    "bs['mRNA_bs_seq'] = bs.apply(lambda row: mRNA_sequences[row['mRNA']][row['bs_start']:row['bs_end']], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:20:02.121012Z",
     "start_time": "2020-10-25T14:20:01.486989Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:20:20.279598Z",
     "start_time": "2020-10-25T14:20:20.189820Z"
    },
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def pos_to_chrom(pos, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if pos >= exon_starts[tid][eid] and pos <= exon_ends[tid][eid]:\n",
    "                diff = pos - exon_starts[tid][eid]\n",
    "                chrom_pos = chrom_exon_starts[tid][eid] + diff\n",
    "                return chrom_pos\n",
    "        else: print('Error: Either',tid,'or',eid,'not in exon_starts or exon_ends')\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:20:37.810692Z",
     "start_time": "2020-10-25T14:20:37.716745Z"
    }
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def get_eid(start, end, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if (start >= exon_starts[tid][eid]) and (start <= exon_ends[tid][eid]) and (end >= exon_starts[tid][eid]) and (end <= exon_ends[tid][eid]):\n",
    "                return eid\n",
    "        else: print('Error: Either',tid,'or',eid,'not in exon_starts or exon_ends')\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:21:08.552003Z",
     "start_time": "2020-10-25T14:20:55.441548Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>miRNA_seed</th>\n",
       "      <th>mRNA_bs_seq</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>bs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.005183</td>\n",
       "      <td>-0.035634</td>\n",
       "      <td>-11.206441</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>309</td>\n",
       "      <td>335</td>\n",
       "      <td>CTGTACAGCCTCCTAGCTTTCC</td>\n",
       "      <td></td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.743590</td>\n",
       "      <td>-21.6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.161796</td>\n",
       "      <td>0.055277</td>\n",
       "      <td>-3.793325</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>430</td>\n",
       "      <td>450</td>\n",
       "      <td>CTGTACAGCCTCCTAGCTTTCC</td>\n",
       "      <td></td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.441</td>\n",
       "      <td>3.547385</td>\n",
       "      <td>2.945605</td>\n",
       "      <td>-6.324962</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>130</td>\n",
       "      <td>TGAGGTAGTAGGTTGTGTGGTT</td>\n",
       "      <td></td>\n",
       "      <td>1576707</td>\n",
       "      <td>1578261</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.021616</td>\n",
       "      <td>-0.047693</td>\n",
       "      <td>-18.306612</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>315</td>\n",
       "      <td>336</td>\n",
       "      <td>CTATACAACCTACTGCCTTCCC</td>\n",
       "      <td></td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-15.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.456</td>\n",
       "      <td>3.773200</td>\n",
       "      <td>3.161275</td>\n",
       "      <td>-9.685015</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>41</td>\n",
       "      <td>CTATACAACCTACTGCCTTCCC</td>\n",
       "      <td>GGTGGCGTGGGCCTGTAA</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "      <td>ENSE00002671000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "0  hsa-let-7a-2-3p  ENST00000576537             1.000000   -25.9     0   \n",
       "1  hsa-let-7a-2-3p  ENST00000576537             0.743590   -21.6     1   \n",
       "2    hsa-let-7b-5p  ENST00000576537             0.846154   -18.4     1   \n",
       "3    hsa-let-7b-3p  ENST00000576537             0.615385   -20.6     0   \n",
       "4    hsa-let-7b-3p  ENST00000576537             0.538462   -15.3     0   \n",
       "\n",
       "   accessibility  AU_content  PhyloP_Stem  PyloP_Flanking       m/e   ...  \\\n",
       "0       0.000156       0.338     0.005183       -0.035634 -11.206441  ...   \n",
       "1       0.000002       0.338    -0.161796        0.055277  -3.793325  ...   \n",
       "2       0.000026       0.441     3.547385        2.945605  -6.324962  ...   \n",
       "3       0.000143       0.338    -0.021616       -0.047693 -18.306612  ...   \n",
       "4       0.000020       0.456     3.773200        3.161275  -9.685015  ...   \n",
       "\n",
       "   pairings_in_3prime_end  difference_of_pairings_between_seed_and_3prime_end  \\\n",
       "0                       7                                                  1    \n",
       "1                       3                                                  4    \n",
       "2                       8                                                  1    \n",
       "3                       7                                                  3    \n",
       "4                       5                                                  1    \n",
       "\n",
       "   bs_start  bs_end              miRNA_seed         mRNA_bs_seq  \\\n",
       "0       309     335  CTGTACAGCCTCCTAGCTTTCC                       \n",
       "1       430     450  CTGTACAGCCTCCTAGCTTTCC                       \n",
       "2       105     130  TGAGGTAGTAGGTTGTGTGGTT                       \n",
       "3       315     336  CTATACAACCTACTGCCTTCCC                       \n",
       "4        23      41  CTATACAACCTACTGCCTTCCC  GGTGGCGTGGGCCTGTAA   \n",
       "\n",
       "   chrom_bs_start  chrom_bs_end          exon_id bs_id  \n",
       "0         1578440       1578466  ENSE00002650258     0  \n",
       "1         1578561       1578581  ENSE00002650258     1  \n",
       "2         1576707       1578261             None     2  \n",
       "3         1578446       1578467  ENSE00002650258     3  \n",
       "4         1576625       1576643  ENSE00002671000     4  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#translate genome position relative to transcript to chromosome\n",
    "#fastest, always puts bs start and end even if bs spans several exons, but only puts exonid if only 1 exon\n",
    "bs['chrom_bs_start'] = bs.apply(lambda row: pos_to_chrom(row.bs_start, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['chrom_bs_end'] = bs.apply(lambda row: pos_to_chrom(row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['exon_id'] = bs.apply(lambda row: get_eid(row.bs_start, row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['bs_id'] = bs.index\n",
    "bs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in TCGA data (processed by Xena) from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-26T16:19:10.768960Z",
     "start_time": "2020-10-26T16:19:10.681011Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr17\tHAVANA\tgene\t1650629\t1684882\t.\t-\t.\tgene_id \"ENSG00000174231.15\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"PRPF8\"; level \"1\"; havana_gene \"OTTHUMG00000090553.5\";\n"
     ]
    }
   ],
   "source": [
    "#TODO option 1: read in exon_id annotation file (compare)\n",
    "#import gffutils\n",
    "#fn = gffutils.example_filename(ann_path/\"gencode.v22.annotation.gtf\")\n",
    "#exon_db = gffutils.create_db(fn, dbfn='exon.db', disable_infer_genes=True, disable_infer_transcripts=True, force=True)\n",
    "#gene = 'ENSG00000174231.15' #todo version wichtig!\n",
    "#for i in exon_db.children(gene, featuretype='exon'):#, order_by='start'):\n",
    " #   print(i)\n",
    "#for exon in exon_db.region(\"chr17:1578446-1578633\", strand=\"-\", featuretype='exon'):#, completely_within=True):\n",
    "    #print(exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-10-31 15:05:57--  https://tcga.xenahubs.net/download/unc_v2_exon_hg19_probe_TCGA\n",
      "Resolving tcga.xenahubs.net (tcga.xenahubs.net)... 54.237.213.246, 52.73.198.193\n",
      "Connecting to tcga.xenahubs.net (tcga.xenahubs.net)|54.237.213.246|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 12431049 (12M) [application/octet-stream]\n",
      "Saving to: ‘unc_v2_exon_hg19_probe_TCGA’\n",
      "\n",
      "unc_v2_exon_hg19_pr 100%[===================>]  11,85M   523KB/s    in 17s     \n",
      "\n",
      "2020-10-31 15:06:15 (715 KB/s) - ‘unc_v2_exon_hg19_probe_TCGA’ saved [12431049/12431049]\n",
      "\n",
      "--2020-10-31 15:06:15--  https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/HiSeqV2_exon.gz\n",
      "Resolving tcga.xenahubs.net (tcga.xenahubs.net)... 52.73.198.193, 54.237.213.246\n",
      "Connecting to tcga.xenahubs.net (tcga.xenahubs.net)|52.73.198.193|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 54181940 (52M) [application/gzip]\n",
      "Saving to: ‘HiSeqV2_exon.gz’\n",
      "\n",
      "HiSeqV2_exon.gz     100%[===================>]  51,67M   589KB/s    in 1m 44s  \n",
      "\n",
      "2020-10-31 15:07:59 (510 KB/s) - ‘HiSeqV2_exon.gz’ saved [54181940/54181940]\n",
      "\n",
      "--2020-10-31 15:08:00--  https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/miRNA_HiSeq_gene.gz\n",
      "Resolving tcga.xenahubs.net (tcga.xenahubs.net)... 52.73.198.193, 54.237.213.246\n",
      "Connecting to tcga.xenahubs.net (tcga.xenahubs.net)|52.73.198.193|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 385921 (377K) [application/gzip]\n",
      "Saving to: ‘miRNA_HiSeq_gene.gz’\n",
      "\n",
      "miRNA_HiSeq_gene.gz 100%[===================>] 376,88K   516KB/s    in 0,7s    \n",
      "\n",
      "2020-10-31 15:08:01 (516 KB/s) - ‘miRNA_HiSeq_gene.gz’ saved [385921/385921]\n",
      "\n",
      "downloaded\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-76250f035b5e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'wget https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/miRNA_HiSeq_gene.gz'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'downloaded'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mmirna_counts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtcga_path\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0;34m'miRNA_HiSeq_gene'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelimiter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'    '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0mexon_counts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtcga_path\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0;34m'HiSeqV2_exon'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelimiter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m' '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mtran_exon\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtcga_path\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0;34m'unc_v2_exon_hg19_probe_TCGA'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelimiter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'    '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "#download GDC ChGR37 data from web, once, TCGA Kidney Chromophobe (KICH)\n",
    "#https://xenabrowser.net/datapages/?hub=https://tcga.xenahubs.net:443\n",
    "'''\n",
    "#!wget https://tcga.xenahubs.net/download/unc_v2_exon_hg19_probe_TCGA #unnecessary cause some exons missing\n",
    "!wget https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/HiSeqV2_exon.gz\n",
    "!wget https://tcga.xenahubs.net/download/TCGA.KICH.sampleMap/miRNA_HiSeq_gene.gz\n",
    "!gunzip HiSeqV2_exon.gz\n",
    "!gunzip miRNA_HiSeq_gene.gz\n",
    "!mv HiSeqV2_exon data/KICH\n",
    "!mv miRNA_HiSeq_gene data/KICH\n",
    "!mv unc_v2_exon_hg19_probe_TCGA data/KICH'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirna_counts = pd.read_csv(tcga_path/'miRNA_HiSeq_gene', delimiter='\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-27T13:09:50.676110Z",
     "start_time": "2020-10-27T12:43:13.778366Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "translate exon count positions from anew\n",
      "239322\n",
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      "207695\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'mirna_counts' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-866729b5a1c6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[0;31m#exon_counts2 = exon_counts.merge(df_exon[['exon_id','chromosome_name','strand','chrom_exon_start','chrom_exon_end']], left_on = ['chrom','strand','chromStart','chromEnd'], right_on = ['chromosome_name','strand','chrom_exon_start','chrom_exon_end'], how='left')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 31\u001b[0;31m     \u001b[0mcaseids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmirna_counts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mintersection\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     32\u001b[0m     \u001b[0mexon_counts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_feather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mirna_counts' is not defined"
     ]
    }
   ],
   "source": [
    "f = Path(path/'exon_counts_kidney.feather')\n",
    "if f.is_file() and False: \n",
    "    exon_counts = pd.read_feather(f)\n",
    "else:\n",
    "    print('translate exon count positions from anew')\n",
    "    exon_counts = pd.read_csv(tcga_path/'HiSeqV2_exon', delimiter='\t')\n",
    "    #exon_counts[['chromosome_name','strand','chrom_exon_start','chrom_exon_end']] = exon_counts.id.apply(lambda x: pd.Series(c_split_chrom_column(str(x))))\n",
    "    print(len(exon_counts))\n",
    "    tran_exon = pd.read_csv(tcga_path/'unc_v2_exon_hg19_probe_TCGA', delimiter='\t')\n",
    "    tran_exon = tran_exon.drop_duplicates(keep='first') # 1884 duplicates -> 948 rows less\n",
    "    #save positions to BED input file\n",
    "    tran_exon['score'] = 0\n",
    "    #tran_exon.to_csv('data/exon_hg19_pos.bed', sep='\\t', columns=['chrom','chromStart','chromEnd','id','score','strand'], header=False, index=False)\n",
    "    #translate positions from GRCh37 to GRCh38\n",
    "    !../liftOver data/exon_hg19_pos.bed data/hg19ToHg38.over.chain.gz data/exon_hg38_pos.bed data/unlifted.bed\n",
    "    #read in translated positions from liftOver output BED file\n",
    "    tran_exon_new = pd.read_csv('data/exon_hg38_pos.bed', delimiter='\t', names=['chrom','chromStart','chromEnd','id','score','strand'])\n",
    "    exon_counts = exon_counts.merge(tran_exon[['gene','id']], left_on='sample', right_on='id', how='left',validate='1:1')\n",
    "    exon_counts = exon_counts.merge(tran_exon_new, left_on='sample', right_on='id', how='inner',validate='1:m')\n",
    "    print(len(exon_counts))\n",
    "   # exon_counts.drop(['sample', 'id_x','id_y','score'], axis=1, inplace=True)\n",
    "    #transform strand {-,+} to {-1,+1}\n",
    "    exon_counts['strand'] = exon_counts.apply(lambda row: -1 if row.strand == '-' else 1, axis=1)\n",
    "    exon_counts['chrom'] = exon_counts['chrom'].str[3:]\n",
    "    exon_counts[['chrom','strand']] = exon_counts[['chrom','strand']].astype('str').astype('category')\n",
    "    exon_counts[['chromStart','chromEnd']] = exon_counts[['chromStart','chromEnd']].astype('Int64')\n",
    "        \n",
    "    #TODO option 2: get exonid per exon count from Ensembl (compare)\n",
    "    #exon_counts2 = exon_counts.merge(df_exon[['exon_id','chromosome_name','strand','chrom_exon_start','chrom_exon_end']], left_on = ['chrom','strand','chromStart','chromEnd'], right_on = ['chromosome_name','strand','chrom_exon_start','chrom_exon_end'], how='left')\n",
    "\n",
    "    caseids = list(set(mirna_counts.columns).intersection(list(exon_counts.columns)))\n",
    "    exon_counts.to_feather(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "208022\n",
      "207695\n"
     ]
    }
   ],
   "source": [
    "print(len(tran_exon))\n",
    "print(len(tran_exon_new))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:12776118-12776347:+</td>\n",
       "      <td>AADACL3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12776118</td>\n",
       "      <td>12776347</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:12779477-12779693:+</td>\n",
       "      <td>AADACL3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12779477</td>\n",
       "      <td>12779693</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:12780885-12780948:+</td>\n",
       "      <td>AADACL3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12780885</td>\n",
       "      <td>12780948</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:12785189-12788726:+</td>\n",
       "      <td>AADACL3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12785189</td>\n",
       "      <td>12788726</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:12704566-12704733:+</td>\n",
       "      <td>AADACL4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12704566</td>\n",
       "      <td>12704733</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208965</th>\n",
       "      <td>chrY:2829115-2829687:+</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2829115</td>\n",
       "      <td>2829687</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208966</th>\n",
       "      <td>chrY:2843136-2843285:+</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2843136</td>\n",
       "      <td>2843285</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208967</th>\n",
       "      <td>chrY:2843552-2843695:+</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2843552</td>\n",
       "      <td>2843695</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208968</th>\n",
       "      <td>chrY:2844711-2845221:+</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2844711</td>\n",
       "      <td>2845221</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208969</th>\n",
       "      <td>chrY:2845981-2846121:+</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2845981</td>\n",
       "      <td>2846121</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208022 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              id     gene chrom  chromStart  chromEnd strand\n",
       "0       chr1:12776118-12776347:+  AADACL3  chr1    12776118  12776347      +\n",
       "1       chr1:12779477-12779693:+  AADACL3  chr1    12779477  12779693      +\n",
       "2       chr1:12780885-12780948:+  AADACL3  chr1    12780885  12780948      +\n",
       "3       chr1:12785189-12788726:+  AADACL3  chr1    12785189  12788726      +\n",
       "4       chr1:12704566-12704733:+  AADACL4  chr1    12704566  12704733      +\n",
       "...                          ...      ...   ...         ...       ...    ...\n",
       "208965    chrY:2829115-2829687:+      ZFY  chrY     2829115   2829687      +\n",
       "208966    chrY:2843136-2843285:+      ZFY  chrY     2843136   2843285      +\n",
       "208967    chrY:2843552-2843695:+      ZFY  chrY     2843552   2843695      +\n",
       "208968    chrY:2844711-2845221:+      ZFY  chrY     2844711   2845221      +\n",
       "208969    chrY:2845981-2846121:+      ZFY  chrY     2845981   2846121      +\n",
       "\n",
       "[208022 rows x 6 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tran_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>TCGA-KL-8346-01</th>\n",
       "      <th>TCGA-KN-8422-01</th>\n",
       "      <th>TCGA-KN-8431-11</th>\n",
       "      <th>TCGA-KN-8430-11</th>\n",
       "      <th>TCGA-KM-8440-01</th>\n",
       "      <th>TCGA-KO-8414-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KM-8639-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "      <th>gene</th>\n",
       "      <th>id_x</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>id_y</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr3:52007981-52008646:-</td>\n",
       "      <td>2.1156</td>\n",
       "      <td>1.7396</td>\n",
       "      <td>2.0641</td>\n",
       "      <td>3.1457</td>\n",
       "      <td>3.0436</td>\n",
       "      <td>2.2209</td>\n",
       "      <td>1.8213</td>\n",
       "      <td>1.9092</td>\n",
       "      <td>2.9139</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0617</td>\n",
       "      <td>0.7424</td>\n",
       "      <td>ABHD14B,RP11-155D18.14,RP11-155D18.12</td>\n",
       "      <td>chr3:52007981-52008646:-</td>\n",
       "      <td>3</td>\n",
       "      <td>51973965</td>\n",
       "      <td>51974630</td>\n",
       "      <td>chr3:52007981-52008646:-</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:215901372-215901726:-</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>chr1:215901372-215901726:-</td>\n",
       "      <td>1</td>\n",
       "      <td>215728030</td>\n",
       "      <td>215728384</td>\n",
       "      <td>chr1:215901372-215901726:-</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4:147214081-147214132:-</td>\n",
       "      <td>2.5290</td>\n",
       "      <td>0.7196</td>\n",
       "      <td>1.3094</td>\n",
       "      <td>2.3074</td>\n",
       "      <td>1.6894</td>\n",
       "      <td>1.5829</td>\n",
       "      <td>1.0835</td>\n",
       "      <td>1.9041</td>\n",
       "      <td>1.2131</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4157</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>SLC10A7</td>\n",
       "      <td>chr4:147214081-147214132:-</td>\n",
       "      <td>4</td>\n",
       "      <td>146292929</td>\n",
       "      <td>146292980</td>\n",
       "      <td>chr4:147214081-147214132:-</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3:195599147-195599341:-</td>\n",
       "      <td>2.3839</td>\n",
       "      <td>3.8702</td>\n",
       "      <td>2.5373</td>\n",
       "      <td>2.3464</td>\n",
       "      <td>2.2976</td>\n",
       "      <td>3.0639</td>\n",
       "      <td>2.5666</td>\n",
       "      <td>2.7963</td>\n",
       "      <td>3.0875</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4157</td>\n",
       "      <td>1.8708</td>\n",
       "      <td>TNK2</td>\n",
       "      <td>chr3:195599147-195599341:-</td>\n",
       "      <td>3</td>\n",
       "      <td>195872276</td>\n",
       "      <td>195872470</td>\n",
       "      <td>chr3:195599147-195599341:-</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17:30351730-30351801:+</td>\n",
       "      <td>1.9974</td>\n",
       "      <td>0.6984</td>\n",
       "      <td>0.9011</td>\n",
       "      <td>0.6827</td>\n",
       "      <td>1.2425</td>\n",
       "      <td>0.9715</td>\n",
       "      <td>1.6701</td>\n",
       "      <td>1.3011</td>\n",
       "      <td>1.3712</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>LRRC37B</td>\n",
       "      <td>chr17:30351730-30351801:+</td>\n",
       "      <td>17</td>\n",
       "      <td>32024711</td>\n",
       "      <td>32024782</td>\n",
       "      <td>chr17:30351730-30351801:+</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207690</th>\n",
       "      <td>chr3:52012275-52012390:+</td>\n",
       "      <td>2.4455</td>\n",
       "      <td>5.1078</td>\n",
       "      <td>2.9468</td>\n",
       "      <td>5.4212</td>\n",
       "      <td>5.1299</td>\n",
       "      <td>4.5371</td>\n",
       "      <td>3.5954</td>\n",
       "      <td>4.4245</td>\n",
       "      <td>4.4484</td>\n",
       "      <td>...</td>\n",
       "      <td>7.3459</td>\n",
       "      <td>4.1126</td>\n",
       "      <td>ABHD14A,ACY1,ABHD14A-ACY1</td>\n",
       "      <td>chr3:52012275-52012390:+</td>\n",
       "      <td>3</td>\n",
       "      <td>51978259</td>\n",
       "      <td>51978374</td>\n",
       "      <td>chr3:52012275-52012390:+</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207691</th>\n",
       "      <td>chr6:43324864-43325503:-</td>\n",
       "      <td>1.5701</td>\n",
       "      <td>0.5500</td>\n",
       "      <td>0.3451</td>\n",
       "      <td>2.2783</td>\n",
       "      <td>1.7366</td>\n",
       "      <td>0.6813</td>\n",
       "      <td>0.5851</td>\n",
       "      <td>1.4460</td>\n",
       "      <td>2.4188</td>\n",
       "      <td>...</td>\n",
       "      <td>1.2410</td>\n",
       "      <td>0.2725</td>\n",
       "      <td>ZNF318</td>\n",
       "      <td>chr6:43324864-43325503:-</td>\n",
       "      <td>6</td>\n",
       "      <td>43357126</td>\n",
       "      <td>43357765</td>\n",
       "      <td>chr6:43324864-43325503:-</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207692</th>\n",
       "      <td>chr17:34495988-34496053:-</td>\n",
       "      <td>0.4810</td>\n",
       "      <td>0.4219</td>\n",
       "      <td>0.3687</td>\n",
       "      <td>0.7313</td>\n",
       "      <td>0.7666</td>\n",
       "      <td>0.7646</td>\n",
       "      <td>2.4586</td>\n",
       "      <td>0.9557</td>\n",
       "      <td>1.1452</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.9102</td>\n",
       "      <td>TBC1D3B</td>\n",
       "      <td>chr17:34495988-34496053:-</td>\n",
       "      <td>17</td>\n",
       "      <td>36168630</td>\n",
       "      <td>36168695</td>\n",
       "      <td>chr17:34495988-34496053:-</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207693</th>\n",
       "      <td>chr1:148010884-148011056:-</td>\n",
       "      <td>2.6944</td>\n",
       "      <td>2.2891</td>\n",
       "      <td>1.8186</td>\n",
       "      <td>4.0291</td>\n",
       "      <td>4.0015</td>\n",
       "      <td>3.7968</td>\n",
       "      <td>2.3984</td>\n",
       "      <td>2.6564</td>\n",
       "      <td>4.2824</td>\n",
       "      <td>...</td>\n",
       "      <td>2.3518</td>\n",
       "      <td>1.1468</td>\n",
       "      <td>NBPF14</td>\n",
       "      <td>chr1:148010884-148011056:-</td>\n",
       "      <td>1</td>\n",
       "      <td>120833602</td>\n",
       "      <td>120833774</td>\n",
       "      <td>chr1:148010884-148011056:-</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207694</th>\n",
       "      <td>chr18:12353004-12353157:-</td>\n",
       "      <td>6.8409</td>\n",
       "      <td>5.8767</td>\n",
       "      <td>6.3420</td>\n",
       "      <td>6.0730</td>\n",
       "      <td>5.8084</td>\n",
       "      <td>6.7076</td>\n",
       "      <td>5.0113</td>\n",
       "      <td>6.8595</td>\n",
       "      <td>6.8408</td>\n",
       "      <td>...</td>\n",
       "      <td>4.9554</td>\n",
       "      <td>3.4175</td>\n",
       "      <td>AFG3L2</td>\n",
       "      <td>chr18:12353004-12353157:-</td>\n",
       "      <td>18</td>\n",
       "      <td>12353005</td>\n",
       "      <td>12353158</td>\n",
       "      <td>chr18:12353004-12353157:-</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>207695 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            sample  TCGA-KN-8419-01  TCGA-KL-8346-01  \\\n",
       "0         chr3:52007981-52008646:-           2.1156           1.7396   \n",
       "1       chr1:215901372-215901726:-           0.0000           0.0000   \n",
       "2       chr4:147214081-147214132:-           2.5290           0.7196   \n",
       "3       chr3:195599147-195599341:-           2.3839           3.8702   \n",
       "4        chr17:30351730-30351801:+           1.9974           0.6984   \n",
       "...                            ...              ...              ...   \n",
       "207690    chr3:52012275-52012390:+           2.4455           5.1078   \n",
       "207691    chr6:43324864-43325503:-           1.5701           0.5500   \n",
       "207692   chr17:34495988-34496053:-           0.4810           0.4219   \n",
       "207693  chr1:148010884-148011056:-           2.6944           2.2891   \n",
       "207694   chr18:12353004-12353157:-           6.8409           5.8767   \n",
       "\n",
       "        TCGA-KN-8422-01  TCGA-KN-8431-11  TCGA-KN-8430-11  TCGA-KM-8440-01  \\\n",
       "0                2.0641           3.1457           3.0436           2.2209   \n",
       "1                0.0000           0.0251           0.0233           0.0000   \n",
       "2                1.3094           2.3074           1.6894           1.5829   \n",
       "3                2.5373           2.3464           2.2976           3.0639   \n",
       "4                0.9011           0.6827           1.2425           0.9715   \n",
       "...                 ...              ...              ...              ...   \n",
       "207690           2.9468           5.4212           5.1299           4.5371   \n",
       "207691           0.3451           2.2783           1.7366           0.6813   \n",
       "207692           0.3687           0.7313           0.7666           0.7646   \n",
       "207693           1.8186           4.0291           4.0015           3.7968   \n",
       "207694           6.3420           6.0730           5.8084           6.7076   \n",
       "\n",
       "        TCGA-KO-8414-01  TCGA-KL-8323-01  TCGA-KM-8639-01  ...  \\\n",
       "0                1.8213           1.9092           2.9139  ...   \n",
       "1                0.0000           0.0000           0.0000  ...   \n",
       "2                1.0835           1.9041           1.2131  ...   \n",
       "3                2.5666           2.7963           3.0875  ...   \n",
       "4                1.6701           1.3011           1.3712  ...   \n",
       "...                 ...              ...              ...  ...   \n",
       "207690           3.5954           4.4245           4.4484  ...   \n",
       "207691           0.5851           1.4460           2.4188  ...   \n",
       "207692           2.4586           0.9557           1.1452  ...   \n",
       "207693           2.3984           2.6564           4.2824  ...   \n",
       "207694           5.0113           6.8595           6.8408  ...   \n",
       "\n",
       "        TCGA-KL-8332-11  TCGA-KL-8327-01  \\\n",
       "0                3.0617           0.7424   \n",
       "1                0.0000           0.0000   \n",
       "2                1.4157           0.0000   \n",
       "3                1.4157           1.8708   \n",
       "4                0.0000           0.0000   \n",
       "...                 ...              ...   \n",
       "207690           7.3459           4.1126   \n",
       "207691           1.2410           0.2725   \n",
       "207692           0.0000           1.9102   \n",
       "207693           2.3518           1.1468   \n",
       "207694           4.9554           3.4175   \n",
       "\n",
       "                                         gene                        id_x  \\\n",
       "0       ABHD14B,RP11-155D18.14,RP11-155D18.12    chr3:52007981-52008646:-   \n",
       "1                                       USH2A  chr1:215901372-215901726:-   \n",
       "2                                     SLC10A7  chr4:147214081-147214132:-   \n",
       "3                                        TNK2  chr3:195599147-195599341:-   \n",
       "4                                     LRRC37B   chr17:30351730-30351801:+   \n",
       "...                                       ...                         ...   \n",
       "207690              ABHD14A,ACY1,ABHD14A-ACY1    chr3:52012275-52012390:+   \n",
       "207691                                 ZNF318    chr6:43324864-43325503:-   \n",
       "207692                                TBC1D3B   chr17:34495988-34496053:-   \n",
       "207693                                 NBPF14  chr1:148010884-148011056:-   \n",
       "207694                                 AFG3L2   chr18:12353004-12353157:-   \n",
       "\n",
       "        chrom  chromStart   chromEnd                        id_y  score  \\\n",
       "0           3    51973965   51974630    chr3:52007981-52008646:-      0   \n",
       "1           1   215728030  215728384  chr1:215901372-215901726:-      0   \n",
       "2           4   146292929  146292980  chr4:147214081-147214132:-      0   \n",
       "3           3   195872276  195872470  chr3:195599147-195599341:-      0   \n",
       "4          17    32024711   32024782   chr17:30351730-30351801:+      0   \n",
       "...       ...         ...        ...                         ...    ...   \n",
       "207690      3    51978259   51978374    chr3:52012275-52012390:+      0   \n",
       "207691      6    43357126   43357765    chr6:43324864-43325503:-      0   \n",
       "207692     17    36168630   36168695   chr17:34495988-34496053:-      0   \n",
       "207693      1   120833602  120833774  chr1:148010884-148011056:-      0   \n",
       "207694     18    12353005   12353158   chr18:12353004-12353157:-      0   \n",
       "\n",
       "        strand  \n",
       "0           -1  \n",
       "1           -1  \n",
       "2           -1  \n",
       "3           -1  \n",
       "4            1  \n",
       "...        ...  \n",
       "207690       1  \n",
       "207691      -1  \n",
       "207692      -1  \n",
       "207693       1  \n",
       "207694      -1  \n",
       "\n",
       "[207695 rows x 100 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>TCGA-KL-8346-01</th>\n",
       "      <th>TCGA-KN-8422-01</th>\n",
       "      <th>TCGA-KN-8431-11</th>\n",
       "      <th>TCGA-KN-8430-11</th>\n",
       "      <th>TCGA-KM-8440-01</th>\n",
       "      <th>TCGA-KO-8414-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KM-8639-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8417-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8340-01</th>\n",
       "      <th>TCGA-KO-8406-01</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8443-01</th>\n",
       "      <th>TCGA-KM-8442-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17:49268962-49270377:-</td>\n",
       "      <td>0.3884</td>\n",
       "      <td>0.2219</td>\n",
       "      <td>0.2050</td>\n",
       "      <td>0.9168</td>\n",
       "      <td>0.6069</td>\n",
       "      <td>0.4547</td>\n",
       "      <td>0.1571</td>\n",
       "      <td>0.3857</td>\n",
       "      <td>0.7535</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4021</td>\n",
       "      <td>0.1127</td>\n",
       "      <td>0.3816</td>\n",
       "      <td>0.5932</td>\n",
       "      <td>0.1602</td>\n",
       "      <td>0.3507</td>\n",
       "      <td>0.7154</td>\n",
       "      <td>0.4353</td>\n",
       "      <td>0.0966</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:16074041-16074475:+</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4721</td>\n",
       "      <td>0.1284</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0397</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0189</td>\n",
       "      <td>0.1334</td>\n",
       "      <td>4.8648</td>\n",
       "      <td>0.0900</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr14:106521424-106521433:-</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr11:65620753-65621170:+</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2804</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0630</td>\n",
       "      <td>0.3368</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1703</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4093</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0614</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0491</td>\n",
       "      <td>0.0390</td>\n",
       "      <td>0.1386</td>\n",
       "      <td>0.1424</td>\n",
       "      <td>0.0240</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr15:31008518-31008545:+</td>\n",
       "      <td>1.1151</td>\n",
       "      <td>0.4858</td>\n",
       "      <td>0.2285</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0212</td>\n",
       "      <td>0.2724</td>\n",
       "      <td>0.7350</td>\n",
       "      <td>0.4937</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2830</td>\n",
       "      <td>0.2826</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2700</td>\n",
       "      <td>0.6008</td>\n",
       "      <td>1.7890</td>\n",
       "      <td>0.8953</td>\n",
       "      <td>0.2553</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239233</th>\n",
       "      <td>chr1:15798485-15798585:+</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2261</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0849</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239271</th>\n",
       "      <td>chr14:69555924-69557873:-</td>\n",
       "      <td>0.3462</td>\n",
       "      <td>0.4367</td>\n",
       "      <td>0.0106</td>\n",
       "      <td>0.1936</td>\n",
       "      <td>0.3415</td>\n",
       "      <td>0.4876</td>\n",
       "      <td>0.2449</td>\n",
       "      <td>0.5533</td>\n",
       "      <td>0.2155</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0311</td>\n",
       "      <td>0.5334</td>\n",
       "      <td>0.4010</td>\n",
       "      <td>0.4566</td>\n",
       "      <td>0.0474</td>\n",
       "      <td>0.5502</td>\n",
       "      <td>0.4836</td>\n",
       "      <td>0.2087</td>\n",
       "      <td>0.1523</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239279</th>\n",
       "      <td>chr1:45249259-45250155:-</td>\n",
       "      <td>0.2589</td>\n",
       "      <td>0.5108</td>\n",
       "      <td>0.1468</td>\n",
       "      <td>0.2811</td>\n",
       "      <td>0.4449</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>0.6869</td>\n",
       "      <td>0.0198</td>\n",
       "      <td>0.1730</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1036</td>\n",
       "      <td>0.6545</td>\n",
       "      <td>0.2105</td>\n",
       "      <td>0.2620</td>\n",
       "      <td>0.4969</td>\n",
       "      <td>0.1978</td>\n",
       "      <td>0.4603</td>\n",
       "      <td>0.2030</td>\n",
       "      <td>0.0660</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239304</th>\n",
       "      <td>chr11:93211641-93212450:-</td>\n",
       "      <td>4.0361</td>\n",
       "      <td>3.7048</td>\n",
       "      <td>2.1906</td>\n",
       "      <td>4.6267</td>\n",
       "      <td>4.3716</td>\n",
       "      <td>1.5919</td>\n",
       "      <td>3.8414</td>\n",
       "      <td>4.4238</td>\n",
       "      <td>4.2922</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6441</td>\n",
       "      <td>2.7758</td>\n",
       "      <td>4.0947</td>\n",
       "      <td>4.4172</td>\n",
       "      <td>3.2479</td>\n",
       "      <td>2.9461</td>\n",
       "      <td>4.0862</td>\n",
       "      <td>4.1306</td>\n",
       "      <td>1.3293</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239313</th>\n",
       "      <td>chr2:242751562-242751722:-</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1583</td>\n",
       "      <td>0.0509</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0940</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31627 rows × 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             sample  TCGA-KN-8419-01  TCGA-KL-8346-01  \\\n",
       "2         chr17:49268962-49270377:-           0.3884           0.2219   \n",
       "4          chr1:16074041-16074475:+           0.0000           0.0000   \n",
       "7       chr14:106521424-106521433:-           0.0000           0.0000   \n",
       "20        chr11:65620753-65621170:+           0.0000           0.2804   \n",
       "44        chr15:31008518-31008545:+           1.1151           0.4858   \n",
       "...                             ...              ...              ...   \n",
       "239233     chr1:15798485-15798585:+           0.0000           0.0000   \n",
       "239271    chr14:69555924-69557873:-           0.3462           0.4367   \n",
       "239279     chr1:45249259-45250155:-           0.2589           0.5108   \n",
       "239304    chr11:93211641-93212450:-           4.0361           3.7048   \n",
       "239313   chr2:242751562-242751722:-           0.0000           0.0000   \n",
       "\n",
       "        TCGA-KN-8422-01  TCGA-KN-8431-11  TCGA-KN-8430-11  TCGA-KM-8440-01  \\\n",
       "2                0.2050           0.9168           0.6069           0.4547   \n",
       "4                0.0000           1.4721           0.1284           0.0000   \n",
       "7                0.0000           0.0000           0.0000           0.0000   \n",
       "20               0.0000           0.0630           0.3368           0.0000   \n",
       "44               0.2285           0.0000           0.0000           1.0212   \n",
       "...                 ...              ...              ...              ...   \n",
       "239233           0.0000           0.0000           0.0000           0.0000   \n",
       "239271           0.0106           0.1936           0.3415           0.4876   \n",
       "239279           0.1468           0.2811           0.4449           0.0456   \n",
       "239304           2.1906           4.6267           4.3716           1.5919   \n",
       "239313           0.0000           0.1583           0.0509           0.0000   \n",
       "\n",
       "        TCGA-KO-8414-01  TCGA-KL-8323-01  TCGA-KM-8639-01  ...  \\\n",
       "2                0.1571           0.3857           0.7535  ...   \n",
       "4                0.0000           0.0000           0.0374  ...   \n",
       "7                0.0000           0.0000           0.0000  ...   \n",
       "20               0.1703           0.0000           0.4093  ...   \n",
       "44               0.2724           0.7350           0.4937  ...   \n",
       "...                 ...              ...              ...  ...   \n",
       "239233           0.0000           0.0000           0.2261  ...   \n",
       "239271           0.2449           0.5533           0.2155  ...   \n",
       "239279           0.6869           0.0198           0.1730  ...   \n",
       "239304           3.8414           4.4238           4.2922  ...   \n",
       "239313           0.0000           0.0000           0.0000  ...   \n",
       "\n",
       "        TCGA-KO-8417-01  TCGA-KM-8438-01  TCGA-KL-8340-01  TCGA-KO-8406-01  \\\n",
       "2                0.4021           0.1127           0.3816           0.5932   \n",
       "4                0.0397           0.0000           0.0000           0.0000   \n",
       "7                0.0000           0.0000           0.0000           0.0000   \n",
       "20               0.0615           0.0614           0.0000           0.0000   \n",
       "44               0.2830           0.2826           0.0000           0.2700   \n",
       "...                 ...              ...              ...              ...   \n",
       "239233           0.0000           0.0000           0.0849           0.0000   \n",
       "239271           0.0311           0.5334           0.4010           0.4566   \n",
       "239279           0.1036           0.6545           0.2105           0.2620   \n",
       "239304           3.6441           2.7758           4.0947           4.4172   \n",
       "239313           0.0000           0.0000           0.0000           0.0000   \n",
       "\n",
       "        TCGA-KO-8408-01  TCGA-KM-8443-01  TCGA-KM-8442-01  TCGA-KL-8332-11  \\\n",
       "2                0.1602           0.3507           0.7154           0.4353   \n",
       "4                0.0000           0.0189           0.1334           4.8648   \n",
       "7                0.0000           0.0000           0.0000           0.0000   \n",
       "20               0.0491           0.0390           0.1386           0.1424   \n",
       "44               0.6008           1.7890           0.8953           0.2553   \n",
       "...                 ...              ...              ...              ...   \n",
       "239233           0.0000           0.0000           0.0000           0.0000   \n",
       "239271           0.0474           0.5502           0.4836           0.2087   \n",
       "239279           0.4969           0.1978           0.4603           0.2030   \n",
       "239304           3.2479           2.9461           4.0862           4.1306   \n",
       "239313           0.0000           0.0000           0.0000           0.0940   \n",
       "\n",
       "        TCGA-KL-8327-01     _merge  \n",
       "2                0.0966  left_only  \n",
       "4                0.0900  left_only  \n",
       "7                0.0000  left_only  \n",
       "20               0.0240  left_only  \n",
       "44               0.0000  left_only  \n",
       "...                 ...        ...  \n",
       "239233           0.0000  left_only  \n",
       "239271           0.1523  left_only  \n",
       "239279           0.0660  left_only  \n",
       "239304           1.3293  left_only  \n",
       "239313           0.0000  left_only  \n",
       "\n",
       "[31627 rows x 93 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#exon_counts_old 239.322, exon_counts 207.695\n",
    "#all rows different in old and mergedwith2: 31.627\n",
    "#samples only in old \n",
    "missing = exon_counts_old.merge(exon_counts['sample'], how = 'outer' ,indicator=True).loc[lambda x : x['_merge']=='left_only']\n",
    "missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>TCGA-KL-8346-01</th>\n",
       "      <th>TCGA-KN-8422-01</th>\n",
       "      <th>TCGA-KN-8431-11</th>\n",
       "      <th>TCGA-KN-8430-11</th>\n",
       "      <th>TCGA-KM-8440-01</th>\n",
       "      <th>TCGA-KO-8414-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KM-8639-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KN-8421-01</th>\n",
       "      <th>TCGA-KO-8417-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8340-01</th>\n",
       "      <th>TCGA-KO-8406-01</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8443-01</th>\n",
       "      <th>TCGA-KM-8442-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr3:52007981-52008646:-</td>\n",
       "      <td>2.1156</td>\n",
       "      <td>1.7396</td>\n",
       "      <td>2.0641</td>\n",
       "      <td>3.1457</td>\n",
       "      <td>3.0436</td>\n",
       "      <td>2.2209</td>\n",
       "      <td>1.8213</td>\n",
       "      <td>1.9092</td>\n",
       "      <td>2.9139</td>\n",
       "      <td>...</td>\n",
       "      <td>1.8941</td>\n",
       "      <td>2.8727</td>\n",
       "      <td>2.1821</td>\n",
       "      <td>2.2543</td>\n",
       "      <td>2.2040</td>\n",
       "      <td>2.2369</td>\n",
       "      <td>2.4951</td>\n",
       "      <td>2.1010</td>\n",
       "      <td>3.0617</td>\n",
       "      <td>0.7424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:215901372-215901726:-</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17:49268962-49270377:-</td>\n",
       "      <td>0.3884</td>\n",
       "      <td>0.2219</td>\n",
       "      <td>0.2050</td>\n",
       "      <td>0.9168</td>\n",
       "      <td>0.6069</td>\n",
       "      <td>0.4547</td>\n",
       "      <td>0.1571</td>\n",
       "      <td>0.3857</td>\n",
       "      <td>0.7535</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2138</td>\n",
       "      <td>0.4021</td>\n",
       "      <td>0.1127</td>\n",
       "      <td>0.3816</td>\n",
       "      <td>0.5932</td>\n",
       "      <td>0.1602</td>\n",
       "      <td>0.3507</td>\n",
       "      <td>0.7154</td>\n",
       "      <td>0.4353</td>\n",
       "      <td>0.0966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4:147214081-147214132:-</td>\n",
       "      <td>2.5290</td>\n",
       "      <td>0.7196</td>\n",
       "      <td>1.3094</td>\n",
       "      <td>2.3074</td>\n",
       "      <td>1.6894</td>\n",
       "      <td>1.5829</td>\n",
       "      <td>1.0835</td>\n",
       "      <td>1.9041</td>\n",
       "      <td>1.2131</td>\n",
       "      <td>...</td>\n",
       "      <td>2.1785</td>\n",
       "      <td>1.6327</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.5862</td>\n",
       "      <td>1.7814</td>\n",
       "      <td>0.6381</td>\n",
       "      <td>1.5227</td>\n",
       "      <td>1.7786</td>\n",
       "      <td>1.4157</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:16074041-16074475:+</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4721</td>\n",
       "      <td>0.1284</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0397</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0189</td>\n",
       "      <td>0.1334</td>\n",
       "      <td>4.8648</td>\n",
       "      <td>0.0900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239317</th>\n",
       "      <td>chr3:52012275-52012390:+</td>\n",
       "      <td>2.4455</td>\n",
       "      <td>5.1078</td>\n",
       "      <td>2.9468</td>\n",
       "      <td>5.4212</td>\n",
       "      <td>5.1299</td>\n",
       "      <td>4.5371</td>\n",
       "      <td>3.5954</td>\n",
       "      <td>4.4245</td>\n",
       "      <td>4.4484</td>\n",
       "      <td>...</td>\n",
       "      <td>4.3723</td>\n",
       "      <td>4.6508</td>\n",
       "      <td>4.2172</td>\n",
       "      <td>4.4777</td>\n",
       "      <td>4.4666</td>\n",
       "      <td>5.0518</td>\n",
       "      <td>4.9067</td>\n",
       "      <td>3.8206</td>\n",
       "      <td>7.3459</td>\n",
       "      <td>4.1126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239318</th>\n",
       "      <td>chr6:43324864-43325503:-</td>\n",
       "      <td>1.5701</td>\n",
       "      <td>0.5500</td>\n",
       "      <td>0.3451</td>\n",
       "      <td>2.2783</td>\n",
       "      <td>1.7366</td>\n",
       "      <td>0.6813</td>\n",
       "      <td>0.5851</td>\n",
       "      <td>1.4460</td>\n",
       "      <td>2.4188</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4057</td>\n",
       "      <td>1.2708</td>\n",
       "      <td>0.7898</td>\n",
       "      <td>0.8588</td>\n",
       "      <td>1.3135</td>\n",
       "      <td>0.8498</td>\n",
       "      <td>1.0774</td>\n",
       "      <td>0.8774</td>\n",
       "      <td>1.2410</td>\n",
       "      <td>0.2725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239319</th>\n",
       "      <td>chr17:34495988-34496053:-</td>\n",
       "      <td>0.4810</td>\n",
       "      <td>0.4219</td>\n",
       "      <td>0.3687</td>\n",
       "      <td>0.7313</td>\n",
       "      <td>0.7666</td>\n",
       "      <td>0.7646</td>\n",
       "      <td>2.4586</td>\n",
       "      <td>0.9557</td>\n",
       "      <td>1.1452</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0783</td>\n",
       "      <td>0.7170</td>\n",
       "      <td>0.7943</td>\n",
       "      <td>0.3544</td>\n",
       "      <td>0.6077</td>\n",
       "      <td>1.3415</td>\n",
       "      <td>1.8842</td>\n",
       "      <td>0.8646</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.9102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239320</th>\n",
       "      <td>chr1:148010884-148011056:-</td>\n",
       "      <td>2.6944</td>\n",
       "      <td>2.2891</td>\n",
       "      <td>1.8186</td>\n",
       "      <td>4.0291</td>\n",
       "      <td>4.0015</td>\n",
       "      <td>3.7968</td>\n",
       "      <td>2.3984</td>\n",
       "      <td>2.6564</td>\n",
       "      <td>4.2824</td>\n",
       "      <td>...</td>\n",
       "      <td>2.9737</td>\n",
       "      <td>2.2911</td>\n",
       "      <td>1.5489</td>\n",
       "      <td>2.4209</td>\n",
       "      <td>3.6234</td>\n",
       "      <td>3.0631</td>\n",
       "      <td>1.8938</td>\n",
       "      <td>2.9015</td>\n",
       "      <td>2.3518</td>\n",
       "      <td>1.1468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239321</th>\n",
       "      <td>chr18:12353004-12353157:-</td>\n",
       "      <td>6.8409</td>\n",
       "      <td>5.8767</td>\n",
       "      <td>6.3420</td>\n",
       "      <td>6.0730</td>\n",
       "      <td>5.8084</td>\n",
       "      <td>6.7076</td>\n",
       "      <td>5.0113</td>\n",
       "      <td>6.8595</td>\n",
       "      <td>6.8408</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0800</td>\n",
       "      <td>6.9124</td>\n",
       "      <td>5.7598</td>\n",
       "      <td>6.3939</td>\n",
       "      <td>6.6215</td>\n",
       "      <td>6.0402</td>\n",
       "      <td>6.0639</td>\n",
       "      <td>6.8604</td>\n",
       "      <td>4.9554</td>\n",
       "      <td>3.4175</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>239322 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            sample  TCGA-KN-8419-01  TCGA-KL-8346-01  \\\n",
       "0         chr3:52007981-52008646:-           2.1156           1.7396   \n",
       "1       chr1:215901372-215901726:-           0.0000           0.0000   \n",
       "2        chr17:49268962-49270377:-           0.3884           0.2219   \n",
       "3       chr4:147214081-147214132:-           2.5290           0.7196   \n",
       "4         chr1:16074041-16074475:+           0.0000           0.0000   \n",
       "...                            ...              ...              ...   \n",
       "239317    chr3:52012275-52012390:+           2.4455           5.1078   \n",
       "239318    chr6:43324864-43325503:-           1.5701           0.5500   \n",
       "239319   chr17:34495988-34496053:-           0.4810           0.4219   \n",
       "239320  chr1:148010884-148011056:-           2.6944           2.2891   \n",
       "239321   chr18:12353004-12353157:-           6.8409           5.8767   \n",
       "\n",
       "        TCGA-KN-8422-01  TCGA-KN-8431-11  TCGA-KN-8430-11  TCGA-KM-8440-01  \\\n",
       "0                2.0641           3.1457           3.0436           2.2209   \n",
       "1                0.0000           0.0251           0.0233           0.0000   \n",
       "2                0.2050           0.9168           0.6069           0.4547   \n",
       "3                1.3094           2.3074           1.6894           1.5829   \n",
       "4                0.0000           1.4721           0.1284           0.0000   \n",
       "...                 ...              ...              ...              ...   \n",
       "239317           2.9468           5.4212           5.1299           4.5371   \n",
       "239318           0.3451           2.2783           1.7366           0.6813   \n",
       "239319           0.3687           0.7313           0.7666           0.7646   \n",
       "239320           1.8186           4.0291           4.0015           3.7968   \n",
       "239321           6.3420           6.0730           5.8084           6.7076   \n",
       "\n",
       "        TCGA-KO-8414-01  TCGA-KL-8323-01  TCGA-KM-8639-01  ...  \\\n",
       "0                1.8213           1.9092           2.9139  ...   \n",
       "1                0.0000           0.0000           0.0000  ...   \n",
       "2                0.1571           0.3857           0.7535  ...   \n",
       "3                1.0835           1.9041           1.2131  ...   \n",
       "4                0.0000           0.0000           0.0374  ...   \n",
       "...                 ...              ...              ...  ...   \n",
       "239317           3.5954           4.4245           4.4484  ...   \n",
       "239318           0.5851           1.4460           2.4188  ...   \n",
       "239319           2.4586           0.9557           1.1452  ...   \n",
       "239320           2.3984           2.6564           4.2824  ...   \n",
       "239321           5.0113           6.8595           6.8408  ...   \n",
       "\n",
       "        TCGA-KN-8421-01  TCGA-KO-8417-01  TCGA-KM-8438-01  TCGA-KL-8340-01  \\\n",
       "0                1.8941           2.8727           2.1821           2.2543   \n",
       "1                0.0000           0.0000           0.0000           0.0000   \n",
       "2                0.2138           0.4021           0.1127           0.3816   \n",
       "3                2.1785           1.6327           0.0000           1.5862   \n",
       "4                0.0000           0.0397           0.0000           0.0000   \n",
       "...                 ...              ...              ...              ...   \n",
       "239317           4.3723           4.6508           4.2172           4.4777   \n",
       "239318           0.4057           1.2708           0.7898           0.8588   \n",
       "239319           1.0783           0.7170           0.7943           0.3544   \n",
       "239320           2.9737           2.2911           1.5489           2.4209   \n",
       "239321           6.0800           6.9124           5.7598           6.3939   \n",
       "\n",
       "        TCGA-KO-8406-01  TCGA-KO-8408-01  TCGA-KM-8443-01  TCGA-KM-8442-01  \\\n",
       "0                2.2040           2.2369           2.4951           2.1010   \n",
       "1                0.0000           0.0000           0.0000           0.0000   \n",
       "2                0.5932           0.1602           0.3507           0.7154   \n",
       "3                1.7814           0.6381           1.5227           1.7786   \n",
       "4                0.0000           0.0000           0.0189           0.1334   \n",
       "...                 ...              ...              ...              ...   \n",
       "239317           4.4666           5.0518           4.9067           3.8206   \n",
       "239318           1.3135           0.8498           1.0774           0.8774   \n",
       "239319           0.6077           1.3415           1.8842           0.8646   \n",
       "239320           3.6234           3.0631           1.8938           2.9015   \n",
       "239321           6.6215           6.0402           6.0639           6.8604   \n",
       "\n",
       "        TCGA-KL-8332-11  TCGA-KL-8327-01  \n",
       "0                3.0617           0.7424  \n",
       "1                0.0000           0.0000  \n",
       "2                0.4353           0.0966  \n",
       "3                1.4157           0.0000  \n",
       "4                4.8648           0.0900  \n",
       "...                 ...              ...  \n",
       "239317           7.3459           4.1126  \n",
       "239318           1.2410           0.2725  \n",
       "239319           0.0000           1.9102  \n",
       "239320           2.3518           1.1468  \n",
       "239321           4.9554           3.4175  \n",
       "\n",
       "[239322 rows x 92 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts_old = pd.read_csv(tcga_path/'HiSeqV2_exon', delimiter='\t')\n",
    "#exon_counts[exon_counts.gene.isna()]#5059 , 208970 - 208643 = 327\n",
    "exon_counts_old#[exon_counts.duplicated(subset=['sample'],keep=False)]#[exon_counts.id.isna()] #left 240270 -> 31627 NA\n",
    "#counts after duplicates: inner:207695 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-26T20:17:28.402273Z",
     "start_time": "2020-10-26T20:17:28.250365Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>TCGA-C4-A0F6-01</th>\n",
       "      <th>TCGA-CU-A0YO-01</th>\n",
       "      <th>TCGA-BT-A0S7-01</th>\n",
       "      <th>TCGA-CU-A0YR-01</th>\n",
       "      <th>TCGA-BL-A0C8-01</th>\n",
       "      <th>TCGA-C4-A0F0-01</th>\n",
       "      <th>TCGA-BL-A13J-01</th>\n",
       "      <th>TCGA-BT-A0YX-01</th>\n",
       "      <th>TCGA-CU-A0YN-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-AG-A020-01</th>\n",
       "      <th>TCGA-AG-A01Y-01</th>\n",
       "      <th>TCGA-AG-A01W-01</th>\n",
       "      <th>TCGA-AG-3726-01</th>\n",
       "      <th>TCGA-AG-3605-01</th>\n",
       "      <th>TCGA-AG-3584-01</th>\n",
       "      <th>TCGA-AG-3599-01</th>\n",
       "      <th>TCGA-AG-3583-01</th>\n",
       "      <th>TCGA-AG-3598-01</th>\n",
       "      <th>TCGA-AG-3586-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.91</td>\n",
       "      <td>3.02</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.85</td>\n",
       "      <td>2.70</td>\n",
       "      <td>2.50</td>\n",
       "      <td>1.22</td>\n",
       "      <td>1.57</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.91</td>\n",
       "      <td>1.18</td>\n",
       "      <td>1.29</td>\n",
       "      <td>1.02</td>\n",
       "      <td>0.16</td>\n",
       "      <td>2.09</td>\n",
       "      <td>1.59</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-3p</td>\n",
       "      <td>5.08</td>\n",
       "      <td>5.99</td>\n",
       "      <td>6.03</td>\n",
       "      <td>4.23</td>\n",
       "      <td>5.06</td>\n",
       "      <td>5.05</td>\n",
       "      <td>5.43</td>\n",
       "      <td>4.64</td>\n",
       "      <td>5.09</td>\n",
       "      <td>...</td>\n",
       "      <td>5.91</td>\n",
       "      <td>5.75</td>\n",
       "      <td>5.40</td>\n",
       "      <td>4.45</td>\n",
       "      <td>4.61</td>\n",
       "      <td>4.54</td>\n",
       "      <td>4.91</td>\n",
       "      <td>4.38</td>\n",
       "      <td>4.75</td>\n",
       "      <td>4.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>14.12</td>\n",
       "      <td>14.35</td>\n",
       "      <td>15.97</td>\n",
       "      <td>14.96</td>\n",
       "      <td>15.16</td>\n",
       "      <td>15.33</td>\n",
       "      <td>14.95</td>\n",
       "      <td>14.67</td>\n",
       "      <td>14.53</td>\n",
       "      <td>...</td>\n",
       "      <td>14.47</td>\n",
       "      <td>14.83</td>\n",
       "      <td>14.70</td>\n",
       "      <td>14.39</td>\n",
       "      <td>14.73</td>\n",
       "      <td>14.50</td>\n",
       "      <td>14.13</td>\n",
       "      <td>14.04</td>\n",
       "      <td>14.50</td>\n",
       "      <td>13.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>2.95</td>\n",
       "      <td>3.68</td>\n",
       "      <td>5.40</td>\n",
       "      <td>3.92</td>\n",
       "      <td>4.26</td>\n",
       "      <td>4.38</td>\n",
       "      <td>4.33</td>\n",
       "      <td>3.42</td>\n",
       "      <td>4.87</td>\n",
       "      <td>...</td>\n",
       "      <td>4.52</td>\n",
       "      <td>3.24</td>\n",
       "      <td>4.08</td>\n",
       "      <td>4.73</td>\n",
       "      <td>3.58</td>\n",
       "      <td>4.43</td>\n",
       "      <td>4.89</td>\n",
       "      <td>4.55</td>\n",
       "      <td>5.42</td>\n",
       "      <td>3.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>12.75</td>\n",
       "      <td>12.55</td>\n",
       "      <td>15.08</td>\n",
       "      <td>12.90</td>\n",
       "      <td>12.54</td>\n",
       "      <td>14.03</td>\n",
       "      <td>13.23</td>\n",
       "      <td>12.83</td>\n",
       "      <td>12.89</td>\n",
       "      <td>...</td>\n",
       "      <td>11.67</td>\n",
       "      <td>11.78</td>\n",
       "      <td>12.56</td>\n",
       "      <td>12.09</td>\n",
       "      <td>12.38</td>\n",
       "      <td>12.22</td>\n",
       "      <td>12.47</td>\n",
       "      <td>12.89</td>\n",
       "      <td>12.58</td>\n",
       "      <td>12.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>738</th>\n",
       "      <td>hsa-miR-888-5p</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>739</th>\n",
       "      <td>hsa-miR-890</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>hsa-miR-891b</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>741</th>\n",
       "      <td>hsa-miR-892b</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>hsa-miR-892c-3p</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>743 rows × 10825 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample  TCGA-C4-A0F6-01  TCGA-CU-A0YO-01  TCGA-BT-A0S7-01  \\\n",
       "0    hsa-let-7a-2-3p             0.99             1.91             3.02   \n",
       "1      hsa-let-7a-3p             5.08             5.99             6.03   \n",
       "2      hsa-let-7a-5p            14.12            14.35            15.97   \n",
       "3      hsa-let-7b-3p             2.95             3.68             5.40   \n",
       "4      hsa-let-7b-5p            12.75            12.55            15.08   \n",
       "..               ...              ...              ...              ...   \n",
       "738   hsa-miR-888-5p             0.00             0.00             0.36   \n",
       "739      hsa-miR-890             0.00             0.00             0.00   \n",
       "740     hsa-miR-891b             0.00             0.00             0.00   \n",
       "741     hsa-miR-892b             0.00             0.00             0.00   \n",
       "742  hsa-miR-892c-3p             0.00             0.00             0.00   \n",
       "\n",
       "     TCGA-CU-A0YR-01  TCGA-BL-A0C8-01  TCGA-C4-A0F0-01  TCGA-BL-A13J-01  \\\n",
       "0               0.85             0.85             2.70             2.50   \n",
       "1               4.23             5.06             5.05             5.43   \n",
       "2              14.96            15.16            15.33            14.95   \n",
       "3               3.92             4.26             4.38             4.33   \n",
       "4              12.90            12.54            14.03            13.23   \n",
       "..               ...              ...              ...              ...   \n",
       "738             0.00             0.00             0.00             0.00   \n",
       "739             0.00             0.00             0.00             0.00   \n",
       "740             0.00             0.00             0.00             0.00   \n",
       "741             0.00             0.00             0.26             0.00   \n",
       "742             0.00             0.00             0.26             0.00   \n",
       "\n",
       "     TCGA-BT-A0YX-01  TCGA-CU-A0YN-01  ...  TCGA-AG-A020-01  TCGA-AG-A01Y-01  \\\n",
       "0               1.22             1.57  ...            -0.91             1.18   \n",
       "1               4.64             5.09  ...             5.91             5.75   \n",
       "2              14.67            14.53  ...            14.47            14.83   \n",
       "3               3.42             4.87  ...             4.52             3.24   \n",
       "4              12.83            12.89  ...            11.67            11.78   \n",
       "..               ...              ...  ...              ...              ...   \n",
       "738             0.25             0.00  ...             0.00             0.00   \n",
       "739             0.00             0.00  ...             0.00             0.00   \n",
       "740             0.00             0.00  ...             0.00             0.00   \n",
       "741             0.00             0.00  ...             0.00             0.00   \n",
       "742             0.00             0.00  ...             0.00             0.00   \n",
       "\n",
       "     TCGA-AG-A01W-01  TCGA-AG-3726-01  TCGA-AG-3605-01  TCGA-AG-3584-01  \\\n",
       "0               1.29             1.02             0.16             2.09   \n",
       "1               5.40             4.45             4.61             4.54   \n",
       "2              14.70            14.39            14.73            14.50   \n",
       "3               4.08             4.73             3.58             4.43   \n",
       "4              12.56            12.09            12.38            12.22   \n",
       "..               ...              ...              ...              ...   \n",
       "738             0.00             0.00             0.00             0.00   \n",
       "739             0.00             0.00             0.00             0.00   \n",
       "740             0.00             0.00             0.00             0.00   \n",
       "741             0.00             0.00             0.00             0.00   \n",
       "742             0.00             0.00             0.00             0.00   \n",
       "\n",
       "     TCGA-AG-3599-01  TCGA-AG-3583-01  TCGA-AG-3598-01  TCGA-AG-3586-01  \n",
       "0               1.59             0.74             1.56             1.35  \n",
       "1               4.91             4.38             4.75             4.73  \n",
       "2              14.13            14.04            14.50            13.98  \n",
       "3               4.89             4.55             5.42             3.96  \n",
       "4              12.47            12.89            12.58            12.72  \n",
       "..               ...              ...              ...              ...  \n",
       "738             0.00             0.00             0.00             0.00  \n",
       "739             0.00             0.00             0.00             0.00  \n",
       "740             0.00             0.00             0.00             0.00  \n",
       "741             0.00             0.00             0.00             0.00  \n",
       "742             0.00             0.00             0.00             0.00  \n",
       "\n",
       "[743 rows x 10825 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-a0ed4d67ab35>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mexon_counts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchrom\u001b[0m \u001b[0;34m==\u001b[0m\u001b[0;34m'17'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'PRPF8'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.conda/envs/thesis/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5137\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5138\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5139\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5140\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5141\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'gene'"
     ]
    }
   ],
   "source": [
    "exon_counts[(exon_counts.chrom =='17') & (exon_counts.gene == 'PRPF8')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-45-9c13fa3f8aff>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-45-9c13fa3f8aff>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    exon_counts[(exon_counts.chrom =='17') & (exon_counts.chromEnd >= ) & (exon_counts.chromStart <=) & (exon_counts.strand ==-1)]\u001b[0m\n\u001b[0m                                                                      ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "exon_counts[(exon_counts.chrom =='17') & (exon_counts.chromEnd >= ) & (exon_counts.chromStart <=) & (exon_counts.strand ==-1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T21:37:48.346942Z",
     "start_time": "2020-10-25T21:37:48.234006Z"
    }
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'chromosome_name'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-51-005bafcbbdb6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchromosome_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.conda/envs/thesis/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5137\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5138\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5139\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5140\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5141\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'chromosome_name'"
     ]
    }
   ],
   "source": [
    "len(exon_counts[exon_counts.chromosome_name.isna()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map miRNA to counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:22:01.482945Z",
     "start_time": "2020-10-25T14:22:01.013851Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576707</td>\n",
       "      <td>1578261</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0000063</td>\n",
       "      <td>let-7</td>\n",
       "      <td>12.820202</td>\n",
       "      <td>...</td>\n",
       "      <td>12.829037</td>\n",
       "      <td>12.382240</td>\n",
       "      <td>13.638682</td>\n",
       "      <td>14.190721</td>\n",
       "      <td>14.086433</td>\n",
       "      <td>13.729646</td>\n",
       "      <td>13.115391</td>\n",
       "      <td>13.531265</td>\n",
       "      <td>13.564590</td>\n",
       "      <td>14.094524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>let-7</td>\n",
       "      <td>4.104356</td>\n",
       "      <td>...</td>\n",
       "      <td>4.210218</td>\n",
       "      <td>3.770706</td>\n",
       "      <td>4.646930</td>\n",
       "      <td>4.295710</td>\n",
       "      <td>5.294043</td>\n",
       "      <td>5.520265</td>\n",
       "      <td>4.097683</td>\n",
       "      <td>4.664511</td>\n",
       "      <td>4.351934</td>\n",
       "      <td>4.063494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>let-7</td>\n",
       "      <td>4.104356</td>\n",
       "      <td>...</td>\n",
       "      <td>4.210218</td>\n",
       "      <td>3.770706</td>\n",
       "      <td>4.646930</td>\n",
       "      <td>4.295710</td>\n",
       "      <td>5.294043</td>\n",
       "      <td>5.520265</td>\n",
       "      <td>4.097683</td>\n",
       "      <td>4.664511</td>\n",
       "      <td>4.351934</td>\n",
       "      <td>4.063494</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end  \\\n",
       "0      0  hsa-let-7a-2-3p  ENST00000576537         1578440       1578466   \n",
       "1      1  hsa-let-7a-2-3p  ENST00000576537         1578561       1578581   \n",
       "2      2    hsa-let-7b-5p  ENST00000576537         1576707       1578261   \n",
       "3      3    hsa-let-7b-3p  ENST00000576537         1578446       1578467   \n",
       "4      4    hsa-let-7b-3p  ENST00000576537         1576625       1576643   \n",
       "\n",
       "  chromosome_name  strand    mature_acc mirna_family  TCGA-KN-8426-01  ...  \\\n",
       "0              17      -1  MIMAT0010195        let-7         2.893871  ...   \n",
       "1              17      -1  MIMAT0010195        let-7         2.893871  ...   \n",
       "2              17      -1  MIMAT0000063        let-7        12.820202  ...   \n",
       "3              17      -1  MIMAT0004482        let-7         4.104356  ...   \n",
       "4              17      -1  MIMAT0004482        let-7         4.104356  ...   \n",
       "\n",
       "   TCGA-KO-8408-01  TCGA-KM-8438-01  TCGA-KL-8323-01  TCGA-KN-8429-11  \\\n",
       "0         2.829471         2.015161         2.594806         2.644861   \n",
       "1         2.829471         2.015161         2.594806         2.644861   \n",
       "2        12.829037        12.382240        13.638682        14.190721   \n",
       "3         4.210218         3.770706         4.646930         4.295710   \n",
       "4         4.210218         3.770706         4.646930         4.295710   \n",
       "\n",
       "   TCGA-KO-8413-01  TCGA-KO-8411-01  TCGA-KL-8332-11  TCGA-KL-8335-01  \\\n",
       "0         1.786991         1.854876         2.824365         2.395032   \n",
       "1         1.786991         1.854876         2.824365         2.395032   \n",
       "2        14.086433        13.729646        13.115391        13.531265   \n",
       "3         5.294043         5.520265         4.097683         4.664511   \n",
       "4         5.294043         5.520265         4.097683         4.664511   \n",
       "\n",
       "   TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "0         2.791374         2.819371  \n",
       "1         2.791374         2.819371  \n",
       "2        13.564590        14.094524  \n",
       "3         4.351934         4.063494  \n",
       "4         4.351934         4.063494  \n",
       "\n",
       "[5 rows x 98 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get subset counts of bs and merge with df_gene\n",
    "counts = bs[['bs_id', 'miRNA', 'mRNA', 'chrom_bs_start', 'chrom_bs_end']].copy()\n",
    "counts = counts.merge(df_gene[['ensembl_transcript_id','chromosome_name','strand']], left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "\n",
    "#mapping mature ID to miRNA family\n",
    "trans = pd.read_feather(path/'mature2families.feather')\n",
    "counts = counts.merge(trans, left_on='miRNA',right_on='mature_name', how='left')\n",
    "\n",
    "#alternative 0: dont join with mirnaids\n",
    "#counts.drop(['ensembl_transcript_id', 'mature_name'], axis=1, inplace=True)\n",
    "\n",
    "#! alternative 1: for all caseids at the same time !\n",
    "counts = counts.merge(mirna_counts, left_on='mature_acc', right_on='sample', how='left')\n",
    "counts.drop(['ensembl_transcript_id', 'mature_name', 'sample'], axis=1, inplace=True)\n",
    "\n",
    "# alternative 2: for 1 caseids at a time\n",
    "for caseid in []:#[caseids[0]]:\n",
    "    counts = counts.merge(mirna_counts[['sample',caseid]], left_on='mature_acc', right_on='sample', how='left')\n",
    "    counts.drop(['ensembl_transcript_id', 'mature_name', 'sample'], axis=1, inplace=True)\n",
    "\n",
    "#counts[['chromosome_name','strand','miRNA','mRNA']] = counts[['chromosome_name','strand','miRNA','mRNA']].astype('str')#.astype('category')\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T14:29:43.996902Z",
     "start_time": "2020-10-23T14:29:43.876970Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "658"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#TODO put in presentation, analysis mirna, how many in xena data\n",
    "#TODO find out : if na should i use it as 0????\n",
    "counts['TCGA-KN-8426-01'].isna().sum() #for caseid TCGA-KN-8426-01: 56035 bs are thrown away \n",
    "len(list(counts[counts['TCGA-KN-8426-01'].isna()].mature_acc.unique())) #we don't have the counts for 1948 mirnas in bs\n",
    "len(list(counts[~ counts['TCGA-KN-8426-01'].isna()].mature_acc.unique())) #we have counts for 658 mirnas\n",
    "\n",
    "len(list(mirna_counts['sample'].unique())) #1917 = absolute nr mirnas from TCGA-KN-8426-01\n",
    "len(list(mirna_counts[~ mirna_counts['TCGA-KN-8426-01'].isna()]['sample'].unique())) #1236 = not na nr mirnas from TCGA-KN-8426-01"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map exons to counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:23:02.738089Z",
     "start_time": "2020-10-25T14:22:50.007963Z"
    }
   },
   "outputs": [],
   "source": [
    "#for a large data set, you will likely get a significant speed increase by creating an index for any column(s) used in the join condition. \n",
    "#https://stackoverflow.com/questions/30627968/merge-pandas-dataframes-where-one-value-is-between-two-others\n",
    "#instead use SQL\n",
    "conn = sqlite3.connect('db.db')#':memory:') #Make the database in memory\n",
    "c = conn.cursor()\n",
    "\n",
    "#write the tables\n",
    "counts.to_sql('counts', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "exon_counts.to_sql('exon_counts', conn, index=True, index_label='exon_id', dtype={caseid : 'INTEGER' for caseid in caseids}.update({\"chromStart\": 'INTEGER', \"chromEnd\": 'INTEGER'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T14:54:28.603611Z",
     "start_time": "2020-10-23T14:54:21.994290Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576707</td>\n",
       "      <td>1578261</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0000063</td>\n",
       "      <td>let-7</td>\n",
       "      <td>12.820202</td>\n",
       "      <td>...</td>\n",
       "      <td>12.829037</td>\n",
       "      <td>12.382240</td>\n",
       "      <td>13.638682</td>\n",
       "      <td>14.190721</td>\n",
       "      <td>14.086433</td>\n",
       "      <td>13.729646</td>\n",
       "      <td>13.115391</td>\n",
       "      <td>13.531265</td>\n",
       "      <td>13.564590</td>\n",
       "      <td>14.094524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>let-7</td>\n",
       "      <td>4.104356</td>\n",
       "      <td>...</td>\n",
       "      <td>4.210218</td>\n",
       "      <td>3.770706</td>\n",
       "      <td>4.646930</td>\n",
       "      <td>4.295710</td>\n",
       "      <td>5.294043</td>\n",
       "      <td>5.520265</td>\n",
       "      <td>4.097683</td>\n",
       "      <td>4.664511</td>\n",
       "      <td>4.351934</td>\n",
       "      <td>4.063494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>let-7</td>\n",
       "      <td>4.104356</td>\n",
       "      <td>...</td>\n",
       "      <td>4.210218</td>\n",
       "      <td>3.770706</td>\n",
       "      <td>4.646930</td>\n",
       "      <td>4.295710</td>\n",
       "      <td>5.294043</td>\n",
       "      <td>5.520265</td>\n",
       "      <td>4.097683</td>\n",
       "      <td>4.664511</td>\n",
       "      <td>4.351934</td>\n",
       "      <td>4.063494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73188</th>\n",
       "      <td>73188</td>\n",
       "      <td>hsa-miR-12128</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38479688</td>\n",
       "      <td>38479728</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049022</td>\n",
       "      <td>hsa-miR-12128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73189</th>\n",
       "      <td>73189</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38481326</td>\n",
       "      <td>38481371</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049023</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73190</th>\n",
       "      <td>73190</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38479605</td>\n",
       "      <td>38479631</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049023</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73191</th>\n",
       "      <td>73191</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38479693</td>\n",
       "      <td>38479708</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049023</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73192</th>\n",
       "      <td>73192</td>\n",
       "      <td>hsa-miR-12133</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38482754</td>\n",
       "      <td>38482772</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049027</td>\n",
       "      <td>hsa-miR-12133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73193 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end  \\\n",
       "0          0  hsa-let-7a-2-3p  ENST00000576537         1578440       1578466   \n",
       "1          1  hsa-let-7a-2-3p  ENST00000576537         1578561       1578581   \n",
       "2          2    hsa-let-7b-5p  ENST00000576537         1576707       1578261   \n",
       "3          3    hsa-let-7b-3p  ENST00000576537         1578446       1578467   \n",
       "4          4    hsa-let-7b-3p  ENST00000576537         1576625       1576643   \n",
       "...      ...              ...              ...             ...           ...   \n",
       "73188  73188    hsa-miR-12128  ENST00000216014        38479688      38479728   \n",
       "73189  73189    hsa-miR-12129  ENST00000216014        38481326      38481371   \n",
       "73190  73190    hsa-miR-12129  ENST00000216014        38479605      38479631   \n",
       "73191  73191    hsa-miR-12129  ENST00000216014        38479693      38479708   \n",
       "73192  73192    hsa-miR-12133  ENST00000216014        38482754      38482772   \n",
       "\n",
       "      chromosome_name  strand    mature_acc   mirna_family  TCGA-KN-8426-01  \\\n",
       "0                  17      -1  MIMAT0010195          let-7         2.893871   \n",
       "1                  17      -1  MIMAT0010195          let-7         2.893871   \n",
       "2                  17      -1  MIMAT0000063          let-7        12.820202   \n",
       "3                  17      -1  MIMAT0004482          let-7         4.104356   \n",
       "4                  17      -1  MIMAT0004482          let-7         4.104356   \n",
       "...               ...     ...           ...            ...              ...   \n",
       "73188              22       1  MIMAT0049022  hsa-miR-12128              NaN   \n",
       "73189              22       1  MIMAT0049023  hsa-miR-12129              NaN   \n",
       "73190              22       1  MIMAT0049023  hsa-miR-12129              NaN   \n",
       "73191              22       1  MIMAT0049023  hsa-miR-12129              NaN   \n",
       "73192              22       1  MIMAT0049027  hsa-miR-12133              NaN   \n",
       "\n",
       "       ...  TCGA-KO-8408-01  TCGA-KM-8438-01  TCGA-KL-8323-01  \\\n",
       "0      ...         2.829471         2.015161         2.594806   \n",
       "1      ...         2.829471         2.015161         2.594806   \n",
       "2      ...        12.829037        12.382240        13.638682   \n",
       "3      ...         4.210218         3.770706         4.646930   \n",
       "4      ...         4.210218         3.770706         4.646930   \n",
       "...    ...              ...              ...              ...   \n",
       "73188  ...              NaN              NaN              NaN   \n",
       "73189  ...              NaN              NaN              NaN   \n",
       "73190  ...              NaN              NaN              NaN   \n",
       "73191  ...              NaN              NaN              NaN   \n",
       "73192  ...              NaN              NaN              NaN   \n",
       "\n",
       "       TCGA-KN-8429-11  TCGA-KO-8413-01  TCGA-KO-8411-01  TCGA-KL-8332-11  \\\n",
       "0             2.644861         1.786991         1.854876         2.824365   \n",
       "1             2.644861         1.786991         1.854876         2.824365   \n",
       "2            14.190721        14.086433        13.729646        13.115391   \n",
       "3             4.295710         5.294043         5.520265         4.097683   \n",
       "4             4.295710         5.294043         5.520265         4.097683   \n",
       "...                ...              ...              ...              ...   \n",
       "73188              NaN              NaN              NaN              NaN   \n",
       "73189              NaN              NaN              NaN              NaN   \n",
       "73190              NaN              NaN              NaN              NaN   \n",
       "73191              NaN              NaN              NaN              NaN   \n",
       "73192              NaN              NaN              NaN              NaN   \n",
       "\n",
       "       TCGA-KL-8335-01  TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "0             2.395032         2.791374         2.819371  \n",
       "1             2.395032         2.791374         2.819371  \n",
       "2            13.531265        13.564590        14.094524  \n",
       "3             4.664511         4.351934         4.063494  \n",
       "4             4.664511         4.351934         4.063494  \n",
       "...                ...              ...              ...  \n",
       "73188              NaN              NaN              NaN  \n",
       "73189              NaN              NaN              NaN  \n",
       "73190              NaN              NaN              NaN  \n",
       "73191              NaN              NaN              NaN  \n",
       "73192              NaN              NaN              NaN  \n",
       "\n",
       "[73193 rows x 98 columns]"
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print sqlite3 table \n",
    "qry = '''\n",
    "    select * from counts\n",
    "    '''\n",
    "c.execute(qry)\n",
    "conn.commit()\n",
    "c.fetchall()\n",
    "pd.read_sql_query(qry,conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:22:28.922131Z",
     "start_time": "2020-10-25T14:22:28.509946Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#DELETE table from sqlite3 DB\n",
    "c.execute(\"DROP TABLE counts;\")\n",
    "conn.commit()\n",
    "c.execute(\"DROP TABLE exon_counts;\")\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-26T16:05:50.874511Z",
     "start_time": "2020-10-26T16:05:50.672629Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>2.829471</td>\n",
       "      <td>2.015161</td>\n",
       "      <td>2.594806</td>\n",
       "      <td>2.644861</td>\n",
       "      <td>1.786991</td>\n",
       "      <td>1.854876</td>\n",
       "      <td>2.824365</td>\n",
       "      <td>2.395032</td>\n",
       "      <td>2.791374</td>\n",
       "      <td>2.819371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>let-7</td>\n",
       "      <td>4.104356</td>\n",
       "      <td>...</td>\n",
       "      <td>4.210218</td>\n",
       "      <td>3.770706</td>\n",
       "      <td>4.646930</td>\n",
       "      <td>4.295710</td>\n",
       "      <td>5.294043</td>\n",
       "      <td>5.520265</td>\n",
       "      <td>4.097683</td>\n",
       "      <td>4.664511</td>\n",
       "      <td>4.351934</td>\n",
       "      <td>4.063494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>hsa-let-7c-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578453</td>\n",
       "      <td>1578471</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0026472</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.695914</td>\n",
       "      <td>...</td>\n",
       "      <td>2.714853</td>\n",
       "      <td>2.455017</td>\n",
       "      <td>2.155998</td>\n",
       "      <td>3.673442</td>\n",
       "      <td>2.676288</td>\n",
       "      <td>3.043974</td>\n",
       "      <td>4.407776</td>\n",
       "      <td>2.500585</td>\n",
       "      <td>3.844001</td>\n",
       "      <td>2.977714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>hsa-let-7d-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578473</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004484</td>\n",
       "      <td>let-7</td>\n",
       "      <td>7.990090</td>\n",
       "      <td>...</td>\n",
       "      <td>8.056507</td>\n",
       "      <td>8.939957</td>\n",
       "      <td>8.391837</td>\n",
       "      <td>7.818937</td>\n",
       "      <td>7.722212</td>\n",
       "      <td>8.106506</td>\n",
       "      <td>7.446075</td>\n",
       "      <td>8.217851</td>\n",
       "      <td>8.154153</td>\n",
       "      <td>8.335863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578579</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004485</td>\n",
       "      <td>let-7</td>\n",
       "      <td>3.147645</td>\n",
       "      <td>...</td>\n",
       "      <td>2.523778</td>\n",
       "      <td>3.836892</td>\n",
       "      <td>2.846723</td>\n",
       "      <td>4.056371</td>\n",
       "      <td>2.882109</td>\n",
       "      <td>3.414638</td>\n",
       "      <td>3.685095</td>\n",
       "      <td>2.885547</td>\n",
       "      <td>3.698584</td>\n",
       "      <td>3.274872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384</th>\n",
       "      <td>3384</td>\n",
       "      <td>hsa-miR-12124</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578599</td>\n",
       "      <td>1578616</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0049018</td>\n",
       "      <td>hsa-miR-12124</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3385</th>\n",
       "      <td>3385</td>\n",
       "      <td>hsa-miR-12124</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578488</td>\n",
       "      <td>1578508</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0049018</td>\n",
       "      <td>hsa-miR-12124</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3388</th>\n",
       "      <td>3388</td>\n",
       "      <td>hsa-miR-12126</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578447</td>\n",
       "      <td>1578480</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0049020</td>\n",
       "      <td>hsa-miR-12126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3390</th>\n",
       "      <td>3390</td>\n",
       "      <td>hsa-miR-12128</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578517</td>\n",
       "      <td>1578566</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0049022</td>\n",
       "      <td>hsa-miR-12128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3393</th>\n",
       "      <td>3393</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578564</td>\n",
       "      <td>1578616</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0049023</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1127 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end  \\\n",
       "1         1  hsa-let-7a-2-3p  ENST00000576537         1578561       1578581   \n",
       "3         3    hsa-let-7b-3p  ENST00000576537         1578446       1578467   \n",
       "6         6    hsa-let-7c-3p  ENST00000576537         1578453       1578471   \n",
       "9         9    hsa-let-7d-3p  ENST00000576537         1578446       1578473   \n",
       "12       12    hsa-let-7e-3p  ENST00000576537         1578561       1578579   \n",
       "...     ...              ...              ...             ...           ...   \n",
       "3384   3384    hsa-miR-12124  ENST00000576537         1578599       1578616   \n",
       "3385   3385    hsa-miR-12124  ENST00000576537         1578488       1578508   \n",
       "3388   3388    hsa-miR-12126  ENST00000576537         1578447       1578480   \n",
       "3390   3390    hsa-miR-12128  ENST00000576537         1578517       1578566   \n",
       "3393   3393    hsa-miR-12129  ENST00000576537         1578564       1578616   \n",
       "\n",
       "     chromosome_name  strand    mature_acc   mirna_family  TCGA-KN-8426-01  \\\n",
       "1                 17      -1  MIMAT0010195          let-7         2.893871   \n",
       "3                 17      -1  MIMAT0004482          let-7         4.104356   \n",
       "6                 17      -1  MIMAT0026472          let-7         2.695914   \n",
       "9                 17      -1  MIMAT0004484          let-7         7.990090   \n",
       "12                17      -1  MIMAT0004485          let-7         3.147645   \n",
       "...              ...     ...           ...            ...              ...   \n",
       "3384              17      -1  MIMAT0049018  hsa-miR-12124              NaN   \n",
       "3385              17      -1  MIMAT0049018  hsa-miR-12124              NaN   \n",
       "3388              17      -1  MIMAT0049020  hsa-miR-12126              NaN   \n",
       "3390              17      -1  MIMAT0049022  hsa-miR-12128              NaN   \n",
       "3393              17      -1  MIMAT0049023  hsa-miR-12129              NaN   \n",
       "\n",
       "      ...  TCGA-KO-8408-01  TCGA-KM-8438-01  TCGA-KL-8323-01  TCGA-KN-8429-11  \\\n",
       "1     ...         2.829471         2.015161         2.594806         2.644861   \n",
       "3     ...         4.210218         3.770706         4.646930         4.295710   \n",
       "6     ...         2.714853         2.455017         2.155998         3.673442   \n",
       "9     ...         8.056507         8.939957         8.391837         7.818937   \n",
       "12    ...         2.523778         3.836892         2.846723         4.056371   \n",
       "...   ...              ...              ...              ...              ...   \n",
       "3384  ...              NaN              NaN              NaN              NaN   \n",
       "3385  ...              NaN              NaN              NaN              NaN   \n",
       "3388  ...              NaN              NaN              NaN              NaN   \n",
       "3390  ...              NaN              NaN              NaN              NaN   \n",
       "3393  ...              NaN              NaN              NaN              NaN   \n",
       "\n",
       "      TCGA-KO-8413-01  TCGA-KO-8411-01  TCGA-KL-8332-11  TCGA-KL-8335-01  \\\n",
       "1            1.786991         1.854876         2.824365         2.395032   \n",
       "3            5.294043         5.520265         4.097683         4.664511   \n",
       "6            2.676288         3.043974         4.407776         2.500585   \n",
       "9            7.722212         8.106506         7.446075         8.217851   \n",
       "12           2.882109         3.414638         3.685095         2.885547   \n",
       "...               ...              ...              ...              ...   \n",
       "3384              NaN              NaN              NaN              NaN   \n",
       "3385              NaN              NaN              NaN              NaN   \n",
       "3388              NaN              NaN              NaN              NaN   \n",
       "3390              NaN              NaN              NaN              NaN   \n",
       "3393              NaN              NaN              NaN              NaN   \n",
       "\n",
       "      TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "1            2.791374         2.819371  \n",
       "3            4.351934         4.063494  \n",
       "6            3.844001         2.977714  \n",
       "9            8.154153         8.335863  \n",
       "12           3.698584         3.274872  \n",
       "...               ...              ...  \n",
       "3384              NaN              NaN  \n",
       "3385              NaN              NaN  \n",
       "3388              NaN              NaN  \n",
       "3390              NaN              NaN  \n",
       "3393              NaN              NaN  \n",
       "\n",
       "[1127 rows x 98 columns]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts[(counts.chrom_bs_start > 1578445) & (counts.chrom_bs_end < 1578634)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T19:43:15.504195Z",
     "start_time": "2020-10-23T19:43:15.416229Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>TCGA-KL-8346-01</th>\n",
       "      <th>TCGA-KN-8422-01</th>\n",
       "      <th>TCGA-KN-8431-11</th>\n",
       "      <th>TCGA-KN-8430-11</th>\n",
       "      <th>TCGA-KM-8440-01</th>\n",
       "      <th>TCGA-KO-8414-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KM-8639-01</th>\n",
       "      <th>TCGA-KO-8415-11</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8443-01</th>\n",
       "      <th>TCGA-KM-8442-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45710</th>\n",
       "      <td>4.922</td>\n",
       "      <td>2.6972</td>\n",
       "      <td>2.0686</td>\n",
       "      <td>4.7578</td>\n",
       "      <td>4.705</td>\n",
       "      <td>4.5074</td>\n",
       "      <td>3.3096</td>\n",
       "      <td>4.5863</td>\n",
       "      <td>4.6372</td>\n",
       "      <td>4.7683</td>\n",
       "      <td>...</td>\n",
       "      <td>2.5759</td>\n",
       "      <td>4.1899</td>\n",
       "      <td>4.4041</td>\n",
       "      <td>3.6642</td>\n",
       "      <td>1.3794</td>\n",
       "      <td>PRPF8</td>\n",
       "      <td>chr17</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578633</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TCGA-KN-8419-01  TCGA-KL-8346-01  TCGA-KN-8422-01  TCGA-KN-8431-11  \\\n",
       "45710            4.922           2.6972           2.0686           4.7578   \n",
       "\n",
       "       TCGA-KN-8430-11  TCGA-KM-8440-01  TCGA-KO-8414-01  TCGA-KL-8323-01  \\\n",
       "45710            4.705           4.5074           3.3096           4.5863   \n",
       "\n",
       "       TCGA-KM-8639-01  TCGA-KO-8415-11  ...  TCGA-KO-8408-01  \\\n",
       "45710           4.6372           4.7683  ...           2.5759   \n",
       "\n",
       "       TCGA-KM-8443-01  TCGA-KM-8442-01  TCGA-KL-8332-11  TCGA-KL-8327-01  \\\n",
       "45710           4.1899           4.4041           3.6642           1.3794   \n",
       "\n",
       "        gene  chrom  chromStart  chromEnd  strand  \n",
       "45710  PRPF8  chr17     1578446   1578633      -1  \n",
       "\n",
       "[1 rows x 96 columns]"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts[(exon_counts.chromStart > 1578445) & (exon_counts.chromEnd < 1578634)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T14:10:08.661651Z",
     "start_time": "2020-10-25T14:09:05.000Z"
    }
   },
   "outputs": [],
   "source": [
    "#TODO change column name case_id to column name case_id_mirna so it doesnt overlap with exon\n",
    "#instead of joining like i do & repeating the exon info -> only add exon_id + pivot / addup per exon + add exon counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2020-10-25T14:13:28.400Z"
    }
   },
   "outputs": [],
   "source": [
    "#TODO right now I hardcoded TCGA-KN-8419-01!!\n",
    "#exon_counts.chrom, exon_counts.chromStart, exon_counts.chromEnd, exon_counts.strand, exon_counts.'TCGA-KN-8419-01', exon_counts.exon_id, counts.bs_id, counts.miRNA, counts.mRNA, counts.chrom_bs_start, counts.chrom_bs_end\n",
    "\n",
    "qry = '''\n",
    "    CREATE TABLE n_counts AS\n",
    "    select\n",
    "        *\n",
    "    from\n",
    "        counts inner join (\n",
    "        select exon_counts.chrom, exon_counts.chromStart, exon_counts.chromEnd, exon_counts.strand, exon_counts.exon_id\n",
    "        from exon_counts )\n",
    "        on\n",
    "        counts.chromosome_name = exon_counts.chrom and counts.strand = exon_counts.strand and counts.chrom_bs_start >= exon_counts.chromStart and counts.chrom_bs_end <= exon_counts.chromEnd\n",
    "    '''\n",
    "c.execute(qry)\n",
    "conn.commit()\n",
    "print('n_counts done.')\n",
    "qry = '''\n",
    "    select * from n_counts\n",
    "    '''\n",
    "c.execute(qry)\n",
    "conn.commit()\n",
    "c.fetchall()\n",
    "#n_counts = pd.read_sql_query(qry,conn) #TODO somehow empty idk why\n",
    "\n",
    "#todo google pandas 3D\n",
    "#look at old code idk why\n",
    "\n",
    "#only returns bs with minimum 1 exon_id cause inner join (else left join)\n",
    "#automatically filters out all binding sites that have more than 1 exon_id\n",
    "#if u dont want that use outer join -> actually wrong, it uses the first instead of leaving it out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-25T12:39:09.497431Z",
     "start_time": "2020-10-25T12:39:09.067541Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>mature_acc</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8443-01</th>\n",
       "      <th>TCGA-KM-8442-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0010195</td>\n",
       "      <td>let-7</td>\n",
       "      <td>2.893871</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576707</td>\n",
       "      <td>1578261</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0000063</td>\n",
       "      <td>let-7</td>\n",
       "      <td>12.820202</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>let-7</td>\n",
       "      <td>4.104356</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "      <td>17</td>\n",
       "      <td>-1</td>\n",
       "      <td>MIMAT0004482</td>\n",
       "      <td>let-7</td>\n",
       "      <td>4.104356</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73188</th>\n",
       "      <td>73188</td>\n",
       "      <td>hsa-miR-12128</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38479688</td>\n",
       "      <td>38479728</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049022</td>\n",
       "      <td>hsa-miR-12128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73189</th>\n",
       "      <td>73189</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38481326</td>\n",
       "      <td>38481371</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049023</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73190</th>\n",
       "      <td>73190</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38479605</td>\n",
       "      <td>38479631</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049023</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73191</th>\n",
       "      <td>73191</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38479693</td>\n",
       "      <td>38479708</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049023</td>\n",
       "      <td>hsa-miR-12129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73192</th>\n",
       "      <td>73192</td>\n",
       "      <td>hsa-miR-12133</td>\n",
       "      <td>ENST00000216014</td>\n",
       "      <td>38482754</td>\n",
       "      <td>38482772</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>MIMAT0049027</td>\n",
       "      <td>hsa-miR-12133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73193 rows × 195 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end  \\\n",
       "0          0  hsa-let-7a-2-3p  ENST00000576537         1578440       1578466   \n",
       "1          1  hsa-let-7a-2-3p  ENST00000576537         1578561       1578581   \n",
       "2          2    hsa-let-7b-5p  ENST00000576537         1576707       1578261   \n",
       "3          3    hsa-let-7b-3p  ENST00000576537         1578446       1578467   \n",
       "4          4    hsa-let-7b-3p  ENST00000576537         1576625       1576643   \n",
       "...      ...              ...              ...             ...           ...   \n",
       "73188  73188    hsa-miR-12128  ENST00000216014        38479688      38479728   \n",
       "73189  73189    hsa-miR-12129  ENST00000216014        38481326      38481371   \n",
       "73190  73190    hsa-miR-12129  ENST00000216014        38479605      38479631   \n",
       "73191  73191    hsa-miR-12129  ENST00000216014        38479693      38479708   \n",
       "73192  73192    hsa-miR-12133  ENST00000216014        38482754      38482772   \n",
       "\n",
       "      chromosome_name  strand    mature_acc   mirna_family  TCGA-KN-8426-01  \\\n",
       "0                  17      -1  MIMAT0010195          let-7         2.893871   \n",
       "1                  17      -1  MIMAT0010195          let-7         2.893871   \n",
       "2                  17      -1  MIMAT0000063          let-7        12.820202   \n",
       "3                  17      -1  MIMAT0004482          let-7         4.104356   \n",
       "4                  17      -1  MIMAT0004482          let-7         4.104356   \n",
       "...               ...     ...           ...            ...              ...   \n",
       "73188              22       1  MIMAT0049022  hsa-miR-12128              NaN   \n",
       "73189              22       1  MIMAT0049023  hsa-miR-12129              NaN   \n",
       "73190              22       1  MIMAT0049023  hsa-miR-12129              NaN   \n",
       "73191              22       1  MIMAT0049023  hsa-miR-12129              NaN   \n",
       "73192              22       1  MIMAT0049027  hsa-miR-12133              NaN   \n",
       "\n",
       "       ...  TCGA-KO-8408-01  TCGA-KM-8443-01  TCGA-KM-8442-01  \\\n",
       "0      ...             None             None             None   \n",
       "1      ...             None             None             None   \n",
       "2      ...             None             None             None   \n",
       "3      ...             None             None             None   \n",
       "4      ...             None             None             None   \n",
       "...    ...              ...              ...              ...   \n",
       "73188  ...             None             None             None   \n",
       "73189  ...             None             None             None   \n",
       "73190  ...             None             None             None   \n",
       "73191  ...             None             None             None   \n",
       "73192  ...             None             None             None   \n",
       "\n",
       "       TCGA-KL-8332-11  TCGA-KL-8327-01  gene  chrom  chromStart  chromEnd  \\\n",
       "0                 None             None  None   None        None      None   \n",
       "1                 None             None  None   None        None      None   \n",
       "2                 None             None  None   None        None      None   \n",
       "3                 None             None  None   None        None      None   \n",
       "4                 None             None  None   None        None      None   \n",
       "...                ...              ...   ...    ...         ...       ...   \n",
       "73188             None             None  None   None        None      None   \n",
       "73189             None             None  None   None        None      None   \n",
       "73190             None             None  None   None        None      None   \n",
       "73191             None             None  None   None        None      None   \n",
       "73192             None             None  None   None        None      None   \n",
       "\n",
       "       strand  \n",
       "0        None  \n",
       "1        None  \n",
       "2        None  \n",
       "3        None  \n",
       "4        None  \n",
       "...       ...  \n",
       "73188    None  \n",
       "73189    None  \n",
       "73190    None  \n",
       "73191    None  \n",
       "73192    None  \n",
       "\n",
       "[73193 rows x 195 columns]"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T19:09:00.074489Z",
     "start_time": "2020-10-23T18:55:59.395843Z"
    }
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-293-e616a8564058>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;31m#takes long! for 1 case id: 2 min\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[1;31m#exon_counts = all_exon_dfs['aba26e6b-f11c-4ae6-a0d6-85bdd2060e8f'] #df#all_exon_dfs[case_id]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 13\u001b[1;33m \u001b[0mcounts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'exon'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcounts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mis_in_exon\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     14\u001b[0m \u001b[0mbs_exons\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcounts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexons\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m','\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcounts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbs_id\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[0mbs_exons\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mbs_exons\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m{\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;34m\"exon_raw_count\"\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36mapply\u001b[1;34m(self, func, axis, raw, result_type, args, **kwds)\u001b[0m\n\u001b[0;32m   6876\u001b[0m             \u001b[0mkwds\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   6877\u001b[0m         )\n\u001b[1;32m-> 6878\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   6879\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   6880\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mapplymap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m->\u001b[0m \u001b[1;34m\"DataFrame\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\apply.py\u001b[0m in \u001b[0;36mget_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    184\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    185\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 186\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    187\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    188\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mapply_empty_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    311\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    312\u001b[0m         \u001b[1;31m# compute the result using the series generator\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 313\u001b[1;33m         \u001b[0mresults\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    314\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    315\u001b[0m         \u001b[1;31m# wrap results\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\apply.py\u001b[0m in \u001b[0;36mapply_series_generator\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    339\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    340\u001b[0m             \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mv\u001b[0m \u001b[1;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 341\u001b[1;33m                 \u001b[0mresults\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mv\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    342\u001b[0m                 \u001b[0mkeys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mv\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    343\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-293-e616a8564058>\u001b[0m in \u001b[0;36m<lambda>\u001b[1;34m(row)\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;31m#takes long! for 1 case id: 2 min\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[1;31m#exon_counts = all_exon_dfs['aba26e6b-f11c-4ae6-a0d6-85bdd2060e8f'] #df#all_exon_dfs[case_id]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 13\u001b[1;33m \u001b[0mcounts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'exon'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcounts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mis_in_exon\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     14\u001b[0m \u001b[0mbs_exons\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcounts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexons\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m','\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcounts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbs_id\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[0mbs_exons\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mbs_exons\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m{\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;34m\"exon_raw_count\"\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-293-e616a8564058>\u001b[0m in \u001b[0;36mis_in_exon\u001b[1;34m(row, exon_counts)\u001b[0m\n\u001b[0;32m      4\u001b[0m     \u001b[0mbs_start\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchrom_bs_start\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mbs_end\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchrom_bs_end\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 6\u001b[1;33m     \u001b[0mhere\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mexon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchrom\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mrow\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchromosome_name\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m&\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'strand'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mrow\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstrand\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m&\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchromStart\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mbs_start\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m&\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mexon_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchromEnd\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mbs_end\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      7\u001b[0m     \u001b[1;31m#print(here)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m     \u001b[0mcount_list\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhere\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'TCGA-KN-8419-01'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\common.py\u001b[0m in \u001b[0;36mnew_method\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     62\u001b[0m         \u001b[0mother\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mitem_from_zerodim\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mother\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 64\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mother\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     65\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     66\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mnew_method\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\__init__.py\u001b[0m in \u001b[0;36mwrapper\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m    527\u001b[0m         \u001b[0mrvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mextract_array\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mother\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mextract_numpy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    528\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 529\u001b[1;33m         \u001b[0mres_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcomparison_op\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    530\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    531\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0m_construct_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mres_values\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mres_name\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py\u001b[0m in \u001b[0;36mcomparison_op\u001b[1;34m(left, right, op)\u001b[0m\n\u001b[0;32m    245\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    246\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlvalues\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 247\u001b[1;33m         \u001b[0mres_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcomp_method_OBJECT_ARRAY\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mop\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrvalues\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    248\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    249\u001b[0m     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Miniconda3\\envs\\thesis\\lib\\site-packages\\pandas\\core\\ops\\array_ops.py\u001b[0m in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[1;34m(op, x, y)\u001b[0m\n\u001b[0;32m     56\u001b[0m     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     57\u001b[0m         \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlibops\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mscalar_compare\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 58\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     59\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     60\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#old\n",
    "#TODO takes fuuuuuckin long, instead join on strand, chromosome_name, start, end\n",
    "#merge bs with exon counts, real slow, TODO cython, 22.34 - lookup when\n",
    "def is_in_exon(row, exon_counts):\n",
    "    bs_start = row.chrom_bs_start\n",
    "    bs_end = row.chrom_bs_end\n",
    "    here = exon_counts.loc[(exon_counts.chrom == row.chromosome_name) & (exon_counts['strand'] == row.strand) & (exon_counts.chromStart == bs_start) & (exon_counts.chromEnd == bs_end)]\n",
    "    #print(here)\n",
    "    count_list = str(list(here['TCGA-KN-8419-01']))[1:-1]\n",
    "    return count_list if not count_list == '' else None\n",
    "\n",
    "#takes long! for 1 case id: 2 min \n",
    "#exon_counts = all_exon_dfs['aba26e6b-f11c-4ae6-a0d6-85bdd2060e8f'] #df#all_exon_dfs[case_id]\n",
    "counts['exon'] = counts.apply(lambda row: is_in_exon(row,exon_counts), axis=1)\n",
    "bs_exons = pd.DataFrame(counts.exons.str.split(',').tolist(),index=counts.bs_id)\n",
    "bs_exons = bs_exons.rename(columns={0: \"exon_raw_count\"})\n",
    "bs_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T00:31:04.522842Z",
     "start_time": "2020-10-13T00:31:04.432995Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#visualize amount exons by bindingsite | amount bindingsites by exon\n",
    "import sqlite3\n",
    "conn = sqlite3.connect(':memory:') #Make the database in memory\n",
    "\n",
    "#write the tables\n",
    "important_counts.to_sql('bs', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "new_exon_counts.to_sql('counts', conn, index=True, index_label='exon_id', dtype={\"exon_id\": 'INTEGER'})\n",
    "qry = '''\n",
    "    select\n",
    "        counts.chromosome_name, counts.chrom_exon_start, counts.chrom_exon_end, counts.strand, counts.raw_counts, counts.exon_id, bs.bs_id, bs.miRNA, bs.mRNA, bs.chrom_bs_start, bs.chrom_bs_end\n",
    "    from\n",
    "        bs left join counts on\n",
    "        bs.chromosome_name = counts.chromosome_name and bs.strand = counts.strand and bs.chrom_bs_start >= counts.chrom_exon_start and bs.chrom_bs_end <= counts.chrom_exon_end\n",
    "    '''\n",
    "bs_exons2 = pd.read_sql_query(qry, conn)\n",
    "bs_exons2.head()\n",
    "#bs_exons2.groupby('bs_id').count().raw_counts.hist()\n",
    "#bs_exons2.groupby('exon_id').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T15:30:24.499780Z",
     "start_time": "2020-10-23T15:30:24.425823Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, chromStart, chromEnd, strand, TCGA-KN-8419-01, exon_id, bs_id, miRNA, mRNA, chrom_bs_start, chrom_bs_end]\n",
       "Index: []"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Put all counts into one & pivot table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T14:36:05.887125Z",
     "start_time": "2020-10-23T14:36:05.763198Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>TCGA-KN-8426-01</th>\n",
       "      <th>TCGA-KN-8419-11</th>\n",
       "      <th>TCGA-KL-8332-01</th>\n",
       "      <th>TCGA-KN-8423-01</th>\n",
       "      <th>TCGA-KO-8405-01</th>\n",
       "      <th>TCGA-KL-8337-01</th>\n",
       "      <th>TCGA-KN-8424-11</th>\n",
       "      <th>TCGA-KN-8422-11</th>\n",
       "      <th>TCGA-KL-8334-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8438-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KN-8429-11</th>\n",
       "      <th>TCGA-KO-8413-01</th>\n",
       "      <th>TCGA-KO-8411-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8335-01</th>\n",
       "      <th>TCGA-KL-8339-11</th>\n",
       "      <th>TCGA-KO-8415-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIMAT0019868</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIMAT0019869</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1644</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.301479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIMAT0019860</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIMAT0019862</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIMAT0019864</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1644</td>\n",
       "      <td>0.1774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1638</td>\n",
       "      <td>0.4246</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2075</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample  TCGA-KN-8426-01  TCGA-KN-8419-11  TCGA-KL-8332-01  \\\n",
       "0  MIMAT0019868              NaN              NaN              NaN   \n",
       "1  MIMAT0019869              NaN           0.1644              NaN   \n",
       "2  MIMAT0019860              NaN              NaN              NaN   \n",
       "3  MIMAT0019862              NaN              NaN              NaN   \n",
       "4  MIMAT0019864              NaN           0.1644           0.1774   \n",
       "\n",
       "   TCGA-KN-8423-01  TCGA-KO-8405-01  TCGA-KL-8337-01  TCGA-KN-8424-11  \\\n",
       "0              NaN              NaN              NaN              NaN   \n",
       "1           0.1544              NaN              NaN              NaN   \n",
       "2              NaN              NaN              NaN              NaN   \n",
       "3           0.1544              NaN              NaN              NaN   \n",
       "4              NaN              NaN           0.1638           0.4246   \n",
       "\n",
       "   TCGA-KN-8422-11  TCGA-KL-8334-01  ...  TCGA-KO-8408-01  TCGA-KM-8438-01  \\\n",
       "0              NaN              NaN  ...              NaN              NaN   \n",
       "1         0.301479              NaN  ...              NaN              NaN   \n",
       "2              NaN              NaN  ...              NaN              NaN   \n",
       "3              NaN              NaN  ...              NaN              NaN   \n",
       "4              NaN           0.2075  ...              NaN           0.2143   \n",
       "\n",
       "   TCGA-KL-8323-01  TCGA-KN-8429-11  TCGA-KO-8413-01  TCGA-KO-8411-01  \\\n",
       "0              NaN              NaN              NaN              NaN   \n",
       "1              NaN              NaN              NaN              NaN   \n",
       "2              NaN              NaN              NaN              NaN   \n",
       "3              NaN              NaN              NaN              NaN   \n",
       "4              NaN              NaN              NaN              NaN   \n",
       "\n",
       "   TCGA-KL-8332-11  TCGA-KL-8335-01  TCGA-KL-8339-11  TCGA-KO-8415-01  \n",
       "0              NaN              NaN              NaN              NaN  \n",
       "1              NaN              NaN              NaN              NaN  \n",
       "2              NaN              NaN              NaN           0.4087  \n",
       "3           0.2042              NaN              NaN              NaN  \n",
       "4              NaN           0.1803              NaN              NaN  \n",
       "\n",
       "[5 rows x 90 columns]"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-23T15:24:42.060464Z",
     "start_time": "2020-10-23T15:24:41.970516Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-KN-8419-01</th>\n",
       "      <th>TCGA-KL-8346-01</th>\n",
       "      <th>TCGA-KN-8422-01</th>\n",
       "      <th>TCGA-KN-8431-11</th>\n",
       "      <th>TCGA-KN-8430-11</th>\n",
       "      <th>TCGA-KM-8440-01</th>\n",
       "      <th>TCGA-KO-8414-01</th>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <th>TCGA-KM-8639-01</th>\n",
       "      <th>TCGA-KO-8415-11</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-KO-8408-01</th>\n",
       "      <th>TCGA-KM-8443-01</th>\n",
       "      <th>TCGA-KM-8442-01</th>\n",
       "      <th>TCGA-KL-8332-11</th>\n",
       "      <th>TCGA-KL-8327-01</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.1156</td>\n",
       "      <td>1.7396</td>\n",
       "      <td>2.0641</td>\n",
       "      <td>3.1457</td>\n",
       "      <td>3.0436</td>\n",
       "      <td>2.2209</td>\n",
       "      <td>1.8213</td>\n",
       "      <td>1.9092</td>\n",
       "      <td>2.9139</td>\n",
       "      <td>3.1049</td>\n",
       "      <td>...</td>\n",
       "      <td>2.2369</td>\n",
       "      <td>2.4951</td>\n",
       "      <td>2.1010</td>\n",
       "      <td>3.0617</td>\n",
       "      <td>0.7424</td>\n",
       "      <td>ABHD14B,RP11-155D18.14,RP11-155D18.12</td>\n",
       "      <td>chr3</td>\n",
       "      <td>52007981</td>\n",
       "      <td>52008646</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>215901372</td>\n",
       "      <td>215901726</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.5290</td>\n",
       "      <td>0.7196</td>\n",
       "      <td>1.3094</td>\n",
       "      <td>2.3074</td>\n",
       "      <td>1.6894</td>\n",
       "      <td>1.5829</td>\n",
       "      <td>1.0835</td>\n",
       "      <td>1.9041</td>\n",
       "      <td>1.2131</td>\n",
       "      <td>1.7214</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6381</td>\n",
       "      <td>1.5227</td>\n",
       "      <td>1.7786</td>\n",
       "      <td>1.4157</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>SLC10A7</td>\n",
       "      <td>chr4</td>\n",
       "      <td>147214081</td>\n",
       "      <td>147214132</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.3839</td>\n",
       "      <td>3.8702</td>\n",
       "      <td>2.5373</td>\n",
       "      <td>2.3464</td>\n",
       "      <td>2.2976</td>\n",
       "      <td>3.0639</td>\n",
       "      <td>2.5666</td>\n",
       "      <td>2.7963</td>\n",
       "      <td>3.0875</td>\n",
       "      <td>2.0227</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0412</td>\n",
       "      <td>3.5838</td>\n",
       "      <td>3.7551</td>\n",
       "      <td>1.4157</td>\n",
       "      <td>1.8708</td>\n",
       "      <td>TNK2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>195599147</td>\n",
       "      <td>195599341</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.9974</td>\n",
       "      <td>0.6984</td>\n",
       "      <td>0.9011</td>\n",
       "      <td>0.6827</td>\n",
       "      <td>1.2425</td>\n",
       "      <td>0.9715</td>\n",
       "      <td>1.6701</td>\n",
       "      <td>1.3011</td>\n",
       "      <td>1.3712</td>\n",
       "      <td>1.5433</td>\n",
       "      <td>...</td>\n",
       "      <td>1.1410</td>\n",
       "      <td>1.2345</td>\n",
       "      <td>1.7403</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>LRRC37B</td>\n",
       "      <td>chr17</td>\n",
       "      <td>30351730</td>\n",
       "      <td>30351801</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   TCGA-KN-8419-01  TCGA-KL-8346-01  TCGA-KN-8422-01  TCGA-KN-8431-11  \\\n",
       "0           2.1156           1.7396           2.0641           3.1457   \n",
       "1           0.0000           0.0000           0.0000           0.0251   \n",
       "2           2.5290           0.7196           1.3094           2.3074   \n",
       "3           2.3839           3.8702           2.5373           2.3464   \n",
       "4           1.9974           0.6984           0.9011           0.6827   \n",
       "\n",
       "   TCGA-KN-8430-11  TCGA-KM-8440-01  TCGA-KO-8414-01  TCGA-KL-8323-01  \\\n",
       "0           3.0436           2.2209           1.8213           1.9092   \n",
       "1           0.0233           0.0000           0.0000           0.0000   \n",
       "2           1.6894           1.5829           1.0835           1.9041   \n",
       "3           2.2976           3.0639           2.5666           2.7963   \n",
       "4           1.2425           0.9715           1.6701           1.3011   \n",
       "\n",
       "   TCGA-KM-8639-01  TCGA-KO-8415-11  ...  TCGA-KO-8408-01  TCGA-KM-8443-01  \\\n",
       "0           2.9139           3.1049  ...           2.2369           2.4951   \n",
       "1           0.0000           0.0000  ...           0.0000           0.0000   \n",
       "2           1.2131           1.7214  ...           0.6381           1.5227   \n",
       "3           3.0875           2.0227  ...           3.0412           3.5838   \n",
       "4           1.3712           1.5433  ...           1.1410           1.2345   \n",
       "\n",
       "   TCGA-KM-8442-01  TCGA-KL-8332-11  TCGA-KL-8327-01  \\\n",
       "0           2.1010           3.0617           0.7424   \n",
       "1           0.0000           0.0000           0.0000   \n",
       "2           1.7786           1.4157           0.0000   \n",
       "3           3.7551           1.4157           1.8708   \n",
       "4           1.7403           0.0000           0.0000   \n",
       "\n",
       "                                    gene  chrom  chromStart   chromEnd  strand  \n",
       "0  ABHD14B,RP11-155D18.14,RP11-155D18.12   chr3    52007981   52008646      -1  \n",
       "1                                  USH2A   chr1   215901372  215901726      -1  \n",
       "2                                SLC10A7   chr4   147214081  147214132      -1  \n",
       "3                                   TNK2   chr3   195599147  195599341      -1  \n",
       "4                                LRRC37B  chr17    30351730   30351801       1  \n",
       "\n",
       "[5 rows x 96 columns]"
      ]
     },
     "execution_count": 279,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO put all different caseids sqlite tables together into one\n",
    "#INSERT INTO artists_backup \n",
    "#SELECT ArtistId, Name\n",
    "#FROM artists;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.071562Z",
     "start_time": "2020-10-18T22:50:38.700Z"
    }
   },
   "outputs": [],
   "source": [
    "#output: for each case_id for each bs_id: interesting miRNA expression + exon expression\n",
    "#unstack family counts from 2 columns to several columns (mirna_family=name,mirna_read_count=value)\n",
    "pivoted = bs_mirnas.copy()\n",
    "pivoted = pivoted.pivot_table(values='mirna_read_count', index=['exon_id', 'exon_raw_counts'], columns='mirna_family', aggfunc='sum', fill_value=0)\n",
    "pivoted#.head()\n",
    "#TODO debug does this do it right OR use sqllite3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO pivot in sqlite3 \n",
    "#https://modern-sql.com/use-case/pivot\n",
    "#https://stackoverflow.com/questions/1237068/how-to-pivot-in-sqlite-or-i-e-select-in-wide-format-a-table-stored-in-long-form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T01:30:28.905033Z",
     "start_time": "2020-10-13T01:30:28.803071Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_raw_counts</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>mirna_read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>let-7</td>\n",
       "      <td>55647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7d-5p</td>\n",
       "      <td>hsa-let-7d</td>\n",
       "      <td>let-7</td>\n",
       "      <td>3767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>let-7</td>\n",
       "      <td>11424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>let-7</td>\n",
       "      <td>11424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-15a-3p</td>\n",
       "      <td>hsa-mir-15a</td>\n",
       "      <td>mir-15</td>\n",
       "      <td>2258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12562</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-5692a</td>\n",
       "      <td>hsa-mir-5692a-2</td>\n",
       "      <td>mir-5692</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12563</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-548az-5p</td>\n",
       "      <td>hsa-mir-548az</td>\n",
       "      <td>mir-548</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12566</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-8062</td>\n",
       "      <td>hsa-mir-8062</td>\n",
       "      <td>hsa-miR-8062</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12571</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-8064</td>\n",
       "      <td>hsa-mir-8064</td>\n",
       "      <td>hsa-miR-8064</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12578</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1257</td>\n",
       "      <td>hsa-mir-1257</td>\n",
       "      <td>mir-1257</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2129 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_id  exon_raw_counts             miRNA         miRNA_ID  \\\n",
       "7       216595                0     hsa-let-7b-5p       hsa-let-7b   \n",
       "14      216595                0     hsa-let-7d-5p       hsa-let-7d   \n",
       "25      216595                0     hsa-let-7e-3p       hsa-let-7e   \n",
       "33      216595                0     hsa-let-7e-3p       hsa-let-7e   \n",
       "46      216595                0    hsa-miR-15a-3p      hsa-mir-15a   \n",
       "...        ...              ...               ...              ...   \n",
       "12562   216595                0     hsa-miR-5692a  hsa-mir-5692a-2   \n",
       "12563   216595                0  hsa-miR-548az-5p    hsa-mir-548az   \n",
       "12566   216595                0      hsa-miR-8062     hsa-mir-8062   \n",
       "12571   216595                0      hsa-miR-8064     hsa-mir-8064   \n",
       "12578   216595                0      hsa-miR-1257     hsa-mir-1257   \n",
       "\n",
       "       mirna_family  mirna_read_count  \n",
       "7             let-7             55647  \n",
       "14            let-7              3767  \n",
       "25            let-7             11424  \n",
       "33            let-7             11424  \n",
       "46           mir-15              2258  \n",
       "...             ...               ...  \n",
       "12562      mir-5692                 0  \n",
       "12563       mir-548                 0  \n",
       "12566  hsa-miR-8062                 0  \n",
       "12571  hsa-miR-8064                 0  \n",
       "12578      mir-1257                 0  \n",
       "\n",
       "[2129 rows x 6 columns]"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs_mirnas[bs_mirnas.exon_id == 216595]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T01:29:10.977928Z",
     "start_time": "2020-10-13T01:29:10.892977Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_raw_counts</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>mirna_read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3365</th>\n",
       "      <td>216603</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371</th>\n",
       "      <td>216596</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exon_id  exon_raw_counts            miRNA      miRNA_ID  \\\n",
       "3365   216603                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3368   216595                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3370   216595                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3371   216596                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "\n",
       "         mirna_family  mirna_read_count  \n",
       "3365  hsa-miR-1229-5p                 4  \n",
       "3368  hsa-miR-1229-5p                 4  \n",
       "3370  hsa-miR-1229-5p                 4  \n",
       "3371  hsa-miR-1229-5p                 4  "
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs_mirnas[bs_mirnas.mirna_family =='hsa-miR-1229-5p']\n",
    "\n",
    "# 1 right? count is added for exonid 216595, but do i already add it before ?\n",
    "# 2 also : exon_id + mirna double (3368 & 3370) wrong!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualize amount exons by bindingsite\n",
    "#bs_exons.groupby('bs_id').count().raw_counts.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:40:44.254071Z",
     "start_time": "2020-10-13T08:40:44.166104Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8, 1680) (8,)\n"
     ]
    }
   ],
   "source": [
    "#TODO check - is nr mirna downloaded also here? if not: mistake\n",
    "print(X.shape,y[0].shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Elastic Net Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.081556Z",
     "start_time": "2020-10-18T22:50:43.000Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.getsizeof(pivoted)#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:42:05.340886Z",
     "start_time": "2020-10-13T08:42:05.198969Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0. 0. 0. ... 0. 0. 0.]\n",
      "0.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Lena\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\linear_model\\_coordinate_descent.py:529: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 0.0, tolerance: 0.0\n",
      "  model = cd_fast.enet_coordinate_descent(\n"
     ]
    }
   ],
   "source": [
    "#regression btw miRNA expression, exon expression\n",
    "import sklearn\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.datasets import make_regression\n",
    "#faster if np.array(x, order='F')\n",
    "#If you are interested in controlling the L1 and L2 penalty separately, keep in mind that this is equivalent to:\n",
    "#a * L1 + b * L2 where: alpha = a + b and l1_ratio = a / (a + b)\n",
    "X = np.array(pivoted)\n",
    "y = np.array(pivoted.index.get_level_values('exon_raw_counts'))\n",
    "\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "print(regr.coef_)\n",
    "print(regr.intercept_)"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "Seqvec basic Seq ligand binding.ipynb",
   "provenance": []
  },
  "hide_input": false,
  "kernelspec": {
   "display_name": "thesis",
   "language": "python",
   "name": "thesis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "581.6px",
    "left": "235px",
    "top": "158.533px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
