{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "center"
   },
   "source": [
    "# Assemble the final Dataframe from tarpmir bindingsites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in bindingsite & Ensembl data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:06:17.213245Z",
     "start_time": "2020-10-21T23:06:16.478962Z"
    },
    "cell_style": "center"
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from helper_fcts import *\n",
    "import sqlite3\n",
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "path = Path('C:/Users/Lena/Documents/auf mega gesyncte Dokumente/Uni/Master/Semester 5/Masterarbeit/alt-splicing-mirna-reg')\n",
    "path_tarp = Path(path/'tarp-bs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:06:21.289688Z",
     "start_time": "2020-10-21T23:06:17.962018Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id_version</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>ensembl_transcript_id_version</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id ensembl_gene_id_version ensembl_transcript_id  \\\n",
       "0  ENSG00000210049       ENSG00000210049.1       ENST00000387314   \n",
       "1  ENSG00000211459       ENSG00000211459.2       ENST00000389680   \n",
       "2  ENSG00000210077       ENSG00000210077.1       ENST00000387342   \n",
       "3  ENSG00000210082       ENSG00000210082.2       ENST00000387347   \n",
       "4  ENSG00000209082       ENSG00000209082.1       ENST00000386347   \n",
       "\n",
       "  ensembl_transcript_id_version chromosome_name  strand  start_position  \\\n",
       "0             ENST00000387314.1              MT       1             577   \n",
       "1             ENST00000389680.2              MT       1             648   \n",
       "2             ENST00000387342.1              MT       1            1602   \n",
       "3             ENST00000387347.2              MT       1            1671   \n",
       "4             ENST00000386347.1              MT       1            3230   \n",
       "\n",
       "   end_position  \n",
       "0           647  \n",
       "1          1601  \n",
       "2          1670  \n",
       "3          3229  \n",
       "4          3304  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#executed once, read in from pickled files\n",
    "ids = read_in_ids(path/'all_mapping_ids.fasta') # mapping of transcript ids to exon ids\n",
    "chrom_exon_starts, exon_starts, exon_ends = calc_exon_data(path) # exon starts and ends per transcript + exon id\n",
    "#read in df_gene: gene_id, g_id_v, transcript_id, t_id_v, chromosome, strand, start, end (1-N)\n",
    "df_gene = pd.read_csv(path/'gene_infos.csv', dtype={'ensembl_gene_id':str,'ensembl_gene_id_version':str,'ensembl_transcript_id':str, 'ensembl_transcript_id_version':str, 'chromosome_name':'category', 'strand':'int8', 'start_position':int, 'end_position':int})\n",
    "del df_gene['Unnamed: 0']\n",
    "df_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T22:31:40.140140Z",
     "start_time": "2020-10-21T22:31:40.070177Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#set bs from feather\n",
    "bs = pd.read_feather(path/'read_in_bs.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:08:53.720453Z",
     "start_time": "2020-10-21T23:08:52.933037Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All predicted bindingsites were read into pandas.\n"
     ]
    }
   ],
   "source": [
    "#read in bs\n",
    "seperate_bs = []\n",
    "for filename in os.listdir(path_tarp):\n",
    "    binding_sites = parse_tarp_bs(path_tarp/filename)\n",
    "    seperate_bs.append(binding_sites)\n",
    "bs = pd.concat(seperate_bs, axis=0, ignore_index=True) #14 388 + 10 326 = 24 714 rows\n",
    "bs.to_feather(path/'read_in_bs.feather')\n",
    "print('All predicted bindingsites were read into pandas.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:07:50.384154Z",
     "start_time": "2020-10-21T23:07:44.980201Z"
    }
   },
   "outputs": [],
   "source": [
    "#parse miRNA and mRNA seed from sequence\n",
    "miRNA_sequences = parse_seq(path/'input_miRNA.fasta')#2661 miRNAs \n",
    "bs['miRNA_seed'] = bs.apply(lambda row: miRNA_sequences[row['miRNA']], axis=1)\n",
    "mRNA_sequences = parse_seq(path/'cdna6247.fasta') #TODO change to bigger file with all mRNAs or iterate over all files\n",
    "bs['mRNA_bs_seq'] = bs.apply(lambda row: mRNA_sequences[row['mRNA']][row['bs_start']:row['bs_end']], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:08:01.223613Z",
     "start_time": "2020-10-21T23:08:00.694183Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:08:13.441303Z",
     "start_time": "2020-10-21T23:08:10.958988Z"
    },
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def pos_to_chrom(pos, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if pos >= exon_starts[tid][eid] and pos <= exon_ends[tid][eid]:\n",
    "                diff = pos - exon_starts[tid][eid]\n",
    "                chrom_pos = chrom_exon_starts[tid][eid] + diff\n",
    "                return chrom_pos\n",
    "        else: print('Error: Either',tid,'or',eid,'not in exon_starts or exon_ends')\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:09:15.553079Z",
     "start_time": "2020-10-21T23:09:15.478121Z"
    }
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def get_eid(start, end, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        if tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]:\n",
    "            if (start >= exon_starts[tid][eid]) and (start <= exon_ends[tid][eid]) and (end >= exon_starts[tid][eid]) and (end <= exon_ends[tid][eid]):\n",
    "                return eid\n",
    "        else: print('Error: Either',tid,'or',eid,'not in exon_starts or exon_ends')\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:10:26.174074Z",
     "start_time": "2020-10-21T23:10:12.923699Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>binding_region_length</th>\n",
       "      <th>longest_consecutive_pairings</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.005183</td>\n",
       "      <td>-0.035634</td>\n",
       "      <td>-11.206441</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>309</td>\n",
       "      <td>335</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.743590</td>\n",
       "      <td>-21.6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.161796</td>\n",
       "      <td>0.055277</td>\n",
       "      <td>-3.793325</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>430</td>\n",
       "      <td>450</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.441</td>\n",
       "      <td>3.547385</td>\n",
       "      <td>2.945605</td>\n",
       "      <td>-6.324962</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>130</td>\n",
       "      <td>1576707</td>\n",
       "      <td>1578261</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.021616</td>\n",
       "      <td>-0.047693</td>\n",
       "      <td>-18.306612</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>315</td>\n",
       "      <td>336</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-15.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.456</td>\n",
       "      <td>3.773200</td>\n",
       "      <td>3.161275</td>\n",
       "      <td>-9.685015</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>41</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "      <td>ENSE00002671000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "0  hsa-let-7a-2-3p  ENST00000576537             1.000000   -25.9     0   \n",
       "1  hsa-let-7a-2-3p  ENST00000576537             0.743590   -21.6     1   \n",
       "2    hsa-let-7b-5p  ENST00000576537             0.846154   -18.4     1   \n",
       "3    hsa-let-7b-3p  ENST00000576537             0.615385   -20.6     0   \n",
       "4    hsa-let-7b-3p  ENST00000576537             0.538462   -15.3     0   \n",
       "\n",
       "   accessibility  AU_content  PhyloP_Stem  PyloP_Flanking       m/e   ...  \\\n",
       "0       0.000156       0.338     0.005183       -0.035634 -11.206441  ...   \n",
       "1       0.000002       0.338    -0.161796        0.055277  -3.793325  ...   \n",
       "2       0.000026       0.441     3.547385        2.945605  -6.324962  ...   \n",
       "3       0.000143       0.338    -0.021616       -0.047693 -18.306612  ...   \n",
       "4       0.000020       0.456     3.773200        3.161275  -9.685015  ...   \n",
       "\n",
       "   binding_region_length  longest_consecutive_pairings  \\\n",
       "0                     21                            16   \n",
       "1                     20                             6   \n",
       "2                     25                            11   \n",
       "3                     21                            12   \n",
       "4                     18                             8   \n",
       "\n",
       "   position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "0                                         4                       7   \n",
       "1                                         8                       3   \n",
       "2                                         5                       8   \n",
       "3                                         9                       7   \n",
       "4                                         7                       5   \n",
       "\n",
       "   difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \\\n",
       "0                                                  1        309     335   \n",
       "1                                                  4        430     450   \n",
       "2                                                  1        105     130   \n",
       "3                                                  3        315     336   \n",
       "4                                                  1         23      41   \n",
       "\n",
       "   chrom_bs_start  chrom_bs_end          exon_id  \n",
       "0         1578440       1578466  ENSE00002650258  \n",
       "1         1578561       1578581  ENSE00002650258  \n",
       "2         1576707       1578261             None  \n",
       "3         1578446       1578467  ENSE00002650258  \n",
       "4         1576625       1576643  ENSE00002671000  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#translate genome position relative to transcript to chromosome\n",
    "#fastest, TODO check for huge files\n",
    "bs['chrom_bs_start'] = bs.apply(lambda row: pos_to_chrom(row.bs_start, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['chrom_bs_end'] = bs.apply(lambda row: pos_to_chrom(row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "bs['exon_id'] = bs.apply(lambda row: get_eid(row.bs_start, row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1)\n",
    "#bs['bs_id'] = bs.index\n",
    "bs.head() #4.69, 4.9, 4.48s | 13.3s, 13.3s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:10:49.351291Z",
     "start_time": "2020-10-21T23:10:49.296326Z"
    }
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def pos_to_chrom_new(start, end, tid, ids, exon_starts, exon_ends, chrom_exon_starts):\n",
    "    for j, eid in enumerate(ids[tid]):\n",
    "        ids_exist = tid in exon_starts and eid in exon_starts[tid] and tid in exon_ends and eid in exon_ends[tid]\n",
    "        both_in_exon = (start >= exon_starts[tid][eid]) and (start <= exon_ends[tid][eid]) and (end >= exon_starts[tid][eid]) and (end <= exon_ends[tid][eid])\n",
    "        if ids_exist and both_in_exon:\n",
    "            chrom_start_pos = chrom_exon_starts[tid][eid] + (start - exon_starts[tid][eid])\n",
    "            chrom_end_pos = chrom_exon_starts[tid][eid] + (end - exon_starts[tid][eid])\n",
    "            return (eid, chrom_start_pos, chrom_end_pos)\n",
    "   # if not both_in_exon:\n",
    "    #    print('Error: The bindingsite spans more than 1 exon with bs length',end-start+1)\n",
    "    #if not ids_exist:\n",
    "     #   print('Error: Either',tid,'or',eid,'not in exon_starts or exon_ends')\n",
    "    return (2,2,2)#None, None, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:11:18.325098Z",
     "start_time": "2020-10-21T23:10:59.589879Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>binding_region_length</th>\n",
       "      <th>longest_consecutive_pairings</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.005183</td>\n",
       "      <td>-0.035634</td>\n",
       "      <td>-11.206441</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>309</td>\n",
       "      <td>335</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.743590</td>\n",
       "      <td>-21.6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.161796</td>\n",
       "      <td>0.055277</td>\n",
       "      <td>-3.793325</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>430</td>\n",
       "      <td>450</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.441</td>\n",
       "      <td>3.547385</td>\n",
       "      <td>2.945605</td>\n",
       "      <td>-6.324962</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>130</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.021616</td>\n",
       "      <td>-0.047693</td>\n",
       "      <td>-18.306612</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>315</td>\n",
       "      <td>336</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-15.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.456</td>\n",
       "      <td>3.773200</td>\n",
       "      <td>3.161275</td>\n",
       "      <td>-9.685015</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>41</td>\n",
       "      <td>ENSE00002671000</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "0  hsa-let-7a-2-3p  ENST00000576537             1.000000   -25.9     0   \n",
       "1  hsa-let-7a-2-3p  ENST00000576537             0.743590   -21.6     1   \n",
       "2    hsa-let-7b-5p  ENST00000576537             0.846154   -18.4     1   \n",
       "3    hsa-let-7b-3p  ENST00000576537             0.615385   -20.6     0   \n",
       "4    hsa-let-7b-3p  ENST00000576537             0.538462   -15.3     0   \n",
       "\n",
       "   accessibility  AU_content  PhyloP_Stem  PyloP_Flanking       m/e   ...  \\\n",
       "0       0.000156       0.338     0.005183       -0.035634 -11.206441  ...   \n",
       "1       0.000002       0.338    -0.161796        0.055277  -3.793325  ...   \n",
       "2       0.000026       0.441     3.547385        2.945605  -6.324962  ...   \n",
       "3       0.000143       0.338    -0.021616       -0.047693 -18.306612  ...   \n",
       "4       0.000020       0.456     3.773200        3.161275  -9.685015  ...   \n",
       "\n",
       "   binding_region_length  longest_consecutive_pairings  \\\n",
       "0                     21                            16   \n",
       "1                     20                             6   \n",
       "2                     25                            11   \n",
       "3                     21                            12   \n",
       "4                     18                             8   \n",
       "\n",
       "   position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "0                                         4                       7   \n",
       "1                                         8                       3   \n",
       "2                                         5                       8   \n",
       "3                                         9                       7   \n",
       "4                                         7                       5   \n",
       "\n",
       "   difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \\\n",
       "0                                                  1        309     335   \n",
       "1                                                  4        430     450   \n",
       "2                                                  1        105     130   \n",
       "3                                                  3        315     336   \n",
       "4                                                  1         23      41   \n",
       "\n",
       "           exon_id chrom_bs_start  chrom_bs_end  \n",
       "0  ENSE00002650258        1578440       1578466  \n",
       "1  ENSE00002650258        1578561       1578581  \n",
       "2                2              2             2  \n",
       "3  ENSE00002650258        1578446       1578467  \n",
       "4  ENSE00002671000        1576625       1576643  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#middle\n",
    "bs['exon_id'] = np.nan\n",
    "bs['chrom_bs_start'] = np.nan\n",
    "bs['chrom_bs_end'] = np.nan\n",
    "bs[['exon_id','chrom_bs_start','chrom_bs_end']] = bs.apply(lambda row: pos_to_chrom_new(row.bs_start, row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1, result_type='expand')\n",
    "#bs['bs_id'] = bs.index\n",
    "bs.head() #6.77s, 6.79, 7.32 | 18.7s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:12:11.284053Z",
     "start_time": "2020-10-21T23:12:11.217077Z"
    }
   },
   "outputs": [],
   "source": [
    "bs = pd.read_feather(path/'read_in_bs.feather')#reset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:12:00.528228Z",
     "start_time": "2020-10-21T23:11:42.193774Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>binding_region_length</th>\n",
       "      <th>longest_consecutive_pairings</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.005183</td>\n",
       "      <td>-0.035634</td>\n",
       "      <td>-11.206441</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>309</td>\n",
       "      <td>335</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.743590</td>\n",
       "      <td>-21.6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.161796</td>\n",
       "      <td>0.055277</td>\n",
       "      <td>-3.793325</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>430</td>\n",
       "      <td>450</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.441</td>\n",
       "      <td>3.547385</td>\n",
       "      <td>2.945605</td>\n",
       "      <td>-6.324962</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>130</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.021616</td>\n",
       "      <td>-0.047693</td>\n",
       "      <td>-18.306612</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>315</td>\n",
       "      <td>336</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-15.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.456</td>\n",
       "      <td>3.773200</td>\n",
       "      <td>3.161275</td>\n",
       "      <td>-9.685015</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>41</td>\n",
       "      <td>ENSE00002671000</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "0  hsa-let-7a-2-3p  ENST00000576537             1.000000   -25.9     0   \n",
       "1  hsa-let-7a-2-3p  ENST00000576537             0.743590   -21.6     1   \n",
       "2    hsa-let-7b-5p  ENST00000576537             0.846154   -18.4     1   \n",
       "3    hsa-let-7b-3p  ENST00000576537             0.615385   -20.6     0   \n",
       "4    hsa-let-7b-3p  ENST00000576537             0.538462   -15.3     0   \n",
       "\n",
       "   accessibility  AU_content  PhyloP_Stem  PyloP_Flanking       m/e   ...  \\\n",
       "0       0.000156       0.338     0.005183       -0.035634 -11.206441  ...   \n",
       "1       0.000002       0.338    -0.161796        0.055277  -3.793325  ...   \n",
       "2       0.000026       0.441     3.547385        2.945605  -6.324962  ...   \n",
       "3       0.000143       0.338    -0.021616       -0.047693 -18.306612  ...   \n",
       "4       0.000020       0.456     3.773200        3.161275  -9.685015  ...   \n",
       "\n",
       "   binding_region_length  longest_consecutive_pairings  \\\n",
       "0                     21                            16   \n",
       "1                     20                             6   \n",
       "2                     25                            11   \n",
       "3                     21                            12   \n",
       "4                     18                             8   \n",
       "\n",
       "   position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "0                                         4                       7   \n",
       "1                                         8                       3   \n",
       "2                                         5                       8   \n",
       "3                                         9                       7   \n",
       "4                                         7                       5   \n",
       "\n",
       "   difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \\\n",
       "0                                                  1        309     335   \n",
       "1                                                  4        430     450   \n",
       "2                                                  1        105     130   \n",
       "3                                                  3        315     336   \n",
       "4                                                  1         23      41   \n",
       "\n",
       "           exon_id chrom_bs_start  chrom_bs_end  \n",
       "0  ENSE00002650258        1578440       1578466  \n",
       "1  ENSE00002650258        1578561       1578581  \n",
       "2                2              2             2  \n",
       "3  ENSE00002650258        1578446       1578467  \n",
       "4  ENSE00002671000        1576625       1576643  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#middle\n",
    "bs = bs.join(bs.apply(lambda row: pos_to_chrom_new(row.bs_start, row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts), axis=1, result_type='expand'))\n",
    "#bs['bs_id'] = bs.index\n",
    "bs = bs.rename(columns={0: 'exon_id',1:'chrom_bs_start',2:'chrom_bs_end'})\n",
    "bs.head() #6.60s, 6.67s, 7.59s | 18.3s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-21T23:13:00.145582Z",
     "start_time": "2020-10-21T23:12:21.675697Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>...</th>\n",
       "      <th>binding_region_length</th>\n",
       "      <th>longest_consecutive_pairings</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.005183</td>\n",
       "      <td>-0.035634</td>\n",
       "      <td>-11.206441</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>309</td>\n",
       "      <td>335</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578440</td>\n",
       "      <td>1578466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.743590</td>\n",
       "      <td>-21.6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.161796</td>\n",
       "      <td>0.055277</td>\n",
       "      <td>-3.793325</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>430</td>\n",
       "      <td>450</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578561</td>\n",
       "      <td>1578581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-18.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.441</td>\n",
       "      <td>3.547385</td>\n",
       "      <td>2.945605</td>\n",
       "      <td>-6.324962</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>130</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.021616</td>\n",
       "      <td>-0.047693</td>\n",
       "      <td>-18.306612</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>315</td>\n",
       "      <td>336</td>\n",
       "      <td>ENSE00002650258</td>\n",
       "      <td>1578446</td>\n",
       "      <td>1578467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000576537</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>-15.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.456</td>\n",
       "      <td>3.773200</td>\n",
       "      <td>3.161275</td>\n",
       "      <td>-9.685015</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>41</td>\n",
       "      <td>ENSE00002671000</td>\n",
       "      <td>1576625</td>\n",
       "      <td>1576643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             miRNA             mRNA  binding_probability  energy  seed  \\\n",
       "0  hsa-let-7a-2-3p  ENST00000576537             1.000000   -25.9     0   \n",
       "1  hsa-let-7a-2-3p  ENST00000576537             0.743590   -21.6     1   \n",
       "2    hsa-let-7b-5p  ENST00000576537             0.846154   -18.4     1   \n",
       "3    hsa-let-7b-3p  ENST00000576537             0.615385   -20.6     0   \n",
       "4    hsa-let-7b-3p  ENST00000576537             0.538462   -15.3     0   \n",
       "\n",
       "   accessibility  AU_content  PhyloP_Stem  PyloP_Flanking       m/e   ...  \\\n",
       "0       0.000156       0.338     0.005183       -0.035634 -11.206441  ...   \n",
       "1       0.000002       0.338    -0.161796        0.055277  -3.793325  ...   \n",
       "2       0.000026       0.441     3.547385        2.945605  -6.324962  ...   \n",
       "3       0.000143       0.338    -0.021616       -0.047693 -18.306612  ...   \n",
       "4       0.000020       0.456     3.773200        3.161275  -9.685015  ...   \n",
       "\n",
       "   binding_region_length  longest_consecutive_pairings  \\\n",
       "0                     21                            16   \n",
       "1                     20                             6   \n",
       "2                     25                            11   \n",
       "3                     21                            12   \n",
       "4                     18                             8   \n",
       "\n",
       "   position_of_longest_consecutive_pairings  pairings_in_3prime_end  \\\n",
       "0                                         4                       7   \n",
       "1                                         8                       3   \n",
       "2                                         5                       8   \n",
       "3                                         9                       7   \n",
       "4                                         7                       5   \n",
       "\n",
       "   difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \\\n",
       "0                                                  1        309     335   \n",
       "1                                                  4        430     450   \n",
       "2                                                  1        105     130   \n",
       "3                                                  3        315     336   \n",
       "4                                                  1         23      41   \n",
       "\n",
       "           exon_id chrom_bs_start  chrom_bs_end  \n",
       "0  ENSE00002650258        1578440       1578466  \n",
       "1  ENSE00002650258        1578561       1578581  \n",
       "2                2              2             2  \n",
       "3  ENSE00002650258        1578446       1578467  \n",
       "4  ENSE00002671000        1576625       1576643  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#slowest\n",
    "bs['exon_id'] = np.nan\n",
    "bs['chrom_bs_start'] = np.nan\n",
    "bs['chrom_bs_end'] = np.nan\n",
    "bs[['exon_id','chrom_bs_start','chrom_bs_end']] = bs.apply(lambda row: pd.Series(pos_to_chrom_new(row.bs_start, row.bs_end, row.mRNA, ids, exon_starts, exon_ends, chrom_exon_starts)), axis=1)\n",
    "\n",
    "#bs['bs_id'] = bs.index\n",
    "bs.head() #9.49s, 9.59s | 39s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get TCGA data from Xena"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T08:31:50.258992Z",
     "start_time": "2020-10-20T08:31:48.250552Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Lena\\Miniconda3\\envs\\thesis\\lib\\site-packages\\xenaPython\\__init__.py:110: FutureWarning: Possible nested set at position 7\n",
      "  re.sub(r\"^[^[]+[[]([^]]*)[]].*$\", r\"\\1\", query, flags=re.DOTALL))\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'get_fields' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-30-58c9d40d0b28>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0msamples\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;36m10\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mdataset\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena'\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 7\u001b[1;33m \u001b[0mexpression\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mget_fields\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mhost\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdataset\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msamples\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mGENES\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m# list of lists.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      8\u001b[0m \u001b[0mexpression_by_gene\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mzip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mGENES\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpression\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m      \u001b[1;31m# index by gene.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mexpression_by_gene\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mGENES\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpression_by_gene\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mGENES\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;36m10\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'get_fields' is not defined"
     ]
    }
   ],
   "source": [
    "import xenaPython as xena\n",
    "cohort = 'TCGA PanCanAtlas'\n",
    "host = xena.PUBLIC_HUBS['pancanAtlasHub']\n",
    "samples = xena.cohort_samples(host, cohort, None)\n",
    "samples[0: 10]\n",
    "dataset = 'EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena'\n",
    "expression = get_fields(host, dataset, samples, GENES) # list of lists.\n",
    "expression_by_gene = dict(zip(GENES, expression))      # index by gene.\n",
    "[expression_by_gene.keys(), GENES[0], expression_by_gene[GENES[0]][0:10]]\n",
    "#instead of loading everything, only load for every row in bs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T11:55:18.837342Z",
     "start_time": "2020-10-20T11:55:16.985418Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3.469]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hub = \"https://toil.xenahubs.net\"\n",
    "dataset = \"tcga_RSEM_gene_tpm\"\n",
    "\n",
    "xena.dataset_samples(hub, dataset, None)\n",
    "xena.dataset_gene_probe_avg(hub, dataset, ['TCGA-61-1910-01'], ['DUSP7'])[0][\"scores\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T08:51:21.793402Z",
     "start_time": "2020-10-20T08:51:18.426833Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TCGA-KN-8426-01', 'TCGA-KN-8419-11', 'TCGA-KL-8332-01', 'TCGA-KN-8423-01', 'TCGA-KO-8405-01', 'TCGA-KL-8337-01', 'TCGA-KN-8424-11', 'TCGA-KN-8422-11', 'TCGA-KL-8334-01', 'TCGA-KO-8404-01']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['NaN', 'NaN', 'NaN', 'NaN', 'NaN', 'NaN', 'NaN', 'NaN', 'NaN', 'NaN']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hub = \"https://tcga.xenahubs.net\"\n",
    "dataset = \"TCGA.KICH.sampleMap/miRNA_HiSeq_gene\"\n",
    "#samples = xena.dataset_samples (hub, dataset, None)\n",
    "samples = xena.dataset_samples (hub, dataset, 10)\n",
    "print(samples)\n",
    "#samples = [\"TCGA-02-0047-01\",\"TCGA-02-0055-01\",\"TCGA-02-2483-01\"]\n",
    "probes = ['ENSG00000282740.1', 'ENSG00000000005.5']\n",
    "[position, [ENSG00000282740_1, ENSG00000000005_5]] = xena.dataset_probe_values (hub, dataset, samples, probes)\n",
    "ENSG00000282740_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:16:01.774905Z",
     "start_time": "2020-10-18T22:16:00.756680Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext Cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:16:41.184687Z",
     "start_time": "2020-10-18T22:16:01.300Z"
    }
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "def c_split_chrom_column(chrom):\n",
    "    chrom_list = chrom.split(':')\n",
    "    chromosome_name = str(chrom_list[0][3:])\n",
    "    strand = 1 if chrom_list[2] == '+' else -1\n",
    "    chrom_exon_start = int(chrom_list[1].split('-')[0])\n",
    "    chrom_exon_end = int(chrom_list[1].split('-')[1])\n",
    "    return chromosome_name,strand,chrom_exon_start,chrom_exon_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:16:41.200240Z",
     "start_time": "2020-10-18T22:16:01.700Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def list_to_df(list_, nr_columns, data_type=None):\n",
    "    data = []\n",
    "    remove_last_columns = 2 if data_type != None else 0\n",
    "    nr_rows = int((len(list_)-1)/nr_columns)\n",
    "    for row in range(nr_rows):\n",
    "        start = row*nr_columns\n",
    "        end = (row+1)*nr_columns\n",
    "        data.append(list_[start:end-remove_last_columns])\n",
    "    df = pd.DataFrame(data[1:], columns = data[0]) #use first line as header\n",
    "    if data_type=='miRNA':\n",
    "        df = df.astype({'miRNA_ID':str, 'read_count':int})#, 'reads_per_million_miRNA_mapped':float, 'cross-mapped': 'category'})\n",
    "    elif data_type=='exon':\n",
    "        df = df.astype({'exon':str, 'raw_counts':int})#, 'reads_per_million_miRNA_mapped':float, 'cross-mapped': 'category'})\n",
    "        #TODO change this to seperate lines cause far faster (see calc_chrom_pos)\n",
    "        df[['chromosome_name','strand','chrom_exon_start','chrom_exon_end']] = df.exon.apply(lambda x: pd.Series(c_split_chrom_column(str(x))))\n",
    "        df[['raw_counts','chrom_exon_start','chrom_exon_end']] = df[['raw_counts','chrom_exon_start','chrom_exon_end']].apply(pd.to_numeric,downcast='unsigned')\n",
    "        df[['chromosome_name','strand']] = df[['chromosome_name','strand']].astype('str').astype('category')\n",
    "        del df['exon']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.092550Z",
     "start_time": "2020-10-18T22:50:52.800Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#UCSC Xena API, get list of files which contain miRNA expression per case_id there\n",
    "\n",
    "primary_site='Kidney'\n",
    "download_files_from_legacy('miRNA',3000,primary_site,write=True) #2029 #70min\n",
    "#download_files_from_legacy('exon',2000,primary_site,write=True) #1588 #35h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:12:30.124823Z",
     "start_time": "2020-10-18T22:11:13.316631Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1588 2013\n"
     ]
    }
   ],
   "source": [
    "#read in downloaded counts from feather files\n",
    "#TODO remove later when all data there\n",
    "n_all_mirna_dfs = {}\n",
    "dir_ = Path(path/'mirna')\n",
    "for i, file in enumerate(os.listdir(dir_)):\n",
    "    #print(file)\n",
    "    filename = ('.').join(file.split('.')[0:3])\n",
    "    n_all_mirna_dfs[filename] = pd.read_feather(os.path.join(dir_,file))\n",
    "    \n",
    "n_all_exon_dfs = {}\n",
    "dir_ = Path(path/'exon')\n",
    "for i, file in enumerate(os.listdir(dir_)):\n",
    "    #print(file)\n",
    "    filename = ('.').join(file.split('.')[0:3])\n",
    "    n_all_exon_dfs[filename] = pd.read_feather(os.path.join(dir_,file))\n",
    "#mirna_counts[['mirna_read_count']].astype('int') #TODO do this before saving while getting it from API\n",
    "#how to get from API? per caseid?\n",
    "print(len(n_all_exon_dfs),len(n_all_mirna_dfs)) #yes this is right, read in from path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T23:38:25.124503Z",
     "start_time": "2020-10-20T23:38:25.044562Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>58346872</td>\n",
       "      <td>58346895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>58346310</td>\n",
       "      <td>58346332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>58345845</td>\n",
       "      <td>58345870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>58345558</td>\n",
       "      <td>58345576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>58345880</td>\n",
       "      <td>58345898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   bs_id            miRNA             mRNA  chrom_bs_start  chrom_bs_end\n",
       "0      0    hsa-let-7a-5p  ENST00000596924        58346872      58346895\n",
       "1      1    hsa-let-7a-5p  ENST00000596924        58346310      58346332\n",
       "2      2    hsa-let-7a-5p  ENST00000596924        58345845      58345870\n",
       "3      3  hsa-let-7a-2-3p  ENST00000596924        58345558      58345576\n",
       "4      4  hsa-let-7a-2-3p  ENST00000596924        58345880      58345898"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get subset counts of bs and merge with df_gene\n",
    "counts = bs[['bs_id', 'miRNA', 'mRNA', 'chrom_bs_start', 'chrom_bs_end']].copy()\n",
    "counts = counts.merge(df_gene[['ensembl_transcript_id','chromosome_name','strand']], left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "counts[['chromosome_name','strand','miRNA','mRNA']] = counts[['chromosome_name','strand','miRNA','mRNA']].astype('str').astype('category')\n",
    "del counts['ensembl_transcript_id']\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mapping miRNA ID to mature ID\n",
    "mirna_counts = all_mirna_dfs[case_id]\n",
    "\n",
    "trans = pd.DataFrame.from_dict(read_in_miRNA(path/'miRNA2mature.fasta'), orient='index')#,columns=[\"mirna\", \"mature\"])\n",
    "trans = trans.rename(columns={0: \"miRNA_ID\"})\n",
    "trans[['miRNA_ID']] = trans[['miRNA_ID']].astype('str').astype('category')\n",
    "\n",
    "#mapping miRNA to miRNA family\n",
    "mirna_families = read_in_mirna_families(path/'mapping_miRNA_families.fasta')\n",
    "trans = trans.join(mirna_families,how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for filename in os.listdir(caseids):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Execute this for every caseid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map miRNA to counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.058588Z",
     "start_time": "2020-10-18T22:52:42.986240Z"
    }
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'24ea3cd5-c2e6-44f7-9658-78af7befeef7'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-29-ecc3fe5af74e>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m#map miRNA ID to mature ID\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mmirna_counts\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mall_mirna_dfs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mcase_id\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mtrans\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfrom_dict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mread_in_miRNA\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpath2\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m'miRNA2mature.fasta'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0morient\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'index'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;31m#,columns=[\"mirna\", \"mature\"])\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mtrans\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtrans\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m{\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;34m\"miRNA_ID\"\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: '24ea3cd5-c2e6-44f7-9658-78af7befeef7'"
     ]
    }
   ],
   "source": [
    "counts = counts.join(trans,on='miRNA').merge(mirna_counts[['miRNA_ID','read_count']],on='miRNA_ID')\n",
    "#bs_mirnas.drop(['miRNA','miRNA_ID'], axis=1, inplace=True)\n",
    "counts.head()\n",
    "\n",
    "#2693 with hsa in miRNA2mature / 49156 overall in miRNA2mature (also other species eg cel)\n",
    "#2656 if inner join cause only 2656 families\n",
    "#TODO check if miRNA counts from TCGA are more different miRNAs?\n",
    "#TODO is inner right? are some in important counts None?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T01:30:28.905033Z",
     "start_time": "2020-10-13T01:30:28.803071Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_raw_counts</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>mirna_read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>let-7</td>\n",
       "      <td>55647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7d-5p</td>\n",
       "      <td>hsa-let-7d</td>\n",
       "      <td>let-7</td>\n",
       "      <td>3767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>let-7</td>\n",
       "      <td>11424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7e-3p</td>\n",
       "      <td>hsa-let-7e</td>\n",
       "      <td>let-7</td>\n",
       "      <td>11424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-15a-3p</td>\n",
       "      <td>hsa-mir-15a</td>\n",
       "      <td>mir-15</td>\n",
       "      <td>2258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12562</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-5692a</td>\n",
       "      <td>hsa-mir-5692a-2</td>\n",
       "      <td>mir-5692</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12563</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-548az-5p</td>\n",
       "      <td>hsa-mir-548az</td>\n",
       "      <td>mir-548</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12566</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-8062</td>\n",
       "      <td>hsa-mir-8062</td>\n",
       "      <td>hsa-miR-8062</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12571</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-8064</td>\n",
       "      <td>hsa-mir-8064</td>\n",
       "      <td>hsa-miR-8064</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12578</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1257</td>\n",
       "      <td>hsa-mir-1257</td>\n",
       "      <td>mir-1257</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2129 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       exon_id  exon_raw_counts             miRNA         miRNA_ID  \\\n",
       "7       216595                0     hsa-let-7b-5p       hsa-let-7b   \n",
       "14      216595                0     hsa-let-7d-5p       hsa-let-7d   \n",
       "25      216595                0     hsa-let-7e-3p       hsa-let-7e   \n",
       "33      216595                0     hsa-let-7e-3p       hsa-let-7e   \n",
       "46      216595                0    hsa-miR-15a-3p      hsa-mir-15a   \n",
       "...        ...              ...               ...              ...   \n",
       "12562   216595                0     hsa-miR-5692a  hsa-mir-5692a-2   \n",
       "12563   216595                0  hsa-miR-548az-5p    hsa-mir-548az   \n",
       "12566   216595                0      hsa-miR-8062     hsa-mir-8062   \n",
       "12571   216595                0      hsa-miR-8064     hsa-mir-8064   \n",
       "12578   216595                0      hsa-miR-1257     hsa-mir-1257   \n",
       "\n",
       "       mirna_family  mirna_read_count  \n",
       "7             let-7             55647  \n",
       "14            let-7              3767  \n",
       "25            let-7             11424  \n",
       "33            let-7             11424  \n",
       "46           mir-15              2258  \n",
       "...             ...               ...  \n",
       "12562      mir-5692                 0  \n",
       "12563       mir-548                 0  \n",
       "12566  hsa-miR-8062                 0  \n",
       "12571  hsa-miR-8064                 0  \n",
       "12578      mir-1257                 0  \n",
       "\n",
       "[2129 rows x 6 columns]"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs_mirnas[bs_mirnas.exon_id == 216595]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T01:29:10.977928Z",
     "start_time": "2020-10-13T01:29:10.892977Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_raw_counts</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>mirna_read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3365</th>\n",
       "      <td>216603</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>216595</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371</th>\n",
       "      <td>216596</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>hsa-mir-1229</td>\n",
       "      <td>hsa-miR-1229-5p</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      exon_id  exon_raw_counts            miRNA      miRNA_ID  \\\n",
       "3365   216603                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3368   216595                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3370   216595                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "3371   216596                0  hsa-miR-1229-5p  hsa-mir-1229   \n",
       "\n",
       "         mirna_family  mirna_read_count  \n",
       "3365  hsa-miR-1229-5p                 4  \n",
       "3368  hsa-miR-1229-5p                 4  \n",
       "3370  hsa-miR-1229-5p                 4  \n",
       "3371  hsa-miR-1229-5p                 4  "
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs_mirnas[bs_mirnas.mirna_family =='hsa-miR-1229-5p']\n",
    "\n",
    "# 1 right? count is added for exonid 216595, but do i already add it before ?\n",
    "# 2 also : exon_id + mirna double (3368 & 3370) wrong!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map exons to counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get exonid for exon instead of introducing exon_id\n",
    "#TODO this is not necessary anymore if the new exon counts work!!!!\n",
    "#instead try directly with ensembl start & end match - if it doesnt work \n",
    "#-> try download from https://www.gencodegenes.org/human/ (GRCh38.p13) if same reference genome\n",
    "bs['exon_id'] = np.nan "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T22:49:28.421098Z",
     "start_time": "2020-10-20T22:49:28.362112Z"
    }
   },
   "outputs": [],
   "source": [
    "bs.to_sql('bs', con=conn)\n",
    "conn = sqlite3.connect('db.db')\n",
    "c = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T22:52:07.052993Z",
     "start_time": "2020-10-20T22:52:06.446321Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>binding_probability</th>\n",
       "      <th>energy</th>\n",
       "      <th>seed</th>\n",
       "      <th>accessibility</th>\n",
       "      <th>AU_content</th>\n",
       "      <th>PhyloP_Stem</th>\n",
       "      <th>PyloP_Flanking</th>\n",
       "      <th>m/e</th>\n",
       "      <th>number_of_pairings</th>\n",
       "      <th>binding_region_length</th>\n",
       "      <th>longest_consecutive_pairings</th>\n",
       "      <th>position_of_longest_consecutive_pairings</th>\n",
       "      <th>pairings_in_3prime_end</th>\n",
       "      <th>difference_of_pairings_between_seed_and_3prime_end</th>\n",
       "      <th>bs_start</th>\n",
       "      <th>bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-22.3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.489720e-02</td>\n",
       "      <td>0.441</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.849714</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1695</td>\n",
       "      <td>1718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>-20.9</td>\n",
       "      <td>1</td>\n",
       "      <td>1.141923e-01</td>\n",
       "      <td>0.471</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-5.850804</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1133</td>\n",
       "      <td>1155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>0.846154</td>\n",
       "      <td>-20.1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.216239e-03</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.993964</td>\n",
       "      <td>19</td>\n",
       "      <td>25</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>668</td>\n",
       "      <td>693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-23.2</td>\n",
       "      <td>0</td>\n",
       "      <td>9.643462e-07</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-13.665600</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>381</td>\n",
       "      <td>399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000596924</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-21.4</td>\n",
       "      <td>0</td>\n",
       "      <td>7.641573e-06</td>\n",
       "      <td>0.456</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-8.968752</td>\n",
       "      <td>15</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>703</td>\n",
       "      <td>721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24709</th>\n",
       "      <td>24709</td>\n",
       "      <td>hsa-miR-12131</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-25.1</td>\n",
       "      <td>1</td>\n",
       "      <td>2.668984e-04</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.333260</td>\n",
       "      <td>0.191636</td>\n",
       "      <td>-5.939295</td>\n",
       "      <td>17</td>\n",
       "      <td>21</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>555</td>\n",
       "      <td>576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24710</th>\n",
       "      <td>24710</td>\n",
       "      <td>hsa-miR-12131</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>0.743590</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>1</td>\n",
       "      <td>1.509081e-02</td>\n",
       "      <td>0.250</td>\n",
       "      <td>-0.148416</td>\n",
       "      <td>0.599113</td>\n",
       "      <td>-4.170624</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1090</td>\n",
       "      <td>1107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24711</th>\n",
       "      <td>24711</td>\n",
       "      <td>hsa-miR-12131</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>0.884615</td>\n",
       "      <td>-20.7</td>\n",
       "      <td>1</td>\n",
       "      <td>1.197306e-05</td>\n",
       "      <td>0.353</td>\n",
       "      <td>0.143539</td>\n",
       "      <td>0.380902</td>\n",
       "      <td>-5.529864</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>615</td>\n",
       "      <td>638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24712</th>\n",
       "      <td>24712</td>\n",
       "      <td>hsa-miR-12133</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>0.717949</td>\n",
       "      <td>-20.2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.061500e-06</td>\n",
       "      <td>0.353</td>\n",
       "      <td>-0.147900</td>\n",
       "      <td>0.224627</td>\n",
       "      <td>-4.906073</td>\n",
       "      <td>19</td>\n",
       "      <td>30</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1561</td>\n",
       "      <td>1591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24713</th>\n",
       "      <td>24713</td>\n",
       "      <td>hsa-miR-12133</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>-16.4</td>\n",
       "      <td>1</td>\n",
       "      <td>1.459486e-06</td>\n",
       "      <td>0.206</td>\n",
       "      <td>0.327759</td>\n",
       "      <td>0.085926</td>\n",
       "      <td>-12.990572</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1052</td>\n",
       "      <td>1073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24714 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index            miRNA             mRNA  binding_probability  energy  \\\n",
       "0          0    hsa-let-7a-5p  ENST00000596924             1.000000   -22.3   \n",
       "1          1    hsa-let-7a-5p  ENST00000596924             0.923077   -20.9   \n",
       "2          2    hsa-let-7a-5p  ENST00000596924             0.846154   -20.1   \n",
       "3          3  hsa-let-7a-2-3p  ENST00000596924             1.000000   -23.2   \n",
       "4          4  hsa-let-7a-2-3p  ENST00000596924             1.000000   -21.4   \n",
       "...      ...              ...              ...                  ...     ...   \n",
       "24709  24709    hsa-miR-12131  ENST00000263100             1.000000   -25.1   \n",
       "24710  24710    hsa-miR-12131  ENST00000263100             0.743590   -20.6   \n",
       "24711  24711    hsa-miR-12131  ENST00000263100             0.884615   -20.7   \n",
       "24712  24712    hsa-miR-12133  ENST00000263100             0.717949   -20.2   \n",
       "24713  24713    hsa-miR-12133  ENST00000263100             0.615385   -16.4   \n",
       "\n",
       "       seed  accessibility  AU_content  PhyloP_Stem  PyloP_Flanking  \\\n",
       "0         1   1.489720e-02       0.441     0.000000        0.000000   \n",
       "1         1   1.141923e-01       0.471     0.000000        0.000000   \n",
       "2         1   8.216239e-03       0.382     0.000000        0.000000   \n",
       "3         0   9.643462e-07       0.382     0.000000        0.000000   \n",
       "4         0   7.641573e-06       0.456     0.000000        0.000000   \n",
       "...     ...            ...         ...          ...             ...   \n",
       "24709     1   2.668984e-04       0.382     0.333260        0.191636   \n",
       "24710     1   1.509081e-02       0.250    -0.148416        0.599113   \n",
       "24711     1   1.197306e-05       0.353     0.143539        0.380902   \n",
       "24712     1   1.061500e-06       0.353    -0.147900        0.224627   \n",
       "24713     1   1.459486e-06       0.206     0.327759        0.085926   \n",
       "\n",
       "            m/e   number_of_pairings  binding_region_length  \\\n",
       "0      -4.849714                  14                     16   \n",
       "1      -5.850804                  19                     22   \n",
       "2      -4.993964                  19                     25   \n",
       "3     -13.665600                  14                     18   \n",
       "4      -8.968752                  15                     18   \n",
       "...          ...                 ...                    ...   \n",
       "24709  -5.939295                  17                     21   \n",
       "24710  -4.170624                  15                     17   \n",
       "24711  -5.529864                  18                     23   \n",
       "24712  -4.906073                  19                     30   \n",
       "24713 -12.990572                  14                     21   \n",
       "\n",
       "       longest_consecutive_pairings  position_of_longest_consecutive_pairings  \\\n",
       "0                                14                                         1   \n",
       "1                                12                                         1   \n",
       "2                                 8                                         1   \n",
       "3                                11                                         1   \n",
       "4                                 9                                         9   \n",
       "...                             ...                                       ...   \n",
       "24709                            11                                         1   \n",
       "24710                             8                                         1   \n",
       "24711                             8                                         1   \n",
       "24712                             8                                         1   \n",
       "24713                             6                                         6   \n",
       "\n",
       "       pairings_in_3prime_end  \\\n",
       "0                           2   \n",
       "1                           7   \n",
       "2                           6   \n",
       "3                           5   \n",
       "4                           4   \n",
       "...                       ...   \n",
       "24709                       6   \n",
       "24710                       3   \n",
       "24711                       6   \n",
       "24712                       5   \n",
       "24713                       4   \n",
       "\n",
       "       difference_of_pairings_between_seed_and_3prime_end  bs_start  bs_end  \n",
       "0                                                      5       1695    1718  \n",
       "1                                                      1       1133    1155  \n",
       "2                                                      2        668     693  \n",
       "3                                                      2        381     399  \n",
       "4                                                      2        703     721  \n",
       "...                                                  ...        ...     ...  \n",
       "24709                                                  1        555     576  \n",
       "24710                                                  5       1090    1107  \n",
       "24711                                                  2        615     638  \n",
       "24712                                                  3       1561    1591  \n",
       "24713                                                  4       1052    1073  \n",
       "\n",
       "[24714 rows x 19 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print sqlite3 table \n",
    "qry = '''\n",
    "    select * from bs_tap\n",
    "    '''\n",
    "c.execute(qry)\n",
    "conn.commit()\n",
    "c.fetchall()\n",
    "pd.read_sql_query(qry,conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T22:54:19.697800Z",
     "start_time": "2020-10-20T22:54:19.602240Z"
    },
    "cell_style": "split",
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#DELETE table from sqlite3 DB\n",
    "c.execute(\"DROP TABLE bs_tap;\")\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:26:35.739119Z",
     "start_time": "2020-10-18T22:26:35.569777Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>raw_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>1</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12595</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12646</td>\n",
       "      <td>12697</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name  chrom_exon_start  chrom_exon_end strand  raw_counts\n",
       "index                                                                     \n",
       "0                   1             11874           12227      1        15.0\n",
       "1                   1             12595           12721      1         3.0\n",
       "2                   1             12613           12721      1         3.0\n",
       "3                   1             12646           12697      1         2.0\n",
       "4                   1             13221           14409      1        67.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts = n_all_exon_dfs\n",
    "exon_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = counts.rename(columns={'read_count': \"mirna_read_count\",'raw_counts':'exon_raw_counts','miRNA family name':'mirna_family'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:51:10.878073Z",
     "start_time": "2020-10-18T22:50:11.327509Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>raw_counts</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>chrom_bs_start</th>\n",
       "      <th>chrom_bs_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>58352928</td>\n",
       "      <td>58353197</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216603</td>\n",
       "      <td>2</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58352987</td>\n",
       "      <td>58353006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>58351391</td>\n",
       "      <td>58351687</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216600</td>\n",
       "      <td>3</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58351458</td>\n",
       "      <td>58351475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>58352283</td>\n",
       "      <td>58352555</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216602</td>\n",
       "      <td>4</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58352400</td>\n",
       "      <td>58352420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>58351391</td>\n",
       "      <td>58351687</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216600</td>\n",
       "      <td>5</td>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58351648</td>\n",
       "      <td>58351684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>58351391</td>\n",
       "      <td>58351687</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>216600</td>\n",
       "      <td>8</td>\n",
       "      <td>hsa-let-7d-5p</td>\n",
       "      <td>ENST00000600966</td>\n",
       "      <td>58351399</td>\n",
       "      <td>58351420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome_name  chrom_exon_start  chrom_exon_end strand  raw_counts  \\\n",
       "0              19          58352928        58353197     -1           0   \n",
       "1              19          58351391        58351687     -1           0   \n",
       "2              19          58352283        58352555     -1           0   \n",
       "3              19          58351391        58351687     -1           0   \n",
       "4              19          58351391        58351687     -1           0   \n",
       "\n",
       "   exon_id  bs_id          miRNA             mRNA  chrom_bs_start  \\\n",
       "0   216603      2  hsa-let-7b-5p  ENST00000600966        58352987   \n",
       "1   216600      3  hsa-let-7b-5p  ENST00000600966        58351458   \n",
       "2   216602      4  hsa-let-7b-5p  ENST00000600966        58352400   \n",
       "3   216600      5  hsa-let-7b-3p  ENST00000600966        58351648   \n",
       "4   216600      8  hsa-let-7d-5p  ENST00000600966        58351399   \n",
       "\n",
       "   chrom_bs_end  \n",
       "0      58353006  \n",
       "1      58351475  \n",
       "2      58352420  \n",
       "3      58351684  \n",
       "4      58351420  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#for a large data set, you will likely get a significant speed increase by creating an index for any column(s) used in the join condition. \n",
    "#https://stackoverflow.com/questions/30627968/merge-pandas-dataframes-where-one-value-is-between-two-others\n",
    "#instead use SQL\n",
    "conn = sqlite3.connect(':memory:') #Make the database in memory\n",
    "\n",
    "#write the tables\n",
    "counts.to_sql('bs', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "exon_counts.to_sql('counts', conn, index=True, index_label='exon_id', dtype={\"raw_counts\": 'INTEGER', \"exon_id\": 'INTEGER'})\n",
    "qry = '''\n",
    "    select\n",
    "        counts.chromosome_name, counts.chrom_exon_start, counts.chrom_exon_end, counts.strand, counts.raw_counts, counts.exon_id, bs.bs_id, bs.miRNA, bs.mRNA, bs.chrom_bs_start, bs.chrom_bs_end\n",
    "    from\n",
    "        bs inner join counts on\n",
    "        bs.chromosome_name = counts.chromosome_name and bs.strand = counts.strand and bs.chrom_bs_start >= counts.chrom_exon_start and bs.chrom_bs_end <= counts.chrom_exon_end\n",
    "    '''\n",
    "#only return bs with minimum 1 exon_id cause inner join (else left join)\n",
    "#automatically filters out all binding sites that have more than 1 exon_id, if u sont want that use outer join -> actually wrong, it uses the first instead of leaving it out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T00:31:04.522842Z",
     "start_time": "2020-10-13T00:31:04.432995Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#visualize amount exons by bindingsite | amount bindingsites by exon\n",
    "import sqlite3\n",
    "conn = sqlite3.connect(':memory:') #Make the database in memory\n",
    "\n",
    "#write the tables\n",
    "important_counts.to_sql('bs', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "new_exon_counts.to_sql('counts', conn, index=True, index_label='exon_id', dtype={\"exon_id\": 'INTEGER'})\n",
    "qry = '''\n",
    "    select\n",
    "        counts.chromosome_name, counts.chrom_exon_start, counts.chrom_exon_end, counts.strand, counts.raw_counts, counts.exon_id, bs.bs_id, bs.miRNA, bs.mRNA, bs.chrom_bs_start, bs.chrom_bs_end\n",
    "    from\n",
    "        bs left join counts on\n",
    "        bs.chromosome_name = counts.chromosome_name and bs.strand = counts.strand and bs.chrom_bs_start >= counts.chrom_exon_start and bs.chrom_bs_end <= counts.chrom_exon_end\n",
    "    '''\n",
    "bs_exons2 = pd.read_sql_query(qry, conn)\n",
    "bs_exons2.head()\n",
    "#bs_exons2.groupby('bs_id').count().raw_counts.hist()\n",
    "#bs_exons2.groupby('exon_id').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Put all counts into one & pivot table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO put all different caseids sqlite tables together into one\n",
    "#INSERT INTO artists_backup \n",
    "#SELECT ArtistId, Name\n",
    "#FROM artists;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.071562Z",
     "start_time": "2020-10-18T22:50:38.700Z"
    }
   },
   "outputs": [],
   "source": [
    "#output: for each case_id for each bs_id: interesting miRNA expression + exon expression\n",
    "#unstack family counts from 2 columns to several columns (mirna_family=name,mirna_read_count=value)\n",
    "pivoted = bs_mirnas.copy()\n",
    "pivoted = pivoted.pivot_table(values='mirna_read_count', index=['exon_id', 'exon_raw_counts'], columns='mirna_family', aggfunc='sum', fill_value=0)\n",
    "pivoted#.head()\n",
    "#TODO debug does this do it right OR use sqllite3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO pivot in sqlite3 \n",
    "#https://modern-sql.com/use-case/pivot\n",
    "#https://stackoverflow.com/questions/1237068/how-to-pivot-in-sqlite-or-i-e-select-in-wide-format-a-table-stored-in-long-form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#visualize amount exons by bindingsite\n",
    "#bs_exons.groupby('bs_id').count().raw_counts.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:40:44.254071Z",
     "start_time": "2020-10-13T08:40:44.166104Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8, 1680) (8,)\n"
     ]
    }
   ],
   "source": [
    "#TODO check - is nr mirna downloaded also here? if not: mistake\n",
    "print(X.shape,y[0].shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Elastic Net Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-18T22:52:43.081556Z",
     "start_time": "2020-10-18T22:50:43.000Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.getsizeof(pivoted)#TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-13T08:42:05.340886Z",
     "start_time": "2020-10-13T08:42:05.198969Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0. 0. 0. ... 0. 0. 0.]\n",
      "0.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Lena\\Miniconda3\\envs\\thesis\\lib\\site-packages\\sklearn\\linear_model\\_coordinate_descent.py:529: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 0.0, tolerance: 0.0\n",
      "  model = cd_fast.enet_coordinate_descent(\n"
     ]
    }
   ],
   "source": [
    "#regression btw miRNA expression, exon expression\n",
    "import sklearn\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.datasets import make_regression\n",
    "#faster if np.array(x, order='F')\n",
    "#If you are interested in controlling the L1 and L2 penalty separately, keep in mind that this is equivalent to:\n",
    "#a * L1 + b * L2 where: alpha = a + b and l1_ratio = a / (a + b)\n",
    "X = np.array(pivoted)\n",
    "y = np.array(pivoted.index.get_level_values('exon_raw_counts'))\n",
    "\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "print(regr.coef_)\n",
    "print(regr.intercept_)"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "Seqvec basic Seq ligand binding.ipynb",
   "provenance": []
  },
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python (thesis)",
   "language": "python",
   "name": "thesis"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "581.6px",
    "left": "235px",
    "top": "158.533px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
