{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in bindingsite & Ensembl data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-27T12:43:04.072027Z",
     "start_time": "2020-10-27T12:43:03.334014Z"
    },
    "cell_style": "center"
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from helper_fcts import *\n",
    "import sqlite3\n",
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "test = False\n",
    "path = Path('data')\n",
    "tcga_path = Path(path/'KICH')\n",
    "path_tarp = Path(path/'tarp-bs') if test else Path('../data')\n",
    "path_input = Path(path/'tarp-bs/cdna') if test else Path('../data/all_cdna_subsets')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test bindingsite big\n",
    "bs = pd.read_parquet(path/'bs_chrom_pos1.parquet')\n",
    "bs = bs[['miRNA','mRNA','chrom_bs_start','chrom_bs_end']]\n",
    "bs['miRNA'] = pd.Categorical(bs.miRNA)\n",
    "bs['mRNA'] = pd.Categorical(bs.mRNA)\n",
    "bs['bs_id'] = bs.index\n",
    "df_gene = pd.read_parquet(path/'ensembl_transcript_info.parquet')\n",
    "bs = bs.merge(df_gene, left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "del bs['ensembl_transcript_id']\n",
    "bs['Strand'] = '-'\n",
    "bs.loc[bs['strand'] == '1', 'Strand'] = '+'\n",
    "del bs['strand']\n",
    "bs['Strand'] = pd.Categorical(bs.Strand)\n",
    "#bs[bs.isnull().values] #problem: 20 572 214 / 99 999 996 chromosome NaN -> cause they spanned more than 1 exon\n",
    "bs = bs.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_counts = pd.read_parquet(path/'KICH'/'temp_exon_counts.parquet') \n",
    "exon_counts['temp'] = '-'\n",
    "exon_counts.loc[exon_counts['strand'] == '1', 'temp'] = '+'\n",
    "exon_counts['strand'] = exon_counts['temp'] \n",
    "exon_counts['strand'] = pd.Categorical(exon_counts.strand)\n",
    "exon_counts = exon_counts.drop(['temp','ensembl_transcript_id','exon_count','exon_id'],axis=1)\n",
    "exon_counts['e_id'] = exon_counts.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirna_counts = pd.read_csv(tcga_path/'miRNA_HiSeq_gene', delimiter='\t')\n",
    "caseids = list(set(mirna_counts.columns).intersection(list(exon_counts.columns)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sqlite_join_1():\n",
    "    conn = sqlite3.connect('speed.db') #Make the database in memory\n",
    "    c = conn.cursor()\n",
    "    #write the tables\n",
    "    bs.to_sql('counts', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "    exon_counts[['chromosome_name','strand','chrom_exon_start','chrom_exon_end','patient_id','exon_count','exon_id']].to_sql('exon_counts', conn, dtype={caseid : 'INTEGER' for caseid in caseids}.update({\"chrom_exon_start\": 'INTEGER', \"chrom_exon_end\": 'INTEGER'}))\n",
    "    qry = '''\n",
    "        CREATE TABLE n_counts AS\n",
    "        select\n",
    "            exon_id, bs_id, miRNA, patient_id, exon_count\n",
    "        from\n",
    "            exon_counts inner join counts\n",
    "            on\n",
    "            counts.chromosome_name = exon_counts.chromosome_name and counts.strand = exon_counts.strand and counts.chrom_bs_start >= exon_counts.chrom_exon_start and counts.chrom_bs_end <= exon_counts.chrom_exon_end\n",
    "        '''\n",
    "    print('start query')\n",
    "    return pd.read_sql_query(qry,conn)\n",
    "    #only returns bs with minimum 1 exon_id cause inner join (else left join)\n",
    "    \n",
    "def sqlite_join_2():\n",
    "    conn = sqlite3.connect('speed.db') #Make the database in memory\n",
    "    c = conn.cursor()\n",
    "    #write the tables\n",
    "    bs.to_sql('counts', conn, index=False, dtype={\"bs_id\": 'INTEGER'})\n",
    "    exon_counts[['ensembl_transcript_id','chrom_exon_start','chrom_exon_end','patient_id','exon_count','exon_id']].to_sql('exon_counts', conn, dtype={caseid : 'INTEGER' for caseid in caseids}.update({\"chrom_exon_start\": 'INTEGER', \"chrom_exon_end\": 'INTEGER'}))\n",
    "    qry = '''\n",
    "        CREATE TABLE n_counts AS\n",
    "        select\n",
    "            exon_id, bs_id, miRNA, patient_id, exon_count\n",
    "        from\n",
    "            exon_counts inner join counts\n",
    "            on\n",
    "            counts.mRNA = exon_counts.ensembl_transcript_id and counts.chrom_bs_start >= exon_counts.chrom_exon_start and counts.chrom_bs_end <= exon_counts.chrom_exon_end\n",
    "        '''\n",
    "    print('start query')\n",
    "    return pd.read_sql_query(qry,conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start query\n"
     ]
    }
   ],
   "source": [
    "%timeit new_bs = sqlite_join_1() \n",
    "#30 - 56 GB - 34 \n",
    "#2.18 - 2.24 (start query) - > 1 hour"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_bs.to_parquet(path/'bs_chrom_exon_pos1.parquet')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Line magic function `%memit` not found.\n"
     ]
    }
   ],
   "source": [
    "%memit new_bs = sqlite_join()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>exon_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>51973965</td>\n",
       "      <td>51974630</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>2.1156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>215728030</td>\n",
       "      <td>215728384</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>51191601</td>\n",
       "      <td>51193016</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>0.3884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>146292929</td>\n",
       "      <td>146292980</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>2.5290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>15747546</td>\n",
       "      <td>15747980</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325745</th>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>51978259</td>\n",
       "      <td>51978374</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>4.1126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325746</th>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>43357126</td>\n",
       "      <td>43357765</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>0.2725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325747</th>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>36168630</td>\n",
       "      <td>36168695</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>1.9102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325748</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>120833602</td>\n",
       "      <td>120833774</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>1.1468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325749</th>\n",
       "      <td>18</td>\n",
       "      <td>-</td>\n",
       "      <td>12353005</td>\n",
       "      <td>12353158</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>3.4175</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30325750 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chromosome_name strand  chrom_exon_start  chrom_exon_end  \\\n",
       "0                      3      -          51973965        51974630   \n",
       "1                      1      -         215728030       215728384   \n",
       "2                     17      -          51191601        51193016   \n",
       "3                      4      -         146292929       146292980   \n",
       "4                      1      +          15747546        15747980   \n",
       "...                  ...    ...               ...             ...   \n",
       "30325745               3      +          51978259        51978374   \n",
       "30325746               6      -          43357126        43357765   \n",
       "30325747              17      -          36168630        36168695   \n",
       "30325748               1      +         120833602       120833774   \n",
       "30325749              18      -          12353005        12353158   \n",
       "\n",
       "               patient_id  exon_count  \n",
       "0         TCGA-KN-8419-01      2.1156  \n",
       "1         TCGA-KN-8419-01      0.0000  \n",
       "2         TCGA-KN-8419-01      0.3884  \n",
       "3         TCGA-KN-8419-01      2.5290  \n",
       "4         TCGA-KN-8419-01      0.0000  \n",
       "...                   ...         ...  \n",
       "30325745  TCGA-KL-8327-01      4.1126  \n",
       "30325746  TCGA-KL-8327-01      0.2725  \n",
       "30325747  TCGA-KL-8327-01      1.9102  \n",
       "30325748  TCGA-KL-8327-01      1.1468  \n",
       "30325749  TCGA-KL-8327-01      3.4175  \n",
       "\n",
       "[30325750 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs = parse_tarp_bs(path/'tarp-bs'/'input_miRNA.fasta_cdna6247.fasta.bp')\n",
    "bs = bs[['miRNA','mRNA','bs_start','bs_end']]\n",
    "bs['miRNA'] = pd.Categorical(bs.miRNA)\n",
    "bs['mRNA'] = pd.Categorical(bs.mRNA)\n",
    "bs['bs_id'] = bs.index\n",
    "#bs.info() # 2.6GB\n",
    "df_gene = pd.read_parquet(path/'ensembl_transcript_info.parquet')\n",
    "bs = bs.merge(df_gene, left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "bs['mRNA'] = pd.Categorical(bs.mRNA)\n",
    "del bs['ensembl_transcript_id']\n",
    "bs['Strand'] = '-'\n",
    "bs.loc[bs['strand'] == '1', 'Strand'] = '+'\n",
    "del bs['strand']\n",
    "bs['Strand'] = pd.Categorical(bs.Strand)\n",
    "\n",
    "exon_counts = pd.read_parquet(path/'KICH'/'temp_exon_counts.parquet') \n",
    "exon_counts['temp'] = '-'\n",
    "exon_counts.loc[exon_counts['strand'] == '1', 'temp'] = '+'\n",
    "exon_counts['strand'] = exon_counts['temp'] \n",
    "exon_counts['strand'] = pd.Categorical(exon_counts.strand)\n",
    "exon_counts = exon_counts.drop(['temp','ensembl_transcript_id','exon_id'],axis=1)\n",
    "exon_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bedtool_join():    \n",
    "    import pybedtools\n",
    "    from pybedtools import BedTool\n",
    "    print('convert data to BedTool')\n",
    "    bs_bed = bs[['chromosome_name', 'bs_start', 'bs_end', 'bs_id','miRNA','Strand']].copy()\n",
    "    exon_bed = exon_counts[['chromosome_name', 'chrom_exon_start', 'chrom_exon_end', 'patient_id','exon_count','strand']].copy()\n",
    "    bs_bed = pybedtools.BedTool.from_dataframe(bs_bed, header=['chrom','start', 'end', 'name', 'score','strand'])\n",
    "    exon_bed = pybedtools.BedTool.from_dataframe(exon_bed, header=['chrom','start', 'end', 'name', 'score','strand'])\n",
    "    print('start join')\n",
    "    return bs_bed.intersect(exon_bed)#, loj=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, stop, name, score, strand]\n",
       "Index: []"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#new_bs = bedtool_join() \n",
    "jo = pd.read_table(new_bs.fn, names=['chrom', 'start', 'stop', 'name', 'score', 'strand'])\n",
    "jo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_bs = bedtool_join().to_dataframe()\n",
    "new_bs.to_parquet(path/'mini_bs_exon_bedtool.parquet')\n",
    "#34 - \n",
    "#2.39-2.41 to convert to bed files\n",
    "#3.20-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#small bs #TODO chromosoooomeeee\n",
    "bs = parse_tarp_bs(path/'tarp-bs'/'input_miRNA.fasta_cdna6247.fasta.bp')\n",
    "bs['bs_id'] = bs.index\n",
    "df_gene = pd.read_parquet(path/'ensembl_transcript_info.parquet')\n",
    "bs = bs.merge(df_gene, left_on='mRNA', right_on='ensembl_transcript_id', how='left')\n",
    "bs = bs[['miRNA','mRNA','bs_start','bs_end','bs_id','chromosome_name','strand']]\n",
    "new_row = {'miRNA':'hsa-let-7a-2-3p', 'mRNA':'ENST00000439513', 'bs_start':1517259.0, 'bs_end':1517304.0, 'bs_id':73193,'chromosome_name': '1','strand': '1'}\n",
    "bs = bs.append(new_row, ignore_index=True)\n",
    "bs['Strand'] = '-'\n",
    "bs.loc[bs['strand'] == '1', 'Strand'] = '+'\n",
    "del bs['strand']\n",
    "bs['chromosome_name'] = pd.Categorical(bs.chromosome_name)\n",
    "bs['Strand'] = pd.Categorical(bs.Strand)\n",
    "#bs[bs.isnull().values] #problem: 20 572 214 / 99 999 996 chromosome NaN -> cause they spanned more than 1 exon\n",
    "bs = bs.dropna()\n",
    "\n",
    "exon_counts = pd.read_parquet(path/'KICH'/'temp_exon_counts.parquet') \n",
    "exon_counts['temp'] = '-'\n",
    "exon_counts.loc[exon_counts['strand'] == '1', 'temp'] = '+'\n",
    "exon_counts['strand'] = exon_counts['temp'] \n",
    "exon_counts['strand'] = pd.Categorical(exon_counts.strand)\n",
    "exon_counts = exon_counts.drop(['temp','ensembl_transcript_id','exon_id'],axis=1)#'exon_count'\n",
    "#exon_counts = exon_counts.dropna() #no NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>chrom_exon_start</th>\n",
       "      <th>chrom_exon_end</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>exon_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>51973965</td>\n",
       "      <td>51974630</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>2.1156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>215728030</td>\n",
       "      <td>215728384</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>51191601</td>\n",
       "      <td>51193016</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>0.3884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>146292929</td>\n",
       "      <td>146292980</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>2.5290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>15747546</td>\n",
       "      <td>15747980</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325745</th>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>51978259</td>\n",
       "      <td>51978374</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>4.1126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325746</th>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>43357126</td>\n",
       "      <td>43357765</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>0.2725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325747</th>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>36168630</td>\n",
       "      <td>36168695</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>1.9102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325748</th>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>120833602</td>\n",
       "      <td>120833774</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>1.1468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30325749</th>\n",
       "      <td>18</td>\n",
       "      <td>-</td>\n",
       "      <td>12353005</td>\n",
       "      <td>12353158</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>3.4175</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30325750 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chromosome_name strand  chrom_exon_start  chrom_exon_end  \\\n",
       "0                      3      -          51973965        51974630   \n",
       "1                      1      -         215728030       215728384   \n",
       "2                     17      -          51191601        51193016   \n",
       "3                      4      -         146292929       146292980   \n",
       "4                      1      +          15747546        15747980   \n",
       "...                  ...    ...               ...             ...   \n",
       "30325745               3      +          51978259        51978374   \n",
       "30325746               6      -          43357126        43357765   \n",
       "30325747              17      -          36168630        36168695   \n",
       "30325748               1      +         120833602       120833774   \n",
       "30325749              18      -          12353005        12353158   \n",
       "\n",
       "               patient_id  exon_count  \n",
       "0         TCGA-KN-8419-01      2.1156  \n",
       "1         TCGA-KN-8419-01      0.0000  \n",
       "2         TCGA-KN-8419-01      0.3884  \n",
       "3         TCGA-KN-8419-01      2.5290  \n",
       "4         TCGA-KN-8419-01      0.0000  \n",
       "...                   ...         ...  \n",
       "30325745  TCGA-KL-8327-01      4.1126  \n",
       "30325746  TCGA-KL-8327-01      0.2725  \n",
       "30325747  TCGA-KL-8327-01      1.9102  \n",
       "30325748  TCGA-KL-8327-01      1.1468  \n",
       "30325749  TCGA-KL-8327-01      3.4175  \n",
       "\n",
       "[30325750 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start join\n",
      "13 : 28 - 13 : 27 = 0 : 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Strand_b</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>exon_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>2.9478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KL-8346-01</td>\n",
       "      <td>2.3182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KN-8422-01</td>\n",
       "      <td>2.3118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KN-8431-11</td>\n",
       "      <td>2.9785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KN-8430-11</td>\n",
       "      <td>3.1038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KO-8408-01</td>\n",
       "      <td>2.5630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KM-8443-01</td>\n",
       "      <td>2.3830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KM-8442-01</td>\n",
       "      <td>2.0297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KL-8332-11</td>\n",
       "      <td>1.8481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>ENST00000439513</td>\n",
       "      <td>1517259</td>\n",
       "      <td>1517304</td>\n",
       "      <td>73193</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1517167</td>\n",
       "      <td>1517412</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>1.0967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              miRNA             mRNA    Start      End  bs_id Chromosome  \\\n",
       "0   hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "1   hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "2   hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "3   hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "4   hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "..              ...              ...      ...      ...    ...        ...   \n",
       "86  hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "87  hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "88  hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "89  hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "90  hsa-let-7a-2-3p  ENST00000439513  1517259  1517304  73193          1   \n",
       "\n",
       "   Strand Strand_b  Start_b    End_b       patient_id  exon_count  \n",
       "0       +        +  1517167  1517412  TCGA-KN-8419-01      2.9478  \n",
       "1       +        +  1517167  1517412  TCGA-KL-8346-01      2.3182  \n",
       "2       +        +  1517167  1517412  TCGA-KN-8422-01      2.3118  \n",
       "3       +        +  1517167  1517412  TCGA-KN-8431-11      2.9785  \n",
       "4       +        +  1517167  1517412  TCGA-KN-8430-11      3.1038  \n",
       "..    ...      ...      ...      ...              ...         ...  \n",
       "86      +        +  1517167  1517412  TCGA-KO-8408-01      2.5630  \n",
       "87      +        +  1517167  1517412  TCGA-KM-8443-01      2.3830  \n",
       "88      +        +  1517167  1517412  TCGA-KM-8442-01      2.0297  \n",
       "89      +        +  1517167  1517412  TCGA-KL-8332-11      1.8481  \n",
       "90      +        +  1517167  1517412  TCGA-KL-8327-01      1.0967  \n",
       "\n",
       "[91 rows x 12 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def pyranges_join():\n",
    "    import pyranges as pr\n",
    "    #TODO for parallel ray \n",
    "    \n",
    "    # Add column names to the dfs so PyRanges knows which \n",
    "    # represent Chromosomes, Starts, Ends and optionally Strand\n",
    "    ranges_bs = bs.copy()\n",
    "    ranges_bs.columns = \"miRNA mRNA Start End bs_id Chromosome Strand\".split()\n",
    "    ranges_exon = exon_counts.copy()\n",
    "    ranges_exon.columns = \"Chromosome Strand Start End patient_id exon_count\".split()\n",
    "    import datetime\n",
    "    now = datetime.datetime.now()\n",
    "    hour = int(now.strftime('%H'))\n",
    "    minu = int(now.strftime('%M'))\n",
    "    # create PyRanges-objects from the dfs\n",
    "    gr1, gr2 = pr.PyRanges(ranges_bs), pr.PyRanges(ranges_exon)\n",
    "    print('start join')\n",
    "    # intersect the two\n",
    "    gr = gr1.join(gr2)\n",
    "    #gr = gr1.intersect(gr2)#, nb_cpu=2)#, how='containment') returns all overlapping bs\n",
    "\n",
    "    now = datetime.datetime.now()\n",
    "    hour2 = int(now.strftime('%H'))\n",
    "    minu2 = int(now.strftime('%M'))\n",
    "    print(hour2,':',minu2,'-',hour,':',minu,'=',hour2-hour,':',minu2-minu)\n",
    "    # get the result as a pandas df\n",
    "    return gr.df\n",
    "new_bs = pyranges_join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>bs_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>exon_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KN-8419-01</td>\n",
       "      <td>2.9478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KL-8346-01</td>\n",
       "      <td>2.3182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KN-8422-01</td>\n",
       "      <td>2.3118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KN-8431-11</td>\n",
       "      <td>2.9785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KN-8430-11</td>\n",
       "      <td>3.1038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KO-8408-01</td>\n",
       "      <td>2.5630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KM-8443-01</td>\n",
       "      <td>2.3830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KM-8442-01</td>\n",
       "      <td>2.0297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KL-8332-11</td>\n",
       "      <td>1.8481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>73193</td>\n",
       "      <td>TCGA-KL-8327-01</td>\n",
       "      <td>1.0967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              miRNA  bs_id       patient_id  exon_count\n",
       "0   hsa-let-7a-2-3p  73193  TCGA-KN-8419-01      2.9478\n",
       "1   hsa-let-7a-2-3p  73193  TCGA-KL-8346-01      2.3182\n",
       "2   hsa-let-7a-2-3p  73193  TCGA-KN-8422-01      2.3118\n",
       "3   hsa-let-7a-2-3p  73193  TCGA-KN-8431-11      2.9785\n",
       "4   hsa-let-7a-2-3p  73193  TCGA-KN-8430-11      3.1038\n",
       "..              ...    ...              ...         ...\n",
       "86  hsa-let-7a-2-3p  73193  TCGA-KO-8408-01      2.5630\n",
       "87  hsa-let-7a-2-3p  73193  TCGA-KM-8443-01      2.3830\n",
       "88  hsa-let-7a-2-3p  73193  TCGA-KM-8442-01      2.0297\n",
       "89  hsa-let-7a-2-3p  73193  TCGA-KL-8332-11      1.8481\n",
       "90  hsa-let-7a-2-3p  73193  TCGA-KL-8327-01      1.0967\n",
       "\n",
       "[91 rows x 4 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_bs = pyranges_join()\n",
    "new_bs.to_parquet(path/'bs_pos1_chrom_exon_pyranges.parquet')\n",
    "new_bs = new_bs.df.drop(['mRNA','Start','End','Chromosome','Strand','Strand_b','Start_b','End_b'],axis=1)\n",
    "new_bs.to_parquet(path/'bs_pos1_chrom_exon_subset_pyranges.parquet')\n",
    "new_bs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "Seqvec basic Seq ligand binding.ipynb",
   "provenance": []
  },
  "hide_input": false,
  "kernelspec": {
   "display_name": "thesis",
   "language": "python",
   "name": "thesis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "581.6px",
    "left": "235px",
    "top": "158.533px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
